---
title: "BEPmetagenomic_majorRevision_github"
author: "Lishi"
date: "2024-10-22"
output: html_document
---

# about this markdown
This markdown combines all extra statistic steps (major revision) of metagenomic analysis we did for detecting the BEP effect on gut microbiol composition and functionality.
Metagenomic sequencing reads mapping and gene-level profiling were performed using inStrain in Python. Therefore these scripts were not included in this markdown, but the demo for use are available at https://github.com/MrOlm/inStrain.
Therefore, this markdown contains visualization of relative abundance, analysis of alpha diversity, beta diversity, taxonomic differential abundance analysis, mediation analysis and gene set enrichment analysis. The source data are available along with this markdown.

# setup
## working directory
First, set the working directory to your folder and set random seed.

```{r echo=FALSE}
# set the work directory
knitr::opts_knit$set(
  root.dir = '~/.../...', 
  echo=TRUE
  )
message(paste0("The current working directory is set to: ", getwd()))

set.seed(123)
```

## packages
Load necessary packages before analysis.

```{r echo=FALSE, message=FALSE}
library(ANCOMBC)
library(clusterProfiler)
library(ComplexHeatmap)
library(dbarts)
library(DESeq2)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(haven)
library(hrbrthemes)
library(lme4)
library(medoutcon)
library(mixtools)
library(permFDP)
library(phyloseq)
library(plotly)
library(randomForest)
library(sl3)
library(SuperLearner)
library(tibble)
library(tidyr)
library(tmle)
library(treeclimbR)
library(UpSetR)
library(merTools)
library(broom.mixed) # For tidying model output
library(scales)
library(igraph)
library(SpiecEasi)
library(dbarts)
library(doFuture)
library(DT)
options(DT.options = list(
  initComplete = JS("function(settings, json) {",
  "$(this.api().table().header()).css({'background-color': 
  '#000', 'color': '#fff'});","}")))


```


# import data
## read data
Save the data files provided first, and read them:
```{r}
metadata <- read.csv("./metadata.csv", row.names = 1) %>%
  mutate(bep_group = ifelse(timepoint=="incl" & code=="A Y", "Mother_incl_IFA",
                     ifelse(timepoint=="incl" & code=="B Z", "Mother_incl_BEP",
                     ifelse(timepoint=="tri3" & code=="A Y", "Mother_tri3_IFA",
                     ifelse(timepoint=="tri3" & code=="B Z", "Mother_tri3_BEP",
                     ifelse(timepoint=="pn12" & child=="no" & code=="A Y", "Mother_pn12_IFA",
                     ifelse(timepoint=="pn12" & child=="no" & code=="B Z", "Mother_pn12_BEP",
                     ifelse(timepoint=="pn12" & child=="yes" & code=="A Y", "Infant_pn12_IFA",
                     ifelse(timepoint=="pn12" & child=="yes" & code=="B Z", "Infant_pn12_BEP",
                     ifelse(timepoint=="pn56" & child=="no" & code=="A Y", "Mother_pn56_IFA",
                     ifelse(timepoint=="pn56" & child=="no" & code=="B Z", "Mother_pn56_BEP",
                     ifelse(timepoint=="pn56" & child=="yes" & code=="A Y", "Infant_pn56_IFA",
                     ifelse(timepoint=="pn56" & child=="yes" & code=="B Z", "Infant_pn56_BEP","other"))))))))))))) # grouping, dyads and time point information

taxonomy <- as.matrix(read.csv("./taxonomy_clean.csv", row.names = 1))
absolute_counts <- read.csv("./absolute_counts_table.csv", row.names = 1)

# read gene tables
gene_counts <- read.csv("./gene_counts_clean.csv", row.names = 1)
gene_anno <- read.csv("./gene_annotations_clean.csv", row.names = 1)
gene_to_pathway <- read.csv("./k2pathway.txt", sep="\t", header=FALSE)
names(gene_to_pathway) <- c("gene", "pathway")
gene_to_pathway$gene <- gsub(pattern = "^ko:", "", gene_to_pathway$gene)
gene_to_pathway$pathway <- gsub(pattern = "^path:", "", gene_to_pathway$pathway)
pathway_to_gene <- gene_to_pathway[, c("pathway", "gene")]

pathway_to_description <- read.csv("./keggmaps_to_name.txt", sep="\t", header=FALSE, colClasses = c("character", "character"))
pathway_to_description$V1 <- gsub(pattern = "^", "map", pathway_to_description$V1)

microbial_pathways <- read.csv("./microbial_ko_table.tsv", sep="\t")

# the data of sample collection dates
pn12e_se1g <- haven::read_dta("/Users/lishideng/Library/CloudStorage/OneDrive-SharedLibraries-UGent/BioSpé - General/From DropBox (7. Misame-III BioSpé)/data/cleandata/pn12/pn12e/pn12e.dta")

pn12m_se1g <- haven::read_dta("/Users/lishideng/Library/CloudStorage/OneDrive-SharedLibraries-UGent/BioSpé - General/From DropBox (7. Misame-III BioSpé)/data/cleandata/pn12/pn12m/pn12m.dta")

pn56e_se1g <- haven::read_dta("/Users/lishideng/Library/CloudStorage/OneDrive-SharedLibraries-UGent/BioSpé - General/From DropBox (7. Misame-III BioSpé)/data/cleandata/pn56/pn56e/pn56e.dta")

pn56m_se1g <- haven::read_dta("/Users/lishideng/Library/CloudStorage/OneDrive-SharedLibraries-UGent/BioSpé - General/From DropBox (7. Misame-III BioSpé)/data/cleandata/pn56/pn56m/pn56m.dta")

incl_se1g <- haven::read_dta("/Users/lishideng/Library/CloudStorage/OneDrive-SharedLibraries-UGent/BioSpé - General/From DropBox (7. Misame-III BioSpé)/data/cleandata/incl/incl.dta")

tri3_se1g <- haven::read_dta("/Users/lishideng/Library/CloudStorage/OneDrive-SharedLibraries-UGent/BioSpé - General/From DropBox (7. Misame-III BioSpé)/data/cleandata/tri3/tri3.dta")

```

## data preparation
```{r}
tax_for_tree <- cbind(taxonomy[, 1:ncol(taxonomy)-1], row.names(taxonomy))
proxy_phylogenetic_tree <- TreeSummarizedExperiment::toTree(tax_for_tree)

ps <- phyloseq(
  otu_table=otu_table(absolute_counts, taxa_are_rows=TRUE), 
  tax_table=tax_table(taxonomy),
  sample_data=sample_data(metadata), 
  phy_tree=phy_tree(proxy_phylogenetic_tree)
  )

# normalize data
pslog <- transform_sample_counts(ps, function(x) log(1 + x))
```

## data subsets
```{r}
incl <- subset_samples(ps, timepoint=="incl")
tri3 <- subset_samples(ps, timepoint=="tri3")
pn12 <- subset_samples(ps, timepoint=="pn12")
pn12m <- subset_samples(ps, timepoint=="pn12" & child=="no")
pn12e <- subset_samples(ps, timepoint=="pn12" & child=="yes")
pn56 <- subset_samples(ps, timepoint=="pn56")
pn56m <- subset_samples(ps, timepoint=="pn56" & child=="no")
pn56e <- subset_samples(ps, timepoint=="pn56" & child=="yes")
# add label to "code" and "child: "A Y" == "IFA", "B Z" == "BEP"; "no" == "Mother", "yes" == "Child"
code.labs <- c("A Y" = "IFA", "B Z" = "BEP")
child.labs <- c("no" = "Mother", "yes" = "Child")

#use log for beta diversity
infant <- subset_samples(pslog, child=="yes")
mother <- subset_samples(pslog, child=="no")
prenatal <- subset_samples(mother, timepoint=="incl" | timepoint=="tri3")
postnatal <- subset_samples(mother, timepoint=="pn12" | timepoint=="pn56")

```

# Figure 2. 
## Relative abundance genus

```{r message = FALSE}

genus_name <- c("Prevotella",  "Bifidobacterium", "Escherichia", "RC9", "Bacteroides", "CAG-83", "ER4", "Faecalibacterium",  "Prevotellamassilia", "CAG-170", "Veillonella_A", "C941", "Klebsiella", "Dialister", "Agathobacter", "Other")
genus_color <- c("#2986cc", "#cd0000", "#f1c232", "#6aa84f",  "#228585","#f4cccc", "#00ffff","#e69138", "#cfe2f3", "#d0f679", "orchid4", "salmon", "thistle", "#104E8E", "#8b7355",  "gray90")

ps_genus <- ps %>%
  tax_glom(taxrank = "genus") %>%
  transform_sample_counts(function(x){x/sum(x)}) %>%
  psmelt() %>%
  filter(Abundance >0)

sort.genus.ps <- ps_genus %>%
  dplyr::count(genus, wt = Abundance) %>%
  arrange(desc(n)) %>%
  slice_head(n = 15) %>%
  pull(genus)

sample.order.ps <- ps_genus %>%
  filter(phylum == sort.genus.ps[1]) %>%
  arrange (desc(Abundance)) %>%
  pull(Sample)

ps_genus_plot <- ps_genus %>%
  mutate(Sample = factor(Sample, levels = sample.order.ps)) %>%
  mutate(genus = factor(genus, levels = rev(sort.genus.ps))) %>%
  ggplot(aes(x=bep_group, y=Abundance)) + 
  geom_bar(aes(fill=genus), stat = 'identity', position = 'fill') + 
  theme(axis.text.x = element_text(angle = 90, size = 7), legend.title = element_text(size=8), legend.key.size = unit(0.3, 'cm'), legend.text = element_text(size = 8)) + 
  labs(x="Sample at different timepoints and treatment group", y="Abundance", fill = "Genera") + 
  scale_fill_manual(breaks=genus_name, values = genus_color) + 
  theme(panel.grid = element_blank(), panel.border = element_blank()) +   
  theme(panel.background = element_blank(),
        axis.ticks.x=element_blank())+ 
  guides(fill=guide_legend(ncol = 1))
ps_genus_plot


svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_revision1/ps_genus_plot.svg",
  width=6, height=5
  )
ps_genus_plot
dev.off()

```


### species counts

#### infant

```{r}

infant_species <- infant_ps %>%
  tax_glom(taxrank = "species") %>%
  transform_sample_counts(function(x) x / sum(x)) %>%
  psmelt() %>%
  filter(Abundance > 0)


# Count the number of unique species for each subject
species_counts_BEP <- infant_species %>%
  mutate(timepoint_n = ifelse(timepoint=="pn12", 0, 1)) %>%
  filter(code == "B Z") %>%
  group_by(idbs, timepoint) %>%
  summarise(num_species = n_distinct(OTU),
            timepoint_n = first(timepoint_n))

infantSpeciesNumberBEP <- ggplot(species_counts_BEP, aes(x = num_species, fill = factor(timepoint_n))) +
  geom_density(alpha = 0.5) +
  labs(x = "Number of Species", y = "Density", fill = "Timepoint", title = "Intervention Group") +
  scale_fill_manual(values = c("skyblue", "#f1c232"), labels = c("Pn12", "Pn56")) +
  theme_minimal() +
  theme(panel.border = element_rect(color = "black", fill = NA),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) + xlim (2, 37)
infantSpeciesNumberBEP

# Count the number of unique species for each subject
species_counts_IFA <- infant_species %>%
  mutate(timepoint_n = ifelse(timepoint=="pn12", 0, 1)) %>%
  filter(code == "A Y") %>%
  group_by(idbs, timepoint) %>%
  summarise(num_species = n_distinct(OTU),
            timepoint_n = first(timepoint_n))

infantSpeciesNumberIFA <- ggplot(species_counts_IFA, aes(x = num_species, fill = factor(timepoint_n))) +
  geom_density(alpha = 0.5) +
  labs(x = "Number of Species", y = "Density", fill = "Timepoint", title = "Control Group") +
  scale_fill_manual(values = c("skyblue", "#f1c232"), labels = c("Pn12", "Pn56")) +
  theme_minimal() +
  theme(panel.border = element_rect(color = "black", fill = NA),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) + xlim (2, 37)
infantSpeciesNumberIFA


svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/infantSpeciesNumberBEP.svg",
  width=4, height=3
  )
infantSpeciesNumberBEP
dev.off()

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/infantSpeciesNumberIFA.svg",
  width=4, height=3
  )
infantSpeciesNumberIFA
dev.off()


```


#### mother

```{r}

mother_species <- mother_ps %>%
  tax_glom(taxrank = "species") %>%
  transform_sample_counts(function(x) x / sum(x)) %>%
  psmelt() %>%
  filter(Abundance > 0)


# Count the number of unique species for each subject
species_counts_BEP <- mother_species %>%
  mutate(timepoint_n = case_when(
    timepoint == "incl" ~ 0,
    timepoint == "tri3" ~ 1,
    timepoint == "pn12" ~ 3,
    timepoint == "pn56" ~ 4,
    TRUE ~ NA_real_  # Assign NA for any other values or missing values
  )) %>%  
  filter(code == "B Z") %>%
  group_by(idbs, timepoint) %>%
  summarise(num_species = n_distinct(OTU),
            timepoint_n = first(timepoint_n))

motherSpeciesNumberBEP <- ggplot(species_counts_BEP, aes(x = num_species, fill = factor(timepoint_n))) +
  geom_density(alpha = 0.5) +
  labs(x = "Number of Species", y = "Density", fill = "Timepoint", title = "Intervention Group") +
  scale_fill_manual(values = c("#00A98F", "#C00000", "skyblue", "#f1c232"), labels = c("Tri2", "Tri3", "Pn12", "Pn56")) +
  theme_minimal() +
  theme(panel.border = element_rect(color = "black", fill = NA),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) + xlim(66, 460)
motherSpeciesNumberBEP

# Count the number of unique species for each subject
species_counts_IFA <- mother_species %>%
  mutate(timepoint_n = case_when(
    timepoint == "incl" ~ 0,
    timepoint == "tri3" ~ 1,
    timepoint == "pn12" ~ 3,
    timepoint == "pn56" ~ 4,
    TRUE ~ NA_real_  # Assign NA for any other values or missing values
  )) %>%  
  filter(code == "A Y") %>%
  group_by(idbs, timepoint) %>%
  summarise(num_species = n_distinct(OTU),
            timepoint_n = first(timepoint_n))

motherSpeciesNumberIFA <- ggplot(species_counts_IFA, aes(x = num_species, fill = factor(timepoint_n))) +
  geom_density(alpha = 0.5) +
  labs(x = "Number of Species", y = "Density", fill = "Timepoint", title = "Control Group") +
  scale_fill_manual(values = c("#00A98F", "#C00000", "skyblue", "#f1c232"), labels = c("Tri2", "Tri3", "Pn12", "Pn56")) +
  theme_minimal() +
  theme(panel.border = element_rect(color = "black", fill = NA),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) + xlim(66, 460)
motherSpeciesNumberIFA

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/motherSpeciesNumberBEP.svg",
  width=5, height=3
  )
motherSpeciesNumberBEP
dev.off()

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/motherSpeciesNumberIFA.svg",
  width=5, height=3
  )
motherSpeciesNumberIFA
dev.off()

```


## Alpha diversity
Alpha diversity metrics are single number summaries of the diversity of microbial communities within a single sample. These are based on the **richness** (number of different taxa or species present) and/or **evenness** (describes how evenly the abundance of each species is distributed) within the sample. We describe
**Observed**: the simplest richness index
**Shannon**: measures the evenness
**Simpson**: takes into account the richness and evenness

### color scheme
```{r}
COLOR_SCHEME <- c(
  "#00A98F",
  "#FFC000",
  "#1565A9",
  "#C00000",
  "#0D0D0D",
  "#6aa84f",
  "tomato4",
  "#8b7355"
)


```


### observed, shannon, simpson
```{r calc diversity}
# calculate richness
alpha_div <- phyloseq::estimate_richness(
  ps,
  split = TRUE,
  measures = c("Observed", "Shannon", "Simpson")
  )

# merge with metadata to split on groups
alpha_div <- merge(
  x=alpha_div, 
  y=metadata[, c("idbs", "child", "timepoint", "code")],
  by=0
  )
DT::datatable(alpha_div) 
#write.csv(alpha_div, "/Users/lishideng/Library/CloudStorage/OneDrive-SharedLibraries-UGent/BioLizard - General/scripts/metagenomic/alpha_diversity_for_Yuri.csv")

group <- c()
for (i in 1:nrow(alpha_div)){
  if (alpha_div[i, "child"] == "yes"){
    child_str <- "Infant"
  } else {
    child_str <- "Mother"
  }
  if (alpha_div[i, "code"] == "B Z"){
    trt_str <- "BEP"
  } else {
    trt_str <- "IFA"
  }
  # add group
  group_str <- paste(child_str, trt_str, alpha_div[i, "timepoint"], sep = " ")
  alpha_div[i, "Group"] <- group_str
  # add tooltips for interactive plots
  text_str <- paste0(
    "mother ID: ", alpha_div[i, "idbs"], "\n", 
    "timepoint of sample: ", alpha_div[i, "timepoint"], "\n",
    "treatment group: ", trt_str, "\n"
    )
  if (alpha_div[i, "child"] == "yes"){
    text_str <- paste0("infant of ", text_str)
  }
  alpha_div[i, "Text"] <- text_str
}

# set as factor to preserve order
alpha_div$Group <- factor(
  alpha_div$Group, 
  levels=c(
    "Mother BEP incl",
    "Mother IFA incl",
    "Mother BEP tri3",
    "Mother IFA tri3",
    "Mother BEP pn12",
    "Mother IFA pn12",
    "Mother BEP pn56",
    "Mother IFA pn56",
    "Infant BEP pn12",
    "Infant IFA pn12",
    "Infant BEP pn56", 
    "Infant IFA pn56"
    )
  )

 alpha_div <- alpha_div %>%
    mutate(code = ifelse(code == "B Z", "BEP", "IFA"))
```


#### visual shannon
```{r create shannon diversity plot, echo = FALSE}
# set up tooltip for interactive figure
alpha_div <- alpha_div %>%
  dplyr::mutate(
    sha_tooltip = paste0(
      Text,
      "Shannon diversity index: ", round(Shannon, digits=4), "\n"
      )
    )

alpha_shannon <- ggplot2::ggplot(data = alpha_div) +
  ggplot2::geom_jitter(
    size = 1, 
    alpha = 0.5,
    ggplot2::aes(x=Group, y=Shannon, color=code, shape=timepoint, text=sha_tooltip)
    ) +
  ggplot2::theme_bw() +
  #ggplot2::coord_flip() +
  # Facet grid with free x-axis scales and free space
  facet_grid(. ~ child, scales = "free_x", space = "free", labeller = labeller(child = facet_labels)) +
  ggplot2::scale_color_manual(
    name="code", 
    labels=c("Intervention", "Control"),
    values = c("steelblue", "darkred")) +
  ggplot2::scale_shape_manual(
    name = "timepoint", 
    labels = c("Tri2", "Pn12", "Pn56", "Tri3"), 
    values=c(5, 16, 17, 3)
    ) +
  ggplot2::geom_boxplot(
    data = alpha_div,
    ggplot2::aes(x = Group, y = Shannon, color = code),
    alpha = 0.1
    ) +
  ggplot2::labs(
    title = "",
    x = "",
    y = "Shannon diversity index"
   ) +
  ggplot2::theme(
    plot.title = ggplot2::element_text(face = "bold"),
    axis.text.x = ggplot2::element_text(vjust = 0.4, hjust = 1),
    axis.text.y = ggplot2::element_text(vjust = 0.4, hjust = 1),
    legend.title = ggplot2::element_blank(),
    legend.position = "right",
    text = ggplot2::element_text(size = 10)
    )

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/alpha_shannon2.svg",
  width=5.6, height=4
  )
alpha_shannon
dev.off()

#ggsave("alpha_shannon.png", width = 5, height = 5, units = "in", bg = "white")

```

#### Wilcoxon rank sum test

```{r}

wilcox_result <- wilcox.test(Shannon ~ prePost, data = alpha_div2)
wilcox_result

```


### TMLE
Targeted Maximum Likelihood Estimation (TMLE) is a general approach for constructing an effectient double-robust semi-parametric substitution estimator of a causal effect parameter or statistical association measure.

TMLE estimate of effect of BEP supplementation on the microbial diversity in the mother prenatally (across incl and tri3) and postnally (across pn12 and pn56) and in infant (across pn12 and pn56).
It tries to estimate the average outcome (Shannon diversity of the gut microbiome). 
**Additive Effect among the Treated**: predicts that for if every women that actually received the treatment.
**Additive Effect among the Contrils**: predicts for if the women in the control group had received the additive.
**Additive Effect**: summarizes the expected increase in Shannon diversity between the two treatments.

```{r}
row.names(alpha_div) <- alpha_div$Row.names
alpha_div$Row.names <- NULL
# merge with metadata to split on groups
alpha_div_extended <- merge(
  x=alpha_div, 
  y=metadata[, c("treatment1_prenatal_n", "m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs",  "csps_n", "w_age", "m_bmiincl", "mddw_10_ave", "ch_sex", "iycf_ebf_age")],
  by=0
  )

alpha_div_extended$Row.names <- NULL
alpha_div_extended <- alpha_div_extended[complete.cases(alpha_div_extended), ]
alpha_div_extended$timepoint <- as.factor(alpha_div_extended$timepoint)
alpha_div_extended$code <- ifelse(alpha_div_extended$code == "BEP", 1, 0)

alpha_div_extended_mat <- alpha_div_extended[alpha_div_extended$child =="no", ]
alpha_div_extended_mat_pre <- alpha_div_extended[alpha_div_extended_mat$timepoint %in% c("incl", "tri3"), ]
alpha_div_extended_mat_post <- alpha_div_extended[alpha_div_extended_mat$timepoint %in% c("pn12", "pn56"), ]

alpha_div_extended_mat_incl <- alpha_div_extended[alpha_div_extended_mat$timepoint == "incl", ]
alpha_div_extended_mat_tri3 <- alpha_div_extended[alpha_div_extended_mat$timepoint == "tri3", ]
alpha_div_extended_mat_pn12 <- alpha_div_extended[alpha_div_extended_mat$timepoint == "pn12", ]
alpha_div_extended_mat_pn56 <- alpha_div_extended[alpha_div_extended_mat$timepoint == "pn56", ]
alpha_div_extended_inf <- alpha_div_extended[alpha_div_extended$child =="yes", ]
alpha_div_extended_inf_pn12 <- alpha_div_extended[alpha_div_extended_inf$timepoint == "pn12", ]
alpha_div_extended_inf_pn56 <- alpha_div_extended[alpha_div_extended_inf$timepoint == "pn56", ]

```


#### incl TMLE
```{r}
set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm', "SL.randomForest", "SL.step","SL.glm.interaction")

# Shannon
TMLE_maternal_incl <- tmle::tmle(
  Y=alpha_div_extended_mat_incl$Shannon, 
  A=alpha_div_extended_mat_incl$code, # must be numeric for the package used
  W=alpha_div_extended_mat_incl[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_maternal_incl)

alpha_div_extended_mat_incl_model <- nlme::lme(
  fixed = Shannon ~ code, 
  data = alpha_div_extended_mat_incl,
  random = ~ 1 | idbs
  )
summary(alpha_div_extended_mat_incl_model)

```


#### tri3 TMLE
```{r}
set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm', "SL.randomForest", "SL.step","SL.glm.interaction")

# Shannon
TMLE_maternal_tri3 <- tmle::tmle(
  Y=alpha_div_extended_mat_tri3$Shannon, 
  A=alpha_div_extended_mat_tri3$code, # must be numeric for the package used
  W=alpha_div_extended_mat_tri3[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_maternal_tri3)

alpha_div_extended_mat_tri3_model <- nlme::lme(
  fixed = Shannon ~ code, 
  data = alpha_div_extended_mat_tri3,
  random = ~ 1 | idbs
  )
summary(alpha_div_extended_mat_tri3_model)

```


#### mat pn12 TMLE
```{r}
set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm')

# Shannon
TMLE_maternal_pn12 <- tmle::tmle(
  Y=alpha_div_extended_mat_pn12$Shannon, 
  A=alpha_div_extended_mat_pn12$code, # must be numeric for the package used
  W=alpha_div_extended_mat_pn12[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_maternal_pn12)

```


#### mat pn56 TMLE
```{r}
set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm')

# Shannon
TMLE_maternal_pn56 <- tmle::tmle(
  Y=alpha_div_extended_mat_pn56$Shannon, 
  A=alpha_div_extended_mat_pn56$code, # must be numeric for the package used
  W=alpha_div_extended_mat_pn56[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_maternal_pn56)

```


#### inf pn12 TMLE
```{r}
set.seed(123)
# Shannon
TMLE_infant_pn12 <- tmle::tmle(
  Y=alpha_div_extended_inf_pn12$Shannon, 
  A=alpha_div_extended_inf_pn12$code, # must be numeric for the package used
  W=alpha_div_extended_inf_pn12[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave", "ch_sex", "iycf_ebf_age")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_infant_pn12)

```

#### inf pn56 TMLE
```{r}
set.seed(123)
# Shannon
TMLE_infant_pn56 <- tmle::tmle(
  Y=alpha_div_extended_inf_pn56$Shannon, 
  A=alpha_div_extended_inf_pn56$code, # must be numeric for the package used
  W=alpha_div_extended_inf_pn56[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave", "ch_sex", "iycf_ebf_age")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_infant_pn56)

```

#### TMLE prenatal
```{r}
set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm', "SL.randomForest", "SL.step","SL.glm.interaction")

TMLE_maternal_pre <- tmle::tmle(
  Y=alpha_div_extended_mat_pre$Shannon, 
  A=alpha_div_extended_mat_pre$treatment1_prenatal_n, # must be numeric for the package used
  W=alpha_div_extended_mat_pre[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave", "timepoint")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_maternal_pre)
```


#### TMLE  postnatal
```{r}
set.seed(123)
TMLE_maternal_post <- tmle::tmle(
  Y=alpha_div_extended_mat_post$Shannon, 
  A=alpha_div_extended_mat_post$treatment1_prenatal_n, # must be numeric for the package used
  W=alpha_div_extended_mat_post[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave", "timepoint")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_maternal_post)
```


#### TMLE infant
```{r}
set.seed(123)

TMLE_infant <- tmle::tmle(
  Y=alpha_div_extended_inf$Shannon, 
  A=alpha_div_extended_inf$treatment1_prenatal_n, # must be numeric for the package used
  W=alpha_div_extended_inf[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave", "ch_sex", "iycf_ebf_age", "timepoint")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_infant)
```



## Beta diversity

### infant 4 groups
```{r message=FALSE}
# Calculate the UniFrac distance matrix
unifrac_dist <- phyloseq::distance(infant, method="unifrac", weighted=TRUE)

# Perform PCoA ordination using the calculated distance matrix
infant_ord <- ordinate(infant, method="PCoA", distance=unifrac_dist)
phyloseq::plot_scree(infant_ord) + geom_bar(stat = "identity", fill="blue") + labs(x="\nAxis", y="Proportion of Variance\n")
head(infant_ord$values$Eigenvalues)

infant_eig1 <- infant_ord$values$Eigenvalues[1]/sum(infant_ord$values$Eigenvalues)
infant_eig2 <- infant_ord$values$Eigenvalues[2]/sum(infant_ord$values$Eigenvalues)
beta_infant <- phyloseq::plot_ordination(infant, infant_ord, color = "bep_group", shape = "timepoint") + 
  scale_color_manual(
    name="bep_group", 
    values = COLOR_SCHEME[1:4]
    ) +
  ggplot2::scale_shape_manual(
    name = "timepoint",
    labels = c("pn12", "pn56"),
    values=c(17, 3)) +
  geom_point(size=1) +
  #coord_fixed(infant_eig1/infant_eig2) +
  stat_ellipse(aes(group=bep_group), linetype=2) +
  theme_bw() +
  theme(legend.position = "bottom") +
  xlim(-0.22, 0.22) +
  ylim(-0.12, 0.12)
beta_infant

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/beta_infant.svg",
  width=4, height=6
  )
beta_infant
dev.off()

infant_dist <-UniFrac(infant, weighted = TRUE, normalized = TRUE, parallel = FALSE, fast=TRUE)
vegan::adonis2(infant_dist ~ phyloseq::sample_data(infant)$code + phyloseq::sample_data(infant)$timepoint )
```


### maternal 8 groups
```{r message=FALSE}

mother_ord <- phyloseq::ordinate(mother, "PCoA", "unifrac", weighted=T)
phyloseq::plot_scree(mother_ord) + geom_bar(stat = "identity", fill="blue") + labs(x="\nAxis", y="Proportion of Variance\n")
head(mother_ord$values$Eigenvalues)

mother_eig1 <- mother_ord$values$Eigenvalues[1]/sum(mother_ord$values$Eigenvalues)
mother_eig2 <- mother_ord$values$Eigenvalues[2]/sum(mother_ord$values$Eigenvalues)
beta_mother <- phyloseq::plot_ordination(mother, mother_ord, color = "bep_group", shape = "timepoint") + 
scale_color_manual(
    name="bep_group", 
    values = COLOR_SCHEME[1:8]
    ) +
  ggplot2::scale_shape_manual(
    name = "timepoint",
    labels = c("incl", "tri3", "pn12", "pn56"),
    values=c(5, 16, 17, 3)) +
  geom_point(size=1) +
  #coord_fixed(mother_eig1/mother_eig2) +
  stat_ellipse(aes(group=bep_group), linetype=2)+
  theme_bw() +
  theme(legend.position = "bottom") +
  xlim(-0.22, 0.22) +
  ylim(-0.12, 0.12) +
  guides(color = guide_legend(nrow = 1), shape = guide_legend(nrow = 1))
beta_mother
#ggsave("beta_mother.png", width = 5, height = 5, units = "in", bg = "white")


svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/beta_mother.svg",
  width=4, height=6
  )
beta_mother
dev.off()


mother_dist <-UniFrac(mother, weighted = T, normalized = TRUE, parallel = FALSE, fast=TRUE)
vegan::adonis2(mother_dist ~ phyloseq::sample_data(mother)$code + phyloseq::sample_data(mother)$timepoint)

```


## Gene richness

```{r}

# Load necessary libraries
library(tidyr)
library(dplyr)

# Assuming your dataset is a data frame named 'data' with rownames as gene names
# Convert rownames to a column for easier manipulation
data_long <- gene_counts %>%
  rownames_to_column(var = "gene") %>%
  pivot_longer(-gene, names_to = "sample", values_to = "count")

# Extract timepoint, sample ID, and other relevant information from the sample column if needed
# Assuming sample names are of the format: timepoint_se1g_idbs_e
data_long <- data_long %>%
  separate(sample, into = c("timepoint", "se1g", "idbs", "e"), sep = "_", remove = FALSE) %>%
  mutate(child = ifelse(is.na(e), "Mother", "Infant"))
metadata$idbs <- as.character(metadata$idbs)
# Calculate gene richness (number of genes with count > 0) for each sample
gene_richness <- data_long %>%
  left_join(metadata %>% dplyr::select(idbs, code), by = "idbs") %>%
  group_by(idbs, code, timepoint, child) %>%
  dplyr::summarize(richness = sum(count > 0)) %>%
  mutate(Group = ifelse(timepoint == "incl" & code == "A Y", "Mother IFA incl",
                     ifelse(timepoint == "incl" & code == "B Z", "Mother BEP incl",
                     ifelse(timepoint == "tri3" & code == "A Y", "Mother IFA tri3",
                     ifelse(timepoint == "tri3" & code == "B Z", "Mother BEP tri3",
                     ifelse(timepoint == "pn12" & child == "Mother" & code == "A Y", "Mother IFA pn12",
                     ifelse(timepoint == "pn12" & child == "Mother" & code == "B Z", "Mother BEP pn12",
                     ifelse(timepoint == "pn12" & child == "Infant" & code == "A Y", "Infant IFA pn12",
                     ifelse(timepoint == "pn12" & child == "Infant" & code == "B Z", "Infant BEP pn12",
                     ifelse(timepoint == "pn56" & child == "Mother" & code == "A Y", "Mother IFA pn56",
                     ifelse(timepoint == "pn56" & child == "Mother" & code == "B Z", "Mother BEP pn56",
                     ifelse(timepoint == "pn56" & child == "Infant" & code == "A Y", "Infant IFA pn56",
                     ifelse(timepoint == "pn56" & child == "Infant" & code == "B Z", "Infant BEP pn56", "other"))))))))))))) %>%
  mutate(Group = factor(Group, levels = c(
    "Mother BEP incl",
    "Mother IFA incl",
    "Mother BEP tri3",
    "Mother IFA tri3",
    "Mother BEP pn12",
    "Mother IFA pn12",
    "Mother BEP pn56",
    "Mother IFA pn56",
    "Infant BEP pn12",
    "Infant IFA pn12",
    "Infant BEP pn56",
    "Infant IFA pn56"
  ))) %>%
  mutate(code = ifelse(code == "B Z", "Intervention", "Control")) %>%
  mutate(code = factor(code, levels = c("Intervention", "Control"))) %>%
  mutate(child = factor(child, levels = c("Mother", "Infant")))

# Generate the plot
gene_richness_plot <- ggplot2::ggplot(data = gene_richness) +
  ggplot2::geom_jitter(
    size = 1, 
    alpha = 0.5,
    ggplot2::aes(x = Group, y = richness, color = code, shape = timepoint)
  ) +
  ggplot2::theme_bw() +
  # Facet grid with free x-axis scales and free space
  facet_grid(. ~ child, scales = "free_x", space = "free") +
  ggplot2::scale_color_manual(
    name = "code", 
    labels = c("Intervention", "Control"),
    values = c("steelblue", "darkred")
  ) +
  ggplot2::scale_shape_manual(
    name = "timepoint", 
    labels = c("Tri2", "Pn12", "Pn56", "Tri3"), 
    values = c(5, 16, 17, 3)
  ) +
  ggplot2::geom_boxplot(
    data = gene_richness,
    ggplot2::aes(x = Group, y = richness, color = code),
    alpha = 0.1
  ) +
  ggplot2::labs(
    title = "",
    x = "",
    y = "Gene richness"
  ) +
  ggplot2::theme(
    plot.title = ggplot2::element_text(face = "bold"),
    axis.text.x = ggplot2::element_text(vjust = 0.4, hjust = 1),
    axis.text.y = ggplot2::element_text(vjust = 0.4, hjust = 1),
    legend.title = ggplot2::element_blank(),
    legend.position = "right",
    text = ggplot2::element_text(size = 10)
  )
gene_richness_plot

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/gene_richness2.svg",
  width=5.6, height=4
  )
gene_richness_plot
dev.off()

```

### gene TMLE incl

```{r}
metadata$idbs <- as.character(metadata$idbs)
# merge with metadata to split on groups
gene_richness_extended <- gene_richness %>%
  left_join(
    metadata %>%
      group_by(idbs) %>%
      slice(1) %>%  # Keep only the first occurrence
      ungroup() %>%
      select(idbs, treatment1_prenatal_n, m_heightincl, m_hbincl, w_school_level, 
             w_marital, asset1comp_10, w_language, w_religion, w_etnicity, 
             w_nr_jobs, csps_n, w_age, m_bmiincl, mddw_10_ave, ch_sex, iycf_ebf_age),
    by = "idbs"
  )


gene_richness_incl <- gene_richness_extended %>%
  filter(timepoint == "incl") %>%
  drop_na(richness, treatment1_prenatal_n, m_heightincl, m_hbincl, w_school_level, 
          w_marital, asset1comp_10, w_language, w_religion, w_etnicity, 
          w_nr_jobs, csps_n, w_age, m_bmiincl, mddw_10_ave)

set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm', "SL.randomForest", "SL.step","SL.glm.interaction")

TMLE_gene_richness_incl <- tmle::tmle(
  Y=gene_richness_incl$richness, 
  A=gene_richness_incl$treatment1_prenatal_n, # must be numeric for the package used
  W=gene_richness_incl[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_gene_richness_incl)

```

### gene TMLE tri3

```{r}
gene_richness_tri3 <- gene_richness_extended %>%
  filter(timepoint == "tri3") %>%
  drop_na(richness, treatment1_prenatal_n, m_heightincl, m_hbincl, w_school_level, 
          w_marital, asset1comp_10, w_language, w_religion, w_etnicity, 
          w_nr_jobs, csps_n, w_age, m_bmiincl, mddw_10_ave)

set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm', "SL.randomForest", "SL.step","SL.glm.interaction")

TMLE_gene_richness_tri3 <- tmle::tmle(
  Y=gene_richness_tri3$richness, 
  A=gene_richness_tri3$treatment1_prenatal_n, # must be numeric for the package used
  W=gene_richness_tri3[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_gene_richness_tri3)

```


### gene TMLE mat pn12

```{r}
gene_richness_pn12m <- gene_richness_extended %>%
  filter(timepoint == "pn12" & child == "Mother") %>%
  drop_na(richness, treatment1_prenatal_n, m_heightincl, m_hbincl, w_school_level, 
          w_marital, asset1comp_10, w_language, w_religion, w_etnicity, 
          w_nr_jobs, csps_n, w_age, m_bmiincl, mddw_10_ave)

set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm', "SL.randomForest", "SL.step","SL.glm.interaction")

TMLE_gene_richness_pn12m <- tmle::tmle(
  Y=gene_richness_pn12m$richness, 
  A=gene_richness_pn12m$treatment1_prenatal_n, # must be numeric for the package used
  W=gene_richness_pn12m[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_gene_richness_pn12m)

```


### gene TMLE mat pn56

```{r}
gene_richness_pn56m <- gene_richness_extended %>%
  filter(timepoint == "pn56" & child == "Mother") %>%
  drop_na(richness, treatment1_prenatal_n, m_heightincl, m_hbincl, w_school_level, 
          w_marital, asset1comp_10, w_language, w_religion, w_etnicity, 
          w_nr_jobs, csps_n, w_age, m_bmiincl, mddw_10_ave)

set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm', "SL.randomForest", "SL.step","SL.glm.interaction")

TMLE_gene_richness_pn56m <- tmle::tmle(
  Y=gene_richness_pn56m$richness, 
  A=gene_richness_pn56m$treatment1_prenatal_n, # must be numeric for the package used
  W=gene_richness_pn56m[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_gene_richness_pn56m)

```


### gene TMLE inf pn12

```{r}
gene_richness_pn12e <- gene_richness_extended %>%
  filter(timepoint == "pn12" & child == "Infant") %>%
  drop_na(richness, treatment1_prenatal_n, m_heightincl, m_hbincl, w_school_level, 
          w_marital, asset1comp_10, w_language, w_religion, w_etnicity, 
          w_nr_jobs, csps_n, w_age, m_bmiincl, mddw_10_ave)

set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm', "SL.randomForest", "SL.step","SL.glm.interaction")

TMLE_gene_richness_pn12e <- tmle::tmle(
  Y=gene_richness_pn12e$richness, 
  A=gene_richness_pn12e$treatment1_prenatal_n, # must be numeric for the package used
  W=gene_richness_pn12e[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_gene_richness_pn12e)

```


### gene TMLE inf pn56

```{r}
gene_richness_pn56e <- gene_richness_extended %>%
  filter(timepoint == "pn12" & child == "Infant") %>%
  drop_na(richness, treatment1_prenatal_n, m_heightincl, m_hbincl, w_school_level, 
          w_marital, asset1comp_10, w_language, w_religion, w_etnicity, 
          w_nr_jobs, csps_n, w_age, m_bmiincl, mddw_10_ave)

set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm', "SL.randomForest", "SL.step","SL.glm.interaction")

TMLE_gene_richness_pn56e <- tmle::tmle(
  Y=gene_richness_pn56e$richness, 
  A=gene_richness_pn56e$treatment1_prenatal_n, # must be numeric for the package used
  W=gene_richness_pn56e[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_gene_richness_pn56e)

```


### gene TMLE prenatal

```{r}

gene_richness_prenatal <- gene_richness_extended %>%
  filter(timepoint == "incl" | timepoint == "tri3") %>%
  drop_na(richness, treatment1_prenatal_n, m_heightincl, m_hbincl, w_school_level, 
          w_marital, asset1comp_10, w_language, w_religion, w_etnicity, 
          w_nr_jobs, csps_n, w_age, m_bmiincl, mddw_10_ave)

set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm', "SL.randomForest", "SL.step","SL.glm.interaction")

TMLE_gene_richness_prenatal <- tmle::tmle(
  Y=gene_richness_prenatal$richness, 
  A=gene_richness_prenatal$treatment1_prenatal_n, # must be numeric for the package used
  W=gene_richness_prenatal[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_gene_richness_prenatal)

```


### gene TMLE postnatal

```{r}

gene_richness_postnatal <- gene_richness_extended %>%
  filter(timepoint == "pn12" | timepoint == "pn56" & child == "Mother") %>%
  drop_na(richness, treatment1_prenatal_n, m_heightincl, m_hbincl, w_school_level, 
          w_marital, asset1comp_10, w_language, w_religion, w_etnicity, 
          w_nr_jobs, csps_n, w_age, m_bmiincl, mddw_10_ave)

set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm', "SL.randomForest", "SL.step","SL.glm.interaction")

TMLE_gene_richness_postnatal <- tmle::tmle(
  Y=gene_richness_postnatal$richness, 
  A=gene_richness_postnatal$treatment1_prenatal_n, # must be numeric for the package used
  W=gene_richness_postnatal[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_gene_richness_postnatal)

```


### gene TMLE infant

```{r}

gene_richness_infant <- gene_richness_extended %>%
  filter(child == "Infant") %>%
  drop_na(richness, treatment1_prenatal_n, m_heightincl, m_hbincl, w_school_level, 
          w_marital, asset1comp_10, w_language, w_religion, w_etnicity, 
          w_nr_jobs, csps_n, w_age, m_bmiincl, mddw_10_ave)

set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm', "SL.randomForest", "SL.step","SL.glm.interaction")

TMLE_gene_richness_infant <- tmle::tmle(
  Y=gene_richness_infant$richness, 
  A=gene_richness_infant$treatment1_prenatal_n, # must be numeric for the package used
  W=gene_richness_infant[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_gene_richness_infant)

```





## Gene beta diversity

```{r}

library(vegan)
library(dplyr)
library(tidyr)
library(ggplot2)

COLOR_SCHEME2 <- c(
  "#0D0D0D",
  "#6aa84f",
  "tomato4",
  "#8b7355",
  "#00A98F",
  "#FFC000",
  "#1565A9",
  "#C00000"
)

# Separate the data for Mother and Infant based on sample ID
mother_samples <- grep("_e$", colnames(gene_counts), invert = TRUE, value = TRUE)
infant_samples <- grep("_e$", colnames(gene_counts), value = TRUE)

# Filter the gene_counts matrix for Mother and Infant
gene_counts_mother <- gene_counts[, mother_samples]
gene_counts_infant <- gene_counts[, infant_samples]

# remove the empty rows
gene_counts_mother <- gene_counts_mother[rowSums(gene_counts_mother != 0) > 0, ]
gene_counts_infant <- gene_counts_infant[rowSums(gene_counts_infant != 0) > 0, ]

# Transpose the data to have samples as rows and genes as columns
data_t_mother <- t(gene_counts_mother)

# Calculate the Bray-Curtis distance matrix
bray_dist_mother <- vegdist(data_t_mother, method = "bray")

# Perform PCoA on the Bray-Curtis distance matrix
pcoa_result_mother <- cmdscale(bray_dist_mother, eig = TRUE, k = 2)

# Extract the coordinates for the first two principal coordinates
pcoa_coords_mother <- as.data.frame(pcoa_result_mother$points) %>%
  mutate(PC1 = V1) %>%
  mutate(PC2 = V2) %>%
  rownames_to_column("sample_id") %>%
  separate(sample_id, into = c("timepoint", "se1g", "idbs", "e"), sep = "_") %>%
  mutate(child = ifelse(is.na(e), "Mother", "Infant")) %>%
  left_join(metadata %>% dplyr::select(idbs, code), by = "idbs") %>%
  mutate(Group = case_when(
    timepoint == "incl" & code == "A Y" ~ "Mother IFA incl",
    timepoint == "incl" & code == "B Z" ~ "Mother BEP incl",
    timepoint == "tri3" & code == "A Y" ~ "Mother IFA tri3",
    timepoint == "tri3" & code == "B Z" ~ "Mother BEP tri3",
    timepoint == "pn12" & child == "Mother" & code == "A Y" ~ "Mother IFA pn12",
    timepoint == "pn12" & child == "Mother" & code == "B Z" ~ "Mother BEP pn12",
    timepoint == "pn56" & child == "Mother" & code == "A Y" ~ "Mother IFA pn56",
    timepoint == "pn56" & child == "Mother" & code == "B Z" ~ "Mother BEP pn56",
    TRUE ~ "other"
  )) %>%
  mutate(Group = factor(Group, levels = c(
    "Mother BEP incl",
    "Mother IFA incl",
    "Mother BEP tri3",
    "Mother IFA tri3",
    "Mother BEP pn12",
    "Mother IFA pn12",
    "Mother BEP pn56",
    "Mother IFA pn56"
  ))) 

# Plot for Mother
gene_beta_mother <- ggplot(pcoa_coords_mother, aes(x = PC1, y = PC2, color = Group, shape = timepoint)) +
  geom_point(size = 1) +
  stat_ellipse(aes(group = Group), linetype = 2) +
  scale_color_manual(
    name = "Group", 
    values = COLOR_SCHEME2[1:8]  # Adjust these colors to fit your needs
  ) +
  scale_shape_manual(
    name = "Timepoint",
    labels = c("incl", "tri3", "pn12", "pn56"),
    values = c(5, 16, 17, 3)
  ) +
  theme_bw() +
  theme(legend.position = "right") +
  guides(color = guide_legend(ncol = 1), shape = guide_legend(ncol = 1)) +
  labs(
    title = "PCoA Plot for Mother",
    x = "PC1",
    y = "PC2"
  ) +
  ylim (-0.6, 0.2) +
  xlim(-0.50, 0.50) 

gene_beta_mother

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/gene_beta_mother2.svg",
  width=5, height=4.7
  )
gene_beta_mother
dev.off()

# Ensure the metadata is available for the samples
metadata_for_adonis <- pcoa_coords_mother %>%
  select(idbs, code, timepoint, Group) %>%
  distinct()

# Run PERMANOVA using the Bray-Curtis distance matrix
adonis_result <- adonis2(bray_dist_mother ~ code + timepoint, data = metadata_for_adonis, permutations = 999)

# View the results
print(adonis_result)


# Transpose the data to have samples as rows and genes as columns
data_t_infant <- t(gene_counts_infant)

# Calculate the Bray-Curtis distance matrix
bray_dist_infant <- vegdist(data_t_infant, method = "bray")

# Perform PCoA on the Bray-Curtis distance matrix
pcoa_result_infant <- cmdscale(bray_dist_infant, eig = TRUE, k = 2)

# Extract the coordinates for the first two principal coordinates
pcoa_coords_infant <- as.data.frame(pcoa_result_infant$points) %>%
  mutate(PC1 = V1) %>%
  mutate(PC2 = V2) %>%
  rownames_to_column("sample_id") %>%
  separate(sample_id, into = c("timepoint", "se1g", "idbs", "e"), sep = "_") %>%
  mutate(child = ifelse(is.na(e), "Mother", "Infant")) %>%
  left_join(metadata %>% dplyr::select(idbs, code), by = "idbs") %>%
  mutate(Group = case_when(
    timepoint == "pn12" & child == "Infant" & code == "A Y" ~ "Infant IFA pn12",
    timepoint == "pn12" & child == "Infant" & code == "B Z" ~ "Infant BEP pn12",
    timepoint == "pn56" & child == "Infant" & code == "A Y" ~ "Infant IFA pn56",
    timepoint == "pn56" & child == "Infant" & code == "B Z" ~ "Infant BEP pn56",
    TRUE ~ "other"
  )) %>%
  mutate(Group = factor(Group, levels = c(
    "Infant BEP pn12",
    "Infant IFA pn12",
    "Infant BEP pn56",
    "Infant IFA pn56"
  ))) 

# Plot for Mother
gene_beta_infant <- ggplot(pcoa_coords_infant, aes(x = PC1, y = PC2, color = Group, shape = timepoint)) +
  geom_point(size = 1) +
  stat_ellipse(aes(group = Group), linetype = 2) +
  scale_color_manual(
    name = "Group", 
    values = COLOR_SCHEME[1:4]  # Adjust these colors to fit your needs
  ) +
  scale_shape_manual(
    name = "Timepoint",
    labels = c("pn12", "pn56"),
    values = c(17, 3)
  ) +
  theme_bw() +
  theme(legend.position = "right") +
  guides(color = guide_legend(ncol = 1), shape = guide_legend(ncol = 1)) +
  labs(
    title = "PCoA Plot for Infant",
    x = "PC1",
    y = "PC2"
  ) +
  ylim (-0.6, 0.2) +
  xlim(-0.50, 0.50) 

gene_beta_infant

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/gene_beta_infant2.svg",
  width=5, height=4.7
  )
gene_beta_infant
dev.off()

# Ensure the metadata is available for the samples
metadata_for_adonis <- pcoa_coords_infant %>%
  select(idbs, code, timepoint, Group) %>%
  distinct()

# Run PERMANOVA using the Bray-Curtis distance matrix
adonis_result <- adonis2(bray_dist_infant ~ code + timepoint, data = metadata_for_adonis, permutations = 999)

# View the results
print(adonis_result)



```


# Figure 3. Stability
## A. recurrence 
```{r}
ps_species <- ps %>%
  tax_glom(taxrank = "species") %>%
  transform_sample_counts(function(x){x/sum(x)}) %>%
  psmelt() %>%
  filter(Abundance > 0)

species_recurrence <- ps_species %>%
  select(OTU, phylum, species, timepoint, child, code, Abundance)

species_recurrence <- species_recurrence %>%
  group_by(OTU, species, code, child) %>%
  dplyr::summarise(recurrence = n_distinct(timepoint)) %>%
  ungroup()

# Create a recurrence category for density plot
species_recurrence <- species_recurrence %>%
  mutate(recurrence_category = case_when(
    child == "no" & recurrence == 2 ~ "Maternal: 2 times",
    child == "no" & recurrence == 3 ~ "Maternal: 3 times",
    child == "no" & recurrence == 4 ~ "Maternal: 4 times",
    child == "yes" & recurrence == 2 ~ "Infant: 2 times",
    TRUE ~ "Other"
  )) %>%
  filter(recurrence_category != "Other")

# Generate density plots
recurrence_plot <- ggplot(species_recurrence, aes(x = recurrence, fill = recurrence_category)) +
  geom_density(alpha = 0.5, size = 0.3) +
  facet_wrap(~ code + child, scales = "free_y", ncol = 2) +
  labs(
    x = "Number of Species",
    y = "Density",
    fill = "Recurrence Category"
  ) +
  xlim(0, 10) +
  ylim(0, 0.6) +
  scale_fill_manual(values = c("#00A98F", "#C00000", "skyblue", "#f1c232")) +
  theme_minimal() +
theme(
    panel.grid.major = element_blank(), # Remove major grid lines
    panel.grid.minor = element_blank(), # Remove minor grid lines
    panel.border = element_rect(colour = "black", fill = NA), # Add box outline
    strip.background = element_blank(), # Remove background of facet labels
    strip.text = element_blank(), # Remove facet labels
    axis.text.x = element_text(angle = 0, hjust = 1), # Angle x-axis labels if needed
    axis.title = element_text(size = 10),
    legend.position = "right" # Position legend at the bottom
  )

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/recurrence.svg",
  width=7, height=4
  )
recurrence_plot
dev.off()

```

```{r}

# Sample dataset creation (Replace this with your actual dataset)
# species_recurrence <- read.csv("path_to_your_dataset.csv")

# Step 1: Add a column indicating the exact number of unique timepoints for each species
species_recurrence_summary <- species_recurrence %>%
  dplyr::group_by(phylum, species, child, code) %>%
  dplyr::summarize(timepoint_count = n_distinct(timepoint), .groups = 'drop')

# Step 2: Calculate the percentage of each timepoint count within each phylum
# Calculate the total number of species per phylum, child, and code combination
total_species <- species_recurrence_summary %>%
  dplyr::group_by(phylum, child, code) %>%
  dplyr::summarize(total_species = n(), .groups = 'drop')

# Merge with the summary to calculate percentages
phylum_percentage <- species_recurrence_summary %>%
  dplyr::left_join(total_species, by = c("phylum", "child", "code")) %>%
  dplyr::group_by(phylum, child, code, timepoint_count) %>%
  dplyr::summarize(count = n(), .groups = 'drop') %>%
  dplyr::group_by(phylum, child, code) %>%
  dplyr::mutate(percentage = count / sum(count))  # Calculate percentage relative to the total count in each phylum

# Combine the data for plotting
combined_plots <- phylum_percentage %>%
  dplyr::mutate(plot_type = paste(child, code, sep = "-"))

# Function to plot the percentage of occurrences with each bar summing to 1
plot_percentage_combined <- function(df, child_status) {
  ggplot(df %>% dplyr::filter(grepl(child_status, plot_type)), aes(x = phylum, y = percentage, fill = as.factor(timepoint_count))) +
    geom_bar(stat = "identity", position = "stack") +
    labs(x = "Phylum", y = "Percentage", title = paste("Phylum Occurrence: Child-", child_status)) +
    scale_y_continuous(labels = scales::percent, limits = c(0, 1)) +  # Ensure y-axis goes from 0 to 1 (0% to 100%)
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
          strip.text = element_text(size = 12)) +  # Adjust facet labels if needed
    scale_fill_manual(values = c("#00A98F", "#C00000", "skyblue", "#f1c232")) +  # Custom colors
    guides(fill = guide_legend(title = "Timepoint Count")) +
    facet_wrap(~code, scales = "fixed") +
    theme(
      axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
      legend.position = "bottom",
      panel.background = element_blank(),
      plot.background = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
    panel.border = element_rect(colour = "black", fill = NA)# Add box outline
    ) 
}

# Generate and display the combined plots
print(plot_percentage_combined(combined_plots, "no"))
print(plot_percentage_combined(combined_plots, "yes"))

plot_percentage_mo <- plot_percentage_combined(combined_plots, "no")
plot_percentage_inf <- plot_percentage_combined(combined_plots, "yes")

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/plot_percentage_mo.svg",
  width=8, height=4.5
  )
plot_percentage_mo
dev.off()

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/plot_percentage_inf.svg",
  width=6, height=4.5
  )
plot_percentage_inf
dev.off()

```


```{r}
library(ggplot2)
library(dplyr)

# Ensure total counts are calculated
total_counts <- phylum_percentage %>%
  dplyr::group_by(phylum, child, code) %>%
  dplyr::summarize(total_count = sum(count), .groups = 'drop')

# Plot heatmaps with specified color gradients
plot_heatmaps <- function(df, child_status) {
  color_gradient <- if (child_status == "yes") {
    scale_fill_gradient(low = "white", high = "steelblue", name = "Total Count")
  } else {
    scale_fill_gradient(low = "white", high = "darkred", name = "Total Count")
  }
  
  ggplot(df %>% dplyr::filter(child == child_status), aes(x = phylum, y = 1, fill = total_count)) +
    geom_tile(color = "white") +  # Add white border to each tile for better separation
    labs(x = "Phylum", y = NULL, title = paste("Heatmap of Total Counts: Child-", child_status)) +
    color_gradient +
    facet_wrap(~code, scales = "fixed") +
    theme_minimal() +
    theme(
      axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
      axis.text.y = element_blank(),  # Remove y-axis text as it's not needed
      axis.ticks.y = element_blank(), # Remove y-axis ticks
      panel.grid = element_blank(),
      strip.text = element_text(size = 12),
      legend.position = "bottom",
      panel.border = element_rect(colour = "black", fill = NA) # Add border around the panel
    )
}

# Generate and display the heatmaps
plot_heatmap_child_no <- plot_heatmaps(total_counts, "no")
plot_heatmap_child_yes <- plot_heatmaps(total_counts, "yes")

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/plot_heatmap_mo.svg",
  width=8, height=2.9
  )
plot_heatmap_child_no
dev.off()

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/plot_heatmap_inf.svg",
  width=6, height=2.9
  )
plot_heatmap_child_yes
dev.off()




```

## B. UPSET

```{r}

#Mother_Tri2 <- subset_samples(incl)
Mother_Tri2 <- subset_samples(incl, code == "B Z")
#Mother_Tri2 <- subset_samples(incl, code == "A Y")
Mother_Tri2 <- otu_table(Mother_Tri2) %>%
  as.data.frame() %>%
  filter(rowSums(across(everything())) != 0)
Mother_Tri2_species <- rownames(Mother_Tri2)

#Mother_Tri3 <- subset_samples(tri3)
Mother_Tri3 <- subset_samples(tri3, code == "B Z")
#Mother_Tri3 <- subset_samples(tri3, code == "A Y")
Mother_Tri3 <- otu_table(Mother_Tri3) %>%
  as.data.frame() %>%
  filter(rowSums(across(everything())) != 0)
Mother_Tri3_species <- rownames(Mother_Tri3)

#Mother_Pn12 <- subset_samples(pn12m)
Mother_Pn12 <- subset_samples(pn12m, code == "B Z")
#Mother_Pn12 <- subset_samples(pn12m, code == "A Y")
Mother_Pn12<- otu_table(Mother_Pn12) %>%
  as.data.frame() %>%
  filter(rowSums(across(everything())) != 0)
Mother_Pn12_species <- rownames(Mother_Pn12)

#Mother_Pn56 <- subset_samples(pn56m)
Mother_Pn56 <- subset_samples(pn56m, code == "B Z")
#Mother_Pn56 <- subset_samples(pn56m, code == "A Y")
Mother_Pn56<- otu_table(Mother_Pn56) %>%
  as.data.frame() %>%
  filter(rowSums(across(everything())) != 0)
Mother_Pn56_species <- rownames(Mother_Pn56)

#Infant_Pn12 <- subset_samples(pn12e)
Infant_Pn12 <- subset_samples(pn12e, code == "B Z")
#Infant_Pn12 <- subset_samples(pn12e, code == "A Y")
Infant_Pn12<- otu_table(Infant_Pn12) %>%
  as.data.frame() %>%
  filter(rowSums(across(everything())) != 0)
Infant_Pn12_species <- rownames(Infant_Pn12)

#Infant_Pn56 <- subset_samples(pn56e)
Infant_Pn56 <- subset_samples(pn56e, code == "B Z")
#Infant_Pn56 <- subset_samples(pn56e, code == "A Y")
Infant_Pn56<- otu_table(Infant_Pn56) %>%
  as.data.frame() %>%
  filter(rowSums(across(everything())) != 0)
Infant_Pn56_species <- rownames(Infant_Pn56)


# Combine all species into a single data frame with a presence/absence matrix
species <- unique(c(Mother_Tri2_species, Mother_Tri3_species, Mother_Pn12_species, Mother_Pn56_species, Infant_Pn12_species, Infant_Pn56_species))
df <- data.frame(
  species = species,
  Mother_Tri2 = as.numeric(species %in% Mother_Tri2_species),
  Mother_Tri3 = as.numeric(species %in% Mother_Tri3_species),
  Mother_Pn12 = as.numeric(species %in% Mother_Pn12_species),
  Mother_Pn56 = as.numeric(species %in% Mother_Pn56_species),
  Infant_Pn12 = as.numeric(species %in% Infant_Pn12_species),
  Infant_Pn56 = as.numeric(species %in% Infant_Pn56_species)
)

# Create an UpSet plot
upset <- upset(
  df, 
  sets = c("Mother_Tri2", "Mother_Tri3", "Mother_Pn12", "Mother_Pn56", "Infant_Pn12", "Infant_Pn56"),
  order.by = "freq",
  main.bar.color = "steelblue",
  sets.bar.color = c("darkred", "#EFC000FF", "#868686FF", "steelblue", "#00A98F", "#6aa84f"),
  keep.order = TRUE
)

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/upset_bep.svg",
  width=7, height=4
  )
upset
dev.off()


```


#### chisq.test
```{r}
# Define the observed values
control_shared <- 165
control_total <- 281
intervention_shared <- 142
intervention_total <- 292

# Create the contingency table
contingency_table <- matrix(c(control_shared, control_total - control_shared,
                              intervention_shared, intervention_total - intervention_shared), 
                            nrow = 2, 
                            byrow = TRUE)

# Set the row and column names for clarity (optional)
rownames(contingency_table) <- c("Control", "Intervention")
colnames(contingency_table) <- c("Shared Species", "Non-Shared Species")

# Perform the chi-square test
test_result <- chisq.test(contingency_table)

# Print the test result
print(test_result)

```



## C. shared species between mat and inf
```{r}

dyads_species <- ps %>%
  tax_glom(taxrank = "species") %>%
  transform_sample_counts(function(x){x/sum(x)}) %>%
  psmelt() %>%
  filter(Abundance > 0)

# First, summarize the data to count distinct values
species_summary <- dyads_species %>%
  filter(timepoint != "incl") %>%
  group_by(OTU, child, code) %>%
  dplyr::summarize(
    n_timepoints = n_distinct(timepoint),
    .groups = 'drop'
  ) %>%
  group_by(OTU, child) %>%
  dplyr::summarize(
    n_codes = n_distinct(code),
    all_codes_have_2_timepoints = all(n_timepoints >= 1),
    .groups = 'drop'
  ) %>%
  group_by(OTU) %>%
  dplyr::summarize(
    n_children = n_distinct(child),
    all_children_have_2_codes_with_2_timepoints = all(n_codes == 1 & all_codes_have_2_timepoints),
    .groups = 'drop'
  ) %>%
  filter(
    n_children == 1,
    all_children_have_2_codes_with_2_timepoints
  )

species_summary <- dyads_species %>%
  filter(timepoint != "incl") %>%
  group_by(species, child, code) %>%
  dplyr::summarize(n_timepoints = n_distinct(timepoint), .groups = 'drop') %>%
  group_by(species, child) %>%
  dplyr::summarize(
    n_codes = n_distinct(code),
    all_codes_have_required_timepoints = ifelse(child == "no", all(n_timepoints >= 3), all(n_timepoints >= 2)),
    .groups = 'drop'
  ) %>%
  group_by(species) %>%
  dplyr::summarize(
    n_children = n_distinct(child),
    all_children_have_required_codes_with_timepoints = all(n_codes == 2 & all_codes_have_required_timepoints),
    .groups = 'drop'
  ) %>%
  filter(
    n_children == 2,
    all_children_have_required_codes_with_timepoints
  )


# Filter original data based on the summarized data
relAbuSpeDyads <- dyads_species %>%
  filter(timepoint != "incl", species %in% species_summary$species)


summary_relAbuSpeDyads <- summarySE(relAbuSpeDyads, measurevar = "Abundance", groupvars = c("species", "timepoint", "code", "child"), na.rm = TRUE) %>%
  mutate(Intervention = if_else(code == "B Z", "Intervention", "Control")) %>%
  mutate(Dyads = if_else(child == "yes", "Infant", "Mother")) %>%
  group_by(species) %>%
  filter(all(N>=5)) %>%
  ungroup()
  

# Calculate the average abundance for each species
avg_abundance <- relAbuSpeDyads %>%
  group_by(species) %>%
  dplyr::summarise(avg_abundance = mean(Abundance, na.rm = TRUE)) %>%
  arrange(desc(avg_abundance))

# Select the top 15 species
top_fifteen_species <- avg_abundance %>%
  filter(species %in% summary_relAbuSpeDyads$species) %>%
  slice_max(order_by = avg_abundance, n=10) %>%
  pull(species)

# Filter the dataset to include only the top 15 species
filtered_data <- summary_relAbuSpeDyads %>%
  filter(species %in% top_fifteen_species) %>%
  mutate(Abundance = Abundance *100) %>%
  mutate(timepoint = factor(timepoint, levels = c("tri3", "pn12", "pn56"))) %>%
  mutate(Dyads = factor(Dyads, levels = c("Mother", "Infant")))%>%
  mutate(se = se*100) %>%
  mutate(sd=sd*100) %>%
  mutate(ci = ci*100)

unique(filtered_data$species)


filtered_data3 <- filtered_data %>%
  filter(species == "Bacteroides fragilis"| species =="Bacteroides fragilis_A" | species =="Escherichia coli" )

SharedInDyadsHigh <- ggplot(filtered_data3, aes(x = timepoint, y = Abundance, color = Intervention, linetype = Dyads, group = interaction(species, Dyads, Intervention))) +
  geom_line(position = position_dodge(0.2), alpha = 0.7) +  # Add lines for each group
  geom_point(position = position_dodge(0.2), alpha = 0.7) +  # Add points for each data point
  #geom_errorbar(aes(ymin = Abundance - se, ymax = Abundance + se), width = 0.05, position = position_dodge(0.2), alpha = 0.5) +  # Add error bars
  labs(title = "Abundance Across Timepoints by Species",
       x = "Timepoint",
       y = "Relative Abundance (%)",
       color = "Intervention",
       linetype = "Dyads") +
  theme_minimal() +
  scale_color_manual(values = c("Intervention" = "steelblue", "Control" = "darkred")) +
  theme(legend.position = "right",  # Improve readability of x-axis labels
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.line = element_line(color = "black"),  # Add axis lines
        axis.ticks = element_line(color = "black"),  # Add axis ticks
        axis.text = element_text(color = "black"),  # Add axis text
        axis.title = element_text(color = "black")) +  # Add axis titles
  facet_wrap(~ species, nrow = 1, strip.position = "bottom", scales = "free_x")

SharedInDyadsHigh

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/SharedInDyadsHigh.svg",
  width=5.5, height=2.5
  )
SharedInDyadsHigh
dev.off()


filtered_data4 <- filtered_data %>%
  filter(species != "Bacteroides fragilis" & species !="Bacteroides fragilis_A" & species !="Escherichia coli")

SharedInDyadsLow <- ggplot(filtered_data4, aes(x = timepoint, y = Abundance, color = Intervention, linetype = Dyads, group = interaction(species, Dyads, Intervention))) +
  geom_line(position = position_dodge(0.2), alpha = 0.7) +  # Add lines for each group
  geom_point(position = position_dodge(0.2), alpha = 0.7) +  # Add points for each data point
  #geom_errorbar(aes(ymin = Abundance - se, ymax = Abundance + se), width = 0.05, position = position_dodge(0.2), alpha = 0.5) +  # Add error bars
  labs(title = "Abundance Across Timepoints by Species",
       x = "Timepoint",
       y = "Relative Abundance (%)",
       color = "Intervention",
       linetype = "Dyads") +
  theme_minimal() +
  scale_color_manual(values = c("Intervention" = "steelblue", "Control" = "darkred")) +
  theme(legend.position = "right",  # Improve readability of x-axis labels
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.line = element_line(color = "black"),  # Add axis lines
        axis.ticks = element_line(color = "black"),  # Add axis ticks
        axis.text = element_text(color = "black"),  # Add axis text
        axis.title = element_text(color = "black")) +  # Add axis titles
  facet_wrap(~ species, nrow = 1, strip.position = "bottom", scales = "free_x")

SharedInDyadsLow

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/SharedInDyadsLow.svg",
  width=10, height=2.5
  )
SharedInDyadsLow
dev.off()


```

## D. top in mother
```{r}

mother_species <- mother_ps %>%
  tax_glom(taxrank = "species") %>%
  transform_sample_counts(function(x){x/sum(x)}) %>%
  psmelt() %>%
  filter(Abundance > 0)
  
relAbuSpeThreeMo <- mother_species %>%
  filter(timepoint != "incl") %>%
  dplyr::group_by(OTU) %>%
  filter(n_distinct(timepoint) == 3 & n_distinct(code) ==2) %>%
  mutate(timepoint = factor(timepoint, levels = desired_order)) %>%
  dplyr::ungroup()

summary_relAbuSpeThreeMo <- summarySE(relAbuSpeThreeMo, measurevar = "Abundance", groupvars = c("species", "timepoint", "code"), na.rm = TRUE) %>%
  mutate(Intervention = if_else(code == "B Z", "Intervention", "Control")) %>%
  group_by(species) %>%
  filter(all(N>5)) %>%
  ungroup()
  

# Calculate the average abundance for each species
avg_abundance <- relAbuSpeThreeMo %>%
  group_by(species) %>%
  dplyr::summarise(avg_abundance = mean(Abundance, na.rm = TRUE)) %>%
  arrange(desc(avg_abundance))

# Select the top 15 species
top_fifteen_species <- avg_abundance %>%
  filter(species %in% summary_relAbuSpeThreeMo$species) %>%
  slice_max(order_by = avg_abundance, n=10) %>%
  pull(species)

# Filter the dataset to include only the top 15 species
filtered_data <- summary_relAbuSpeThreeMo %>%
  filter(species %in% top_fifteen_species) %>%
  mutate(Abundance = Abundance *100) %>%
  mutate(se = se*100) %>%
  mutate(sd=sd*100) %>%
  mutate(ci = ci*100)

plotMoThreeBEPIFA <- ggplot(filtered_data, aes(x = timepoint, y = Abundance, color = Intervention)) +
  geom_errorbar(aes(ymin = Abundance - se, ymax = Abundance +se), width = 0.05, position = position_dodge(0.05), alpha = 0.5) +
  geom_line(aes(group = interaction(species, Intervention), color = Intervention), position = position_dodge(0.05) )+
  geom_point(position = position_dodge(0.05)) +
  facet_wrap(~ species, scales = "free_y", ncol = 5) +
  theme_bw() +
  labs(color = "Intervention", y = "Relative Abundance, %", x = "Timepoint" ) +
  scale_color_manual(values = c("Intervention" = "steelblue", "Control" = "darkred")) +
  expand_limits(y = 0) +  # Ensure y-axis starts at 0
  theme(panel.spacing = unit(0.5, "lines"),
        strip.text = element_text(size = 8))

  plotMoThreeBEPIFA

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/NotsharedSpeciesThreeMoBEPIFA2.svg",
  width=10, height=3
  )
plotMoThreeBEPIFA
dev.off()

```


### mixed effect

```{r}

relAbuSpeMo <- relAbuSpeThreeMo %>%
  filter(timepoint != "incl") %>%
  #group_by(OTU, idbs, code) %>%
  #filter(n_distinct(Sample) == 3) %>%
  mutate(Intervention = if_else(code == "B Z", "2", "1")) %>%
  mutate(timepoint = factor(timepoint, levels = desired_order)) %>%
  mutate(Timepoint = if_else(timepoint == "tri3", "1",
                             if_else(timepoint == "pn12", "2", "3"))) %>%
  #ungroup() %>%
  select(species, Sample, idbs, Intervention, Timepoint, Abundance)

library(lmerTest)
library(emmeans)

Mo_fragilis <- relAbuSpeMo %>% filter(species == "Bacteroides fragilis")
Mo_fragilis_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Mo_fragilis)
summary(Mo_fragilis_model)
confint(Mo_fragilis_model, level = 0.95, method = "profile")
p_values <- summary(Mo_fragilis_model)$coefficients[, "Pr(>|t|)"]
adjusted_p_values <- round(p.adjust(p_values, method = "BH"),3)
adjusted_p_values


Mo_c941 <- relAbuSpeMo %>% filter(species == "C941 sp004557565")
Mo_c941_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Mo_c941)
summary(Mo_c941_model)

Mo_copri <- relAbuSpeMo %>% filter(species == "Prevotella copri") 
Mo_copri_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Mo_copri)
summary(Mo_copri_model)

Mo_sp002251295 <- relAbuSpeMo %>% filter(species == "Prevotella sp002251295") 
Mo_sp002251295_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Mo_sp002251295)
summary(Mo_sp002251295_model)

Mo_sp000434515 <- relAbuSpeMo %>% filter(species == "Prevotella sp000434515") 
Mo_sp000434515_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Mo_sp000434515)
summary(Mo_sp000434515_model)

Mo_sp002299635 <- relAbuSpeMo %>% filter(species == "Prevotella sp002299635") 
Mo_sp002299635_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Mo_sp002299635)
summary(Mo_sp002299635_model)
confint(Mo_sp002299635_model, level = 0.95, method = "profile")
p_values <- summary(Mo_sp002299635_model)$coefficients[, "Pr(>|t|)"]
adjusted_p_values <- round(p.adjust(p_values, method = "BH"),3)
adjusted_p_values

Mo_sp900546535 <- relAbuSpeMo %>% filter(species == "Prevotella sp900546535") 
Mo_sp900546535_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Mo_sp900546535)
summary(Mo_sp900546535_model)

Mo_sp900551275 <- relAbuSpeMo %>% filter(species == "Prevotella sp900551275") 
Mo_sp900551275_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Mo_sp900551275)
summary(Mo_sp900551275_model)

Mo_rc9 <- relAbuSpeMo %>% filter(species == "RC9 sp000433355") 
Mo_rc9_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Mo_rc9 )
summary(Mo_rc9_model)
confint(Mo_rc9_model, level = 0.95, method = "profile")
p_values <- summary(Mo_rc9_model)$coefficients[, "Pr(>|t|)"]
adjusted_p_values <- round(p.adjust(p_values, method = "BH"),3)
adjusted_p_values


Mo_unk <- relAbuSpeMo %>% filter(species == "Unknown_Prevotella") 
Mo_unk_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Mo_unk)
summary(Mo_unk_model)


```


## D. top in infant
```{r}

library(Rmisc) 
infant_species <- infant_ps %>%
  tax_glom(taxrank = "species") %>%
  transform_sample_counts(function(x){x/sum(x)}) %>%
  psmelt() %>%
  filter(Abundance > 0)
  
relAbuSpeInf <- infant_species %>%
  group_by(OTU) %>%
  filter(n_distinct(timepoint) == 2 & n_distinct(code) == 2) %>%
  mutate(timepoint = factor(timepoint, levels = desired_order)) %>%
  ungroup()

summary_relAbuSpeInf <- summarySE(relAbuSpeInf, measurevar = "Abundance", groupvars = c("species", "timepoint", "code"), na.rm = TRUE)%>%
  mutate(Intervention = if_else(code == "B Z", "Intervention", "Control"))%>%
  group_by(species) %>%
  filter(all(N>5)) %>%
  ungroup()

avg_abundance <- relAbuSpeInf %>%
  group_by(species) %>%
  dplyr::summarise(avg_abundance = mean(Abundance, na.rm = TRUE)) %>%
  arrange(desc(avg_abundance))

# Select the top 15 species
top_fifteen_species <- avg_abundance %>%
  filter(species %in% summary_relAbuSpeInf$species) %>%
  slice_max(order_by = avg_abundance, n=10) %>%
  pull(species)

# Filter the dataset to include only the top 15 species
filtered_data <- summary_relAbuSpeInf %>%
  filter(species %in% top_fifteen_species) %>%
  mutate(Abundance = Abundance *100) %>%
  mutate(se = se*100) %>%
  mutate(sd=sd*100) %>%
  mutate(ci = ci*100) 

plotInfBEPIFA <- ggplot(filtered_data, aes(x = timepoint, y = Abundance, color = Intervention)) +
  geom_errorbar(aes(ymin = Abundance - se, ymax = Abundance + se), linetype = "dashed", width = 0.05, position = position_dodge(0.05), alpha = 0.5) +
  geom_line(aes(group = interaction(species, Intervention)), linetype = "dashed", position = position_dodge(0.05)) +  # Set linetype globally to "dashed"
  geom_point(position = position_dodge(0.05)) +
  facet_wrap(~ species, scales = "free_y", ncol = 5) +
  theme_bw() +
  labs(color = "Intervention", y = "Relative Abundance, %", x = "Timepoint") +
  scale_color_manual(values = c("Intervention" = "steelblue", "Control" = "darkred")) +
  expand_limits(y = 0.0) + # Ensure y-axis starts at 0
  theme(panel.spacing = unit(0.5, "lines"),
        strip.text = element_text(size = 8))
  
plotInfBEPIFA

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/NotsharedSpeciesInfBEPIFA2.svg",
  width=10, height=2.8
  )
plotInfBEPIFA
dev.off()

```


### mixed effect

```{r}

relAbuSpeInf2 <- relAbuSpeInf %>%
  #group_by(OTU, idbs, code) %>%
  #filter(n_distinct(Sample) == 2) %>%
  mutate(Intervention = if_else(code == "B Z", "2", "1")) %>%
  mutate(timepoint = factor(timepoint, levels = desired_order)) %>%
  mutate(Timepoint = if_else(timepoint == "pn12", "1", "2")) %>%
  #ungroup() %>%
  select(species, Sample, idbs, Intervention, Timepoint, Abundance)

Inf_fragilis <- relAbuSpeInf2 %>% filter(species == "Bacteroides fragilis_A")
Inf_fragilis_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Inf_fragilis)
summary(Inf_fragilis_model)

Inf_uniformis <- relAbuSpeInf2 %>% filter(species == "Bacteroides uniformis")
Inf_uniformis_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Inf_uniformis)
summary(Inf_uniformis_model)

Inf_bifidum <- relAbuSpeInf2 %>% filter(species == "Bifidobacterium bifidum") 
Inf_bifidum_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Inf_bifidum)
summary(Inf_bifidum_model)

Inf_breve <- relAbuSpeInf2 %>% filter(species == "Bifidobacterium breve") 
Inf_breve_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Inf_breve)
summary(Inf_breve_model)
confint(Inf_breve_model, level = 0.95, method = "profile")

Inf_infantis <- relAbuSpeInf2 %>% filter(species == "Bifidobacterium infantis") 
Inf_infantis_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Inf_infantis)
summary(Inf_infantis_model)

Inf_longum <- relAbuSpeInf2 %>% filter(species == "Bifidobacterium longum") 
Inf_longum_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Inf_longum)
summary(Inf_longum_model)

Inf_coli <- relAbuSpeInf2 %>% filter(species == "Escherichia coli") 
Inf_coli_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Inf_coli)
summary(Inf_coli_model)

Inf_pneumoniae <- relAbuSpeInf2 %>% filter(species == "Klebsiella pneumoniae") 
Inf_pneumoniae_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Inf_pneumoniae)
summary(Inf_pneumoniae_model)

Inf_uli <- relAbuSpeInf2 %>% filter(species == "Parolsenella uli_B") 
Inf_uli_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Inf_uli)
summary(Inf_uli_model)

Inf_sp000431435 <- relAbuSpeInf2 %>% filter(species == "Veillonella_A sp000431435") 
Inf_sp000431435_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Inf_sp000431435)
summary(Inf_sp000431435_model)


```




## E. intra dist ~ days + (1|subject)

### reformate tables 
```{r}
incl_se1g_date <- incl_se1g %>%
  mutate(
    incl_se1g_dadi = na_if(incl_se1g_dadi, ""),
    incl_sa10ab_dtdi = na_if(incl_sa10ab_dtdi, ""),
    incl_sa20ab_dadi = na_if(incl_sa20ab_dadi, ""),
    incl_sapl_dati = na_if(incl_sapl_dati, "")
  ) %>%
  mutate(date = coalesce(substr(incl_se1g_dadi, 1, 8), substr(incl_sa10ab_dtdi, 1, 8), substr(incl_sa20ab_dadi, 1, 8), substr(incl_sapl_dati, 1, 8))) %>%
  mutate(sampleid = paste0("incl_se1g_", incl_idbs)) %>%
  dplyr::select(sampleid, date)

tri3_se1g_date <- tri3_se1g %>%
  mutate(
    tri3_se1g_dadi = na_if(tri3_se1g_dadi, ""),
    tri3_sa10ab_dtdi = na_if(tri3_sa10ab_dtdi, ""),
    tri3_sa20ab_dadi = na_if(tri3_sa20ab_dadi, ""),
    tri3_sapl_dati = na_if(tri3_sapl_dati, "")
  ) %>%
  mutate(date = coalesce(substr(tri3_se1g_dadi, 1, 8), substr(tri3_sa10ab_dtdi, 1, 8), substr(tri3_sa20ab_dadi, 1, 8), substr(tri3_sapl_dati, 1, 8))) %>%
  mutate(sampleid = paste0("tri3_se1g_", tri3_idbs)) %>%
  dplyr::select(sampleid, date)

pn12m_se1g_date <- pn12m_se1g %>%
  mutate(
    pn12m_se1g_dadi = na_if(pn12m_se1g_dadi, ""),
    pn12m_sapl_dati = na_if(pn12m_sapl_dati, "")
  ) %>%
  mutate(date = coalesce(substr(pn12m_se1g_dadi, 1, 8),  substr(pn12m_sapl_dati, 1, 8))) %>%
  mutate(sampleid = paste0("pn12_se1g_", pn12m_idbs)) %>%
  dplyr::select(sampleid, date)

pn56m_se1g_date <- pn56m_se1g %>%
  mutate(
    pn56m_se1g_dadi = na_if(pn56m_se1g_dadi, ""),
    pn56m_sa10ab_dtdi = na_if(pn56m_sa10ab_dtdi, ""),
    pn56m_sa20ab_dadi = na_if(pn56m_sa20ab_dadi, ""),
    pn56m_sapl_dati = na_if(pn56m_sapl_dati, "")
  ) %>%
  mutate(date = coalesce(substr(pn56m_se1g_dadi, 1, 8), substr(pn56m_sa10ab_dtdi, 1, 8), substr(pn56m_sa20ab_dadi, 1, 8), substr(pn56m_sapl_dati, 1, 8))) %>%
  mutate(sampleid = paste0("pn56_se1g_", pn56m_idbs)) %>%
  dplyr::select(sampleid, date)

pn12e_se1g_date <- pn12e_se1g %>%
  mutate(
    pn12e_se1g_dadi = na_if(pn12e_se1g_dadi, ""),
    pn12e_sa10ab_dtdi = na_if(pn12e_sa10ab_dtdi, ""),
    pn12e_sa20ab_dati = na_if(pn12e_sa20ab_dati, ""),
  ) %>%
  mutate(date = coalesce(substr(pn12e_se1g_dadi, 1, 8), substr(pn12e_sa10ab_dtdi, 1, 8), substr(pn12e_sa20ab_dati, 1, 8))) %>%
  mutate(sampleid = paste0("pn12_se1g_", pn12e_idbs, "_e")) %>%
  dplyr::select(sampleid, date)

pn56e_se1g_date <- pn56e_se1g %>%
  mutate(
    pn56e_se1g_dadi = na_if(pn56e_se1g_dadi, ""),
    pn56e_sa10ab_dtdi = na_if(pn56e_sa10ab_dtdi, ""),
    pn56e_sa20ab_dati = na_if(pn56e_sa20ab_dati, ""),
  ) %>%
  mutate(date = coalesce(substr(pn56e_se1g_dadi, 1, 8), substr(pn56e_sa10ab_dtdi, 1, 8), substr(pn56e_sa20ab_dati, 1, 8))) %>%
  mutate(sampleid = paste0("pn56_se1g_", pn56e_idbs, "_e")) %>%
  dplyr::select(sampleid, date)


combined_se1g_date <- bind_rows(
  incl_se1g_date,
  tri3_se1g_date,
  pn12e_se1g_date,
  pn56e_se1g_date,
  pn12m_se1g_date,
  pn56m_se1g_date
)

dist_df_combined <- rbind(dist_df1, dist_df2, dist_df3, dist_df4)
# Convert the date column to Date type if it's not already
combined_se1g_date <- combined_se1g_date %>%
  mutate(date = as.Date(date))

# Print the structure of combined_se1g_date to check column names
print(str(combined_se1g_date))

# Merge datasets to get date1 and date2
dist_df_combined <- dist_df_combined %>%
  # Merge to get date1
  left_join(
    combined_se1g_date %>%
      dplyr::select(sampleid, date) %>%
      mutate(date1 = date) %>%
      dplyr::select(-date),  # Exclude original date column
    by = c("Sample1" = "sampleid")
  ) %>%
  # Merge to get date2
  left_join(
    combined_se1g_date %>%
      dplyr::select(sampleid, date) %>%
      mutate(date2 = date) %>%
      dplyr::select(-date),  # Exclude original date column
    by = c("Sample2" = "sampleid")
  )

# Print the structure of dist_df_combined to check if date1 and date2 are added
print(str(dist_df_combined))

# Ensure date columns are in Date format and calculate the absolute difference
dist_df_combined <- dist_df_combined %>%
  mutate(
    date1 = as.Date(date1),
    date2 = as.Date(date2),
    # Calculate the absolute difference in days
    date_diff = abs(as.numeric(difftime(date1, date2, units = "days")))
  )

# Extract idbs from Sample1 and Sample2
dist_df_combined <- dist_df_combined %>%
  mutate(idbs1 = sub(".*_(\\d+)", "\\1", Sample1),
         idbs2 = sub(".*_(\\d+)", "\\1", Sample2))

# Filter to keep only intra-subject distances
dist_df_combined_intra <- dist_df_combined %>%
  filter(idbs1 == idbs2) %>%
  drop_na()
```



```{r}

# Initialize a list to store results
results <- list()

# Fit the model for each unique group level
for (group_level in unique(dist_df_combined_intra$group)) {
  # Filter the data for the current group level
  group_data <- dist_df_combined_intra %>%
    filter(group == group_level)
  
  # Check if there are enough data points to fit the model
  if (nrow(group_data) > 1) {
    # Rescale variables
    group_data <- group_data %>%
      mutate(
        transformed_dist_scaled = scale(transformed_dist, center = TRUE, scale = TRUE),
        date_diff_scaled = scale(date_diff, center = TRUE, scale = TRUE)
      )
    
    # Fit the mixed-effects model with tryCatch to handle potential errors
    model <- tryCatch(
      lmer(transformed_dist_scaled ~ date_diff_scaled + (1 | idbs1), data = group_data),
      error = function(e) NULL
    )
    
    if (!is.null(model)) {
      # Extract beta coefficients and CI using broom.mixed
      model_summary <- tidy(model, conf.int = TRUE, conf.level = 0.95)
      
      # Filter for date_diff_scaled coefficient
      beta_coefficient <- model_summary %>%
        filter(term == "date_diff_scaled") %>%
        dplyr::select(estimate, conf.low, conf.high)
      
      # Store the result in the list
      results[[group_level]] <- beta_coefficient
    } else {
      warning(paste("Model did not converge for group:", group_level))
    }
  }
}

# Convert results list to a data frame
results_df <- bind_rows(
  lapply(names(results), function(group) {
    data.frame(
      group = group,
      beta_coefficient = results[[group]]$estimate,
      conf_int_lower = results[[group]]$conf.low,
      conf_int_upper = results[[group]]$conf.high
    )
  })
)

# Print the results data frame
print(results_df)

intra_dist_plot <- ggplot(results_df, aes(x = beta_coefficient, y = group, color = group, shape = group)) +
  geom_point(size = 4) +
  geom_errorbarh(aes(xmin = `conf_int_lower`, xmax = `conf_int_upper`), height = 0.2) +
  scale_color_manual(values = c("#00A98F", "#C00000", "skyblue", "#f1c232")) +  # Customize colors
  scale_shape_manual(values = c(16, 17, 18, 15)) +  # Customize shapes
  labs(x = "Beta Coefficient Estimates", y = "") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 12)) +
  xlim(0, 0.7)# Customize y-axis text
intra_dist_plot

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/intra_dist_plot.svg",
  width=6, height=2
  )
intra_dist_plot
dev.off()

model <- lmer(transformed_dist ~ date_diff*group + (1 | idbs1), data = dist_df_combined_intra)

summary(model)

dist_df_combined_intra$group <- relevel(factor(dist_df_combined_intra$group), ref = "Infant-intervention")

```

#### line plot show the change of distance

```{r}
library(dplyr)

dist_df_combined_intra_mo <- dist_df_combined_intra %>%
  filter(group == "Mother-intervention" | group == "Mother-control") %>%
  mutate(Intervention = if_else(group == "Mother-intervention", "Intervention", "Control"))

# Calculate summary statistics including SE and CI
summary_stats_mo <- dist_df_combined_intra_mo %>%
  group_by(Intervention, timediff) %>%
  dplyr::summarise(
    mean_dist = mean(dist, na.rm = TRUE),
    sd_dist = sd(dist, na.rm = TRUE),
    n = n(),  # Number of observations
    se = sd_dist / sqrt(n),  # Standard Error
    ci = se * qt(0.975, df = n - 1)  # 95% Confidence Interval
  )

# Print the summarized data
print(summary_stats_mo)

# Create the plot with error bars
dist_plot_mo <- ggplot(summary_stats_mo, aes(x = timediff, y = mean_dist, color = Intervention, group = Intervention)) +
  geom_line(size = 1) +
  geom_point(size = 3) +
  geom_errorbar(aes(ymin = mean_dist - se, ymax = mean_dist + se), width = 0.2, alpha = 0.5) +
  labs(x = "Time point Interval ", y = "Transformed Distance", color = "Group") +
scale_color_manual(values = c("Intervention" = "steelblue", "Control" = "darkred")) +  theme_bw() +
  theme(axis.text.y = element_text(size = 12),
        axis.text.x = element_text(size = 12),
        legend.title = element_text(size = 12),
        legend.text = element_text(size = 10)) 

# Print the plot
print(dist_plot_mo)

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/dist_plot_mo.svg",
  width=4, height=3
  )
dist_plot_mo
dev.off()

dist_df_combined_intra_mo_model <- lmer(dist ~ timediff * group + (1 | idbs1), data = dist_df_combined_intra_mo)
summary(dist_df_combined_intra_mo_model)


```


## F. line plot 
```{r}

dist_df_combined_intra_inf <- dist_df_combined_intra %>%
  filter(group == "Infant-intervention" | group == "Infant-control") %>%
  mutate(Intervention = if_else(group == "Infant-intervention", "Intervention", "Control"))

# Calculate summary statistics including SE and CI
summary_stats_inf <- dist_df_combined_intra_inf %>%
  group_by(Intervention, timediff) %>%
  dplyr::summarise(
    mean_dist = mean(dist, na.rm = TRUE),
    sd_dist = sd(dist, na.rm = TRUE),
    n = n(),  # Number of observations
    se = sd_dist / sqrt(n),  # Standard Error
    ci = se * qt(0.975, df = n - 1)  # 95% Confidence Interval
  )

# Print the summarized data
print(summary_stats_inf)

# Create the plot with error bars
dist_plot_inf <- ggplot(summary_stats_inf, aes(x = timediff, y = mean_dist, color = Intervention, group = Intervention)) +
  geom_line(size = 1, linetype = "dashed") +
  geom_point(size = 3) +
  geom_errorbar(aes(ymin = mean_dist - se, ymax = mean_dist + se), linetype = "dashed", width = 0.2, alpha = 0.5) +
  labs(x = "Time Point Interval", y = "UniFrac Distance", color = "Group") +
scale_color_manual(values = c("Intervention" = "steelblue", "Control" = "darkred")) +  theme_bw() +
  theme(axis.text.y = element_text(size = 12),
        axis.text.x = element_text(size = 12),
        legend.title = element_text(size = 12),
        legend.text = element_text(size = 10)) # Optionally ensure y-axis starts at 0

# Print the plot
print(dist_plot_inf)

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/dist_plot_inf.svg",
  width=3.5, height=3
  )
dist_plot_inf
dev.off()

model7 <- lmer(transformed_dist ~ date_diff*group + (1 | Sample1), data = dist_df_combined)
summary(model7)


dist_df_combined <- rbind(dist_df1, dist_df2, dist_df3, dist_df4)
model5 <- lmer(transformed_dist ~ timediff*group + (1 | Sample1), data = dist_df_combined)
summary(model5)

# Relevel the group factor so that "Mother-control" is the reference level
dist_df_combined$group <- relevel(factor(dist_df_combined$group), ref = "Mother-control")
model6 <- lmer(transformed_dist ~ timediff * group + (1 | Sample1), data = dist_df_combined)
summary(model6)

# Define a function to extract the required values from a model
extract_model_stats <- function(model, model_name) {
  model_summary <- summary(model)
  conf_intervals <- confint(model, level = 0.95)
  
  data.frame(
    Model = model_name,
    Effect = "timediff",
    Estimate = model_summary$coefficients["timediff", "Estimate"],
    `2.5% CI` = conf_intervals["timediff", "2.5 %"],
    `97.5% CI` = conf_intervals["timediff", "97.5 %"],
    `P-value` = model_summary$coefficients["timediff", "Pr(>|t|)"]
  )
}

# Apply the function to all models
results_list <- list(
  extract_model_stats(model1, "model1"),
  extract_model_stats(model2, "model2"),
  extract_model_stats(model3, "model3"),
  extract_model_stats(model4, "model4")
)

# Combine all results into one data frame
all_results <- do.call(rbind, results_list)
all_results <- all_results %>%
  mutate(Model = case_when(
    Model == "model1" ~ "Mother - intervention",
    Model == "model2" ~ "Mother - control",
    Model == "model3" ~ "Infant - intervention",
    Model == "model4" ~ "Infant - control"
  ))
ggplot(all_results, aes(x = Estimate, y = Model, color = Model, shape = Model)) +
  geom_point(size = 4) +
  geom_errorbarh(aes(xmin = `X2.5..CI`, xmax = `X97.5..CI`), height = 0.2) +
  scale_color_manual(values = c("#00A98F", "#C00000", "skyblue", "#f1c232")) +  # Customize colors
  scale_shape_manual(values = c(16, 17, 18, 15)) +  # Customize shapes
  labs(x = "Beta Coefficient Estimates", y = "") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 12))  # Customize y-axis text

```


# Figure 4. co-occurrence
## incl 
```{r}
# subset data
incl_bep <- subset_samples(incl, code=="B Z")
incl_ifa <- subset_samples(incl, code=="A Y")

# filter data
filtered_incl_bep <- filter_taxa(incl_bep, function(x) sum(x > 0.01) > (0.05 * length(x)), TRUE)
filtered_incl_ifa <- filter_taxa(incl_ifa, function(x) sum(x > 0.01) > (0.05 * length(x)), TRUE)
otu_data_filtered_incl_bep <- otu_table(filtered_incl_bep)
otu_data_filtered_incl_ifa <- otu_table(filtered_incl_ifa)

# model: time-consuming
spiec_result_incl_bep <- spiec.easi(filtered_incl_bep, method = "mb", lambda.min.ratio = 1e-2, nlambda = 20)
spiec_result_incl_ifa <- spiec.easi(filtered_incl_ifa, method = "mb", lambda.min.ratio = 1e-2, nlambda = 20)

# plot and density/diameter
adj_matrix_incl_bep <- as.matrix(symBeta(getOptBeta(spiec_result_incl_bep), mode = "maxabs"))
adj_matrix_incl_bep[adj_matrix_incl_bep < 0] <- 0
net_incl_bep <- graph_from_adjacency_matrix(adj_matrix_incl_bep, mode = "undirected", weighted = TRUE, diag = FALSE)
plot(net_incl_bep, vertex.label = NA, vertex.size = 5, edge.width = E(net_incl_bep)$weight * 5, edge.color = "gray50")
degree_values <- degree(net_incl_bep)
betweenness_values <- betweenness(net_incl_bep)
closeness_values <- closeness(net_incl_bep)

adj_matrix_incl_ifa <- as.matrix(symBeta(getOptBeta(spiec_result_incl_ifa), mode = "maxabs"))
adj_matrix_incl_ifa[adj_matrix_incl_ifa < 0] <- 0
net_incl_ifa <- graph_from_adjacency_matrix(adj_matrix_incl_ifa, mode = "undirected", weighted = TRUE, diag = FALSE)
plot(net_incl_ifa, vertex.label = NA, vertex.size = 5, edge.width = E(net_incl_ifa)$weight * 5, edge.color = "gray50")
degree_values <- degree(net_incl_ifa)
betweenness_values <- betweenness(net_incl_ifa)
closeness_values <- closeness(net_incl_ifa)

#  Community Detection
louvain_communities <- cluster_louvain(net_incl_bep)
plot(louvain_communities, net_incl_bep, vertex.size = 5, vertex.label = NA, main = "Louvain Community Detection")

louvain_communities <- cluster_louvain(net_incl_ifa)
plot(louvain_communities, net_incl_ifa, vertex.size = 5, vertex.label = NA, main = "Louvain Community Detection")

```

### visualize community
```{r}
# Community Detection and Layout for B Z Group
louvain_communities_bep <- cluster_louvain(net_incl_bep)
layout_bep <- layout_with_fr(net_incl_bep, niter = 1000)  # Increase iterations for better distribution
layout_bep_scaled <- layout_bep * 1000  # Multiply the coordinates by a factor to increase spacing between nodes

# Plot the Louvain communities with scaled layout
svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/coOccurrence_incl_bep.svg",
  width=10, height=10
)

plot(louvain_communities_bep, net_incl_bep,
     layout = layout_bep_scaled,        # Apply the scaled layout for longer edges
     vertex.size = degree(net_incl_bep) * 0.2,  # Node size proportional to degree (number of connections)
     vertex.label = NA,                  # No vertex labels
     edge.width = E(net_incl_bep)$weight * 1, # Edge width proportional to weight
     edge.color = "gray",
     mark.groups = NULL,
     main = "Louvain Community Detection (B Z Group - Longer Edges)")

dev.off()


# Community Detection and Layout for A Y Group
louvain_communities_ifa <- cluster_louvain(net_incl_ifa)
layout_ifa <- layout_with_fr(net_incl_ifa, niter = 1000)  # Increase iterations for better distribution
layout_ifa_scaled <- layout_ifa * 1000  # Multiply the coordinates by a factor to increase spacing between nodes

# Plot the Louvain communities with scaled layout
svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/coOccurrence_incl_ifa.svg",
  width=10, height=10
)

plot(louvain_communities_ifa, net_incl_ifa,
     layout = layout_ifa,               # Apply the more spread-out layout
     vertex.size = degree(net_incl_ifa) * 0.2,  # Node size proportional to degree (number of connections)
     vertex.label = NA,                  # No vertex labels
     edge.width = E(net_incl_ifa)$weight * 1, # Edge width proportional to weight
     edge.color = "gray",
     mark.groups = NULL,
     main = "Louvain Community Detection (A Y Group - Larger Layout)")
dev.off()

```


## tri3 
```{r}
# subset data
tri3_bep <- subset_samples(tri3, code=="B Z")
tri3_ifa <- subset_samples(tri3, code=="A Y")

# filter data
filtered_tri3_bep <- filter_taxa(tri3_bep, function(x) sum(x > 0.01) > (0.05 * length(x)), TRUE)
filtered_tri3_ifa <- filter_taxa(tri3_ifa, function(x) sum(x > 0.01) > (0.05 * length(x)), TRUE)
otu_data_filtered_tri3_bep <- otu_table(filtered_tri3_bep)
otu_data_filtered_tri3_ifa <- otu_table(filtered_tri3_ifa)

# model: time-consuming
spiec_result_tri3_bep <- spiec.easi(filtered_tri3_bep, method = "mb", lambda.min.ratio = 1e-2, nlambda = 20)
spiec_result_tri3_ifa <- spiec.easi(filtered_tri3_ifa, method = "mb", lambda.min.ratio = 1e-2, nlambda = 20)

# plot and density/diameter
adj_matrix_tri3_bep <- as.matrix(symBeta(getOptBeta(spiec_result_tri3_bep), mode = "maxabs"))
adj_matrix_tri3_bep[adj_matrix_tri3_bep < 0] <- 0
net_tri3_bep <- graph_from_adjacency_matrix(adj_matrix_tri3_bep, mode = "undirected", weighted = TRUE, diag = FALSE)
degree_values <- degree(net_tri3_bep)
betweenness_values <- betweenness(net_tri3_bep)
closeness_values <- closeness(net_tri3_bep)

adj_matrix_tri3_ifa <- as.matrix(symBeta(getOptBeta(spiec_result_tri3_ifa), mode = "maxabs"))
adj_matrix_tri3_ifa[adj_matrix_tri3_ifa < 0] <- 0
net_tri3_ifa <- graph_from_adjacency_matrix(adj_matrix_tri3_ifa, mode = "undirected", weighted = TRUE, diag = FALSE)
degree_values <- degree(net_tri3_ifa)
betweenness_values <- betweenness(net_tri3_ifa)
closeness_values <- closeness(net_tri3_ifa)

#  Community Detection
louvain_communities <- cluster_louvain(net_tri3_bep)
plot(louvain_communities, net_tri3_bep, vertex.size = 5, vertex.label = NA, main = "Louvain Community Detection")

louvain_communities <- cluster_louvain(net_tri3_ifa)
plot(louvain_communities, net_tri3_ifa, vertex.size = 5, vertex.label = NA, main = "Louvain Community Detection")

```

### visualize community
```{r}
# Community Detection and Layout for B Z Group
louvain_communities_bep <- cluster_louvain(net_tri3_bep)
layout_bep <- layout_with_fr(net_tri3_bep, niter = 1000)  # Increase iterations for better distribution
layout_bep_scaled <- layout_bep * 1000  # Multiply the coordinates by a factor to increase spacing between nodes

# Plot the Louvain communities with scaled layout
svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/coOccurrence_tri3_bep.svg",
  width=10, height=10
)

plot(louvain_communities_bep, net_tri3_bep,
     layout = layout_bep_scaled,        # Apply the scaled layout for longer edges
     vertex.size = degree(net_tri3_bep) * 0.2,  # Node size proportional to degree (number of connections)
     vertex.label = NA,                  # No vertex labels
     edge.width = E(net_tri3_bep)$weight * 1, # Edge width proportional to weight
     edge.color = "gray",
     mark.groups = NULL,
     main = "Louvain Community Detection (B Z Group - Longer Edges)")

dev.off()


# Community Detection and Layout for A Y Group
louvain_communities_ifa <- cluster_louvain(net_tri3_ifa)
layout_ifa <- layout_with_fr(net_tri3_ifa, niter = 1000)  # Increase iterations for better distribution
layout_ifa_scaled <- layout_ifa * 1000  # Multiply the coordinates by a factor to increase spacing between nodes

# Plot the Louvain communities with scaled layout
svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/coOccurrence_tri3_ifa.svg",
  width=10, height=10
)

plot(louvain_communities_ifa, net_tri3_ifa,
     layout = layout_ifa,               # Apply the more spread-out layout
     vertex.size = degree(net_tri3_ifa) * 0.2,  # Node size proportional to degree (number of connections)
     vertex.label = NA,                  # No vertex labels
     edge.width = E(net_tri3_ifa)$weight * 1, # Edge width proportional to weight
     edge.color = "gray",
     mark.groups = NULL,
     main = "Louvain Community Detection (A Y Group - Larger Layout)")
dev.off()

```



## pn12m 
```{r}
# subset data
pn12m_bep <- subset_samples(pn12m, code=="B Z")
pn12m_ifa <- subset_samples(pn12m, code=="A Y")

# filter data
filtered_pn12m_bep <- filter_taxa(pn12m_bep, function(x) sum(x > 0.01) > (0.05 * length(x)), TRUE)
filtered_pn12m_ifa <- filter_taxa(pn12m_ifa, function(x) sum(x > 0.01) > (0.05 * length(x)), TRUE)
otu_data_filtered_pn12m_bep <- otu_table(filtered_pn12m_bep)
otu_data_filtered_pn12m_ifa <- otu_table(filtered_pn12m_ifa)

# model: time-consuming
spiec_result_pn12m_bep <- spiec.easi(filtered_pn12m_bep, method = "mb", lambda.min.ratio = 1e-2, nlambda = 20)
spiec_result_pn12m_ifa <- spiec.easi(filtered_pn12m_ifa, method = "mb", lambda.min.ratio = 1e-2, nlambda = 20)

# plot and density/diameter
adj_matrix_pn12m_bep <- as.matrix(symBeta(getOptBeta(spiec_result_pn12m_bep), mode = "maxabs"))
adj_matrix_pn12m_bep[adj_matrix_pn12m_bep < 0] <- 0
net_pn12m_bep <- graph_from_adjacency_matrix(adj_matrix_pn12m_bep, mode = "undirected", weighted = TRUE, diag = FALSE)
degree_values <- degree(net_pn12m_bep)
betweenness_values <- betweenness(net_pn12m_bep)
closeness_values <- closeness(net_pn12m_bep)

adj_matrix_pn12m_ifa <- as.matrix(symBeta(getOptBeta(spiec_result_pn12m_ifa), mode = "maxabs"))
adj_matrix_pn12m_ifa[adj_matrix_pn12m_ifa < 0] <- 0
net_pn12m_ifa <- graph_from_adjacency_matrix(adj_matrix_pn12m_ifa, mode = "undirected", weighted = TRUE, diag = FALSE)
degree_values <- degree(net_pn12m_ifa)
betweenness_values <- betweenness(net_pn12m_ifa)
closeness_values <- closeness(net_pn12m_ifa)

#  Community Detection
louvain_communities <- cluster_louvain(net_pn12m_bep)
plot(louvain_communities, net_pn12m_bep, vertex.size = 5, vertex.label = NA, main = "Louvain Community Detection")

louvain_communities <- cluster_louvain(net_pn12m_ifa)
plot(louvain_communities, net_pn12m_ifa, vertex.size = 5, vertex.label = NA, main = "Louvain Community Detection")

```

### visualize community
```{r}
# Community Detection and Layout for B Z Group
louvain_communities_bep <- cluster_louvain(net_pn12m_bep)
layout_bep <- layout_with_fr(net_pn12m_bep, niter = 1000)  # Increase iterations for better distribution
layout_bep_scaled <- layout_bep * 1000  # Multiply the coordinates by a factor to increase spacing between nodes

# Plot the Louvain communities with scaled layout
svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/coOccurrence_pn12m_bep.svg",
  width=10, height=10
)

plot(louvain_communities_bep, net_pn12m_bep,
     layout = layout_bep_scaled,        # Apply the scaled layout for longer edges
     vertex.size = degree(net_pn12m_bep) * 0.2,  # Node size proportional to degree (number of connections)
     vertex.label = NA,                  # No vertex labels
     edge.width = E(net_pn12m_bep)$weight * 1, # Edge width proportional to weight
     edge.color = "gray",
     mark.groups = NULL,
     main = "Louvain Community Detection (B Z Group - Longer Edges)")

dev.off()


# Community Detection and Layout for A Y Group
louvain_communities_ifa <- cluster_louvain(net_pn12m_ifa)
layout_ifa <- layout_with_fr(net_pn12m_ifa, niter = 1000)  # Increase iterations for better distribution
layout_ifa_scaled <- layout_ifa * 1000  # Multiply the coordinates by a factor to increase spacing between nodes

# Plot the Louvain communities with scaled layout
svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/coOccurrence_pn12m_ifa.svg",
  width=10, height=10
)

plot(louvain_communities_ifa, net_pn12m_ifa,
     layout = layout_ifa,               # Apply the more spread-out layout
     vertex.size = degree(net_pn12m_ifa) * 0.2,  # Node size proportional to degree (number of connections)
     vertex.label = NA,                  # No vertex labels
     edge.width = E(net_pn12m_ifa)$weight * 1, # Edge width proportional to weight
     edge.color = "gray",
     mark.groups = NULL,
     main = "Louvain Community Detection (A Y Group - Larger Layout)")
dev.off()

```


## pn56m 
```{r}
# subset data
pn56m_bep <- subset_samples(pn56m, code=="B Z")
pn56m_ifa <- subset_samples(pn56m, code=="A Y")

# filter data
filtered_pn56m_bep <- filter_taxa(pn56m_bep, function(x) sum(x > 0.01) > (0.05 * length(x)), TRUE)
filtered_pn56m_ifa <- filter_taxa(pn56m_ifa, function(x) sum(x > 0.01) > (0.05 * length(x)), TRUE)
otu_data_filtered_pn56m_bep <- otu_table(filtered_pn56m_bep)
otu_data_filtered_pn56m_ifa <- otu_table(filtered_pn56m_ifa)

# model: time-consuming
spiec_result_pn56m_bep <- spiec.easi(filtered_pn56m_bep, method = "mb", lambda.min.ratio = 1e-2, nlambda = 20)
spiec_result_pn56m_ifa <- spiec.easi(filtered_pn56m_ifa, method = "mb", lambda.min.ratio = 1e-2, nlambda = 20)

# plot and density/diameter
adj_matrix_pn56m_bep <- as.matrix(symBeta(getOptBeta(spiec_result_pn56m_bep), mode = "maxabs"))
adj_matrix_pn56m_bep[adj_matrix_pn56m_bep < 0] <- 0
net_pn56m_bep <- graph_from_adjacency_matrix(adj_matrix_pn56m_bep, mode = "undirected", weighted = TRUE, diag = FALSE)
degree_values <- degree(net_pn56m_bep)
betweenness_values <- betweenness(net_pn56m_bep)
closeness_values <- closeness(net_pn56m_bep)

adj_matrix_pn56m_ifa <- as.matrix(symBeta(getOptBeta(spiec_result_pn56m_ifa), mode = "maxabs"))
adj_matrix_pn56m_ifa[adj_matrix_pn56m_ifa < 0] <- 0
net_pn56m_ifa <- graph_from_adjacency_matrix(adj_matrix_pn56m_ifa, mode = "undirected", weighted = TRUE, diag = FALSE)
degree_values <- degree(net_pn56m_ifa)
betweenness_values <- betweenness(net_pn56m_ifa)
closeness_values <- closeness(net_pn56m_ifa)

#  Community Detection
louvain_communities <- cluster_louvain(net_pn56m_bep)
plot(louvain_communities, net_pn56m_bep, vertex.size = 5, vertex.label = NA, main = "Louvain Community Detection")

louvain_communities <- cluster_louvain(net_pn56m_ifa)
plot(louvain_communities, net_pn56m_ifa, vertex.size = 5, vertex.label = NA, main = "Louvain Community Detection")

```

### visualize community
```{r}
# Community Detection and Layout for B Z Group
louvain_communities_bep <- cluster_louvain(net_pn56m_bep)
layout_bep <- layout_with_fr(net_pn56m_bep, niter = 1000)  # Increase iterations for better distribution
layout_bep_scaled <- layout_bep * 1000  # Multiply the coordinates by a factor to increase spacing between nodes

# Plot the Louvain communities with scaled layout
svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/coOccurrence_pn56m_bep.svg",
  width=10, height=10
)

plot(louvain_communities_bep, net_pn56m_bep,
     layout = layout_bep_scaled,        # Apply the scaled layout for longer edges
     vertex.size = degree(net_pn56m_bep) * 0.2,  # Node size proportional to degree (number of connections)
     vertex.label = NA,                  # No vertex labels
     edge.width = E(net_pn56m_bep)$weight * 1, # Edge width proportional to weight
     edge.color = "gray",
     mark.groups = NULL,
     main = "Louvain Community Detection (B Z Group - Longer Edges)")

dev.off()


# Community Detection and Layout for A Y Group
louvain_communities_ifa <- cluster_louvain(net_pn56m_ifa)
layout_ifa <- layout_with_fr(net_pn56m_ifa, niter = 1000)  # Increase iterations for better distribution
layout_ifa_scaled <- layout_ifa * 1000  # Multiply the coordinates by a factor to increase spacing between nodes

# Plot the Louvain communities with scaled layout
svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/coOccurrence_pn56m_ifa.svg",
  width=10, height=10
)

plot(louvain_communities_ifa, net_pn56m_ifa,
     layout = layout_ifa,               # Apply the more spread-out layout
     vertex.size = degree(net_pn56m_ifa) * 0.2,  # Node size proportional to degree (number of connections)
     vertex.label = NA,                  # No vertex labels
     edge.width = E(net_pn56m_ifa)$weight * 1, # Edge width proportional to weight
     edge.color = "gray",
     mark.groups = NULL,
     main = "Louvain Community Detection (A Y Group - Larger Layout)")
dev.off()

```


## pn12e 
```{r}
# subset data
pn12e_bep <- subset_samples(pn12e, code=="B Z")
pn12e_ifa <- subset_samples(pn12e, code=="A Y")

# filter data
filtered_pn12e_bep <- filter_taxa(pn12e_bep, function(x) sum(x > 0.01) > (0.05 * length(x)), TRUE)
filtered_pn12e_ifa <- filter_taxa(pn12e_ifa, function(x) sum(x > 0.01) > (0.05 * length(x)), TRUE)
otu_data_filtered_pn12e_bep <- otu_table(filtered_pn12e_bep)
otu_data_filtered_pn12e_ifa <- otu_table(filtered_pn12e_ifa)

# model: time-consuming
spiec_result_pn12e_bep <- spiec.easi(filtered_pn12e_bep, method = "mb", lambda.min.ratio = 1e-2, nlambda = 20)
spiec_result_pn12e_ifa <- spiec.easi(filtered_pn12e_ifa, method = "mb", lambda.min.ratio = 1e-2, nlambda = 20)

# plot and density/diameter
adj_matrix_pn12e_bep <- as.matrix(symBeta(getOptBeta(spiec_result_pn12e_bep), mode = "maxabs"))
adj_matrix_pn12e_bep[adj_matrix_pn12e_bep < 0] <- 0
net_pn12e_bep <- graph_from_adjacency_matrix(adj_matrix_pn12e_bep, mode = "undirected", weighted = TRUE, diag = FALSE)
degree_values <- degree(net_pn12e_bep)
betweenness_values <- betweenness(net_pn12e_bep)
closeness_values <- closeness(net_pn12e_bep)

adj_matrix_pn12e_ifa <- as.matrix(symBeta(getOptBeta(spiec_result_pn12e_ifa), mode = "maxabs"))
adj_matrix_pn12e_ifa[adj_matrix_pn12e_ifa < 0] <- 0
net_pn12e_ifa <- graph_from_adjacency_matrix(adj_matrix_pn12e_ifa, mode = "undirected", weighted = TRUE, diag = FALSE)
degree_values <- degree(net_pn12e_ifa)
betweenness_values <- betweenness(net_pn12e_ifa)
closeness_values <- closeness(net_pn12e_ifa)

#  Community Detection
louvain_communities <- cluster_louvain(net_pn12e_bep)
plot(louvain_communities, net_pn12e_bep, vertex.size = 5, vertex.label = NA, main = "Louvain Community Detection")

louvain_communities <- cluster_louvain(net_pn12e_ifa)
plot(louvain_communities, net_pn12e_ifa, vertex.size = 5, vertex.label = NA, main = "Louvain Community Detection")

```

### visualize community
```{r}
# Community Detection and Layout for B Z Group
louvain_communities_bep <- cluster_louvain(net_pn12e_bep)
layout_bep <- layout_with_fr(net_pn12e_bep, niter = 1000)  # Increase iterations for better distribution
layout_bep_scaled <- layout_bep * 1000  # Multiply the coordinates by a factor to increase spacing between nodes

# Plot the Louvain communities with scaled layout
svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/coOccurrence_pn12e_bep.svg",
  width=5, height=5
)

plot(louvain_communities_bep, net_pn12e_bep,
     layout = layout_bep_scaled,        # Apply the scaled layout for longer edges
     vertex.size = degree(net_pn12e_bep) * 0.2,  # Node size proportional to degree (number of connections)
     vertex.label = NA,                  # No vertex labels
     edge.width = E(net_pn12e_bep)$weight * 2, # Edge width proportional to weight
     edge.color = "gray",
     main = "Louvain Community Detection (B Z Group - Longer Edges)")

dev.off()


# Community Detection and Layout for A Y Group
louvain_communities_ifa <- cluster_louvain(net_pn12e_ifa)
layout_ifa <- layout_with_fr(net_pn12e_ifa, niter = 1000)  # Increase iterations for better distribution
layout_ifa_scaled <- layout_ifa * 1000  # Multiply the coordinates by a factor to increase spacing between nodes

# Plot the Louvain communities with scaled layout
svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/coOccurrence_pn12e_ifa.svg",
  width=5, height=5
)

plot(louvain_communities_ifa, net_pn12e_ifa,
     layout = layout_ifa,               # Apply the more spread-out layout
     vertex.size = degree(net_pn12e_ifa) * 0.2,  # Node size proportional to degree (number of connections)
     vertex.label = NA,                  # No vertex labels
     edge.width = E(net_pn12e_ifa)$weight * 2, # Edge width proportional to weight
     edge.color = "gray",
     main = "Louvain Community Detection (A Y Group - Larger Layout)")
dev.off()

```

### permutation test
```{r}
library(igraph)
intervention_net <- net_pn12e_bep
control_net <- net_pn12e_ifa

# Function to calculate modularity for a given network
get_modularity <- function(net, algorithm = "louvain") {
  if (algorithm == "louvain") {
    communities <- cluster_louvain(net)
  } else if (algorithm == "walktrap") {
    communities <- cluster_walktrap(net)
  }
  return(modularity(communities))
}

# Get adjacency matrices from both networks (intervention and control)
adj_intervention <- as.matrix(as_adjacency_matrix(intervention_net))
adj_control <- as.matrix(as_adjacency_matrix(control_net))

# Combine the two adjacency matrices
combined_adj <- rbind(cbind(adj_intervention, matrix(0, nrow = nrow(adj_intervention), ncol = ncol(adj_control))),
                      cbind(matrix(0, nrow = nrow(adj_control), ncol = ncol(adj_intervention)), adj_control))

# Create a graph from the combined adjacency matrix
combined_net <- graph_from_adjacency_matrix(combined_adj, mode = "undirected", weighted = TRUE)

# Define group labels: intervention and control
group_labels <- c(rep("intervention", nrow(adj_intervention)), rep("control", nrow(adj_control)))

# Calculate the observed modularity difference
modularity_intervention <- get_modularity(intervention_net, algorithm = "louvain")
modularity_control <- get_modularity(control_net, algorithm = "louvain")
observed_diff <- modularity_intervention - modularity_control

# Permutation test: shuffle labels and recalculate the modularity difference
n_permutations <- 1000
permuted_diffs <- numeric(n_permutations)

for (i in 1:n_permutations) {
  # Shuffle group labels
  shuffled_labels <- sample(group_labels)
  
  # Split the combined network by shuffled labels
  shuffled_intervention_net <- induced_subgraph(combined_net, which(shuffled_labels == "intervention"))
  shuffled_control_net <- induced_subgraph(combined_net, which(shuffled_labels == "control"))
  
  # Recalculate modularity for the shuffled groups
  perm_modularity_intervention <- get_modularity(shuffled_intervention_net, algorithm = "louvain")
  perm_modularity_control <- get_modularity(shuffled_control_net, algorithm = "louvain")
  
  # Store the modularity difference
  permuted_diffs[i] <- perm_modularity_intervention - perm_modularity_control
}

# Calculate p-value by comparing the observed difference to the permuted differences
p_value <- mean(abs(permuted_diffs) >= abs(observed_diff))
cat("Permutation Test P-value:", p_value)

```

## pn56e 
```{r}
# subset data
pn56e_bep <- subset_samples(pn56e, code=="B Z")
pn56e_ifa <- subset_samples(pn56e, code=="A Y")

# filter data
filtered_pn56e_bep <- filter_taxa(pn56e_bep, function(x) sum(x > 0.01) > (0.05 * length(x)), TRUE)
filtered_pn56e_ifa <- filter_taxa(pn56e_ifa, function(x) sum(x > 0.01) > (0.05 * length(x)), TRUE)
otu_data_filtered_pn56e_bep <- otu_table(filtered_pn56e_bep)
otu_data_filtered_pn56e_ifa <- otu_table(filtered_pn56e_ifa)

# model: time-consuming
spiec_result_pn56e_bep <- spiec.easi(filtered_pn56e_bep, method = "mb", lambda.min.ratio = 1e-2, nlambda = 20)
spiec_result_pn56e_ifa <- spiec.easi(filtered_pn56e_ifa, method = "mb", lambda.min.ratio = 1e-2, nlambda = 20)

# plot and density/diameter
adj_matrix_pn56e_bep <- as.matrix(symBeta(getOptBeta(spiec_result_pn56e_bep), mode = "maxabs"))
adj_matrix_pn56e_bep[adj_matrix_pn56e_bep < 0] <- 0
net_pn56e_bep <- graph_from_adjacency_matrix(adj_matrix_pn56e_bep, mode = "undirected", weighted = TRUE, diag = FALSE)
degree_values <- degree(net_pn56e_bep)
betweenness_values <- betweenness(net_pn56e_bep)
closeness_values <- closeness(net_pn56e_bep)

adj_matrix_pn56e_ifa <- as.matrix(symBeta(getOptBeta(spiec_result_pn56e_ifa), mode = "maxabs"))
adj_matrix_pn56e_ifa[adj_matrix_pn56e_ifa < 0] <- 0
net_pn56e_ifa <- graph_from_adjacency_matrix(adj_matrix_pn56e_ifa, mode = "undirected", weighted = TRUE, diag = FALSE)
degree_values <- degree(net_pn56e_ifa)
betweenness_values <- betweenness(net_pn56e_ifa)
closeness_values <- closeness(net_pn56e_ifa)

#  Community Detection
louvain_communities <- cluster_louvain(net_pn56e_bep)
plot(louvain_communities, net_pn56e_bep, vertex.size = 5, vertex.label = NA, main = "Louvain Community Detection")

louvain_communities <- cluster_louvain(net_pn56e_ifa)
plot(louvain_communities, net_pn56e_ifa, vertex.size = 5, vertex.label = NA, main = "Louvain Community Detection")

```

### visualize community
```{r}
# Community Detection and Layout for B Z Group
louvain_communities_bep <- cluster_louvain(net_pn56e_bep)
layout_bep <- layout_with_fr(net_pn56e_bep, niter = 1000)  # Increase iterations for better distribution
layout_bep_scaled <- layout_bep * 1000  # Multiply the coordinates by a factor to increase spacing between nodes

# Plot the Louvain communities with scaled layout
svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/coOccurrence_pn56e_bep.svg",
  width=5, height=5
)

plot(louvain_communities_bep, net_pn56e_bep,
     layout = layout_bep_scaled,        # Apply the scaled layout for longer edges
     vertex.size = degree(net_pn56e_bep) * 0.2,  # Node size proportional to degree (number of connections)
     vertex.label = NA,                  # No vertex labels
     edge.width = E(net_pn56e_bep)$weight * 2, # Edge width proportional to weight
     edge.color = "gray",
     main = "Louvain Community Detection (B Z Group - Longer Edges)")

dev.off()


# Community Detection and Layout for A Y Group
louvain_communities_ifa <- cluster_louvain(net_pn56e_ifa)
layout_ifa <- layout_with_fr(net_pn56e_ifa, niter = 1000)  # Increase iterations for better distribution
layout_ifa_scaled <- layout_ifa * 1000  # Multiply the coordinates by a factor to increase spacing between nodes

# Plot the Louvain communities with scaled layout
svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/coOccurrence_pn56e_ifa.svg",
  width=5, height=5
)

plot(louvain_communities_ifa, net_pn56e_ifa,
     layout = layout_ifa,               # Apply the more spread-out layout
     vertex.size = degree(net_pn56e_ifa) * 0.2,  # Node size proportional to degree (number of connections)
     vertex.label = NA,                  # No vertex labels
     edge.width = E(net_pn56e_ifa)$weight * 2, # Edge width proportional to weight
     edge.color = "gray",
     main = "Louvain Community Detection (A Y Group - Larger Layout)")
dev.off()

```


## permutation analysis
### nodes, edges, coefficient, density, diameter, and community
```{r}
vcount(net_incl_bep)
ecount(net_incl_bep)
weighted_clustering_bep <- transitivity(net_incl_bep, type = "global")
cat("Weighted Clustering Coefficient (BEP):", weighted_clustering_bep, "\n")
cat("Network Density: ", edge_density(net_incl_bep), "\n")
cat("Network Diameter: ", diameter(net_incl_bep), "\n")
louvain_communities <- cluster_louvain(net_incl_bep)
cat("Louvain communities:", length(louvain_communities), "\n")
cat("Modularity (Louvain):", modularity(louvain_communities), "\n")
cat("Sizes of Louvain communities:", sizes(louvain_communities), "\n")
avg_path_length_bep <- average.path.length(net_incl_bep, directed = FALSE, unconnected = TRUE)
cat("Average Path Length (Intervention Group):", avg_path_length_bep, "\n")


vcount(net_incl_ifa)
ecount(net_incl_ifa)
weighted_clustering_ifa <- transitivity(net_incl_ifa, type = "global")
cat("Weighted Clustering Coefficient (IFA):", weighted_clustering_ifa, "\n")
cat("Network Density: ", edge_density(net_incl_ifa), "\n")
cat("Network Diameter: ", diameter(net_incl_ifa), "\n")
louvain_communities <- cluster_louvain(net_incl_ifa)
cat("Louvain communities:", length(louvain_communities), "\n")
cat("Modularity (Louvain):", modularity(louvain_communities), "\n")
cat("Sizes of Louvain communities:", sizes(louvain_communities), "\n")
avg_path_length_bep <- average.path.length(net_incl_ifa, directed = FALSE, unconnected = TRUE)
cat("Average Path Length (Intervention Group):", avg_path_length_ifa, "\n")

vcount(net_tri3_bep)
ecount(net_tri3_bep)
weighted_clustering_bep <- transitivity(net_tri3_bep, type = "global")
cat("Weighted Clustering Coefficient (BEP):", weighted_clustering_bep, "\n")
cat("Network Density: ", edge_density(net_tri3_bep), "\n")
cat("Network Diameter: ", diameter(net_tri3_bep), "\n")
louvain_communities <- cluster_louvain(net_tri3_bep)
cat("Louvain communities:", length(louvain_communities), "\n")
cat("Modularity (Louvain):", modularity(louvain_communities), "\n")
cat("Sizes of Louvain communities:", sizes(louvain_communities), "\n")
avg_path_length_bep <- average.path.length(net_tri3_bep, directed = FALSE, unconnected = TRUE)
cat("Average Path Length (Intervention Group):", avg_path_length_bep, "\n")

vcount(net_tri3_ifa)
ecount(net_tri3_ifa)
weighted_clustering_ifa <- transitivity(net_tri3_ifa, type = "global")
cat("Weighted Clustering Coefficient (IFA):", weighted_clustering_ifa, "\n")
cat("Network Density: ", edge_density(net_tri3_ifa), "\n")
cat("Network Diameter: ", diameter(net_tri3_ifa), "\n")
louvain_communities <- cluster_louvain(net_tri3_ifa)
cat("Louvain communities:", length(louvain_communities), "\n")
cat("Modularity (Louvain):", modularity(louvain_communities), "\n")
cat("Sizes of Louvain communities:", sizes(louvain_communities), "\n")
avg_path_length_bep <- average.path.length(net_tri3_ifa, directed = FALSE, unconnected = TRUE)
cat("Average Path Length (Intervention Group):", avg_path_length_ifa, "\n")

vcount(net_pn12m_bep)
ecount(net_pn12m_bep)
weighted_clustering_bep <- transitivity(net_pn12m_bep, type = "global")
cat("Weighted Clustering Coefficient (BEP):", weighted_clustering_bep, "\n")
cat("Network Density: ", edge_density(net_pn12m_bep), "\n")
cat("Network Diameter: ", diameter(net_pn12m_bep), "\n")
louvain_communities <- cluster_louvain(net_pn12m_bep)
cat("Louvain communities:", length(louvain_communities), "\n")
cat("Modularity (Louvain):", modularity(louvain_communities), "\n")
cat("Sizes of Louvain communities:", sizes(louvain_communities), "\n")
avg_path_length_bep <- average.path.length(net_pn12m_bep, directed = FALSE, unconnected = TRUE)
cat("Average Path Length (Intervention Group):", avg_path_length_bep, "\n")

vcount(net_pn12m_ifa)
ecount(net_pn12m_ifa)
weighted_clustering_ifa <- transitivity(net_pn12m_ifa, type = "global")
cat("Weighted Clustering Coefficient (IFA):", weighted_clustering_ifa, "\n")
cat("Network Density: ", edge_density(net_pn12m_ifa), "\n")
cat("Network Diameter: ", diameter(net_pn12m_ifa), "\n")
louvain_communities <- cluster_louvain(net_pn12m_ifa)
cat("Louvain communities:", length(louvain_communities), "\n")
cat("Modularity (Louvain):", modularity(louvain_communities), "\n")
cat("Sizes of Louvain communities:", sizes(louvain_communities), "\n")
avg_path_length_bep <- average.path.length(net_pn12m_ifa, directed = FALSE, unconnected = TRUE)
cat("Average Path Length (Intervention Group):", avg_path_length_ifa, "\n")

vcount(net_pn56m_bep)
ecount(net_pn56m_bep)
weighted_clustering_bep <- transitivity(net_pn56m_bep, type = "global")
cat("Weighted Clustering Coefficient (BEP):", weighted_clustering_bep, "\n")
cat("Network Density: ", edge_density(net_pn56m_bep), "\n")
cat("Network Diameter: ", diameter(net_pn56m_bep), "\n")
louvain_communities <- cluster_louvain(net_pn56m_bep)
cat("Louvain communities:", length(louvain_communities), "\n")
cat("Modularity (Louvain):", modularity(louvain_communities), "\n")
cat("Sizes of Louvain communities:", sizes(louvain_communities), "\n")
avg_path_length_bep <- average.path.length(net_pn56m_bep, directed = FALSE, unconnected = TRUE)
cat("Average Path Length (Intervention Group):", avg_path_length_bep, "\n")

vcount(net_pn56m_ifa)
ecount(net_pn56m_ifa)
weighted_clustering_ifa <- transitivity(net_pn56m_ifa, type = "global")
cat("Weighted Clustering Coefficient (IFA):", weighted_clustering_ifa, "\n")
cat("Network Density: ", edge_density(net_pn56m_ifa), "\n")
cat("Network Diameter: ", diameter(net_pn56m_ifa), "\n")
louvain_communities <- cluster_louvain(net_pn56m_ifa)
cat("Louvain communities:", length(louvain_communities), "\n")
cat("Modularity (Louvain):", modularity(louvain_communities), "\n")
cat("Sizes of Louvain communities:", sizes(louvain_communities), "\n")
avg_path_length_bep <- average.path.length(net_pn56m_ifa, directed = FALSE, unconnected = TRUE)
cat("Average Path Length (Intervention Group):", avg_path_length_ifa, "\n")

vcount(net_pn12e_bep)
ecount(net_pn12e_bep)
weighted_clustering_bep <- transitivity(net_pn12e_bep, type = "global")
cat("Weighted Clustering Coefficient (BEP):", weighted_clustering_bep, "\n")
cat("Network Density: ", edge_density(net_pn12e_bep), "\n")
cat("Network Diameter: ", diameter(net_pn12e_bep), "\n")
louvain_communities <- cluster_louvain(net_pn12e_bep)
cat("Louvain communities:", length(louvain_communities), "\n")
cat("Modularity (Louvain):", modularity(louvain_communities), "\n")
cat("Sizes of Louvain communities:", sizes(louvain_communities), "\n")
avg_path_length_bep <- average.path.length(net_pn12e_bep, directed = FALSE, unconnected = TRUE)
cat("Average Path Length (Intervention Group):", avg_path_length_bep, "\n")

vcount(net_pn12e_ifa)
ecount(net_pn12e_ifa)
weighted_clustering_ifa <- transitivity(net_pn12e_ifa, type = "global")
cat("Weighted Clustering Coefficient (IFA):", weighted_clustering_ifa, "\n")
cat("Network Density: ", edge_density(net_pn12e_ifa), "\n")
cat("Network Diameter: ", diameter(net_pn12e_ifa), "\n")
louvain_communities <- cluster_louvain(net_pn12e_ifa)
cat("Louvain communities:", length(louvain_communities), "\n")
cat("Modularity (Louvain):", modularity(louvain_communities), "\n")
cat("Sizes of Louvain communities:", sizes(louvain_communities), "\n")
avg_path_length_bep <- average.path.length(net_pn12e_ifa, directed = FALSE, unconnected = TRUE)
cat("Average Path Length (Intervention Group):", avg_path_length_ifa, "\n")

vcount(net_pn56e_bep)
ecount(net_pn56e_bep)
weighted_clustering_bep <- transitivity(net_pn56e_bep, type = "global")
cat("Weighted Clustering Coefficient (BEP):", weighted_clustering_bep, "\n")
cat("Network Density: ", edge_density(net_pn56e_bep), "\n")
cat("Network Diameter: ", diameter(net_pn56e_bep), "\n")
louvain_communities <- cluster_louvain(net_pn56e_bep)
cat("Louvain communities:", length(louvain_communities), "\n")
cat("Modularity (Louvain):", modularity(louvain_communities), "\n")
cat("Sizes of Louvain communities:", sizes(louvain_communities), "\n")
avg_path_length_bep <- average.path.length(net_pn56e_bep, directed = FALSE, unconnected = TRUE)
cat("Average Path Length (Intervention Group):", avg_path_length_bep, "\n")

vcount(net_pn56e_ifa)
ecount(net_pn56e_ifa)
weighted_clustering_ifa <- transitivity(net_pn56e_ifa, type = "global")
cat("Weighted Clustering Coefficient (IFA):", weighted_clustering_ifa, "\n")
cat("Network Density: ", edge_density(net_pn56e_ifa), "\n")
cat("Network Diameter: ", diameter(net_pn56e_ifa), "\n")
louvain_communities <- cluster_louvain(net_pn56e_ifa)
cat("Louvain communities:", length(louvain_communities), "\n")
cat("Modularity (Louvain):", modularity(louvain_communities), "\n")
cat("Sizes of Louvain communities:", sizes(louvain_communities), "\n")
avg_path_length_bep <- average.path.length(net_pn56e_ifa, directed = FALSE, unconnected = TRUE)
cat("Average Path Length (Intervention Group):", avg_path_length_ifa, "\n")

```


### density comparison
```{r}
# Function to calculate network density
calculate_density <- function(network) {
  return(edge_density(network))
}

# Function to perform permutation test between two networks
permutation_test_density <- function(net_bep, net_ifa, num_permutations = 1000) {
  # Calculate the observed density difference
  density_bep <- calculate_density(net_bep)
  density_ifa <- calculate_density(net_ifa)
  observed_difference <- density_bep - density_ifa
  
  # Initialize vector to store permuted differences
  permuted_differences <- numeric(num_permutations)
  
  # Combine networks' edges
  all_edges <- cbind(get.edgelist(net_bep), rep("BEP", ecount(net_bep)))  # Label for BEP edges
  all_edges <- rbind(all_edges, cbind(get.edgelist(net_ifa), rep("IFA", ecount(net_ifa))))  # Label for IFA edges
  
  # Perform permutations
  for (i in 1:num_permutations) {
    # Randomly shuffle the labels (BEP and IFA) of the edges
    permuted_labels <- sample(all_edges[, 3])  # Shuffle the labels
    
    # Recreate permuted BEP and IFA networks
    permuted_bep_edges <- all_edges[permuted_labels == "BEP", 1:2]
    permuted_ifa_edges <- all_edges[permuted_labels == "IFA", 1:2]
    
    # Create graphs for permuted groups
    permuted_net_bep <- graph_from_edgelist(as.matrix(permuted_bep_edges), directed = FALSE)
    permuted_net_ifa <- graph_from_edgelist(as.matrix(permuted_ifa_edges), directed = FALSE)
    
    # Calculate the permuted density difference
    permuted_density_bep <- calculate_density(permuted_net_bep)
    permuted_density_ifa <- calculate_density(permuted_net_ifa)
    permuted_differences[i] <- permuted_density_bep - permuted_density_ifa
  }
  
  # Calculate p-value
  p_value <- mean(abs(permuted_differences) >= abs(observed_difference))
  
  # Return results
  return(list(
    observed_difference = observed_difference,
    permuted_differences = permuted_differences,
    p_value = p_value
  ))
}


# Run the permutation test for network density comparison between BEP and IFA
set.seed(123)
result <- permutation_test_density(net_incl_bep, net_incl_ifa)
cat("Observed Density Difference (BEP - IFA):", result$observed_difference, "\n")
cat("Permutation Test P-value:", result$p_value, "\n")

result <- permutation_test_density(net_tri3_bep, net_tri3_ifa)
cat("Observed Density Difference (BEP - IFA):", result$observed_difference, "\n")
cat("Permutation Test P-value:", result$p_value, "\n")

result <- permutation_test_density(net_pn12m_bep, net_pn12m_ifa)
cat("Observed Density Difference (BEP - IFA):", result$observed_difference, "\n")
cat("Permutation Test P-value:", result$p_value, "\n")

result <- permutation_test_density(net_pn56m_bep, net_pn56m_ifa)
cat("Observed Density Difference (BEP - IFA):", result$observed_difference, "\n")
cat("Permutation Test P-value:", result$p_value, "\n")

result <- permutation_test_density(net_pn12e_bep, net_pn12e_ifa)
cat("Observed Density Difference (BEP - IFA):", result$observed_difference, "\n")
cat("Permutation Test P-value:", result$p_value, "\n")

result <- permutation_test_density(net_pn56e_bep, net_pn56e_ifa)
cat("Observed Density Difference (BEP - IFA):", result$observed_difference, "\n")
cat("Permutation Test P-value:", result$p_value, "\n")


```

### diameter comparison
```{r}
# Function to calculate network diameter
calculate_diameter <- function(network) {
  return(diameter(network))
}

# Function to perform permutation test between two networks for diameter
permutation_test_diameter <- function(net_bep, net_ifa, num_permutations = 1000) {
  # Calculate the observed diameter difference
  diameter_bep <- calculate_diameter(net_bep)
  diameter_ifa <- calculate_diameter(net_ifa)
  observed_difference <- diameter_bep - diameter_ifa
  
  # Initialize vector to store permuted differences
  permuted_differences <- numeric(num_permutations)
  
  # Combine networks' edges
  all_edges <- cbind(get.edgelist(net_bep), rep("BEP", ecount(net_bep)))  # Label for BEP edges
  all_edges <- rbind(all_edges, cbind(get.edgelist(net_ifa), rep("IFA", ecount(net_ifa))))  # Label for IFA edges
  
  # Perform permutations
  for (i in 1:num_permutations) {
    # Randomly shuffle the labels (BEP and IFA) of the edges
    permuted_labels <- sample(all_edges[, 3])  # Shuffle the labels
    
    # Recreate permuted BEP and IFA networks
    permuted_bep_edges <- all_edges[permuted_labels == "BEP", 1:2]
    permuted_ifa_edges <- all_edges[permuted_labels == "IFA", 1:2]
    
    # Create graphs for permuted groups
    permuted_net_bep <- graph_from_edgelist(as.matrix(permuted_bep_edges), directed = FALSE)
    permuted_net_ifa <- graph_from_edgelist(as.matrix(permuted_ifa_edges), directed = FALSE)
    
    # Calculate the permuted diameter difference
    permuted_diameter_bep <- calculate_diameter(permuted_net_bep)
    permuted_diameter_ifa <- calculate_diameter(permuted_net_ifa)
    permuted_differences[i] <- permuted_diameter_bep - permuted_diameter_ifa
  }
  
  # Calculate p-value
  p_value <- mean(abs(permuted_differences) >= abs(observed_difference))
  
  # Return results
  return(list(
    observed_difference = observed_difference,
    permuted_differences = permuted_differences,
    p_value = p_value
  ))
}

# Run the permutation test for network diameter comparison between BEP and IFA
set.seed(123)
result_diameter <- permutation_test_diameter(net_incl_bep, net_incl_ifa)
cat("Observed Diameter Difference (BEP - IFA):", result_diameter$observed_difference, "\n")
cat("Permutation Test P-value:", result_diameter$p_value, "\n")

result_diameter <- permutation_test_diameter(net_tri3_bep, net_tri3_ifa)
cat("Observed Diameter Difference (BEP - IFA):", result_diameter$observed_difference, "\n")
cat("Permutation Test P-value:", result_diameter$p_value, "\n")

result_diameter <- permutation_test_diameter(net_pn12m_bep, net_pn12m_ifa)
cat("Observed Diameter Difference (BEP - IFA):", result_diameter$observed_difference, "\n")
cat("Permutation Test P-value:", result_diameter$p_value, "\n")

result_diameter <- permutation_test_diameter(net_pn56m_bep, net_pn56m_ifa)
cat("Observed Diameter Difference (BEP - IFA):", result_diameter$observed_difference, "\n")
cat("Permutation Test P-value:", result_diameter$p_value, "\n")

result_diameter <- permutation_test_diameter(net_pn12e_bep, net_pn12e_ifa)
cat("Observed Diameter Difference (BEP - IFA):", result_diameter$observed_difference, "\n")
cat("Permutation Test P-value:", result_diameter$p_value, "\n")

result_diameter <- permutation_test_diameter(net_pn56e_bep, net_pn56e_ifa)
cat("Observed Diameter Difference (BEP - IFA):", result_diameter$observed_difference, "\n")
cat("Permutation Test P-value:", result_diameter$p_value, "\n")


```

### modularity comparison
```{r}
# Function to calculate modularity
calculate_modularity <- function(network) {
  # Perform Louvain community detection
  louvain_communities <- cluster_louvain(network)
  # Calculate modularity of the community structure
  return(modularity(louvain_communities))
}

# Function to perform permutation test for modularity between two networks
permutation_test_modularity <- function(net_bep, net_ifa, num_permutations = 1000) {
  # Calculate the observed modularity difference
  modularity_bep <- calculate_modularity(net_bep)
  modularity_ifa <- calculate_modularity(net_ifa)
  observed_difference <- modularity_bep - modularity_ifa
  
  # Initialize vector to store permuted differences
  permuted_differences <- numeric(num_permutations)
  
  # Combine networks' edges
  all_edges <- cbind(get.edgelist(net_bep), rep("BEP", ecount(net_bep)))  # Label for BEP edges
  all_edges <- rbind(all_edges, cbind(get.edgelist(net_ifa), rep("IFA", ecount(net_ifa))))  # Label for IFA edges
  
  # Perform permutations
  for (i in 1:num_permutations) {
    # Randomly shuffle the labels (BEP and IFA) of the edges
    permuted_labels <- sample(all_edges[, 3])  # Shuffle the labels
    
    # Recreate permuted BEP and IFA networks
    permuted_bep_edges <- all_edges[permuted_labels == "BEP", 1:2]
    permuted_ifa_edges <- all_edges[permuted_labels == "IFA", 1:2]
    
    # Create graphs for permuted groups
    permuted_net_bep <- graph_from_edgelist(as.matrix(permuted_bep_edges), directed = FALSE)
    permuted_net_ifa <- graph_from_edgelist(as.matrix(permuted_ifa_edges), directed = FALSE)
    
    # Calculate the permuted modularity difference
    permuted_modularity_bep <- calculate_modularity(permuted_net_bep)
    permuted_modularity_ifa <- calculate_modularity(permuted_net_ifa)
    permuted_differences[i] <- permuted_modularity_bep - permuted_modularity_ifa
  }
  
  # Calculate p-value
  p_value <- mean(abs(permuted_differences) >= abs(observed_difference))
  
  # Return results
  return(list(
    observed_difference = observed_difference,
    permuted_differences = permuted_differences,
    p_value = p_value
  ))
}

# Run the permutation test for modularity comparison between BEP and IFA
set.seed(123)
result_modularity <- permutation_test_modularity(net_incl_bep, net_incl_ifa)
cat("Observed Modularity Difference (BEP - IFA):", result_modularity$observed_difference, "\n")
cat("Permutation Test P-value:", result_modularity$p_value, "\n")

result_modularity <- permutation_test_modularity(net_tri3_bep, net_tri3_ifa)
cat("Observed Modularity Difference (BEP - IFA):", result_modularity$observed_difference, "\n")
cat("Permutation Test P-value:", result_modularity$p_value, "\n")

result_modularity <- permutation_test_modularity(net_pn12m_bep, net_pn12m_ifa)
cat("Observed Modularity Difference (BEP - IFA):", result_modularity$observed_difference, "\n")
cat("Permutation Test P-value:", result_modularity$p_value, "\n")

result_modularity <- permutation_test_modularity(net_pn56m_bep, net_pn56m_ifa)
cat("Observed Modularity Difference (BEP - IFA):", result_modularity$observed_difference, "\n")
cat("Permutation Test P-value:", result_modularity$p_value, "\n")

result_modularity <- permutation_test_modularity(net_pn12e_bep, net_pn12e_ifa)
cat("Observed Modularity Difference (BEP - IFA):", result_modularity$observed_difference, "\n")
cat("Permutation Test P-value:", result_modularity$p_value, "\n")

result_modularity <- permutation_test_modularity(net_pn56e_bep, net_pn56e_ifa)
cat("Observed Modularity Difference (BEP - IFA):", result_modularity$observed_difference, "\n")
cat("Permutation Test P-value:", result_modularity$p_value, "\n")

```

### clustering coefficient
```{r}
# Function to calculate global clustering coefficient (transitivity)
calculate_clustering <- function(network) {
  return(transitivity(network, type = "global"))
}

# Function to perform permutation test for clustering coefficient between two networks
permutation_test_clustering <- function(net_bep, net_ifa, num_permutations = 1000) {
  # Calculate the observed clustering coefficient difference
  clustering_bep <- calculate_clustering(net_bep)
  clustering_ifa <- calculate_clustering(net_ifa)
  observed_difference <- clustering_bep - clustering_ifa
  
  # Initialize vector to store permuted differences
  permuted_differences <- numeric(num_permutations)
  
  # Combine networks' edges
  all_edges <- cbind(get.edgelist(net_bep), rep("BEP", ecount(net_bep)))  # Label for BEP edges
  all_edges <- rbind(all_edges, cbind(get.edgelist(net_ifa), rep("IFA", ecount(net_ifa))))  # Label for IFA edges
  
  # Perform permutations
  for (i in 1:num_permutations) {
    # Randomly shuffle the labels (BEP and IFA) of the edges
    permuted_labels <- sample(all_edges[, 3])  # Shuffle the labels
    
    # Recreate permuted BEP and IFA networks
    permuted_bep_edges <- all_edges[permuted_labels == "BEP", 1:2]
    permuted_ifa_edges <- all_edges[permuted_labels == "IFA", 1:2]
    
    # Create graphs for permuted groups
    permuted_net_bep <- graph_from_edgelist(as.matrix(permuted_bep_edges), directed = FALSE)
    permuted_net_ifa <- graph_from_edgelist(as.matrix(permuted_ifa_edges), directed = FALSE)
    
    # Calculate the permuted clustering coefficient difference
    permuted_clustering_bep <- calculate_clustering(permuted_net_bep)
    permuted_clustering_ifa <- calculate_clustering(permuted_net_ifa)
    permuted_differences[i] <- permuted_clustering_bep - permuted_clustering_ifa
  }
  
  # Calculate p-value
  p_value <- mean(abs(permuted_differences) >= abs(observed_difference))
  
  # Return results
  return(list(
    observed_difference = observed_difference,
    permuted_differences = permuted_differences,
    p_value = p_value
  ))
}

# Run the permutation test for clustering coefficient comparison between BEP and IFA
set.seed(123)
result_clustering <- permutation_test_clustering(net_incl_bep, net_incl_ifa)
cat("Observed Clustering Coefficient Difference (BEP - IFA):", result_clustering$observed_difference, "\n")
cat("Permutation Test P-value:", result_clustering$p_value, "\n")

result_clustering <- permutation_test_clustering(net_tri3_bep, net_tri3_ifa)
cat("Observed Clustering Coefficient Difference (BEP - IFA):", result_clustering$observed_difference, "\n")
cat("Permutation Test P-value:", result_clustering$p_value, "\n")

result_clustering <- permutation_test_clustering(net_pn12m_bep, net_pn12m_ifa)
cat("Observed Clustering Coefficient Difference (BEP - IFA):", result_clustering$observed_difference, "\n")
cat("Permutation Test P-value:", result_clustering$p_value, "\n")

result_clustering <- permutation_test_clustering(net_pn56m_bep, net_pn56m_ifa)
cat("Observed Clustering Coefficient Difference (BEP - IFA):", result_clustering$observed_difference, "\n")
cat("Permutation Test P-value:", result_clustering$p_value, "\n")

result_clustering <- permutation_test_clustering(net_pn12e_bep, net_pn12e_ifa)
cat("Observed Clustering Coefficient Difference (BEP - IFA):", result_clustering$observed_difference, "\n")
cat("Permutation Test P-value:", result_clustering$p_value, "\n")

result_clustering <- permutation_test_clustering(net_pn56e_bep, net_pn56e_ifa)
cat("Observed Clustering Coefficient Difference (BEP - IFA):", result_clustering$observed_difference, "\n")
cat("Permutation Test P-value:", result_clustering$p_value, "\n")

```

### average path length
```{r}
library(igraph)

# Function to calculate the average path length for a network
calculate_avg_path_length <- function(network) {
  return(average.path.length(network, directed = FALSE, unconnected = TRUE))
}

# Function to perform permutation test for average path length
permutation_test_path_length <- function(net_bep, net_ifa, num_permutations = 1000) {
  # Calculate the observed average path length difference
  avg_path_length_bep <- calculate_avg_path_length(net_bep)
  avg_path_length_ifa <- calculate_avg_path_length(net_ifa)
  observed_difference <- avg_path_length_bep - avg_path_length_ifa
  
  # Initialize a vector to store permuted differences
  permuted_differences <- numeric(num_permutations)
  
  # Combine networks' edges
  all_edges <- cbind(get.edgelist(net_bep), rep("BEP", ecount(net_bep)))  # Label for BEP edges
  all_edges <- rbind(all_edges, cbind(get.edgelist(net_ifa), rep("IFA", ecount(net_ifa))))  # Label for IFA edges
  
  # Perform permutations
  for (i in 1:num_permutations) {
    # Randomly shuffle the group labels of the edges
    permuted_labels <- sample(all_edges[, 3])  # Shuffle the labels
    
    # Recreate permuted BEP and IFA networks
    permuted_bep_edges <- all_edges[permuted_labels == "BEP", 1:2]
    permuted_ifa_edges <- all_edges[permuted_labels == "IFA", 1:2]
    
    # Create graphs for permuted groups
    permuted_net_bep <- graph_from_edgelist(as.matrix(permuted_bep_edges), directed = FALSE)
    permuted_net_ifa <- graph_from_edgelist(as.matrix(permuted_ifa_edges), directed = FALSE)
    
    # Calculate the permuted average path length difference
    permuted_avg_path_length_bep <- calculate_avg_path_length(permuted_net_bep)
    permuted_avg_path_length_ifa <- calculate_avg_path_length(permuted_net_ifa)
    permuted_differences[i] <- permuted_avg_path_length_bep - permuted_avg_path_length_ifa
  }
  
  # Calculate p-value
  p_value <- mean(abs(permuted_differences) >= abs(observed_difference))
  
  # Return results
  return(list(
    observed_difference = observed_difference,
    permuted_differences = permuted_differences,
    p_value = p_value
  ))
}

# Run the permutation test for average path length
set.seed(123)
result_path_length <- permutation_test_path_length(net_incl_bep, net_incl_ifa, num_permutations = 1000)
cat("Observed Average Path Length Difference (BEP - IFA):", result_path_length$observed_difference, "\n")
cat("Permutation Test P-value:", result_path_length$p_value, "\n")

result_path_length <- permutation_test_path_length(net_tri3_bep, net_tri3_ifa, num_permutations = 1000)
cat("Observed Average Path Length Difference (BEP - IFA):", result_path_length$observed_difference, "\n")
cat("Permutation Test P-value:", result_path_length$p_value, "\n")

result_path_length <- permutation_test_path_length(net_pn12m_bep, net_pn12m_ifa, num_permutations = 1000)
cat("Observed Average Path Length Difference (BEP - IFA):", result_path_length$observed_difference, "\n")
cat("Permutation Test P-value:", result_path_length$p_value, "\n")

result_path_length <- permutation_test_path_length(net_pn56m_bep, net_pn56m_ifa, num_permutations = 1000)
cat("Observed Average Path Length Difference (BEP - IFA):", result_path_length$observed_difference, "\n")
cat("Permutation Test P-value:", result_path_length$p_value, "\n")

result_path_length <- permutation_test_path_length(net_pn12e_bep, net_pn12e_ifa, num_permutations = 1000)
cat("Observed Average Path Length Difference (BEP - IFA):", result_path_length$observed_difference, "\n")
cat("Permutation Test P-value:", result_path_length$p_value, "\n")

result_path_length <- permutation_test_path_length(net_pn56e_bep, net_pn56e_ifa, num_permutations = 1000)
cat("Observed Average Path Length Difference (BEP - IFA):", result_path_length$observed_difference, "\n")
cat("Permutation Test P-value:", result_path_length$p_value, "\n")

```



# Figure 5. ANCOMBC 
## Combined TP

### mother 4 TP

```{r}

mother_all_repeated_ancombc <- ancombc2(data=mother_ps, 
                        assay_name = "counts", tax_level = "species",
                        fix_formula = "code + timepoint", rand_formula = "(1 | idbs)",
                        p_adj_method = "BH", pseudo_sens = TRUE,
                        prv_cut = 0.1,
                        lib_cut = 1000, s0_perc = 0.05,
                        group = "timepoint", struc_zero = TRUE, neg_lb = TRUE,
                        alpha = 0.05, n_cl = 2, verbose = TRUE,
                        global = TRUE, pairwise = TRUE, dunnet = TRUE, trend = TRUE,
                        iter_control = list(tol=1e-2, max_iter=20, verbose=TRUE),
                        em_control = list(tol=1e-5, max_iter=100),
                        lme_control = lme4::lmerControl(),
                        mdfdr_control = list(fwer_ctrl_method = "BH", B=1000),
                        trend_control = list(contrast=list(matrix(c(1,-1,0,0,1,-1,-1,0,1), 
                                                                  nrow = 3, byrow = TRUE)),
                                             node=list(2),
                                             solver="ECOS",
                                             B=10)) 
mother_all_repeated_ancombc_output <- mother_all_repeated_ancombc$res

```


### mother 3 TP

```{r}

# mother dataset excluding incl
mother_ps@sam_data$timepoint <- factor(mother_ps@sam_data$timepoint, levels = c("incl", "tri3", "pn12", "pn56"))

mother_repeated <- subset_samples(mother_ps, timepoint !="incl") 

mother_repeated_ancombc <- ancombc2(data=mother_repeated, 
                        assay_name = "counts", tax_level = "species",
                        fix_formula = "code + timepoint", rand_formula = "(1 | idbs)",
                        p_adj_method = "BH", pseudo_sens = TRUE,
                        prv_cut = 0.1,
                        lib_cut = 1000, s0_perc = 0.05,
                        group = "timepoint", struc_zero = TRUE, neg_lb = TRUE,
                        alpha = 0.05, n_cl = 2, verbose = TRUE,
                        global = TRUE, pairwise = TRUE, dunnet = TRUE, trend = TRUE,
                        iter_control = list(tol=1e-2, max_iter=20, verbose=TRUE),
                        em_control = list(tol=1e-5, max_iter=100),
                        lme_control = lme4::lmerControl(),
                        mdfdr_control = list(fwer_ctrl_method = "BH", B=1000),
                        trend_control = list(contrast=list(matrix(c(1,0,-1,1),
                                                                  nrow = 2,
                                                                  byrow = TRUE)),
                                             node=list(2),
                                             solver="ECOS",
                                             B=10)) 
# ancombc output
mother_repeated_ancombc_output <- mother_repeated_ancombc$res %>%
  mutate_if(is.numeric, function(x) round(x, 2))
mother_repeated_ancombc_output$taxon[mother_repeated_ancombc_output$`diff_codeB Z`==TRUE] #0
mother_repeated_ancombc_output$taxon[mother_repeated_ancombc_output$diff_timepointpn56==TRUE & mother_repeated_ancombc_output$passed_ss_timepointpn56==TRUE] 
mother_repeated_ancombc_output$taxon[mother_repeated_ancombc_output$diff_timepointpn12==TRUE & mother_repeated_ancombc_output$passed_ss_timepointpn12==TRUE]

# further look into each
mother_repeated_ancombc_res_global <- mother_repeated_ancombc$res_global %>%
  mutate_if(is.numeric, function(x) round(x, 2))

mother_repeated_ancombc_res_pair <- mother_repeated_ancombc$res_pair 
#write.csv(mother_repeated_ancombc_res_pair, "../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/mother_repeated_ancom.csv")

mother_repeated_ancombc_res_pair <- merge(mother_repeated_ancombc_res_pair, taxonomy, by.x = "taxon", by.y = "species", all.x = TRUE)
library(openxlsx)
write.xlsx(mother_repeated_ancombc_res_pair, "../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/mother_repeated_ancom.xlsx")

mother_repeated_ancombc_res_trend <- mother_repeated_ancombc$res_trend %>%
  mutate_if(is.numeric, function(x) round(x, 2))



mother_repeatedTri3PN12 <- mother_repeated_ancombc_res_pair |>
  mutate(`q_timepointpn12`=-log10(`q_timepointpn12`)) |>
  mutate(color=if_else(`lfc_timepointpn12`>0, "steelblue", "darkred")) |>
  mutate(label=if_else(abs(`lfc_timepointpn12`)>0.5 & `q_timepointpn12` >1.3 & `passed_ss_timepointpn12` == TRUE, taxon, NA_character_)) |>
  mutate(color=if_else(abs(`lfc_timepointpn12`)<0.5 & `q_timepointpn12` <1.3, "#000000", color)) |>
  mutate(color=if_else(`passed_ss_timepointpn12` == FALSE, "grey55", color)) |>
  ggplot(mapping = aes(`lfc_timepointpn12`, `q_timepointpn12`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 0.5) +
  geom_vline(linetype = "dotted", xintercept = -0.5) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "Mother - PN12 vs. Tri3",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~P)) +
  theme_bw() + xlim(-2, 2) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
mother_repeatedTri3PN12

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/mother_repeatedTri3PN12.svg",
  width=4, height=4
  )
mother_repeatedTri3PN12
dev.off()

mother_repeatedTri3PN56 <- mother_repeated_ancombc_res_pair |>
  mutate(`q_timepointpn56`=-log10(`q_timepointpn56`)) |>
  mutate(color=if_else(`lfc_timepointpn56`>0, "steelblue", "darkred")) |>
  mutate(label=if_else(abs(`lfc_timepointpn56`)>0.5 & `q_timepointpn56` >1.3 & `passed_ss_timepointpn56` == TRUE, taxon, NA_character_)) |>
  mutate(color=if_else(abs(`lfc_timepointpn56`)<0.5 & `q_timepointpn56` <1.3, "#000000", color)) |>
  mutate(color=if_else(`passed_ss_timepointpn56` == FALSE, "grey55", color)) |>
  ggplot(mapping = aes(`lfc_timepointpn56`, `q_timepointpn56`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 0.5) +
  geom_vline(linetype = "dotted", xintercept = -0.5) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "Mother - PN56 vs. Tri3",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~P)) +
  theme_bw() + xlim(-2, 2) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
mother_repeatedTri3PN56

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/mother_repeatedTri3PN56.svg",
  width=4, height=4
  )
mother_repeatedTri3PN56
dev.off()

mother_repeatedPN12PN56 <- mother_repeated_ancombc_res_pair |>
  mutate(`q_timepointpn56_timepointpn12`=-log10(`q_timepointpn56_timepointpn12`)) |>
  mutate(color=if_else(`lfc_timepointpn56_timepointpn12`>0, "steelblue", "darkred")) |>
  mutate(label=if_else(abs(`lfc_timepointpn56_timepointpn12`)>0.5 & `q_timepointpn56_timepointpn12` >1.3 & `passed_ss_timepointpn56_timepointpn12` == TRUE, taxon, NA_character_)) |>
  mutate(color=if_else(abs(`lfc_timepointpn56_timepointpn12`)<0.5 & `q_timepointpn56_timepointpn12` <1.3, "#000000", color)) |>
  mutate(color=if_else(`passed_ss_timepointpn56_timepointpn12` == FALSE, "grey55", color)) |>
  ggplot(mapping = aes(`lfc_timepointpn56_timepointpn12`, `q_timepointpn56_timepointpn12`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 0.5) +
  geom_vline(linetype = "dotted", xintercept = -0.5) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "Mother - PN56 vs. PN12",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~P)) +
  theme_bw() + xlim(-2, 2) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
mother_repeatedPN12PN56

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/mother_repeatedPN12PN56.svg",
  width=4, height=4
  )
mother_repeatedPN12PN56
dev.off()

mother_repeatedBEP <- mother_repeated_ancombc_output |>
  mutate(`q_codeB Z`=-log10(`q_codeB Z`)) |>
  mutate(color=if_else(`lfc_codeB Z`>0, "steelblue", "darkred")) |>
  mutate(label=if_else(abs(`lfc_codeB Z`)>0.5 & `q_codeB Z` >1.3 & `passed_ss_codeB Z` == TRUE, taxon, NA_character_)) |>
  mutate(color=if_else(abs(`lfc_codeB Z`)<0.5 & `q_codeB Z` <1.3, "#000000", color)) |>
  mutate(color=if_else(`passed_ss_codeB Z` == FALSE, "grey55", color)) |>
  ggplot(mapping = aes(`lfc_codeB Z`, `q_codeB Z`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 0.5) +
  geom_vline(linetype = "dotted", xintercept = -0.5) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "Mother - Interv. vs. Contr.",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~P)) +
  theme_bw() + xlim(-2, 2) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
mother_repeatedBEP

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/mother_repeatedBEP.svg",
  width=4, height=4
  )
mother_repeatedBEP
dev.off()


```


### infant


```{r}

# infant
infant_repeated_ancombc <- ancombc2(data=infant_ps, 
                        assay_name = "counts", tax_level = "species",
                        fix_formula = "code + timepoint", rand_formula = "(1 | idbs)",
                        p_adj_method = "BH", pseudo_sens = TRUE,
                        prv_cut = 0.1,
                        lib_cut = 1000, s0_perc = 0.05,
                        group = "timepoint", struc_zero = TRUE, neg_lb = TRUE,
                        alpha = 0.05, n_cl = 2, verbose = TRUE,
                        global = TRUE, pairwise = TRUE, dunnet = TRUE, trend = TRUE,
                        iter_control = list(tol=1e-2, max_iter=20, verbose=TRUE),
                        em_control = list(tol=1e-5, max_iter=100),
                        lme_control = lme4::lmerControl(),
                        mdfdr_control = list(fwer_ctrl_method = "BH", B=1000),
                        trend_control = list(contrast=list(matrix(c(1,0,-1,1),
                                                                  nrow = 2,
                                                                  byrow = TRUE)),
                                             node=list(2),
                                             solver="ECOS",
                                             B=10)) 
# ancombc output
infant_repeated_ancombc_output <- infant_repeated_ancombc$res
infant_repeated_ancombc_output$taxon[infant_repeated_ancombc_output$`diff_codeB Z`==TRUE] #0
infant_repeated_ancombc_output$taxon[infant_repeated_ancombc_output$diff_timepointpn56==TRUE] #"Escherichia coli"
infant_repeated_ancombc_output$taxon[infant_repeated_ancombc_output$diff_timepointpn56==TRUE & infant_repeated_ancombc_output$passed_ss_timepointpn56==TRUE] #"Escherichia coli"

infant_repeated_ancombc_output <- merge(infant_repeated_ancombc_output, taxonomy, by.x = "taxon", by.y = "species", all.x = TRUE)

library(openxlsx)
write.xlsx(infant_repeated_ancombc_output, "../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/infant_repeated_ancombc_output.xlsx")


infant_repeatedPN12PN56 <- infant_repeated_ancombc_output |>
  mutate(`q_timepointpn56`=-log10(`q_timepointpn56`)) |>
  mutate(color=if_else(`lfc_timepointpn56`>0, "steelblue", "darkred")) |>
  mutate(label=if_else(abs(`lfc_timepointpn56`)>0.5 & `q_timepointpn56` >1.3 & `passed_ss_timepointpn56` == TRUE, taxon, NA_character_)) |>
  mutate(color=if_else(abs(`lfc_timepointpn56`)<0.5 & `q_timepointpn56` <1.3, "#000000", color)) |>
  mutate(color=if_else(`passed_ss_timepointpn56` == FALSE, "grey55", color)) |>
  ggplot(mapping = aes(`lfc_timepointpn56`, `q_timepointpn56`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 0.5) +
  geom_vline(linetype = "dotted", xintercept = -0.5) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "Infant - PN56 vs. PN12",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~P)) +
  theme_bw() + xlim(-2, 2) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
infant_repeatedPN12PN56

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/infant_repeatedPN12PN56.svg",
  width=4, height=4
  )
infant_repeatedPN12PN56
dev.off()


infant_repeatedBEP <- infant_repeated_ancombc_output |>
  mutate(`q_codeB Z`=-log10(`q_codeB Z`)) |>
  mutate(color=if_else(`lfc_codeB Z`>0, "steelblue", "darkred")) |>
  mutate(label=if_else(abs(`lfc_codeB Z`)>0.5 & `q_codeB Z` >1.3 & `passed_ss_codeB Z` == TRUE, taxon, NA_character_)) |>
  mutate(color=if_else(abs(`lfc_codeB Z`)<0.5 & `q_codeB Z` <1.3, "#000000", color)) |>
  mutate(color=if_else(`passed_ss_codeB Z` == FALSE, "grey55", color)) |>
  ggplot(mapping = aes(`lfc_codeB Z`, `q_codeB Z`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 0.5) +
  geom_vline(linetype = "dotted", xintercept = -0.5) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "Infant - Interv. vs. Contr.",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~P)) +
  theme_bw() + xlim(-2, 2) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
mother_repeatedBEP

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/infant_repeatedBEP.svg",
  width=4, height=4
  )
infant_repeatedBEP
dev.off()

```


##  Separated TP

```{r}

mother_incl <- incl_primres |>
  mutate(`q_codeB Z`=-log10(`q_codeB Z`)) |>
  mutate(color=if_else(`lfc_codeB Z`>0, "steelblue", "darkred")) |>
  mutate(label=if_else(abs(`lfc_codeB Z`)>0.5 & `q_codeB Z` >1.3 & `passed_ss_codeB Z` == TRUE, taxon, NA_character_)) |>
  mutate(color=if_else(abs(`lfc_codeB Z`)<0.5 & `q_codeB Z` <1.3, "#000000", color)) |>
  mutate(color=if_else(`passed_ss_codeB Z` == FALSE, "grey55", color)) |>
  ggplot(mapping = aes(`lfc_codeB Z`, `q_codeB Z`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 0.5) +
  geom_vline(linetype = "dotted", xintercept = -0.5) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "Mother - Tri2. - Interv. vs. Contr.",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~P)) +
  theme_bw() + xlim(-2, 2) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
mother_incl

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/mother_incl.svg",
  width=4, height=4
  )
mother_incl
dev.off()

mother_tri3 <- tri3_primres |>
  mutate(`q_codeB Z`=-log10(`q_codeB Z`)) |>
  mutate(color=if_else(`lfc_codeB Z`>0, "steelblue", "darkred")) |>
  mutate(label=if_else(abs(`lfc_codeB Z`)>0.5 & `q_codeB Z` >1.3 & `passed_ss_codeB Z` == TRUE, species, NA_character_)) |>
  mutate(color=if_else(abs(`lfc_codeB Z`)<0.5 & `q_codeB Z` <1.3, "#000000", color)) |>
  mutate(color=if_else(`passed_ss_codeB Z` == FALSE, "grey55", color)) |>
  ggplot(mapping = aes(`lfc_codeB Z`, `q_codeB Z`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 0.5) +
  geom_vline(linetype = "dotted", xintercept = -0.5) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "Mother - Tri3. - Interv. vs. Contr.",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~P)) +
  theme_bw() + xlim(-2, 2) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
mother_tri3

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/mother_tri3.svg",
  width=4, height=4
  )
mother_tri3
dev.off()

mother_pn12m <- pn12m_primres |>
  mutate(`q_codeB Z`=-log10(`q_codeB Z`)) |>
  mutate(color=if_else(`lfc_codeB Z`>0, "steelblue", "darkred")) |>
  mutate(label=if_else(abs(`lfc_codeB Z`)>0.5 & `q_codeB Z` >1.3 & `passed_ss_codeB Z` == TRUE, species, NA_character_)) |>
  mutate(color=if_else(abs(`lfc_codeB Z`)<0.5 & `q_codeB Z` <1.3, "#000000", color)) |>
  mutate(color=if_else(`passed_ss_codeB Z` == FALSE, "grey55", color)) |>
  ggplot(mapping = aes(`lfc_codeB Z`, `q_codeB Z`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 0.5) +
  geom_vline(linetype = "dotted", xintercept = -0.5) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "Mother - PN12. - Interv. vs. Contr.",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~P)) +
  theme_bw() + xlim(-2, 2) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
mother_pn12m

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/mother_pn12m.svg",
  width=4, height=4
  )
mother_pn12m
dev.off()

mother_pn56m <- pn56m_primres |>
  mutate(`q_codeB Z`=-log10(`q_codeB Z`)) |>
  mutate(color=if_else(`lfc_codeB Z`>0, "steelblue", "darkred")) |>
  mutate(label=if_else(abs(`lfc_codeB Z`)>0.5 & `q_codeB Z` >1.3 & `passed_ss_codeB Z` == TRUE, species, NA_character_)) |>
  mutate(color=if_else(abs(`lfc_codeB Z`)<0.5 & `q_codeB Z` <1.3, "#000000", color)) |>
  mutate(color=if_else(`passed_ss_codeB Z` == FALSE, "grey55", color)) |>
  ggplot(mapping = aes(`lfc_codeB Z`, `q_codeB Z`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 0.5) +
  geom_vline(linetype = "dotted", xintercept = -0.5) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "Mother - PN56. - Interv. vs. Contr.",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~P)) +
  theme_bw() + xlim(-2, 2) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
mother_pn56m

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/mother_pn56m.svg",
  width=4, height=4
  )
mother_pn56m
dev.off()


infant_pn12e <- pn12e_primres |>
  mutate(`q_codeB Z`=-log10(`q_codeB Z`)) |>
  mutate(color=if_else(`lfc_codeB Z`>0, "steelblue", "darkred")) |>
  mutate(label=if_else(abs(`lfc_codeB Z`)>0.5 & `q_codeB Z` >1.3 & `passed_ss_codeB Z` == TRUE, species, NA_character_)) |>
  mutate(color=if_else(abs(`lfc_codeB Z`)<0.5 & `q_codeB Z` <1.3, "#000000", color)) |>
  mutate(color=if_else(`passed_ss_codeB Z` == FALSE, "grey55", color)) |>
  ggplot(mapping = aes(`lfc_codeB Z`, `q_codeB Z`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 0.5) +
  geom_vline(linetype = "dotted", xintercept = -0.5) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "Infant - PN12. - Interv. vs. Contr.",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~P)) +
  theme_bw() + xlim(-2, 2) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
infant_pn12e

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/infant_pn12e.svg",
  width=4, height=4
  )
infant_pn12e
dev.off()

infant_pn56e <- pn56e_primres |>
  mutate(`q_codeB Z`=-log10(`q_codeB Z`)) |>
  mutate(color=if_else(`lfc_codeB Z`>0, "steelblue", "darkred")) |>
  mutate(label=if_else(abs(`lfc_codeB Z`)>0.5 & `q_codeB Z` >1.3 & `passed_ss_codeB Z` == TRUE, species, NA_character_)) |>
  mutate(color=if_else(abs(`lfc_codeB Z`)<0.5 & `q_codeB Z` <1.3, "#000000", color)) |>
  mutate(color=if_else(`passed_ss_codeB Z` == FALSE, "grey55", color)) |>
  ggplot(mapping = aes(`lfc_codeB Z`, `q_codeB Z`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 0.5) +
  geom_vline(linetype = "dotted", xintercept = -0.5) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "Mother - PN56. - Interv. vs. Contr.",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~P)) +
  theme_bw() + xlim(-2, 2) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
infant_pn56e

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/infant_pn56e.svg",
  width=4, height=4
  )
infant_pn56e
dev.off()


```


# Figure 8. HDmediation
## mediation on taxonomic data

### load packages
```{r}

source(paste(wd,"/HDmediation/b.R", sep=""))
source(paste(wd,"/HDmediation/c.R", sep=""))
source(paste(wd,"/HDmediation/crossfit.R", sep=""))
source(paste(wd,"/HDmediation/e.R", sep=""))
source(paste(wd,"/HDmediation/folds.R", sep=""))
source(paste(wd,"/HDmediation/g.R", sep=""))
source(paste(wd,"/HDmediation/h_m.R", sep=""))
source(paste(wd,"/HDmediation/h_z.R", sep=""))
source(paste(wd,"/HDmediation/not_transported.R", sep=""))
source(paste(wd,"/HDmediation/npsem.R", sep=""))
source(paste(wd,"/HDmediation/transported.R", sep=""))
source(paste(wd,"/HDmediation/u.R", sep=""))
source(paste(wd,"/HDmediation/utils.R", sep=""))
source(paste(wd,"/HDmediation/v.R", sep=""))
source(paste(wd,"/HDmediation/mediation.R", sep=""))


```

### load function
```{r}
set.seed(123)
sl_library <- c("SL.glm", "SL.mean")

extract_vars <- function(timepoint_ps, outcome_var, taxonomic_level="species", postnatal=FALSE, impute=FALSE){
  stopifnot(taxonomic_level %in% phyloseq::rank_names(timepoint_ps))
  # Extract the exposure
  A <- phyloseq::sample_data(timepoint_ps)$treatment1_prenatal_n
  # Extract the outcome
  Y <- phyloseq::sample_data(timepoint_ps)[[outcome_var]]
  # Extract the mediators
  tax_level_ps <- phyloseq::tax_glom(timepoint_ps, taxonomic_level)
  M <- t(phyloseq::otu_table(tax_level_ps))
  colnames(M) <- phyloseq::tax_table(tax_level_ps)[, taxonomic_level]
  # prevalence filtering (cutoff 20%)
  M <- as.matrix(M[, apply(M, 2, function(col) mean(col != 0)) >= 0.2])
  # extract the covariates
  covars_prenatal <- c(
    "m_heightincl", "m_hbincl", "w_marital", "asset1comp_10", "w_language", "w_religion",
    "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave"
    )
  covars_postnatal <- c(covars_prenatal, "iycf_ebf_age")
  W <- data.frame(phyloseq::sample_data(timepoint_ps)) %>%
    select(
      ifelse(postnatal, covars_postnatal, covars_prenatal)
      )
  if (impute){
    W <- W %>%
      mutate(across(everything(), impute_lowest))
  }
  # take complete cases
  complete_cases <- complete.cases(A, Y, M, W)
  if (sum(complete_cases) != length(A)){
    message("Removed ", length(A) - sum(complete_cases), " observations due to missing values.")
  }
  A <- A[complete_cases]
  Y <- Y[complete_cases]
  M <- M[complete_cases, ]
  W <- W[complete_cases, ]
  # Convert covariates to numeric
  W <- data.frame(lapply(W, function(x) as.numeric(as.character(x)))) %>%
    as.matrix()
  message(
    paste0(
      "Extracted exposure A with ", length(A), " observations.\n",
      "Extracted outcome Y with ", length(Y), " observations.\n",
      "Extracted mediators M with ", dim(M)[1], " observations and ", dim(M)[2], " features.\n",
      "Extracted covariates W with ", dim(W)[1], " observations and ", dim(W)[2], " features."
      )
    )
  vars_df <- data.frame(A, Y, M, W)
  return(
    list(
      df=vars_df,
      mediator_names=names(vars_df)[3:(ncol(vars_df)-ncol(W))],
      covariate_names=names(vars_df)[(ncol(vars_df)-ncol(W)+1):ncol(vars_df)]
    )
  )
}

impute_lowest <- function(x) {
  if (any(!is.na(x))) {
    min_val <- min(x, na.rm = TRUE)
    x[is.na(x)] <- min_val
  }
  return(x)
}
```

### Infant microbiome mediation
#### of the BEP->HAZ6 effect pn12
```{r mat pn12 haz, message=FALSE}
set.seed(123)
library(foreach)
library(doFuture)
vars = extract_vars(
  pn12e,
  "c_haz6",
  taxonomic_level = "species",
  postnatal = TRUE,
  impute = FALSE
  )
mediation_results_pn12_haz <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 3, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_pn12_haz <- rbind(mediation_results_pn12_haz, ans)
}
write.csv(mediation_results_pn12_haz, "./results/tax_mediation/species/mediation_results_pn12_haz.csv")

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_pn12_haz, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_pn12_haz6 <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_pn12_haz$ci_indirect_low, mediation_results_pn12_haz$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_pn12_haz$ci_indirect_high, mediation_results_pn12_haz$ci_direct_high)),
                width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "6m: HAZ", y = "PN12: Infant gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_pn12_haz6

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_pn12_haz6.svg",
  width=5, height=4
  )
plot_pn12_haz6
dev.off()

```

#### of the BEP->HAZ6 effect pn56
```{r mat pn56 haz, message=FALSE}
set.seed(123)
vars = extract_vars(
  pn56e,
  "c_haz6",
  taxonomic_level = "species",
  postnatal = TRUE,
  impute = FALSE
  )
mediation_results_pn56_haz <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_pn56_haz <- rbind(mediation_results_pn56_haz, ans)
}
write.csv(mediation_results_pn56_haz, "./results/tax_mediation/species/mediation_results_pn56_haz.csv")

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_pn56_haz, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_pn56_haz6 <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_pn56_haz$ci_indirect_low, mediation_results_pn56_haz$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_pn56_haz$ci_indirect_high, mediation_results_pn56_haz$ci_direct_high)),
                width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "6m: HAZ", y = "PN56: Infant gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_pn56_haz6

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_pn56_haz6.svg",
  width=5, height=5
  )
plot_pn56_haz6
dev.off()

```

#### of the BEP->stunting effect pn12
```{r mat pn12 stunting, message=FALSE}
set.seed(123)
library(foreach)
library(doFuture)
vars = extract_vars(
  pn12e,
  "c_stunting6",
  taxonomic_level = "species",
  postnatal = TRUE,
  impute = FALSE
  )
mediation_results_pn12_stunting <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "binomial",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_pn12_stunting <- rbind(mediation_results_pn12_stunting, ans)
}
write.csv(mediation_results_pn12_stunting, "./results/tax_mediation/species/mediation_results_pn12_stunting.csv")

library(reshape2)
# Melt the data frame for ggplot
data_melted <- melt(mediation_results_pn12_stunting, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_pn12_stunting6 <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_pn12_stunting$ci_indirect_low, mediation_results_pn12_stunting$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_pn12_stunting$ci_indirect_high, mediation_results_pn12_stunting$ci_direct_high)),
                width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "6m: Stunting", y = "PN12: Infant gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_pn12_stunting6

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_pn12_stunting6.svg",
  width=5, height=4
  )
plot_pn12_stunting6
dev.off()

```

#### of the BEP->stunting effect pn56
```{r mat pn56 stunting, message=FALSE}
set.seed(123)
library(foreach)
library(doFuture)
vars = extract_vars(
  pn56e,
  "c_stunting6",
  taxonomic_level = "species",
  postnatal = TRUE,
  impute = FALSE
  )
mediation_results_pn56_stunting <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "binomial",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_pn56_stunting <- rbind(mediation_results_pn56_stunting, ans)
}
write.csv(mediation_results_pn56_stunting, "./results/tax_mediation/species/mediation_results_pn56_stunting.csv")

library(reshape2)
# Melt the data frame for ggplot
data_melted <- melt(mediation_results_pn56_stunting, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_pn56_stunting6 <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_pn56_stunting$ci_indirect_low, mediation_results_pn56_stunting$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_pn56_stunting$ci_indirect_high, mediation_results_pn56_stunting$ci_direct_high)),
                width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "6m: Stunting", y = "PN12: Infant gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_pn56_stunting6

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_pn56_stunting6.svg",
  width=5, height=4
  )
plot_pn56_stunting6
dev.off()

```


#### of the BEP->WAZ6 effect pn12
```{r mat pn12 waz, message=FALSE}
set.seed(123)
library(foreach)
library(doFuture)
vars = extract_vars(
  pn12e,
  "c_waz6",
  taxonomic_level = "species",
  postnatal = TRUE,
  impute = FALSE
  )
mediation_results_pn12_waz <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 3, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_pn12_waz <- rbind(mediation_results_pn12_waz, ans)
}
write.csv(mediation_results_pn12_waz, "./results/tax_mediation/species/mediation_results_pn12_waz.csv")

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_pn12_waz, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_pn12_waz6 <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_pn12_waz$ci_indirect_low, mediation_results_pn12_waz$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_pn12_waz$ci_indirect_high, mediation_results_pn12_waz$ci_direct_high)),
                width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "6m: WAZ", y = "PN12: Infant gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_pn12_waz6

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_pn12_waz6.svg",
  width=5, height=4
  )
plot_pn12_waz6
dev.off()

```

#### of the BEP->WAZ6 effect pn56
```{r mat pn56 waz, message=FALSE}
set.seed(123)
vars = extract_vars(
  pn56e,
  "c_waz6",
  taxonomic_level = "species",
  postnatal = TRUE,
  impute = FALSE
  )
mediation_results_pn56_waz <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_pn56_waz <- rbind(mediation_results_pn56_waz, ans)
}
write.csv(mediation_results_pn56_waz, "./results/tax_mediation/species/mediation_results_pn56_waz.csv")

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_pn56_waz, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_pn56_waz6 <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_pn56_waz$ci_indirect_low, mediation_results_pn56_waz$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_pn56_waz$ci_indirect_high, mediation_results_pn56_waz$ci_direct_high)),
                width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "6m: WAZ", y = "PN56: Infant gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_pn56_waz6

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_pn56_waz6.svg",
  width=5, height=5
  )
plot_pn56_waz6
dev.off()

```


#### of the BEP->WHZ6 effect pn12
```{r mat pn12 whz, message=FALSE}
set.seed(123)
library(foreach)
library(doFuture)
vars = extract_vars(
  pn12e,
  "c_whz6",
  taxonomic_level = "species",
  postnatal = TRUE,
  impute = FALSE
  )
mediation_results_pn12_whz <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 3, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_pn12_whz <- rbind(mediation_results_pn12_whz, ans)
}
write.csv(mediation_results_pn12_whz, "./results/tax_mediation/species/mediation_results_pn12_whz.csv")

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_pn12_whz, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_pn12_whz6 <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_pn12_whz$ci_indirect_low, mediation_results_pn12_whz$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_pn12_whz$ci_indirect_high, mediation_results_pn12_whz$ci_direct_high)),
                width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "6m: WHZ", y = "PN12: Infant gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_pn12_whz6

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_pn12_whz6.svg",
  width=5, height=4
  )
plot_pn12_whz6
dev.off()

```

#### of the BEP->WHZ6 effect pn56
```{r mat pn56 whz, message=FALSE}
set.seed(123)
vars = extract_vars(
  pn56e,
  "c_whz6",
  taxonomic_level = "species",
  postnatal = TRUE,
  impute = FALSE
  )
mediation_results_pn56_whz <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_pn56_whz <- rbind(mediation_results_pn56_whz, ans)
}
write.csv(mediation_results_pn56_whz, "./results/tax_mediation/species/mediation_results_pn56_whz.csv")

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_pn56_whz, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_pn56_whz6 <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_pn56_whz$ci_indirect_low, mediation_results_pn56_whz$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_pn56_whz$ci_indirect_high, mediation_results_pn56_whz$ci_direct_high)),
                width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "6m: WHZ", y = "PN56: Infant gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_pn56_whz6

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_pn56_whz6.svg",
  width=5, height=5
  )
plot_pn56_whz6
dev.off()

```


#### of the BEP->HAZ3 effect pn12
```{r mat pn12 haz, message=FALSE}
set.seed(123)
library(foreach)
library(doFuture)
vars = extract_vars(
  pn12e,
  "c_haz3",
  taxonomic_level = "species",
  postnatal = TRUE,
  impute = FALSE
  )
mediation_results_pn12_haz <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 3, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_pn12_haz <- rbind(mediation_results_pn12_haz, ans)
}
write.csv(mediation_results_pn12_haz, "./results/tax_mediation/species/mediation_results_pn12_haz3.csv")

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_pn12_haz, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_pn12_haz3 <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_pn12_haz$ci_indirect_low, mediation_results_pn12_haz$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_pn12_haz$ci_indirect_high, mediation_results_pn12_haz$ci_direct_high)),
                width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "3m: HAZ", y = "PN12: Infant gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_pn12_haz3

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_pn12_haz3.svg",
  width=5, height=4
  )
plot_pn12_haz3
dev.off()

```


#### of the BEP->WAZ3 effect pn12
```{r mat pn12 waz, message=FALSE}
set.seed(123)
library(foreach)
library(doFuture)
vars = extract_vars(
  pn12e,
  "c_waz3",
  taxonomic_level = "species",
  postnatal = TRUE,
  impute = FALSE
  )
mediation_results_pn12_waz <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 3, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_pn12_waz <- rbind(mediation_results_pn12_waz, ans)
}
write.csv(mediation_results_pn12_waz, "./results/tax_mediation/species/mediation_results_pn12_waz3.csv")

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_pn12_waz, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_pn12_waz3 <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_pn12_waz$ci_indirect_low, mediation_results_pn12_waz$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_pn12_waz$ci_indirect_high, mediation_results_pn12_waz$ci_direct_high)),
                width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "3m: WAZ", y = "PN12: Infant gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_pn12_waz3

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_pn12_waz3.svg",
  width=5, height=4
  )
plot_pn12_waz3
dev.off()

```


#### of the BEP->WHZ3 effect pn12
```{r mat pn12 whz, message=FALSE}
set.seed(123)
library(foreach)
library(doFuture)
vars = extract_vars(
  pn12e,
  "c_whz3",
  taxonomic_level = "species",
  postnatal = TRUE,
  impute = FALSE
  )

mediation_results_pn12_whz <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 3, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  # Check if the results contain NaNs or missing values
   if (any(is.nan(unlist(ans)))) {
     cat("NaNs produced for mediator:", vars$mediator_names[i], "\n")
   }
  
  ans$mediator <- vars$mediator_names[i]
  mediation_results_pn12_whz <- rbind(mediation_results_pn12_whz, ans)
}
write.csv(mediation_results_pn12_whz, "./results/tax_mediation/species/mediation_results_pn12_whz3.csv")

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_pn12_whz, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_pn12_whz3 <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_pn12_whz$ci_indirect_low, mediation_results_pn12_whz$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_pn12_whz$ci_indirect_high, mediation_results_pn12_whz$ci_direct_high)),
                width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "3m: WHZ", y = "PN12: Infant gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_pn12_whz3

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_pn12_whz3.svg",
  width=5, height=4
  )
plot_pn12_whz3
dev.off()

```


### Maternal microbiome mediation 

#### of the BEP->GA effect
```{r mat incl GA}
set.seed(123)
vars = extract_vars(
  incl,
  "GAbirthweeks",
  taxonomic_level = "species",
  impute = FALSE
  )
mediation_results_incl_GA <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian", folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_incl_GA <- rbind(mediation_results_incl_GA, ans)
}
write.csv(mediation_results_incl_GA, "./results/tax_mediation/species/mediation_results_incl_GA.csv")

mediation_results_incl_GA <- read.csv("./results/tax_mediation/species/mediation_results_incl_GA.csv", row.names = 1)

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_incl_GA, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_incl_GA <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_incl_GA$ci_indirect_low, mediation_results_incl_GA$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_incl_GA$ci_indirect_high, mediation_results_incl_GA$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Gestational age", y = "Tri2: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_incl_GA

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_incl_GA.svg",
  width=5, height=45
  )
plot_incl_GA
dev.off()

filtered_mediation_results_incl_GA <- mediation_results_incl_GA %>%
  filter(direct >=0) %>%
  #filter(sign(ci_indirect_low) == sign(ci_indirect_high)) %>%
  filter(sign(ci_direct_low) == sign(ci_direct_high))

# Melt the data frame for ggplot
data_melted <- melt(filtered_mediation_results_incl_GA, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_incl_GA_filter <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", filtered_mediation_results_incl_GA$ci_indirect_low, filtered_mediation_results_incl_GA$ci_direct_low),
                    xmax = ifelse(variable == "indirect", filtered_mediation_results_incl_GA$ci_indirect_high, filtered_mediation_results_incl_GA$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Gestational age", y = "Tri2: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_incl_GA_filter

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_incl_GA_filter2.svg",
  width=6, height=3
  )
plot_incl_GA_filter
dev.off()


```

```{r mat tri3 GA}
set.seed(123)
vars = extract_vars(
  tri3,
  "GAbirthweeks",
  taxonomic_level = "species",
  impute = FALSE
  )
mediation_results_tri3_GA <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_tri3_GA <- rbind(mediation_results_tri3_GA, ans)
}
write.csv(mediation_results_tri3_GA, "./results/tax_mediation/species/mediation_results_tri3_GA.csv")

mediation_results_tri3_GA <- read.csv("./results/tax_mediation/species/mediation_results_tri3_GA.csv", row.names = 1)

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_tri3_GA, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_tri3_GA <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_tri3_GA$ci_indirect_low, mediation_results_tri3_GA$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_tri3_GA$ci_indirect_high, mediation_results_tri3_GA$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Gestational age", y = "Tri3: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_tri3_GA

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_tri3_GA.svg",
  width=5, height=45
  )
plot_tri3_GA
dev.off()

filtered_mediation_results_tri3_GA <- mediation_results_tri3_GA %>%
  #filter(direct >=0) %>%
  #filter(indirect >=0) %>%
  filter(sign(ci_indirect_low) == sign(ci_indirect_high)) %>%
  filter(sign(ci_direct_low) == sign(ci_direct_high))

# Melt the data frame for ggplot
data_melted <- melt(filtered_mediation_results_tri3_GA, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_tri3_GA_filter <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", filtered_mediation_results_tri3_GA$ci_indirect_low, filtered_mediation_results_tri3_GA$ci_direct_low),
                    xmax = ifelse(variable == "indirect", filtered_mediation_results_tri3_GA$ci_indirect_high, filtered_mediation_results_tri3_GA$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Gestational age", y = "Tri3: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_tri3_GA_filter

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_tri3_GA_filter2.svg",
  width=6, height=3
  )
plot_tri3_GA_filter
dev.off()

```

#### of the BEP->birth weight effect
```{r mat incl weight}
set.seed(123)
vars = extract_vars(
  incl,
  "c_weight0",
  taxonomic_level = "species",
  impute = FALSE
  )
mediation_results_incl_weight <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_incl_weight <- rbind(mediation_results_incl_weight, ans)
}
write.csv(mediation_results_incl_weight, "./results/tax_mediation/species/mediation_results_incl_weight.csv")

mediation_results_incl_weight <- read.csv("./results/tax_mediation/species/mediation_results_incl_weight.csv", row.names = 1)

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_incl_weight, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_incl_weight <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_incl_weight$ci_indirect_low, mediation_results_incl_weight$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_incl_weight$ci_indirect_high, mediation_results_incl_weight$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Birth Weight", y = "Tri2: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_incl_weight

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_incl_weight.svg",
  width=5, height=45
  )
plot_incl_weight
dev.off()

filtered_mediation_results_incl_weight <- mediation_results_incl_weight %>%
  filter(direct >=0) %>%
  filter(sign(ci_indirect_low) == sign(ci_indirect_high))%>%
  filter(sign(ci_direct_low) == sign(ci_direct_high))

# Melt the data frame for ggplot
data_melted <- melt(filtered_mediation_results_incl_weight, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_incl_weight_filter <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", filtered_mediation_results_incl_weight$ci_indirect_low, filtered_mediation_results_incl_weight$ci_direct_low),
                    xmax = ifelse(variable == "indirect", filtered_mediation_results_incl_weight$ci_indirect_high, filtered_mediation_results_incl_weight$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Birth Weight", y = "Tri2: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_incl_weight_filter

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_incl_weight_filter2.svg",
  width=6, height=3
  )
plot_incl_weight_filter
dev.off()

```

```{r mat tri3 weight}
set.seed(123)
vars = extract_vars(
  tri3,
  "c_weight0",
  taxonomic_level = "species",
  impute = FALSE
  )
mediation_results_tri3_weight <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_tri3_weight <- rbind(mediation_results_tri3_weight, ans)
}
write.csv(mediation_results_tri3_weight, "./results/tax_mediation/species/mediation_results_tri3_weight.csv")

mediation_results_tri3_weight <- read.csv("./results/tax_mediation/species/mediation_results_tri3_weight.csv", row.names = 1)

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_tri3_weight, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_tri3_weight <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_tri3_weight$ci_indirect_low, mediation_results_tri3_weight$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_tri3_weight$ci_indirect_high, mediation_results_tri3_weight$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Birth Weight", y = "Tri3: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_tri3_weight

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_tri3_weight.svg",
  width=5, height=45
  )
plot_tri3_weight
dev.off()

filtered_mediation_results_tri3_weight <- mediation_results_tri3_weight %>%
  filter(direct >=0) %>%
  filter(sign(ci_indirect_low) == sign(ci_indirect_high)) %>%
  filter(sign(ci_direct_low) == sign(ci_direct_high))

# Melt the data frame for ggplot
data_melted <- melt(filtered_mediation_results_tri3_weight, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_tri3_weight_filter <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", filtered_mediation_results_tri3_weight$ci_indirect_low, filtered_mediation_results_tri3_weight$ci_direct_low),
                    xmax = ifelse(variable == "indirect", filtered_mediation_results_tri3_weight$ci_indirect_high, filtered_mediation_results_tri3_weight$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Birth Weight", y = "Tri3: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_tri3_weight_filter

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_tri3_weight_filter2.svg",
  width=6, height=3
  )
plot_tri3_weight_filter
dev.off()

```

#### of the BEP->birth height effect
```{r mat incl height}
set.seed(123)
vars = extract_vars(
  incl,
  "c_height0",
  taxonomic_level = "species",
  impute = FALSE
  )
mediation_results_incl_height <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_incl_height <- rbind(mediation_results_incl_height, ans)
}
write.csv(mediation_results_incl_height, "./results/tax_mediation/species/mediation_results_incl_height.csv")

mediation_results_incl_height <- read.csv("./results/tax_mediation/species/mediation_results_incl_height.csv", row.names = 1)

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_incl_height, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_incl_height <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_incl_height$ci_indirect_low, mediation_results_incl_height$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_incl_height$ci_indirect_high, mediation_results_incl_height$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Birth Height", y = "Tri2: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_incl_height

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_incl_height.svg",
  width=5, height=45
  )
plot_incl_height
dev.off()

filtered_mediation_results_incl_height <- mediation_results_incl_height %>%
  filter(direct >=0) %>%
  filter(sign(ci_indirect_low) == sign(ci_indirect_high)) %>%
  filter(sign(ci_direct_low) == sign(ci_direct_high))

# Melt the data frame for ggplot
data_melted <- melt(filtered_mediation_results_incl_height, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_incl_height_filter <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", filtered_mediation_results_incl_height$ci_indirect_low, filtered_mediation_results_incl_height$ci_direct_low),
                    xmax = ifelse(variable == "indirect", filtered_mediation_results_incl_height$ci_indirect_high, filtered_mediation_results_incl_height$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Birth Height", y = "Tri2: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_incl_height_filter

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_incl_height_filter2.svg",
  width=6, height=3
  )
plot_incl_height_filter
dev.off()

```

```{r mat tri3 height}
set.seed(123)
vars = extract_vars(
  tri3,
  "c_height0",
  taxonomic_level = "species",
  impute = FALSE
  )
mediation_results_tri3_height <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_tri3_height <- rbind(mediation_results_tri3_height, ans)
}
write.csv(mediation_results_tri3_height, "./results/tax_mediation/species/mediation_results_tri3_height.csv")

mediation_results_tri3_height <- read.csv("./results/tax_mediation/species/mediation_results_tri3_height.csv", row.names = 1)
# Melt the data frame for ggplot
data_melted <- melt(mediation_results_tri3_height, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_tri3_height <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_tri3_height$ci_indirect_low, mediation_results_tri3_height$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_tri3_height$ci_indirect_high, mediation_results_tri3_height$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Birth Height", y = "Tri3: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_tri3_height

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_tri3_height.svg",
  width=5, height=45
  )
plot_tri3_height
dev.off()

filtered_mediation_results_tri3_height <- mediation_results_tri3_height %>%
  filter(direct >=0) %>%
  filter(sign(ci_indirect_low) == sign(ci_indirect_high)) %>%
  filter(sign(ci_direct_low) == sign(ci_direct_high))

# Melt the data frame for ggplot
data_melted <- melt(filtered_mediation_results_tri3_height, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_tri3_height_filter <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", filtered_mediation_results_tri3_height$ci_indirect_low, filtered_mediation_results_tri3_height$ci_direct_low),
                    xmax = ifelse(variable == "indirect", filtered_mediation_results_tri3_height$ci_indirect_high, filtered_mediation_results_tri3_height$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Birth Height", y = "Tri3: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_tri3_height_filter

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_tri3_height_filter2.svg",
  width=6, height=3
  )
plot_tri3_height_filter
dev.off()
```

#### of the BEP->LBW effect
Didn't work for LBW maybe because of the unbalance of 0 and 1 
```{r mat incl lbw}
set.seed(123)
vars = extract_vars(
  incl,
  "lbw",
  taxonomic_level = "species",
  impute = FALSE
  )
mediation_results_incl_lbw <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "binomial",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_incl_lbw <- rbind(mediation_results_incl_lbw, ans)
}
write.csv(mediation_results_incl_lbw, "./results/tax_mediation/species/mediation_results_incl_lbw.csv")
```

```{r mat tri3 lbw}
set.seed(123)
vars = extract_vars(
  tri3,
  "lbw",
  taxonomic_level = "species",
  impute = FALSE
  )
mediation_results_tri3_lbw <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "binomial",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_tri3_lbw <- rbind(mediation_results_tri3_lbw, ans)
}
write.csv(mediation_results_tri3_lbw, "./results/tax_mediation/species/mediation_results_tri3_lbw.csv")
```

 #### of the BEP->SGA effect
```{r mat incl sga}
set.seed(123)
vars = extract_vars(
  incl,
  "sga_new",
  taxonomic_level = "species",
  impute = FALSE
  )
mediation_results_incl_sga <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "binomial",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_incl_sga <- rbind(mediation_results_incl_sga, ans)
}
write.csv(mediation_results_incl_sga, "./results/tax_mediation/species/mediation_results_incl_sga.csv")

mediation_results_incl_sga <- read.csv("./results/tax_mediation/species/mediation_results_incl_sga.csv", row.names = 1)
# Melt the data frame for ggplot
data_melted <- melt(mediation_results_incl_sga, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_incl_sga <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_incl_sga$ci_indirect_low, mediation_results_incl_sga$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_incl_sga$ci_indirect_high, mediation_results_incl_sga$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Small for gestational age", y = "Tri2: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_incl_sga

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_incl_sga.svg",
  width=5, height=45
  )
plot_incl_sga
dev.off()

filtered_mediation_results_incl_sga <- mediation_results_incl_sga %>%
  filter(sign(ci_indirect_low) == sign(ci_indirect_high)) %>%
  filter(sign(ci_direct_low) == sign(ci_direct_high)) %>%
  filter(direct <= 0)

# Melt the data frame for ggplot
data_melted <- melt(filtered_mediation_results_incl_sga, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_incl_sga_filter <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", filtered_mediation_results_incl_sga$ci_indirect_low, filtered_mediation_results_incl_sga$ci_direct_low),
                    xmax = ifelse(variable == "indirect", filtered_mediation_results_incl_sga$ci_indirect_high, filtered_mediation_results_incl_sga$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Small for gestational age", y = "Tri2: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_incl_sga_filter

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_incl_sga_filter2.svg",
  width=6, height=6
  )
plot_incl_sga_filter
dev.off()
```

```{r mat tri3 sga}
set.seed(123)
vars = extract_vars(
  tri3,
  "sga_new",
  taxonomic_level = "species",
  impute = FALSE
  )
mediation_results_tri3_sga <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "binomial",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_tri3_sga <- rbind(mediation_results_tri3_sga, ans)
}
write.csv(mediation_results_tri3_sga, "./results/tax_mediation/species/mediation_results_tri3_sga.csv")

mediation_results_tri3_sga <- read.csv("./results/tax_mediation/species/mediation_results_tri3_sga.csv", row.names = 1)
# Melt the data frame for ggplot
data_melted <- melt(mediation_results_tri3_sga, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_tri3_sga <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_tri3_sga$ci_indirect_low, mediation_results_tri3_sga$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_tri3_sga$ci_indirect_high, mediation_results_tri3_sga$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Small for gestational age", y = "Tri3: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_tri3_sga

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_tri3_sga.svg",
  width=5, height=45
  )
plot_tri3_sga
dev.off()

filtered_mediation_results_tri3_sga <- mediation_results_tri3_sga %>%
  filter(sign(ci_indirect_low) == sign(ci_indirect_high)) %>%
  filter(sign(ci_direct_low) == sign(ci_direct_high))

# Melt the data frame for ggplot
data_melted <- melt(filtered_mediation_results_tri3_sga, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_tri3_sga_filter <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", filtered_mediation_results_tri3_sga$ci_indirect_low, filtered_mediation_results_tri3_sga$ci_direct_low),
                    xmax = ifelse(variable == "indirect", filtered_mediation_results_tri3_sga$ci_indirect_high, filtered_mediation_results_tri3_sga$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Small for gestational age", y = "Tri3: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_tri3_sga_filter

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_tri3_sga_filter2.svg",
  width=6, height=6
  )
plot_tri3_sga_filter
dev.off()
```


## mediation on gene data
## Formatting 

Before we start modelling, we perform a few formatting steps on the data. 
1. We split the data into subsets to analyse separately (because the treatment effect on specific genes will likely be different for mothers and infants and different at each timepoint).
2. We combine everything into TreeSummarizedExperiment R objects.

```{r splitting data maternal, include=FALSE}
# metadata
metadata_mat <- metadata[metadata$child == "no", ]
metadata_mat_incl <- metadata_mat[metadata_mat$timepoint == "incl", ]
metadata_mat_tri3 <- metadata_mat[metadata_mat$timepoint == "tri3", ]
metadata_mat_pn12 <- metadata_mat[metadata_mat$timepoint == "pn12", ]
metadata_mat_pn56 <- metadata_mat[metadata_mat$timepoint == "pn56", ]

# counts
gene_counts_mat <- gene_counts[, rownames(metadata_mat)]
gene_counts_mat <- gene_counts_mat[rowSums(gene_counts_mat) != 0, ]
gene_counts_mat_incl <- gene_counts_mat[, rownames(metadata_mat_incl)]
gene_counts_mat_tri3 <- gene_counts_mat[, rownames(metadata_mat_tri3)]
gene_counts_mat_pn12 <- gene_counts_mat[, rownames(metadata_mat_pn12)]
gene_counts_mat_pn56 <- gene_counts_mat[, rownames(metadata_mat_pn56)]

# fix order
metadata_mat_incl <- metadata_mat_incl[colnames(gene_counts_mat_incl), ]
metadata_mat_tri3 <- metadata_mat_tri3[colnames(gene_counts_mat_tri3), ]
metadata_mat_pn12 <- metadata_mat_pn12[colnames(gene_counts_mat_pn12), ]
metadata_mat_pn56 <- metadata_mat_pn56[colnames(gene_counts_mat_pn56), ]
```


```{r splitting data infant, include=FALSE}
# metadata
metadata_inf <- metadata[metadata$child == "yes", ]
metadata_inf_pn12 <- metadata_inf[metadata_inf$timepoint == "pn12", ]
metadata_inf_pn56 <- metadata_inf[metadata_inf$timepoint == "pn56", ]

# counts
gene_counts_inf <- gene_counts[, rownames(metadata_inf)]
gene_counts_inf <- gene_counts_inf[rowSums(gene_counts_inf) != 0, ]
gene_counts_inf_pn12 <- gene_counts_inf[, rownames(metadata_inf_pn12)]
gene_counts_inf_pn56 <- gene_counts_inf[, rownames(metadata_inf_pn56)]

# fix order
metadata_inf_pn12 <- metadata_inf_pn12[colnames(gene_counts_inf_pn12), ]
metadata_inf_pn56 <- metadata_inf_pn56[colnames(gene_counts_inf_pn56), ]
```

## functions

```{r functions, echo=FALSE}
run_GSEA <- function(mediation_output, export_path, export_name, top_n = 500, gene_anno=gene_anno, pathway_to_gene=pathway_to_gene, pathway_to_description=pathway_to_description){
  tmp <- data.frame(mediation_output[, c("indirect", "mediator")])
  colnames(tmp) <- c("logFC", "names")
  kegg_genes_tmp <- tmp$names %in% gene_anno[gene_anno$database == "KEGG", "anno"]
  tmp <- tmp[kegg_genes_tmp, ]
  gene_list <- tmp[,1]
  names(gene_list) <- as.character(tmp[, 2])
  gene_list <- sort(gene_list, decreasing = TRUE)
  #GSEA
  kk <- clusterProfiler::GSEA(
    geneList = gene_list,
    minGSSize = 2,
    pvalueCutoff = 0.05,
    TERM2GENE = pathway_to_gene,
    TERM2NAME = pathway_to_description
    )
  # output tables
  enriched_genesets <- base::intersect(
    names(kk@geneSets),
    kk@result$ID
    )
  for (set in enriched_genesets){
    enriched_genes <- as.data.frame(
      na.omit(kk@geneList[unlist(kk@geneSets[set])])
      )
    colnames(enriched_genes) <- "logFC"
    enriched_genes$gene <- rownames(enriched_genes)
    cmap <- c(COLOR_SCHEME[c(3, 1)], "white", COLOR_SCHEME[c(2, 4)])
    lms <- c(-max(abs(enriched_genes$logFC)), max(abs(enriched_genes$logFC)))
    p <- ggplot2::ggplot(enriched_genes, ggplot2::aes(x=1, y=gene, color=logFC)) +
      ggplot2::geom_jitter(size=5) +
      ggplot2::scale_color_gradientn(colors=cmap, limits=lms)
    fname_plot <- paste0(export_path, "figures/", export_name, "_", set, "_enriched_genes.svg")
    ggplot2::ggsave(fname_plot, p)
    fname <- paste0(export_path, export_name, "_", set, "_enriched_genes.csv")
    write.csv(enriched_genes, fname)
  }
  # output plots
  vis <- nrow(kk@result) > 0 # only run plots if there are results
  if (vis) {
    csv_fname <- paste0(export_path, "pathway_enrichment_", export_name, ".csv")
    # svg_fname <- paste0(export_path, "figures/GSEA_", export_name, ".svg")
    write.csv(kk@result, csv_fname)
    # svglite::svglite(svg_fname, width=5, height=6)
    # enrichment_lollipop(kk)
    # dev.off()
    return(kk)
  } else {
    return(NULL)
  }
}

extract_vars <- function(timepoint_metadata, timepoint_genecounts, outcome_var, postnatal=FALSE){
  # Extract the exposure
  A <- timepoint_metadata$treatment1_prenatal_n
  # Extract the outcome
  Y <- timepoint_metadata[[outcome_var]]
  # Extract the mediators
  M <- t(timepoint_genecounts)
  # prevalence filtering (filter columns with less than 20% non-zero values)
  M <- as.matrix(M[, apply(M, 2, function(col) mean(col != 0)) >= 0.2])
  # variance filtering (sort the columns by variance and only take the top n)
  M <- as.matrix(M[, order(apply(M, 2, var), decreasing = TRUE)[1:top_n]])
  # extract the covariates
  if (postnatal){
    W <- timepoint_metadata %>%
      dplyr::select(
        "m_heightincl", "m_hbincl", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave", "iycf_ebf_age"
        )
  } else {
    W <- timepoint_metadata %>%
      dplyr::select(
        "m_heightincl", "m_hbincl", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave"
        )
  }
  row.names(W) <- row.names(timepoint_metadata)
  # take complete cases
  complete_cases <- complete.cases(A, Y, M, W)
  if (sum(complete_cases) != length(A)){
    message("Removed ", length(A) - sum(complete_cases), " observations due to missing values.")
  }
  A <- A[complete_cases]
  Y <- Y[complete_cases]
  M <- M[complete_cases, ]
  W <- W[complete_cases, ]
  # Convert covariates to numeric
  W <- data.frame(lapply(W, function(x) as.numeric(as.character(x)))) %>%
    as.matrix()
  message(
    paste0(
      "Extracted exposure A with ", length(A), " observations.\n",
      "Extracted outcome Y with ", length(Y), " observations.\n",
      "Extracted mediators M with ", dim(M)[1], " observations and ", dim(M)[2], " features.\n",
      "Extracted covariates W with ", dim(W)[1], " observations and ", dim(W)[2], " features."
      )
    )
  vars_df <- data.frame(A, Y, M, W)
  return(
    list(
      df=vars_df,
      mediator_names=names(vars_df)[3:(ncol(vars_df)-ncol(W))],
      covariate_names=names(vars_df)[(ncol(vars_df)-ncol(W)+1):ncol(vars_df)]
    )
  )
}
```

## TMLE mediation on genes as mediators using HDmediation
Targeted Maximum Likelihood Estimation (TMLE) is a general approach for constructing an efficient double-robust
semi-parametric substitution estimator of a causal effect parameter.

```{r mediation setup, echo=FALSE}
set.seed(123)
sl_library <- c("SL.glm","SL.mean")
```

### Maternal microbiome mediation 
#### of the BEP->GA effect
```{r mat incl GA}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_incl_GA.csv")) {
  vars <- extract_vars(
    timepoint_metadata = metadata_mat_incl,
    timepoint_genecounts = gene_counts_mat_incl,
    outcome_var = "GAbirthweeks"
    )
  mediation_results_incl_GA <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "gaussian", folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_incl_GA <- rbind(mediation_results_incl_GA, ans)
  }
  write.csv(mediation_results_incl_GA, "./results/gene_mediation/top500/mediation_results_incl_GA.csv")
} else {
  mediation_results_incl_GA <- read.csv("./results/gene_mediation/top500/mediation_results_incl_GA.csv")
}
```

```{r GSEA mat incl GA, echo=FALSE}
kk_incl_GA <- run_GSEA(
  mediation_output = mediation_results_incl_GA,
  export_path = "./results/gene_mediation/top500/",
  export_name = "maternal_incl_GA",
  top_n = 500,
  gene_anno = gene_anno, 
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_incl_GA)){
  DT::datatable(kk_incl_GA@result)
}
```

```{r mat tri3 GA}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_tri3_GA.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_mat_tri3,
    timepoint_genecounts = gene_counts_mat_tri3,
    outcome_var = "GAbirthweeks"
    )
  mediation_results_tri3_GA <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "gaussian",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_tri3_GA <- rbind(mediation_results_tri3_GA, ans)
  }
  write.csv(mediation_results_tri3_GA, "./results/gene_mediation/top500/mediation_results_tri3_GA.csv")
} else {
  mediation_results_tri3_GA <- read.csv("./results/gene_mediation/top500/mediation_results_tri3_GA.csv")
}
```

```{r GSEA mat tri3 GA, echo=FALSE}
kk_tri3_GA <- run_GSEA(
  mediation_output = mediation_results_tri3_GA,
  export_path = "./results/gene_mediation/top500/",
  export_name = "maternal_tri3_GA",
  top_n = 500,
  gene_anno = gene_anno,
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_tri3_GA)){
  DT::datatable(kk_tri3_GA@result)
}
```

#### of the BEP->birth weight effect
```{r mat incl weight}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_incl_weight.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_mat_incl,
    timepoint_genecounts = gene_counts_mat_incl,
    outcome_var = "c_weight0"
    )
  mediation_results_incl_weight <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "gaussian",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_incl_weight <- rbind(mediation_results_incl_weight, ans)
  }
  write.csv(mediation_results_incl_weight, "./results/gene_mediation/top500/mediation_results_incl_weight.csv")
} else {
  mediation_results_incl_weight <- read.csv("./results/gene_mediation/top500/mediation_results_incl_weight.csv")
}
```

```{r GSEA mat incl weight, echo=FALSE}
kk_incl_weight <- run_GSEA(
  mediation_output = mediation_results_incl_weight,
  export_path = "./results/gene_mediation/top500/",
  export_name = "maternal_incl_weight",
  top_n = 500,
  gene_anno = gene_anno, 
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_incl_weight)){
  DT::datatable(kk_incl_weight@result)
}
```

```{r mat tri3 weight}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_tri3_weight.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_mat_tri3,
    timepoint_genecounts = gene_counts_mat_tri3,
    outcome_var = "c_weight0"
    )
  mediation_results_tri3_weight <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "gaussian",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_tri3_weight <- rbind(mediation_results_tri3_weight, ans)
  }
  write.csv(mediation_results_tri3_weight, "./results/gene_mediation/top500/mediation_results_tri3_weight.csv")
} else {
  mediation_results_tri3_weight <- read.csv("./results/gene_mediation/top500/mediation_results_tri3_weight.csv")
}
```

```{r GSEA mat tri3 weight, echo=FALSE}
kk_tri3_weight <- run_GSEA(
  mediation_output = mediation_results_tri3_weight,
  export_path = "./results/gene_mediation/top500/",
  export_name = "maternal_tri3_weight",
  top_n = 500,
  gene_anno = gene_anno,
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_tri3_weight)){
  DT::datatable(kk_tri3_weight@result)
}
```

#### of the BEP->birth height effect
```{r mat incl height}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_incl_height.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_mat_incl,
    timepoint_genecounts = gene_counts_mat_incl,
    outcome_var = "c_height0"
    )
  mediation_results_incl_height <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "gaussian",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_incl_height <- rbind(mediation_results_incl_height, ans)
  }
  write.csv(mediation_results_incl_height, "./results/gene_mediation/top500/mediation_results_incl_height.csv")
} else {
  mediation_results_incl_height <- read.csv("./results/gene_mediation/top500/mediation_results_incl_height.csv")
}
```

```{r GSEA mat incl height, echo=FALSE}
kk_incl_height <- run_GSEA(
  mediation_output = mediation_results_incl_height,
  export_path = "./results/gene_mediation/top500/",
  export_name = "maternal_incl_height",
  top_n = 500,
  gene_anno = gene_anno, 
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_incl_height)){
  DT::datatable(kk_incl_height@result)
}
```

```{r mat tri3 height}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_tri3_height.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_mat_tri3,
    timepoint_genecounts = gene_counts_mat_tri3,
    outcome_var = "c_height0"
    )
  mediation_results_tri3_height <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "gaussian",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_tri3_height <- rbind(mediation_results_tri3_height, ans)
  }
  write.csv(mediation_results_tri3_height, "./results/gene_mediation/top500/mediation_results_tri3_height.csv")
} else {
  mediation_results_tri3_height <- read.csv("./results/gene_mediation/top500/mediation_results_tri3_height.csv")
}
```

```{r GSEA mat tri3 height, echo=FALSE}
kk_tri3_height <- run_GSEA(
  mediation_output = mediation_results_tri3_height,
  export_path = "./results/gene_mediation/top500/",
  export_name = "maternal_tri3_height",
  top_n = 500,
  gene_anno = gene_anno,
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_tri3_height)){
  DT::datatable(kk_tri3_height@result)
}
```

#### of the BEP->LBW effect
```{r mat incl lbw}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_incl_lbw.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_mat_incl,
    timepoint_genecounts = gene_counts_mat_incl,
    outcome_var = "lbw"
    )
  mediation_results_incl_lbw <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "binomial",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_incl_lbw <- rbind(mediation_results_incl_lbw, ans)
  }
  write.csv(mediation_results_incl_lbw, "./results/gene_mediation/top500/mediation_results_incl_lbw.csv")
} else {
  mediation_results_incl_lbw <- read.csv("./results/gene_mediation/top500/mediation_results_incl_lbw.csv")
}
```

```{r GSEA mat incl lbw, echo=FALSE}
kk_incl_lbw <- run_GSEA(
  mediation_output = mediation_results_incl_lbw,
  export_path = "./results/gene_mediation/top500/",
  export_name = "maternal_incl_lbw",
  top_n = 500,
  gene_anno = gene_anno, 
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_incl_lbw)){
  DT::datatable(kk_incl_lbw@result)
}
```

```{r mat tri3 lbw}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_tri3_lbw.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_mat_tri3,
    timepoint_genecounts = gene_counts_mat_tri3,
    outcome_var = "lbw"
    )
  mediation_results_tri3_lbw <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "binomial",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_tri3_lbw <- rbind(mediation_results_tri3_lbw, ans)
  }
  write.csv(mediation_results_tri3_lbw, "./results/gene_mediation/top500/mediation_results_tri3_lbw.csv")
} else {
  mediation_results_tri3_lbw <- read.csv("./results/gene_mediation/top500/mediation_results_tri3_lbw.csv")
}
```

```{r GSEA mat tri3 lbw, echo=FALSE}
kk_tri3_lbw <- run_GSEA(
  mediation_output = mediation_results_tri3_lbw,
  export_path = "./results/gene_mediation/top500/",
  export_name = "maternal_tri3_lbw",
  top_n = 500,
  gene_anno = gene_anno,
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_tri3_lbw)){
  DT::datatable(kk_tri3_lbw@result)
}
```

#### of the BEP->SGA effect
```{r mat incl sga}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_incl_sga.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_mat_incl,
    timepoint_genecounts = gene_counts_mat_incl,
    outcome_var = "sga_new"
    )
  mediation_results_incl_sga <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "binomial",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_incl_sga <- rbind(mediation_results_incl_sga, ans)
  }
  write.csv(mediation_results_incl_sga, "./results/gene_mediation/top500/mediation_results_incl_sga.csv")
} else {
  mediation_results_incl_sga <- read.csv("./results/gene_mediation/top500/mediation_results_incl_sga.csv")
}
```

```{r GSEA mat incl sga, echo=FALSE}
kk_incl_sga <- run_GSEA(
  mediation_output = mediation_results_incl_sga,
  export_path = "./results/gene_mediation/top500/",
  export_name = "maternal_incl_sga",
  top_n = 500,
  gene_anno = gene_anno, 
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_incl_sga)){
  DT::datatable(kk_incl_sga@result)
}
```

```{r mat tri3 sga}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_tri3_sga.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_mat_tri3,
    timepoint_genecounts = gene_counts_mat_tri3,
    outcome_var = "sga_new"
    )
  mediation_results_tri3_sga <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "binomial",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_tri3_sga <- rbind(mediation_results_tri3_sga, ans)
  }
  write.csv(mediation_results_tri3_sga, "./results/gene_mediation/top500/mediation_results_tri3_sga.csv")
} else {
  mediation_results_tri3_sga <- read.csv("./results/gene_mediation/top500/mediation_results_tri3_sga.csv")
}
```

```{r GSEA mat tri3 sga, echo=FALSE}
kk_tri3_sga <- run_GSEA(
  mediation_output = mediation_results_tri3_sga,
  export_path = "./results/gene_mediation/top500/",
  export_name = "maternal_tri3_sga",
  top_n = 500,
  gene_anno = gene_anno,
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_tri3_sga)){
  DT::datatable(kk_tri3_sga@result)
}
```

### Infant microbiome mediation
#### of the BEP->HAZ effect
```{r inf pn12 haz3}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_pn12_haz3.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_inf_pn12,
    timepoint_genecounts = gene_counts_inf_pn12,
    outcome_var = "c_haz3",
    postnatal = TRUE
    )
  mediation_results_pn12_haz3 <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "gaussian",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_pn12_haz3 <- rbind(mediation_results_pn12_haz3, ans)
  }
  write.csv(mediation_results_pn12_haz3, "./results/gene_mediation/top500/mediation_results_pn12_haz3.csv")
} else {
  mediation_results_pn12_haz3 <- read.csv("./results/gene_mediation/top500/mediation_results_pn12_haz3.csv")
}
```

```{r GSEA inf pn12 haz3, echo=FALSE}
kk_inf_pn12_haz3 <- run_GSEA(
  mediation_output = mediation_results_pn12_haz3,
  export_path = "./results/gene_mediation/top500/",
  export_name = "infant_pn12_haz3",
  top_n = 500,
  gene_anno = gene_anno, 
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_inf_pn12_haz3)){
  DT::datatable(kk_inf_pn12_haz3@result)
}
```

```{r inf pn56 haz6}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_pn56_haz6.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_inf_pn56,
    timepoint_genecounts = gene_counts_inf_pn56,
    outcome_var = "c_haz6",
    postnatal = TRUE
    )
  mediation_results_pn56_haz6 <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    print((i/length(vars$mediator_names))*100)
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "gaussian",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_pn56_haz6 <- rbind(mediation_results_pn56_haz6, ans)
  }
  write.csv(mediation_results_pn56_haz6, "./results/gene_mediation/top500/mediation_results_pn56_haz6.csv")
} else {
  mediation_results_pn56_haz6 <- read.csv("./results/gene_mediation/top500/mediation_results_pn56_haz6.csv")
}
```

```{r GSEA inf pn12 haz6, echo=FALSE}
kk_inf_pn56_haz6 <- run_GSEA(
  mediation_output = mediation_results_pn56_haz6,
  export_path = "./results/gene_mediation/top500/",
  export_name = "infant_pn56_haz6",
  top_n = 500,
  gene_anno = gene_anno,
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_inf_pn56_haz6)){
  DT::datatable(kk_inf_pn56_haz6@result)
}
```

#### of the BEP->WAZ effect
```{r inf pn12 waz3}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_pn12_waz3.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_inf_pn12,
    timepoint_genecounts = gene_counts_inf_pn12,
    outcome_var = "c_waz3",
    postnatal = TRUE
    )
  mediation_results_pn12_waz3 <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "gaussian",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_pn12_waz3 <- rbind(mediation_results_pn12_waz3, ans)
  }
  write.csv(mediation_results_pn12_waz3, "./results/gene_mediation/top500/mediation_results_pn12_waz3.csv")
} else {
  mediation_results_pn12_waz3 <- read.csv("./results/gene_mediation/top500/mediation_results_pn12_waz3.csv")
}
```

```{r GSEA inf pn12 waz3, echo=FALSE}
kk_inf_pn12_waz3 <- run_GSEA(
  mediation_output = mediation_results_pn12_waz3,
  export_path = "./results/gene_mediation/top500/",
  export_name = "infant_pn12_waz3",
  top_n = 500,
  gene_anno = gene_anno,
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_inf_pn12_waz3)){
  DT::datatable(kk_inf_pn12_waz3@result)
}
```

```{r inf pn12 waz6}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_pn12_waz6.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_inf_pn12,
    timepoint_genecounts = gene_counts_inf_pn12,
    outcome_var = "c_waz6",
    postnatal = TRUE
    )
  mediation_results_pn12_waz6 <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "gaussian",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_pn12_waz6 <- rbind(mediation_results_pn12_waz6, ans)
  }
  write.csv(mediation_results_pn12_waz6, "./results/gene_mediation/top500/mediation_results_pn12_waz6.csv")
} else {
  mediation_results_pn12_waz6 <- read.csv("./results/gene_mediation/top500/mediation_results_pn12_waz6.csv")
}
```

```{r GSEA inf pn12 waz6, echo=FALSE}
kk_inf_pn12_waz6 <- run_GSEA(
  mediation_output = mediation_results_pn12_waz6,
  export_path = "./results/gene_mediation/top500/",
  export_name = "infant_pn12_waz6",
  top_n = 500,
  gene_anno = gene_anno,
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_inf_pn12_waz6)){
  DT::datatable(kk_inf_pn12_waz6@result)
}
```

```{r inf pn56 waz6}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_pn56_waz6.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_inf_pn56,
    timepoint_genecounts = gene_counts_inf_pn56,
    outcome_var = "c_waz6",
    postnatal = TRUE
    )
  mediation_results_pn56_waz6 <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "gaussian",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_pn56_waz6 <- rbind(mediation_results_pn56_waz6, ans)
  }
  write.csv(mediation_results_pn56_waz6, "./results/gene_mediation/top500/mediation_results_pn56_waz6.csv")
} else {
  mediation_results_pn56_waz6 <- read.csv("./results/gene_mediation/top500/mediation_results_pn56_waz6.csv")
}
```

```{r GSEA inf pn56 waz6, echo=FALSE}
kk_inf_pn56_waz6 <- run_GSEA(
  mediation_output = mediation_results_pn56_waz6,
  export_path = "./results/gene_mediation/top500/",
  export_name = "infant_pn56_waz6",
  top_n = 500,
  gene_anno = gene_anno,
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_inf_pn56_waz6)){
  DT::datatable(kk_inf_pn56_waz6@result)
}
```

#### of the BEP->WHZ effect
```{r mat pn12 whz3}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_pn12_whz3.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_inf_pn12,
    timepoint_genecounts = gene_counts_inf_pn12,
    outcome_var = "c_whz3",
    postnatal = TRUE
    )
  mediation_results_pn12_whz3 <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "gaussian",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_pn12_whz3 <- rbind(mediation_results_pn12_whz3, ans)
  }
  write.csv(mediation_results_pn12_whz3, "./results/gene_mediation/top500/mediation_results_pn12_whz3.csv")
} else {
  mediation_results_pn12_whz3 <- read.csv("./results/gene_mediation/top500/mediation_results_pn12_whz3.csv")
}
```

```{r GSEA inf pn12 whz3, echo=FALSE}
kk_inf_pn12_whz3 <- run_GSEA(
  mediation_output = mediation_results_pn12_whz3,
  export_path = "./results/gene_mediation/top500/",
  export_name = "infant_pn12_whz3",
  top_n = 500,
  gene_anno = gene_anno,
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_inf_pn12_whz3)){
  DT::datatable(kk_inf_pn12_whz3@result)
}
```

```{r inf pn56 whz6}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_pn56_whz6.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_inf_pn56,
    timepoint_genecounts = gene_counts_inf_pn56,
    outcome_var = "c_whz6",
    postnatal = TRUE
    )
  mediation_results_pn56_whz6 <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "gaussian",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_pn56_whz6 <- rbind(mediation_results_pn56_whz6, ans)
  }
  write.csv(mediation_results_pn56_whz6, "./results/gene_mediation/top500/mediation_results_pn56_whz6.csv")
} else {
  mediation_results_pn56_whz6 <- read.csv("./results/gene_mediation/top500/mediation_results_pn56_whz6.csv")
}
```

```{r GSEA inf pn56 whz6, echo=FALSE}
kk_inf_pn56_whz6 <- run_GSEA(
  mediation_output = mediation_results_pn56_whz6,
  export_path = "./results/gene_mediation/top500/",
  export_name = "infant_pn56_whz6",
  top_n = 500,
  gene_anno = gene_anno,
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_inf_pn56_whz6)){
  DT::datatable(kk_inf_pn56_whz6@result)
}
```

#### of the BEP->stunting effect
```{r inf pn12 stunting}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_pn12_stunting.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_inf_pn12,
    timepoint_genecounts = gene_counts_inf_pn12,
    outcome_var = "c_stunting6",
    postnatal = TRUE
    )
  mediation_results_pn12_stunting <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "binomial",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_pn12_stunting <- rbind(mediation_results_pn12_stunting, ans)
  }
  write.csv(mediation_results_pn12_stunting, "./results/gene_mediation/top500/mediation_results_pn12_stunting.csv")
} else {
  mediation_results_pn12_stunting <- read.csv("./results/gene_mediation/top500/mediation_results_pn12_stunting.csv")
}
```

```{r GSEA inf pn12 stunting, echo=FALSE}
kk_inf_pn12_stunting <- run_GSEA(
  mediation_output = mediation_results_pn12_stunting,
  export_path = "./results/gene_mediation/top500/",
  export_name = "infant_pn12_stunting",
  top_n = 500,
  gene_anno = gene_anno,
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_inf_pn12_stunting)){
  DT::datatable(kk_inf_pn12_stunting@result)
}
```

```{r inf pn56 stunting}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_pn56_stunting.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_inf_pn56,
    timepoint_genecounts = gene_counts_inf_pn56,
    outcome_var = "c_stunting",
    postnatal = TRUE
    )
  mediation_results_pn56_stunting <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "binomial",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_pn56_stunting <- rbind(mediation_results_pn56_stunting, ans)
  }
  write.csv(mediation_results_pn56_stunting, "./results/gene_mediation/top500/mediation_results_pn56_stunting.csv")
} else {
  mediation_results_pn56_stunting <- read.csv("./results/gene_mediation/top500/mediation_results_pn56_stunting.csv")
}
```

```{r GSEA inf pn56 stunting, echo=FALSE}
kk_inf_pn56_stunting <- run_GSEA(
  mediation_output = mediation_results_pn56_stunting,
  export_path = "./results/gene_mediation/top500/",
  export_name = "infant_pn56_stunting",
  top_n = 500,
  gene_anno = gene_anno,
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_inf_pn56_stunting)){
  DT::datatable(kk_inf_pn56_stunting@result)
}
```
