---
title: "BEPmetagenomic_majorRevision_github"
author: "Lishi"
date: "2024-10-22"
output: html_document
---

# about this markdown
This markdown combines all extra statistic steps (major revision) of metagenomic analysis for the manuscript we submited to Nature Communications.
Metagenomic sequencing reads mapping and gene-level profiling were performed using inStrain in Python. Therefore these scripts were not included in this markdown, but the demo for use are available at https://github.com/MrOlm/inStrain.
Therefore, this markdown contains visualization of relative abundance, analysis of alpha diversity, beta diversity, taxonomic differential abundance analysis, mediation analysis and gene set enrichment analysis. The source data are available along with this markdown.

# setup
## working directory
First, set the working directory to your folder and set random seed.

```{r echo=FALSE}
# set the work directory
knitr::opts_knit$set(
  root.dir = '~/.../...', 
  echo=TRUE
  )
message(paste0("The current working directory is set to: ", getwd()))

set.seed(123)
```

## packages
Load necessary packages before analysis.

```{r echo=FALSE, message=FALSE}
library(ANCOMBC)
library(clusterProfiler)
library(ComplexHeatmap)
library(dbarts)
library(DESeq2)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(haven)
library(hrbrthemes)
library(lme4)
library(medoutcon)
library(mixtools)
library(permFDP)
library(phyloseq)
library(plotly)
library(randomForest)
library(sl3)
library(SuperLearner)
library(tibble)
library(tidyr)
library(tmle)
library(treeclimbR)
library(UpSetR)
library(merTools)
library(broom.mixed) # For tidying model output
library(scales)
library(igraph)
library(SpiecEasi)
library(dbarts)
library(doFuture)
library(DT)
library(stringr)
options(DT.options = list(
  initComplete = JS("function(settings, json) {",
  "$(this.api().table().header()).css({'background-color': 
  '#000', 'color': '#fff'});","}")))


```


# import data
## read data
Save the data files provided first, and read them:
```{r}
metadata <- read.csv("./metadata.csv", row.names = 1) %>%
  mutate(bep_group = ifelse(timepoint=="incl" & code=="A Y", "Mother_incl_IFA",
                     ifelse(timepoint=="incl" & code=="B Z", "Mother_incl_BEP",
                     ifelse(timepoint=="tri3" & code=="A Y", "Mother_tri3_IFA",
                     ifelse(timepoint=="tri3" & code=="B Z", "Mother_tri3_BEP",
                     ifelse(timepoint=="pn12" & child=="no" & code=="A Y", "Mother_pn12_IFA",
                     ifelse(timepoint=="pn12" & child=="no" & code=="B Z", "Mother_pn12_BEP",
                     ifelse(timepoint=="pn12" & child=="yes" & code=="A Y", "Infant_pn12_IFA",
                     ifelse(timepoint=="pn12" & child=="yes" & code=="B Z", "Infant_pn12_BEP",
                     ifelse(timepoint=="pn56" & child=="no" & code=="A Y", "Mother_pn56_IFA",
                     ifelse(timepoint=="pn56" & child=="no" & code=="B Z", "Mother_pn56_BEP",
                     ifelse(timepoint=="pn56" & child=="yes" & code=="A Y", "Infant_pn56_IFA",
                     ifelse(timepoint=="pn56" & child=="yes" & code=="B Z", "Infant_pn56_BEP","other"))))))))))))) # grouping, dyads and time point information

taxonomy <- as.matrix(read.csv("./taxonomy_clean.csv", row.names = 1))
absolute_counts <- read.csv("./absolute_counts_table.csv", row.names = 1)

# read gene tables
gene_counts <- read.csv("./gene_counts_clean.csv", row.names = 1)
gene_anno <- read.csv("./gene_annotations_clean.csv", row.names = 1)
gene_to_pathway <- read.csv("./k2pathway.txt", sep="\t", header=FALSE)
names(gene_to_pathway) <- c("gene", "pathway")
gene_to_pathway$gene <- gsub(pattern = "^ko:", "", gene_to_pathway$gene)
gene_to_pathway$pathway <- gsub(pattern = "^path:", "", gene_to_pathway$pathway)
pathway_to_gene <- gene_to_pathway[, c("pathway", "gene")]

pathway_to_description <- read.csv("./keggmaps_to_name.txt", sep="\t", header=FALSE, colClasses = c("character", "character"))
pathway_to_description$V1 <- gsub(pattern = "^", "map", pathway_to_description$V1)

microbial_pathways <- read.csv("./microbial_ko_table.tsv", sep="\t")

# the data of sample collection dates
pn12e_se1g <- read.csv("./pn12e.csv")
pn12m_se1g <- read.csv("./pn12m.csv")
pn56e_se1g <- read.csv("./pn56e.csv")
pn56m_se1g <- read.csv("./pn56m.csv")
incl_se1g <- read.csv("./incl.csv")
tri3_se1g <- read.csv("./tri3.csv")


```

## data preparation
```{r}
tax_for_tree <- cbind(taxonomy[, 1:ncol(taxonomy)-1], row.names(taxonomy))
proxy_phylogenetic_tree <- TreeSummarizedExperiment::toTree(tax_for_tree)

ps <- phyloseq(
  otu_table=otu_table(absolute_counts, taxa_are_rows=TRUE), 
  tax_table=tax_table(taxonomy),
  sample_data=sample_data(metadata), 
  phy_tree=phy_tree(proxy_phylogenetic_tree)
  )

# normalize data
pslog <- transform_sample_counts(ps, function(x) log(1 + x))
```

## data subsets
```{r}
incl <- subset_samples(ps, timepoint=="incl")
tri3 <- subset_samples(ps, timepoint=="tri3")
pn12 <- subset_samples(ps, timepoint=="pn12")
pn12m <- subset_samples(ps, timepoint=="pn12" & child=="no")
pn12e <- subset_samples(ps, timepoint=="pn12" & child=="yes")
pn56 <- subset_samples(ps, timepoint=="pn56")
pn56m <- subset_samples(ps, timepoint=="pn56" & child=="no")
pn56e <- subset_samples(ps, timepoint=="pn56" & child=="yes")
# add label to "code" and "child: "A Y" == "IFA", "B Z" == "BEP"; "no" == "Mother", "yes" == "Child"
code.labs <- c("A Y" = "IFA", "B Z" = "BEP")
child.labs <- c("no" = "Mother", "yes" = "Child")

#use log for beta diversity
infant <- subset_samples(pslog, child=="yes")
mother <- subset_samples(pslog, child=="no")
prenatal <- subset_samples(mother, timepoint=="incl" | timepoint=="tri3")
postnatal <- subset_samples(mother, timepoint=="pn12" | timepoint=="pn56")

```

# Figure 1.
## baseline
```{r}
metadata_distinct <- metadata %>%
  distinct(idbs, .keep_all = TRUE)

metadata_baseline <- metadata_distinct %>%
  dplyr::select(idbs, code, GAbirthweeks, w_age, m_weightincl, m_heightincl, m_bmiincl, hemoglobin_anc3, mddw_10_ave, asset1comp_10)

# Transform the data from wide to long format
metadata_long <- metadata_baseline %>%
  pivot_longer(
    cols = c(GAbirthweeks, w_age, m_weightincl, m_heightincl, m_bmiincl, hemoglobin_anc3, mddw_10_ave, asset1comp_10),
    names_to = "variable",
    values_to = "value"
  ) %>%
    mutate(code = if_else(code== "B Z", "Intervention", "Control")) %>%
  mutate(code = factor(code, levels = c("Intervention", "Control")))

# Faceted plot with specified colors
baseline <- ggplot(metadata_long, aes(x = code, y = value, color = code)) +
  geom_boxplot() +  # Remove color for outliers
  geom_jitter(width = 0.2, alpha = 0.7) +
  facet_wrap(~ variable, scales = "free_y", ncol = 6) +
  scale_color_manual(values = c("Intervention" = "steelblue", "Control" = "darkred")) +
  labs(x = "", y = "") +
  theme(
    axis.text = element_text(size = 8),     # Set axis text font size
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_rect(color = "black", fill = NA)
  ) + theme_bw()


svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/z-baseline.svg",
  width=10, height=3.5
  )
baseline
dev.off()

# Create a summary data frame from your counts
data <- data.frame(
  code = c(rep("Intervention", 2), rep("Control", 2)),
  sga_new = c(0, 1, 0, 1),
  count = c(53, 16, 59, 18)
)%>%
    mutate(code = factor(code, levels = c("Intervention", "Control")))

# Create a stacked bar plot
sga<- ggplot(data, aes(x = code, y = count, fill = factor(sga_new))) +
  geom_bar(stat = "identity") +
  labs(
    x = "",
    y = "",
    fill = "SGA",
    title = ""
  ) +
  scale_fill_manual(values = c("0" = "steelblue", "1" = "darkred")) +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_rect(color = "black", fill = NA)
  )
svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/z-sga.svg",
  width=2.3, height=2
  )
sga
dev.off()


# Create a summary data frame from your counts
data <- data.frame(
  code = c(rep("Intervention", 2), rep("Control", 2)),
  sga_new = c(0, 1, 0, 1),
  count = c(61, 8, 66, 13)
)%>%
    mutate(code = factor(code, levels = c("Intervention", "Control")))

# Create a stacked bar plot
lbw<- ggplot(data, aes(x = code, y = count, fill = factor(sga_new))) +
  geom_bar(stat = "identity") +
  labs(
    x = "",
    y = "",
    fill = "LBW",
    title = ""
  ) +
  scale_fill_manual(values = c("0" = "steelblue", "1" = "darkred")) +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_rect(color = "black", fill = NA)
  )
svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/z-lbw.svg",
  width=2.3, height=2
  )
lbw
dev.off()

```


###infant weight
```{r}
metadata_distinct <- metadata %>%
  distinct(idbs, .keep_all = TRUE)

metadata_weight <- metadata_distinct %>%
  dplyr::select(idbs, code, c_weight0, c_weight1, c_weight2, c_weight3, c_weight4, c_weight5, c_weight6)

# Transform the data from wide to long format
metadata_long <- metadata_weight %>%
  pivot_longer(
    cols = starts_with("c_weight"),
    names_to = "time",
    values_to = "weight"
  )

# Fit the mixed-effects linear regression model
library(lmerTest)
model <- lmer(weight ~ time * code + (1 | idbs), data = metadata_long)
summary(model)

# Convert 'time' to a numeric variable for plotting
metadata_long <- metadata_long %>%
  mutate(time = as.numeric(gsub("c_weight", "", time)))

# Summarize the data
metadata_summary <- metadata_long %>%
  group_by(code, time) %>%
  dplyr::summarize(
    mean_weight = mean(weight, na.rm = TRUE),
    sd_weight = sd(weight, na.rm = TRUE),
    .groups = 'drop'
  ) %>%
  mutate(code=if_else(code=="B Z", "Intervention", "Control"))

# Create the line plot
infant_weight <- ggplot(metadata_summary, aes(x = time, y = mean_weight, color = code, group = code)) +
  geom_line(size = 0.5) +
  geom_point(size = 1) +
  geom_errorbar(aes(ymin = mean_weight - sd_weight, ymax = mean_weight + sd_weight), width = 0.2) +
  labs(x = "Months", y = "Weight, kg", color = "Group") +
  scale_color_manual(values = c("Intervention" = "steelblue", "Control" = "darkred")) +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_rect(color = "black", fill = NA)
  )


svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/z-infant_weight.svg",
  width=6, height=2
  )
infant_weight
dev.off()


```
###infant height
```{r}
metadata_distinct <- metadata %>%
  distinct(idbs, .keep_all = TRUE)

metadata_height <- metadata_distinct %>%
  dplyr::select(idbs, code, c_height0, c_height1, c_height2, c_height3, c_height4, c_height5, c_height6)

# Transform the data from wide to long format
metadata_long <- metadata_height %>%
  pivot_longer(
    cols = starts_with("c_height"),
    names_to = "time",
    values_to = "height"
  )

# Fit the mixed-effects linear regression model
library(lmerTest)
model <- lmer(height ~ time * code + (1 | idbs), data = metadata_long)
summary(model)

# Convert 'time' to a numeric variable for plotting
metadata_long <- metadata_long %>%
  mutate(time = as.numeric(gsub("c_height", "", time)))

# Summarize the data
metadata_summary <- metadata_long %>%
  group_by(code, time) %>%
  dplyr::summarize(
    mean_height = mean(height, na.rm = TRUE),
    sd_height = sd(height, na.rm = TRUE),
    .groups = 'drop'
  ) %>%
  mutate(code=if_else(code=="B Z", "Intervention", "Control"))

# Create the line plot
infant_height <- ggplot(metadata_summary, aes(x = time, y = mean_height, color = code, group = code)) +
  geom_line(size = 0.5) +
  geom_point(size = 1) +
  geom_errorbar(aes(ymin = mean_height - sd_height, ymax = mean_height + sd_height), width = 0.2) +
  labs(x = "Months", y = "Height, cm", color = "Group") +
  scale_color_manual(values = c("Intervention" = "steelblue", "Control" = "darkred")) +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_rect(color = "black", fill = NA)
  )


svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/z-infant_height.svg",
  width=6, height=2
  )
infant_height
dev.off()


```


###infant waz
```{r}
metadata_distinct <- metadata %>%
  distinct(idbs, .keep_all = TRUE)

metadata_waz <- metadata_distinct %>%
  dplyr::select(idbs, code, c_waz0, c_waz1, c_waz2, c_waz3, c_waz4, c_waz5, c_waz6)

# Transform the data from wide to long format
metadata_long <- metadata_waz %>%
  pivot_longer(
    cols = starts_with("c_waz"),
    names_to = "time",
    values_to = "waz"
  )

# Fit the mixed-effects linear regression model
library(lmerTest)
model <- lmer(waz ~ time * code + (1 | idbs), data = metadata_long)
summary(model)

# Convert 'time' to a numeric variable for plotting
metadata_long <- metadata_long %>%
  mutate(time = as.numeric(gsub("c_waz", "", time)))

# Summarize the data
metadata_summary <- metadata_long %>%
  group_by(code, time) %>%
  dplyr::summarize(
    mean_waz = mean(waz, na.rm = TRUE),
    sd_waz = sd(waz, na.rm = TRUE),
    .groups = 'drop'
  ) %>%
  mutate(code=if_else(code=="B Z", "Intervention", "Control"))

# Create the line plot
infant_waz <- ggplot(metadata_summary, aes(x = time, y = mean_waz, color = code, group = code)) +
  geom_line(size = 0.5) +
  geom_point(size = 1) +
  geom_errorbar(aes(ymin = mean_waz - sd_waz, ymax = mean_waz + sd_waz), width = 0.2) +
  labs(x = "Months", y = "WAZ", color = "Group") +
  scale_color_manual(values = c("Intervention" = "steelblue", "Control" = "darkred")) +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_rect(color = "black", fill = NA)
  )


svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/z-infant_waz.svg",
  width=6, height=2
  )
infant_waz
dev.off()


```


###infant haz
```{r}
metadata_distinct <- metadata %>%
  distinct(idbs, .keep_all = TRUE)

metadata_haz <- metadata_distinct %>%
  dplyr::select(idbs, code, c_haz0, c_haz1, c_haz2, c_haz3, c_haz4, c_haz5, c_haz6)

# Fit the mixed-effects linear regression model
library(lmerTest)
model <- lmer(haz ~ time * code + (1 | idbs), data = metadata_long)
summary(model)

# Transform the data from wide to long format
metadata_long <- metadata_haz %>%
  pivot_longer(
    cols = starts_with("c_haz"),
    names_to = "time",
    values_to = "haz"
  )

# Convert 'time' to a numeric variable for plotting
metadata_long <- metadata_long %>%
  mutate(time = as.numeric(gsub("c_haz", "", time)))

# Summarize the data
metadata_summary <- metadata_long %>%
  group_by(code, time) %>%
  dplyr::summarize(
    mean_haz = mean(haz, na.rm = TRUE),
    sd_haz = sd(haz, na.rm = TRUE),
    .groups = 'drop'
  ) %>%
  mutate(code=if_else(code=="B Z", "Intervention", "Control"))

# Create the line plot
infant_haz <- ggplot(metadata_summary, aes(x = time, y = mean_haz, color = code, group = code)) +
  geom_line(size = 0.5) +
  geom_point(size = 1) +
  geom_errorbar(aes(ymin = mean_haz - sd_haz, ymax = mean_haz + sd_haz), width = 0.2) +
  labs(x = "Months", y = "HAZ", color = "Group") +
  scale_color_manual(values = c("Intervention" = "steelblue", "Control" = "darkred")) +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_rect(color = "black", fill = NA)
  )


svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/z-infant_haz.svg",
  width=6, height=2
  )
infant_haz
dev.off()


```


###infant whz
```{r}
metadata_distinct <- metadata %>%
  distinct(idbs, .keep_all = TRUE)

metadata_whz <- metadata_distinct %>%
  dplyr::select(idbs, code, c_whz0, c_whz1, c_whz2, c_whz3, c_whz4, c_whz5, c_whz6)

# Transform the data from wide to long format
metadata_long <- metadata_whz %>%
  pivot_longer(
    cols = starts_with("c_whz"),
    names_to = "time",
    values_to = "whz"
  )

# Fit the mixed-effects linear regression model
library(lmerTest)
model <- lmer(whz ~ time * code + (1 | idbs), data = metadata_long)
summary(model)

# Convert 'time' to a numeric variable for plotting
metadata_long <- metadata_long %>%
  mutate(time = as.numeric(gsub("c_whz", "", time)))

# Summarize the data
metadata_summary <- metadata_long %>%
  group_by(code, time) %>%
  dplyr::summarize(
    mean_whz = mean(whz, na.rm = TRUE),
    sd_whz = sd(whz, na.rm = TRUE),
    .groups = 'drop'
  ) %>%
  mutate(code=if_else(code=="B Z", "Intervention", "Control"))

# Create the line plot
infant_whz <- ggplot(metadata_summary, aes(x = time, y = mean_whz, color = code, group = code)) +
  geom_line(size = 0.5) +
  geom_point(size = 1) +
  geom_errorbar(aes(ymin = mean_whz - sd_whz, ymax = mean_whz + sd_whz), width = 0.2) +
  labs(x = "Months", y = "WHZ", color = "Group") +
  scale_color_manual(values = c("Intervention" = "steelblue", "Control" = "darkred")) +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_rect(color = "black", fill = NA)
  )


svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/z-infant_whz.svg",
  width=6, height=2
  )
infant_whz
dev.off()


```



# Figure 2. 
## A. Relative abundance genus

```{r message = FALSE}

genus_name <- c("Prevotella",  "Bifidobacterium", "Escherichia", "RC9", "Bacteroides", "CAG-83", "ER4", "Faecalibacterium",  "Prevotellamassilia", "CAG-170", "Veillonella_A", "C941", "Klebsiella", "Dialister", "Agathobacter", "Other")
genus_color <- c("#2986cc", "#cd0000", "#f1c232", "#6aa84f",  "#228585","#f4cccc", "#00ffff","#e69138", "#cfe2f3", "#d0f679", "orchid4", "salmon", "thistle", "#104E8E", "#8b7355",  "gray90")

ps_genus <- ps %>%
  tax_glom(taxrank = "genus") %>%
  transform_sample_counts(function(x){x/sum(x)}) %>%
  psmelt() %>%
  filter(Abundance >0)

sort.genus.ps <- ps_genus %>%
  dplyr::count(genus, wt = Abundance) %>%
  arrange(desc(n)) %>%
  slice_head(n = 15) %>%
  pull(genus)

sample.order.ps <- ps_genus %>%
  filter(phylum == sort.genus.ps[1]) %>%
  arrange (desc(Abundance)) %>%
  pull(Sample)

ps_genus_plot <- ps_genus %>%
  mutate(Sample = factor(Sample, levels = sample.order.ps)) %>%
  mutate(genus = factor(genus, levels = rev(sort.genus.ps))) %>%
  ggplot(aes(x=bep_group, y=Abundance)) + 
  geom_bar(aes(fill=genus), stat = 'identity', position = 'fill') + 
  theme(axis.text.x = element_text(angle = 90, size = 7), legend.title = element_text(size=8), legend.key.size = unit(0.3, 'cm'), legend.text = element_text(size = 8)) + 
  labs(x="Sample at different timepoints and treatment group", y="Abundance", fill = "Genera") + 
  scale_fill_manual(breaks=genus_name, values = genus_color) + 
  theme(panel.grid = element_blank(), panel.border = element_blank()) +   
  theme(panel.background = element_blank(),
        axis.ticks.x=element_blank())+ 
  guides(fill=guide_legend(ncol = 1))
ps_genus_plot


svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_revision1/ps_genus_plot.svg",
  width=6, height=5
  )
ps_genus_plot
dev.off()

```


## B. species counts

#### infant

```{r}

infant_species <- infant_ps %>%
  tax_glom(taxrank = "species") %>%
  transform_sample_counts(function(x) x / sum(x)) %>%
  psmelt() %>%
  filter(Abundance > 0)


# Count the number of unique species for each subject
species_counts_BEP <- infant_species %>%
  mutate(timepoint_n = ifelse(timepoint=="pn12", 0, 1)) %>%
  filter(code == "B Z") %>%
  group_by(idbs, timepoint) %>%
  summarise(num_species = n_distinct(OTU),
            timepoint_n = first(timepoint_n))

infantSpeciesNumberBEP <- ggplot(species_counts_BEP, aes(x = num_species, fill = factor(timepoint_n))) +
  geom_density(alpha = 0.5) +
  labs(x = "Number of Species", y = "Density", fill = "Timepoint", title = "Intervention Group") +
  scale_fill_manual(values = c("skyblue", "#f1c232"), labels = c("Pn12", "Pn56")) +
  theme_minimal() +
  theme(panel.border = element_rect(color = "black", fill = NA),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) + xlim (2, 37)
infantSpeciesNumberBEP

# Count the number of unique species for each subject
species_counts_IFA <- infant_species %>%
  mutate(timepoint_n = ifelse(timepoint=="pn12", 0, 1)) %>%
  filter(code == "A Y") %>%
  group_by(idbs, timepoint) %>%
  summarise(num_species = n_distinct(OTU),
            timepoint_n = first(timepoint_n))

infantSpeciesNumberIFA <- ggplot(species_counts_IFA, aes(x = num_species, fill = factor(timepoint_n))) +
  geom_density(alpha = 0.5) +
  labs(x = "Number of Species", y = "Density", fill = "Timepoint", title = "Control Group") +
  scale_fill_manual(values = c("skyblue", "#f1c232"), labels = c("Pn12", "Pn56")) +
  theme_minimal() +
  theme(panel.border = element_rect(color = "black", fill = NA),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) + xlim (2, 37)
infantSpeciesNumberIFA


svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/infantSpeciesNumberBEP.svg",
  width=4, height=3
  )
infantSpeciesNumberBEP
dev.off()

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/infantSpeciesNumberIFA.svg",
  width=4, height=3
  )
infantSpeciesNumberIFA
dev.off()


```


#### mother

```{r}

mother_species <- mother_ps %>%
  tax_glom(taxrank = "species") %>%
  transform_sample_counts(function(x) x / sum(x)) %>%
  psmelt() %>%
  filter(Abundance > 0)


# Count the number of unique species for each subject
species_counts_BEP <- mother_species %>%
  mutate(timepoint_n = case_when(
    timepoint == "incl" ~ 0,
    timepoint == "tri3" ~ 1,
    timepoint == "pn12" ~ 3,
    timepoint == "pn56" ~ 4,
    TRUE ~ NA_real_  # Assign NA for any other values or missing values
  )) %>%  
  filter(code == "B Z") %>%
  group_by(idbs, timepoint) %>%
  summarise(num_species = n_distinct(OTU),
            timepoint_n = first(timepoint_n))

motherSpeciesNumberBEP <- ggplot(species_counts_BEP, aes(x = num_species, fill = factor(timepoint_n))) +
  geom_density(alpha = 0.5) +
  labs(x = "Number of Species", y = "Density", fill = "Timepoint", title = "Intervention Group") +
  scale_fill_manual(values = c("#00A98F", "#C00000", "skyblue", "#f1c232"), labels = c("Tri2", "Tri3", "Pn12", "Pn56")) +
  theme_minimal() +
  theme(panel.border = element_rect(color = "black", fill = NA),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) + xlim(66, 460)
motherSpeciesNumberBEP

# Count the number of unique species for each subject
species_counts_IFA <- mother_species %>%
  mutate(timepoint_n = case_when(
    timepoint == "incl" ~ 0,
    timepoint == "tri3" ~ 1,
    timepoint == "pn12" ~ 3,
    timepoint == "pn56" ~ 4,
    TRUE ~ NA_real_  # Assign NA for any other values or missing values
  )) %>%  
  filter(code == "A Y") %>%
  group_by(idbs, timepoint) %>%
  summarise(num_species = n_distinct(OTU),
            timepoint_n = first(timepoint_n))

motherSpeciesNumberIFA <- ggplot(species_counts_IFA, aes(x = num_species, fill = factor(timepoint_n))) +
  geom_density(alpha = 0.5) +
  labs(x = "Number of Species", y = "Density", fill = "Timepoint", title = "Control Group") +
  scale_fill_manual(values = c("#00A98F", "#C00000", "skyblue", "#f1c232"), labels = c("Tri2", "Tri3", "Pn12", "Pn56")) +
  theme_minimal() +
  theme(panel.border = element_rect(color = "black", fill = NA),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) + xlim(66, 460)
motherSpeciesNumberIFA

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/motherSpeciesNumberBEP.svg",
  width=5, height=3
  )
motherSpeciesNumberBEP
dev.off()

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/motherSpeciesNumberIFA.svg",
  width=5, height=3
  )
motherSpeciesNumberIFA
dev.off()

```


## C. Alpha diversity
Alpha diversity metrics are single number summaries of the diversity of microbial communities within a single sample. These are based on the **richness** (number of different taxa or species present) and/or **evenness** (describes how evenly the abundance of each species is distributed) within the sample. We describe
**Observed**: the simplest richness index
**Shannon**: measures the evenness
**Simpson**: takes into account the richness and evenness

### color scheme
```{r}
COLOR_SCHEME <- c(
  "#00A98F",
  "#FFC000",
  "#1565A9",
  "#C00000",
  "#0D0D0D",
  "#6aa84f",
  "tomato4",
  "#8b7355"
)


```


### observed, shannon, simpson
```{r calc diversity}
# calculate richness
alpha_div <- phyloseq::estimate_richness(
  ps,
  split = TRUE,
  measures = c("Observed", "Shannon", "Simpson")
  )

# merge with metadata to split on groups
alpha_div <- merge(
  x=alpha_div, 
  y=metadata[, c("idbs", "child", "timepoint", "code")],
  by=0
  )
DT::datatable(alpha_div) 

group <- c()
for (i in 1:nrow(alpha_div)){
  if (alpha_div[i, "child"] == "yes"){
    child_str <- "Infant"
  } else {
    child_str <- "Mother"
  }
  if (alpha_div[i, "code"] == "B Z"){
    trt_str <- "BEP"
  } else {
    trt_str <- "IFA"
  }
  # add group
  group_str <- paste(child_str, trt_str, alpha_div[i, "timepoint"], sep = " ")
  alpha_div[i, "Group"] <- group_str
  # add tooltips for interactive plots
  text_str <- paste0(
    "mother ID: ", alpha_div[i, "idbs"], "\n", 
    "timepoint of sample: ", alpha_div[i, "timepoint"], "\n",
    "treatment group: ", trt_str, "\n"
    )
  if (alpha_div[i, "child"] == "yes"){
    text_str <- paste0("infant of ", text_str)
  }
  alpha_div[i, "Text"] <- text_str
}

# set as factor to preserve order
alpha_div$Group <- factor(
  alpha_div$Group, 
  levels=c(
    "Mother BEP incl",
    "Mother IFA incl",
    "Mother BEP tri3",
    "Mother IFA tri3",
    "Mother BEP pn12",
    "Mother IFA pn12",
    "Mother BEP pn56",
    "Mother IFA pn56",
    "Infant BEP pn12",
    "Infant IFA pn12",
    "Infant BEP pn56", 
    "Infant IFA pn56"
    )
  )

 alpha_div <- alpha_div %>%
    mutate(code = ifelse(code == "B Z", "BEP", "IFA"))
```


#### visual shannon
```{r create shannon diversity plot, echo = FALSE}
# set up tooltip for interactive figure
alpha_div <- alpha_div %>%
  dplyr::mutate(
    sha_tooltip = paste0(
      Text,
      "Shannon diversity index: ", round(Shannon, digits=4), "\n"
      )
    )

alpha_shannon <- ggplot2::ggplot(data = alpha_div) +
  ggplot2::geom_jitter(
    size = 1, 
    alpha = 0.5,
    ggplot2::aes(x=Group, y=Shannon, color=code, shape=timepoint, text=sha_tooltip)
    ) +
  ggplot2::theme_bw() +
  #ggplot2::coord_flip() +
  # Facet grid with free x-axis scales and free space
  facet_grid(. ~ child, scales = "free_x", space = "free", labeller = labeller(child = facet_labels)) +
  ggplot2::scale_color_manual(
    name="code", 
    labels=c("Intervention", "Control"),
    values = c("steelblue", "darkred")) +
  ggplot2::scale_shape_manual(
    name = "timepoint", 
    labels = c("Tri2", "Pn12", "Pn56", "Tri3"), 
    values=c(5, 16, 17, 3)
    ) +
  ggplot2::geom_boxplot(
    data = alpha_div,
    ggplot2::aes(x = Group, y = Shannon, color = code),
    alpha = 0.1
    ) +
  ggplot2::labs(
    title = "",
    x = "",
    y = "Shannon diversity index"
   ) +
  ggplot2::theme(
    plot.title = ggplot2::element_text(face = "bold"),
    axis.text.x = ggplot2::element_text(vjust = 0.4, hjust = 1),
    axis.text.y = ggplot2::element_text(vjust = 0.4, hjust = 1),
    legend.title = ggplot2::element_blank(),
    legend.position = "right",
    text = ggplot2::element_text(size = 10)
    )

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/alpha_shannon2.svg",
  width=5.6, height=4
  )
alpha_shannon
dev.off()

#ggsave("alpha_shannon.png", width = 5, height = 5, units = "in", bg = "white")

```

#### Wilcoxon rank sum test

```{r}

wilcox_result <- wilcox.test(Shannon ~ prePost, data = alpha_div2)
wilcox_result

```


### TMLE
Targeted Maximum Likelihood Estimation (TMLE) is a general approach for constructing an effectient double-robust semi-parametric substitution estimator of a causal effect parameter or statistical association measure.

TMLE estimate of effect of BEP supplementation on the microbial diversity in the mother prenatally (across incl and tri3) and postnally (across pn12 and pn56) and in infant (across pn12 and pn56).
It tries to estimate the average outcome (Shannon diversity of the gut microbiome). 
**Additive Effect among the Treated**: predicts that for if every women that actually received the treatment.
**Additive Effect among the Contrils**: predicts for if the women in the control group had received the additive.
**Additive Effect**: summarizes the expected increase in Shannon diversity between the two treatments.

```{r}
row.names(alpha_div) <- alpha_div$Row.names
alpha_div$Row.names <- NULL
# merge with metadata to split on groups
alpha_div_extended <- merge(
  x=alpha_div, 
  y=metadata[, c("treatment1_prenatal_n", "m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs",  "csps_n", "w_age", "m_bmiincl", "mddw_10_ave", "ch_sex", "iycf_ebf_age")],
  by=0
  )

alpha_div_extended$Row.names <- NULL
alpha_div_extended <- alpha_div_extended[complete.cases(alpha_div_extended), ]
alpha_div_extended$timepoint <- as.factor(alpha_div_extended$timepoint)
alpha_div_extended$code <- ifelse(alpha_div_extended$code == "BEP", 1, 0)

alpha_div_extended_mat <- alpha_div_extended[alpha_div_extended$child =="no", ]
alpha_div_extended_mat_pre <- alpha_div_extended[alpha_div_extended_mat$timepoint %in% c("incl", "tri3"), ]
alpha_div_extended_mat_post <- alpha_div_extended[alpha_div_extended_mat$timepoint %in% c("pn12", "pn56"), ]

alpha_div_extended_mat_incl <- alpha_div_extended[alpha_div_extended_mat$timepoint == "incl", ]
alpha_div_extended_mat_tri3 <- alpha_div_extended[alpha_div_extended_mat$timepoint == "tri3", ]
alpha_div_extended_mat_pn12 <- alpha_div_extended[alpha_div_extended_mat$timepoint == "pn12", ]
alpha_div_extended_mat_pn56 <- alpha_div_extended[alpha_div_extended_mat$timepoint == "pn56", ]
alpha_div_extended_inf <- alpha_div_extended[alpha_div_extended$child =="yes", ]
alpha_div_extended_inf_pn12 <- alpha_div_extended[alpha_div_extended_inf$timepoint == "pn12", ]
alpha_div_extended_inf_pn56 <- alpha_div_extended[alpha_div_extended_inf$timepoint == "pn56", ]

```


#### incl TMLE
```{r}
set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm', "SL.randomForest", "SL.step","SL.glm.interaction")

# Shannon
TMLE_maternal_incl <- tmle::tmle(
  Y=alpha_div_extended_mat_incl$Shannon, 
  A=alpha_div_extended_mat_incl$code, # must be numeric for the package used
  W=alpha_div_extended_mat_incl[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_maternal_incl)

alpha_div_extended_mat_incl_model <- nlme::lme(
  fixed = Shannon ~ code, 
  data = alpha_div_extended_mat_incl,
  random = ~ 1 | idbs
  )
summary(alpha_div_extended_mat_incl_model)

```


#### tri3 TMLE
```{r}
set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm', "SL.randomForest", "SL.step","SL.glm.interaction")

# Shannon
TMLE_maternal_tri3 <- tmle::tmle(
  Y=alpha_div_extended_mat_tri3$Shannon, 
  A=alpha_div_extended_mat_tri3$code, # must be numeric for the package used
  W=alpha_div_extended_mat_tri3[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_maternal_tri3)

alpha_div_extended_mat_tri3_model <- nlme::lme(
  fixed = Shannon ~ code, 
  data = alpha_div_extended_mat_tri3,
  random = ~ 1 | idbs
  )
summary(alpha_div_extended_mat_tri3_model)

```


#### mat pn12 TMLE
```{r}
set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm')

# Shannon
TMLE_maternal_pn12 <- tmle::tmle(
  Y=alpha_div_extended_mat_pn12$Shannon, 
  A=alpha_div_extended_mat_pn12$code, # must be numeric for the package used
  W=alpha_div_extended_mat_pn12[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_maternal_pn12)

```


#### mat pn56 TMLE
```{r}
set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm')

# Shannon
TMLE_maternal_pn56 <- tmle::tmle(
  Y=alpha_div_extended_mat_pn56$Shannon, 
  A=alpha_div_extended_mat_pn56$code, # must be numeric for the package used
  W=alpha_div_extended_mat_pn56[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_maternal_pn56)

```


#### inf pn12 TMLE
```{r}
set.seed(123)
# Shannon
TMLE_infant_pn12 <- tmle::tmle(
  Y=alpha_div_extended_inf_pn12$Shannon, 
  A=alpha_div_extended_inf_pn12$code, # must be numeric for the package used
  W=alpha_div_extended_inf_pn12[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave", "ch_sex", "iycf_ebf_age")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_infant_pn12)

```

#### inf pn56 TMLE
```{r}
set.seed(123)
# Shannon
TMLE_infant_pn56 <- tmle::tmle(
  Y=alpha_div_extended_inf_pn56$Shannon, 
  A=alpha_div_extended_inf_pn56$code, # must be numeric for the package used
  W=alpha_div_extended_inf_pn56[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave", "ch_sex", "iycf_ebf_age")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_infant_pn56)

```

#### TMLE prenatal
```{r}
set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm', "SL.randomForest", "SL.step","SL.glm.interaction")

TMLE_maternal_pre <- tmle::tmle(
  Y=alpha_div_extended_mat_pre$Shannon, 
  A=alpha_div_extended_mat_pre$treatment1_prenatal_n, # must be numeric for the package used
  W=alpha_div_extended_mat_pre[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave", "timepoint")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_maternal_pre)
```


#### TMLE  postnatal
```{r}
set.seed(123)
TMLE_maternal_post <- tmle::tmle(
  Y=alpha_div_extended_mat_post$Shannon, 
  A=alpha_div_extended_mat_post$treatment1_prenatal_n, # must be numeric for the package used
  W=alpha_div_extended_mat_post[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave", "timepoint")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_maternal_post)
```


#### TMLE infant
```{r}
set.seed(123)

TMLE_infant <- tmle::tmle(
  Y=alpha_div_extended_inf$Shannon, 
  A=alpha_div_extended_inf$treatment1_prenatal_n, # must be numeric for the package used
  W=alpha_div_extended_inf[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave", "ch_sex", "iycf_ebf_age", "timepoint")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_infant)
```



## D. Beta diversity

### infant 4 groups
```{r message=FALSE}
# Calculate the UniFrac distance matrix
unifrac_dist <- phyloseq::distance(infant, method="unifrac", weighted=TRUE)

# Perform PCoA ordination using the calculated distance matrix
infant_ord <- ordinate(infant, method="PCoA", distance=unifrac_dist)
phyloseq::plot_scree(infant_ord) + geom_bar(stat = "identity", fill="blue") + labs(x="\nAxis", y="Proportion of Variance\n")
head(infant_ord$values$Eigenvalues)

infant_eig1 <- infant_ord$values$Eigenvalues[1]/sum(infant_ord$values$Eigenvalues)
infant_eig2 <- infant_ord$values$Eigenvalues[2]/sum(infant_ord$values$Eigenvalues)
beta_infant <- phyloseq::plot_ordination(infant, infant_ord, color = "bep_group", shape = "timepoint") + 
  scale_color_manual(
    name="bep_group", 
    values = COLOR_SCHEME[1:4]
    ) +
  ggplot2::scale_shape_manual(
    name = "timepoint",
    labels = c("pn12", "pn56"),
    values=c(17, 3)) +
  geom_point(size=1) +
  #coord_fixed(infant_eig1/infant_eig2) +
  stat_ellipse(aes(group=bep_group), linetype=2) +
  theme_bw() +
  theme(legend.position = "bottom") +
  xlim(-0.22, 0.22) +
  ylim(-0.12, 0.12)
beta_infant

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/beta_infant.svg",
  width=4, height=6
  )
beta_infant
dev.off()

infant_dist <-UniFrac(infant, weighted = TRUE, normalized = TRUE, parallel = FALSE, fast=TRUE)
vegan::adonis2(infant_dist ~ phyloseq::sample_data(infant)$code + phyloseq::sample_data(infant)$timepoint )
```


### maternal 8 groups
```{r message=FALSE}

mother_ord <- phyloseq::ordinate(mother, "PCoA", "unifrac", weighted=T)
phyloseq::plot_scree(mother_ord) + geom_bar(stat = "identity", fill="blue") + labs(x="\nAxis", y="Proportion of Variance\n")
head(mother_ord$values$Eigenvalues)

mother_eig1 <- mother_ord$values$Eigenvalues[1]/sum(mother_ord$values$Eigenvalues)
mother_eig2 <- mother_ord$values$Eigenvalues[2]/sum(mother_ord$values$Eigenvalues)
beta_mother <- phyloseq::plot_ordination(mother, mother_ord, color = "bep_group", shape = "timepoint") + 
scale_color_manual(
    name="bep_group", 
    values = COLOR_SCHEME[1:8]
    ) +
  ggplot2::scale_shape_manual(
    name = "timepoint",
    labels = c("incl", "tri3", "pn12", "pn56"),
    values=c(5, 16, 17, 3)) +
  geom_point(size=1) +
  #coord_fixed(mother_eig1/mother_eig2) +
  stat_ellipse(aes(group=bep_group), linetype=2)+
  theme_bw() +
  theme(legend.position = "bottom") +
  xlim(-0.22, 0.22) +
  ylim(-0.12, 0.12) +
  guides(color = guide_legend(nrow = 1), shape = guide_legend(nrow = 1))
beta_mother
#ggsave("beta_mother.png", width = 5, height = 5, units = "in", bg = "white")


svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/beta_mother.svg",
  width=4, height=6
  )
beta_mother
dev.off()


mother_dist <-UniFrac(mother, weighted = T, normalized = TRUE, parallel = FALSE, fast=TRUE)
vegan::adonis2(mother_dist ~ phyloseq::sample_data(mother)$code + phyloseq::sample_data(mother)$timepoint)

```


## E. Gene richness

```{r}

# Load necessary libraries
library(tidyr)
library(dplyr)

# Assuming your dataset is a data frame named 'data' with rownames as gene names
# Convert rownames to a column for easier manipulation
data_long <- gene_counts %>%
  rownames_to_column(var = "gene") %>%
  pivot_longer(-gene, names_to = "sample", values_to = "count")

# Extract timepoint, sample ID, and other relevant information from the sample column if needed
# Assuming sample names are of the format: timepoint_se1g_idbs_e
data_long <- data_long %>%
  separate(sample, into = c("timepoint", "se1g", "idbs", "e"), sep = "_", remove = FALSE) %>%
  mutate(child = ifelse(is.na(e), "Mother", "Infant"))
metadata$idbs <- as.character(metadata$idbs)
# Calculate gene richness (number of genes with count > 0) for each sample
gene_richness <- data_long %>%
  left_join(metadata %>% dplyr::select(idbs, code), by = "idbs") %>%
  group_by(idbs, code, timepoint, child) %>%
  dplyr::summarize(richness = sum(count > 0)) %>%
  mutate(Group = ifelse(timepoint == "incl" & code == "A Y", "Mother IFA incl",
                     ifelse(timepoint == "incl" & code == "B Z", "Mother BEP incl",
                     ifelse(timepoint == "tri3" & code == "A Y", "Mother IFA tri3",
                     ifelse(timepoint == "tri3" & code == "B Z", "Mother BEP tri3",
                     ifelse(timepoint == "pn12" & child == "Mother" & code == "A Y", "Mother IFA pn12",
                     ifelse(timepoint == "pn12" & child == "Mother" & code == "B Z", "Mother BEP pn12",
                     ifelse(timepoint == "pn12" & child == "Infant" & code == "A Y", "Infant IFA pn12",
                     ifelse(timepoint == "pn12" & child == "Infant" & code == "B Z", "Infant BEP pn12",
                     ifelse(timepoint == "pn56" & child == "Mother" & code == "A Y", "Mother IFA pn56",
                     ifelse(timepoint == "pn56" & child == "Mother" & code == "B Z", "Mother BEP pn56",
                     ifelse(timepoint == "pn56" & child == "Infant" & code == "A Y", "Infant IFA pn56",
                     ifelse(timepoint == "pn56" & child == "Infant" & code == "B Z", "Infant BEP pn56", "other"))))))))))))) %>%
  mutate(Group = factor(Group, levels = c(
    "Mother BEP incl",
    "Mother IFA incl",
    "Mother BEP tri3",
    "Mother IFA tri3",
    "Mother BEP pn12",
    "Mother IFA pn12",
    "Mother BEP pn56",
    "Mother IFA pn56",
    "Infant BEP pn12",
    "Infant IFA pn12",
    "Infant BEP pn56",
    "Infant IFA pn56"
  ))) %>%
  mutate(code = ifelse(code == "B Z", "Intervention", "Control")) %>%
  mutate(code = factor(code, levels = c("Intervention", "Control"))) %>%
  mutate(child = factor(child, levels = c("Mother", "Infant")))

# Generate the plot
gene_richness_plot <- ggplot2::ggplot(data = gene_richness) +
  ggplot2::geom_jitter(
    size = 1, 
    alpha = 0.5,
    ggplot2::aes(x = Group, y = richness, color = code, shape = timepoint)
  ) +
  ggplot2::theme_bw() +
  # Facet grid with free x-axis scales and free space
  facet_grid(. ~ child, scales = "free_x", space = "free") +
  ggplot2::scale_color_manual(
    name = "code", 
    labels = c("Intervention", "Control"),
    values = c("steelblue", "darkred")
  ) +
  ggplot2::scale_shape_manual(
    name = "timepoint", 
    labels = c("Tri2", "Pn12", "Pn56", "Tri3"), 
    values = c(5, 16, 17, 3)
  ) +
  ggplot2::geom_boxplot(
    data = gene_richness,
    ggplot2::aes(x = Group, y = richness, color = code),
    alpha = 0.1
  ) +
  ggplot2::labs(
    title = "",
    x = "",
    y = "Gene richness"
  ) +
  ggplot2::theme(
    plot.title = ggplot2::element_text(face = "bold"),
    axis.text.x = ggplot2::element_text(vjust = 0.4, hjust = 1),
    axis.text.y = ggplot2::element_text(vjust = 0.4, hjust = 1),
    legend.title = ggplot2::element_blank(),
    legend.position = "right",
    text = ggplot2::element_text(size = 10)
  )
gene_richness_plot

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/gene_richness2.svg",
  width=5.6, height=4
  )
gene_richness_plot
dev.off()

```

### gene TMLE incl

```{r}
metadata$idbs <- as.character(metadata$idbs)
# merge with metadata to split on groups
gene_richness_extended <- gene_richness %>%
  left_join(
    metadata %>%
      group_by(idbs) %>%
      slice(1) %>%  # Keep only the first occurrence
      ungroup() %>%
      select(idbs, treatment1_prenatal_n, m_heightincl, m_hbincl, w_school_level, 
             w_marital, asset1comp_10, w_language, w_religion, w_etnicity, 
             w_nr_jobs, csps_n, w_age, m_bmiincl, mddw_10_ave, ch_sex, iycf_ebf_age),
    by = "idbs"
  )


gene_richness_incl <- gene_richness_extended %>%
  filter(timepoint == "incl") %>%
  drop_na(richness, treatment1_prenatal_n, m_heightincl, m_hbincl, w_school_level, 
          w_marital, asset1comp_10, w_language, w_religion, w_etnicity, 
          w_nr_jobs, csps_n, w_age, m_bmiincl, mddw_10_ave)

set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm', "SL.randomForest", "SL.step","SL.glm.interaction")

TMLE_gene_richness_incl <- tmle::tmle(
  Y=gene_richness_incl$richness, 
  A=gene_richness_incl$treatment1_prenatal_n, # must be numeric for the package used
  W=gene_richness_incl[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_gene_richness_incl)

```

### gene TMLE tri3

```{r}
gene_richness_tri3 <- gene_richness_extended %>%
  filter(timepoint == "tri3") %>%
  drop_na(richness, treatment1_prenatal_n, m_heightincl, m_hbincl, w_school_level, 
          w_marital, asset1comp_10, w_language, w_religion, w_etnicity, 
          w_nr_jobs, csps_n, w_age, m_bmiincl, mddw_10_ave)

set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm', "SL.randomForest", "SL.step","SL.glm.interaction")

TMLE_gene_richness_tri3 <- tmle::tmle(
  Y=gene_richness_tri3$richness, 
  A=gene_richness_tri3$treatment1_prenatal_n, # must be numeric for the package used
  W=gene_richness_tri3[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_gene_richness_tri3)

```


### gene TMLE mat pn12

```{r}
gene_richness_pn12m <- gene_richness_extended %>%
  filter(timepoint == "pn12" & child == "Mother") %>%
  drop_na(richness, treatment1_prenatal_n, m_heightincl, m_hbincl, w_school_level, 
          w_marital, asset1comp_10, w_language, w_religion, w_etnicity, 
          w_nr_jobs, csps_n, w_age, m_bmiincl, mddw_10_ave)

set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm', "SL.randomForest", "SL.step","SL.glm.interaction")

TMLE_gene_richness_pn12m <- tmle::tmle(
  Y=gene_richness_pn12m$richness, 
  A=gene_richness_pn12m$treatment1_prenatal_n, # must be numeric for the package used
  W=gene_richness_pn12m[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_gene_richness_pn12m)

```


### gene TMLE mat pn56

```{r}
gene_richness_pn56m <- gene_richness_extended %>%
  filter(timepoint == "pn56" & child == "Mother") %>%
  drop_na(richness, treatment1_prenatal_n, m_heightincl, m_hbincl, w_school_level, 
          w_marital, asset1comp_10, w_language, w_religion, w_etnicity, 
          w_nr_jobs, csps_n, w_age, m_bmiincl, mddw_10_ave)

set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm', "SL.randomForest", "SL.step","SL.glm.interaction")

TMLE_gene_richness_pn56m <- tmle::tmle(
  Y=gene_richness_pn56m$richness, 
  A=gene_richness_pn56m$treatment1_prenatal_n, # must be numeric for the package used
  W=gene_richness_pn56m[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_gene_richness_pn56m)

```


### gene TMLE inf pn12

```{r}
gene_richness_pn12e <- gene_richness_extended %>%
  filter(timepoint == "pn12" & child == "Infant") %>%
  drop_na(richness, treatment1_prenatal_n, m_heightincl, m_hbincl, w_school_level, 
          w_marital, asset1comp_10, w_language, w_religion, w_etnicity, 
          w_nr_jobs, csps_n, w_age, m_bmiincl, mddw_10_ave)

set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm', "SL.randomForest", "SL.step","SL.glm.interaction")

TMLE_gene_richness_pn12e <- tmle::tmle(
  Y=gene_richness_pn12e$richness, 
  A=gene_richness_pn12e$treatment1_prenatal_n, # must be numeric for the package used
  W=gene_richness_pn12e[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_gene_richness_pn12e)

```


### gene TMLE inf pn56

```{r}
gene_richness_pn56e <- gene_richness_extended %>%
  filter(timepoint == "pn12" & child == "Infant") %>%
  drop_na(richness, treatment1_prenatal_n, m_heightincl, m_hbincl, w_school_level, 
          w_marital, asset1comp_10, w_language, w_religion, w_etnicity, 
          w_nr_jobs, csps_n, w_age, m_bmiincl, mddw_10_ave)

set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm', "SL.randomForest", "SL.step","SL.glm.interaction")

TMLE_gene_richness_pn56e <- tmle::tmle(
  Y=gene_richness_pn56e$richness, 
  A=gene_richness_pn56e$treatment1_prenatal_n, # must be numeric for the package used
  W=gene_richness_pn56e[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_gene_richness_pn56e)

```


### gene TMLE prenatal

```{r}

gene_richness_prenatal <- gene_richness_extended %>%
  filter(timepoint == "incl" | timepoint == "tri3") %>%
  drop_na(richness, treatment1_prenatal_n, m_heightincl, m_hbincl, w_school_level, 
          w_marital, asset1comp_10, w_language, w_religion, w_etnicity, 
          w_nr_jobs, csps_n, w_age, m_bmiincl, mddw_10_ave)

set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm', "SL.randomForest", "SL.step","SL.glm.interaction")

TMLE_gene_richness_prenatal <- tmle::tmle(
  Y=gene_richness_prenatal$richness, 
  A=gene_richness_prenatal$treatment1_prenatal_n, # must be numeric for the package used
  W=gene_richness_prenatal[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_gene_richness_prenatal)

```


### gene TMLE postnatal

```{r}

gene_richness_postnatal <- gene_richness_extended %>%
  filter(timepoint == "pn12" | timepoint == "pn56" & child == "Mother") %>%
  drop_na(richness, treatment1_prenatal_n, m_heightincl, m_hbincl, w_school_level, 
          w_marital, asset1comp_10, w_language, w_religion, w_etnicity, 
          w_nr_jobs, csps_n, w_age, m_bmiincl, mddw_10_ave)

set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm', "SL.randomForest", "SL.step","SL.glm.interaction")

TMLE_gene_richness_postnatal <- tmle::tmle(
  Y=gene_richness_postnatal$richness, 
  A=gene_richness_postnatal$treatment1_prenatal_n, # must be numeric for the package used
  W=gene_richness_postnatal[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_gene_richness_postnatal)

```


### gene TMLE infant

```{r}

gene_richness_infant <- gene_richness_extended %>%
  filter(child == "Infant") %>%
  drop_na(richness, treatment1_prenatal_n, m_heightincl, m_hbincl, w_school_level, 
          w_marital, asset1comp_10, w_language, w_religion, w_etnicity, 
          w_nr_jobs, csps_n, w_age, m_bmiincl, mddw_10_ave)

set.seed(123)
sl_library <- c('SL.glmnet', 'SL.ranger', 'SL.earth', 'SL.glm', "SL.randomForest", "SL.step","SL.glm.interaction")

TMLE_gene_richness_infant <- tmle::tmle(
  Y=gene_richness_infant$richness, 
  A=gene_richness_infant$treatment1_prenatal_n, # must be numeric for the package used
  W=gene_richness_infant[, c("m_heightincl", "m_hbincl", "w_school_level", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave")], 
  Q.SL.library = sl_library,
  family="gaussian"
  )
summary(TMLE_gene_richness_infant)

```





## F. Gene beta diversity

```{r}

library(vegan)
library(dplyr)
library(tidyr)
library(ggplot2)

COLOR_SCHEME2 <- c(
  "#0D0D0D",
  "#6aa84f",
  "tomato4",
  "#8b7355",
  "#00A98F",
  "#FFC000",
  "#1565A9",
  "#C00000"
)

# Separate the data for Mother and Infant based on sample ID
mother_samples <- grep("_e$", colnames(gene_counts), invert = TRUE, value = TRUE)
infant_samples <- grep("_e$", colnames(gene_counts), value = TRUE)

# Filter the gene_counts matrix for Mother and Infant
gene_counts_mother <- gene_counts[, mother_samples]
gene_counts_infant <- gene_counts[, infant_samples]

# Remove the empty rows
gene_counts_mother <- gene_counts_mother[rowSums(gene_counts_mother != 0) > 0, ]
gene_counts_infant <- gene_counts_infant[rowSums(gene_counts_infant != 0) > 0, ]

# ----------------------- Normalization and Transformation -----------------------
# Normalize with Total Sum Scaling (TSS)
gene_counts_mother_tss <- sweep(gene_counts_mother, 2, colSums(gene_counts_mother), FUN = "/")
gene_counts_infant_tss <- sweep(gene_counts_infant, 2, colSums(gene_counts_infant), FUN = "/")

# Log transform to reduce skewness
gene_counts_mother_log <- log1p(gene_counts_mother_tss)  # log1p(x) = log(1 + x)
gene_counts_infant_log <- log1p(gene_counts_infant_tss)

# ----------------------- PCoA for Mother -----------------------
# Transpose the data to have samples as rows and genes as columns
data_t_mother <- t(gene_counts_mother_log)

# Calculate the Bray-Curtis distance matrix
bray_dist_mother <- vegdist(data_t_mother, method = "bray")

# Perform PCoA on the Bray-Curtis distance matrix
pcoa_result_mother <- cmdscale(bray_dist_mother, eig = TRUE, k = 2)

# Extract the coordinates for the first two principal coordinates
pcoa_coords_mother <- as.data.frame(pcoa_result_mother$points) %>%
  mutate(PC1 = V1) %>%
  mutate(PC2 = V2) %>%
  rownames_to_column("sample_id") %>%
  separate(sample_id, into = c("timepoint", "se1g", "idbs", "e"), sep = "_") %>%
  mutate(child = ifelse(is.na(e), "Mother", "Infant")) %>%
  left_join(metadata %>% dplyr::select(idbs, code), by = "idbs") %>%
  mutate(Group = case_when(
    timepoint == "incl" & code == "A Y" ~ "Mother IFA incl",
    timepoint == "incl" & code == "B Z" ~ "Mother BEP incl",
    timepoint == "tri3" & code == "A Y" ~ "Mother IFA tri3",
    timepoint == "tri3" & code == "B Z" ~ "Mother BEP tri3",
    timepoint == "pn12" & child == "Mother" & code == "A Y" ~ "Mother IFA pn12",
    timepoint == "pn12" & child == "Mother" & code == "B Z" ~ "Mother BEP pn12",
    timepoint == "pn56" & child == "Mother" & code == "A Y" ~ "Mother IFA pn56",
    timepoint == "pn56" & child == "Mother" & code == "B Z" ~ "Mother BEP pn56",
    TRUE ~ "other"
  )) %>%
  mutate(Group = factor(Group, levels = c(
    "Mother BEP incl",
    "Mother IFA incl",
    "Mother BEP tri3",
    "Mother IFA tri3",
    "Mother BEP pn12",
    "Mother IFA pn12",
    "Mother BEP pn56",
    "Mother IFA pn56"
  )))

# Compute the proportion of variance explained by PC1 and PC2
variance_explained_PC1 <- pcoa_result_mother$eig[1] / sum(pcoa_result_mother$eig) * 100
variance_explained_PC2 <- pcoa_result_mother$eig[2] / sum(pcoa_result_mother$eig) * 100

# Print the variance explained
cat("Variance explained by PC1:", round(variance_explained_PC1, 2), "%\n")
cat("Variance explained by PC2:", round(variance_explained_PC2, 2), "%\n")

# Plot for Mother
gene_beta_mother <- ggplot(pcoa_coords_mother, aes(x = PC1, y = PC2, color = Group, shape = timepoint)) +
  geom_point(size = 1) +
  stat_ellipse(aes(group = Group), linetype = 2) +
  scale_color_manual(
    name = "Group", 
    values = COLOR_SCHEME2[1:8]  # Adjust these colors to fit your needs
  ) +
  scale_shape_manual(
    name = "Timepoint",
    labels = c("incl", "tri3", "pn12", "pn56"),
    values = c(5, 16, 17, 3)
  ) +
  theme_bw() +
  theme(legend.position = "right") +
  guides(color = guide_legend(ncol = 1), shape = guide_legend(ncol = 1)) +
  labs(
    title = "PCoA Plot for Mother",
    x = "PC1",
    y = "PC2"
  ) 

gene_beta_mother

svglite::svglite(
  filename="/Users/lishideng/Library/CloudStorage/OneDrive-SharedLibraries-UGent/BioLizard - General/manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/gene_beta_mother2_Jan16.svg",
  width=5, height=4.7
  )
gene_beta_mother
dev.off()

# Ensure the metadata is available for the samples
metadata_for_adonis <- pcoa_coords_mother %>%
  dplyr::select(idbs, code, timepoint, Group) %>%
  distinct()

# Run PERMANOVA using the Bray-Curtis distance matrix
adonis_result <- adonis2(bray_dist_mother ~ code + timepoint, data = metadata_for_adonis, permutations = 999)

# View the results
print(adonis_result)


# ----------------------- Repeat for Infant -----------------------
# Transpose the data to have samples as rows and genes as columns
data_t_infant <- t(gene_counts_infant_log)

# Calculate the Bray-Curtis distance matrix
bray_dist_infant <- vegdist(data_t_infant, method = "bray")

# Perform PCoA on the Bray-Curtis distance matrix
pcoa_result_infant <- cmdscale(bray_dist_infant, eig = TRUE, k = 2)

# Compute the proportion of variance explained by PC1 and PC2
variance_explained_PC1 <- pcoa_result_infant$eig[1] / sum(pcoa_result_infant$eig) * 100
variance_explained_PC2 <- pcoa_result_infant$eig[2] / sum(pcoa_result_infant$eig) * 100

# Print the variance explained
cat("Variance explained by PC1:", round(variance_explained_PC1, 2), "%\n")
cat("Variance explained by PC2:", round(variance_explained_PC2, 2), "%\n")

# Extract the coordinates for the first two principal coordinates
pcoa_coords_infant <- as.data.frame(pcoa_result_infant$points) %>%
  mutate(PC1 = V1) %>%
  mutate(PC2 = V2) %>%
  rownames_to_column("sample_id") %>%
  separate(sample_id, into = c("timepoint", "se1g", "idbs", "e"), sep = "_") %>%
  mutate(child = ifelse(is.na(e), "Mother", "Infant")) %>%
  left_join(metadata %>% dplyr::select(idbs, code), by = "idbs") %>%
  mutate(Group = case_when(
    timepoint == "pn12" & child == "Infant" & code == "A Y" ~ "Infant IFA pn12",
    timepoint == "pn12" & child == "Infant" & code == "B Z" ~ "Infant BEP pn12",
    timepoint == "pn56" & child == "Infant" & code == "A Y" ~ "Infant IFA pn56",
    timepoint == "pn56" & child == "Infant" & code == "B Z" ~ "Infant BEP pn56",
    TRUE ~ "other"
  )) %>%
  mutate(Group = factor(Group, levels = c(
    "Infant BEP pn12",
    "Infant IFA pn12",
    "Infant BEP pn56",
    "Infant IFA pn56"
  )))

# Plot for Mother
gene_beta_infant <- ggplot(pcoa_coords_infant, aes(x = PC1, y = PC2, color = Group, shape = timepoint)) +
  geom_point(size = 1) +
  stat_ellipse(aes(group = Group), linetype = 2) +
  scale_color_manual(
    name = "Group", 
    values = COLOR_SCHEME[1:4]  # Adjust these colors to fit your needs
  ) +
  scale_shape_manual(
    name = "Timepoint",
    labels = c("pn12", "pn56"),
    values = c(17, 3)
  ) +
  theme_bw() +
  theme(legend.position = "right") +
  guides(color = guide_legend(ncol = 1), shape = guide_legend(ncol = 1)) +
  labs(
    title = "PCoA Plot for Infant",
    x = "PC1",
    y = "PC2"
  )
gene_beta_infant

svglite::svglite(
  filename="/Users/lishideng/Library/CloudStorage/OneDrive-SharedLibraries-UGent/BioLizard - General/manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/gene_beta_infant2_Jan16.svg",
  width=4.7, height=4.7
  )
gene_beta_infant
dev.off()

# Ensure the metadata is available for the samples
metadata_for_adonis <- pcoa_coords_infant %>%
  dplyr::select(idbs, code, timepoint, Group) %>%
  distinct()

# Run PERMANOVA using the Bray-Curtis distance matrix
adonis_result <- adonis2(bray_dist_infant ~ code + timepoint, data = metadata_for_adonis, permutations = 999)

# View the results
print(adonis_result)



```


# Figure 3. Stability
## A. recurrence 
```{r}
ps_species <- ps %>%
  tax_glom(taxrank = "species") %>%
  transform_sample_counts(function(x){x/sum(x)}) %>%
  psmelt() %>%
  filter(Abundance > 0)

species_recurrence <- ps_species %>%
  select(OTU, phylum, species, timepoint, child, code, Abundance)

species_recurrence <- species_recurrence %>%
  group_by(OTU, species, code, child) %>%
  dplyr::summarise(recurrence = n_distinct(timepoint)) %>%
  ungroup()

# Create a recurrence category for density plot
species_recurrence <- species_recurrence %>%
  mutate(recurrence_category = case_when(
    child == "no" & recurrence == 2 ~ "Maternal: 2 times",
    child == "no" & recurrence == 3 ~ "Maternal: 3 times",
    child == "no" & recurrence == 4 ~ "Maternal: 4 times",
    child == "yes" & recurrence == 2 ~ "Infant: 2 times",
    TRUE ~ "Other"
  )) %>%
  filter(recurrence_category != "Other")

# Generate density plots
recurrence_plot <- ggplot(species_recurrence, aes(x = recurrence, fill = recurrence_category)) +
  geom_density(alpha = 0.5, size = 0.3) +
  facet_wrap(~ code + child, scales = "free_y", ncol = 2) +
  labs(
    x = "Number of Species",
    y = "Density",
    fill = "Recurrence Category"
  ) +
  xlim(0, 10) +
  ylim(0, 0.6) +
  scale_fill_manual(values = c("#00A98F", "#C00000", "skyblue", "#f1c232")) +
  theme_minimal() +
theme(
    panel.grid.major = element_blank(), # Remove major grid lines
    panel.grid.minor = element_blank(), # Remove minor grid lines
    panel.border = element_rect(colour = "black", fill = NA), # Add box outline
    strip.background = element_blank(), # Remove background of facet labels
    strip.text = element_blank(), # Remove facet labels
    axis.text.x = element_text(angle = 0, hjust = 1), # Angle x-axis labels if needed
    axis.title = element_text(size = 10),
    legend.position = "right" # Position legend at the bottom
  )

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/recurrence.svg",
  width=7, height=4
  )
recurrence_plot
dev.off()

```

```{r}

# Sample dataset creation (Replace this with your actual dataset)
# species_recurrence <- read.csv("path_to_your_dataset.csv")

# Step 1: Add a column indicating the exact number of unique timepoints for each species
species_recurrence_summary <- species_recurrence %>%
  dplyr::group_by(phylum, species, child, code) %>%
  dplyr::summarize(timepoint_count = n_distinct(timepoint), .groups = 'drop')

# Step 2: Calculate the percentage of each timepoint count within each phylum
# Calculate the total number of species per phylum, child, and code combination
total_species <- species_recurrence_summary %>%
  dplyr::group_by(phylum, child, code) %>%
  dplyr::summarize(total_species = n(), .groups = 'drop')

# Merge with the summary to calculate percentages
phylum_percentage <- species_recurrence_summary %>%
  dplyr::left_join(total_species, by = c("phylum", "child", "code")) %>%
  dplyr::group_by(phylum, child, code, timepoint_count) %>%
  dplyr::summarize(count = n(), .groups = 'drop') %>%
  dplyr::group_by(phylum, child, code) %>%
  dplyr::mutate(percentage = count / sum(count))  # Calculate percentage relative to the total count in each phylum

# Combine the data for plotting
combined_plots <- phylum_percentage %>%
  dplyr::mutate(plot_type = paste(child, code, sep = "-"))

# Function to plot the percentage of occurrences with each bar summing to 1
plot_percentage_combined <- function(df, child_status) {
  ggplot(df %>% dplyr::filter(grepl(child_status, plot_type)), aes(x = phylum, y = percentage, fill = as.factor(timepoint_count))) +
    geom_bar(stat = "identity", position = "stack") +
    labs(x = "Phylum", y = "Percentage", title = paste("Phylum Occurrence: Child-", child_status)) +
    scale_y_continuous(labels = scales::percent, limits = c(0, 1)) +  # Ensure y-axis goes from 0 to 1 (0% to 100%)
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
          strip.text = element_text(size = 12)) +  # Adjust facet labels if needed
    scale_fill_manual(values = c("#00A98F", "#C00000", "skyblue", "#f1c232")) +  # Custom colors
    guides(fill = guide_legend(title = "Timepoint Count")) +
    facet_wrap(~code, scales = "fixed") +
    theme(
      axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
      legend.position = "bottom",
      panel.background = element_blank(),
      plot.background = element_blank(),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
    panel.border = element_rect(colour = "black", fill = NA)# Add box outline
    ) 
}

# Generate and display the combined plots
print(plot_percentage_combined(combined_plots, "no"))
print(plot_percentage_combined(combined_plots, "yes"))

plot_percentage_mo <- plot_percentage_combined(combined_plots, "no")
plot_percentage_inf <- plot_percentage_combined(combined_plots, "yes")

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/plot_percentage_mo.svg",
  width=8, height=4.5
  )
plot_percentage_mo
dev.off()

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/plot_percentage_inf.svg",
  width=6, height=4.5
  )
plot_percentage_inf
dev.off()

```


```{r}
library(ggplot2)
library(dplyr)

# Ensure total counts are calculated
total_counts <- phylum_percentage %>%
  dplyr::group_by(phylum, child, code) %>%
  dplyr::summarize(total_count = sum(count), .groups = 'drop')

# Plot heatmaps with specified color gradients
plot_heatmaps <- function(df, child_status) {
  color_gradient <- if (child_status == "yes") {
    scale_fill_gradient(low = "white", high = "steelblue", name = "Total Count")
  } else {
    scale_fill_gradient(low = "white", high = "darkred", name = "Total Count")
  }
  
  ggplot(df %>% dplyr::filter(child == child_status), aes(x = phylum, y = 1, fill = total_count)) +
    geom_tile(color = "white") +  # Add white border to each tile for better separation
    labs(x = "Phylum", y = NULL, title = paste("Heatmap of Total Counts: Child-", child_status)) +
    color_gradient +
    facet_wrap(~code, scales = "fixed") +
    theme_minimal() +
    theme(
      axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
      axis.text.y = element_blank(),  # Remove y-axis text as it's not needed
      axis.ticks.y = element_blank(), # Remove y-axis ticks
      panel.grid = element_blank(),
      strip.text = element_text(size = 12),
      legend.position = "bottom",
      panel.border = element_rect(colour = "black", fill = NA) # Add border around the panel
    )
}

# Generate and display the heatmaps
plot_heatmap_child_no <- plot_heatmaps(total_counts, "no")
plot_heatmap_child_yes <- plot_heatmaps(total_counts, "yes")

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/plot_heatmap_mo.svg",
  width=8, height=2.9
  )
plot_heatmap_child_no
dev.off()

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/plot_heatmap_inf.svg",
  width=6, height=2.9
  )
plot_heatmap_child_yes
dev.off()




```

## B. UPSET

```{r}

#Mother_Tri2 <- subset_samples(incl)
Mother_Tri2 <- subset_samples(incl, code == "B Z")
#Mother_Tri2 <- subset_samples(incl, code == "A Y")
Mother_Tri2 <- otu_table(Mother_Tri2) %>%
  as.data.frame() %>%
  filter(rowSums(across(everything())) != 0)
Mother_Tri2_species <- rownames(Mother_Tri2)

#Mother_Tri3 <- subset_samples(tri3)
Mother_Tri3 <- subset_samples(tri3, code == "B Z")
#Mother_Tri3 <- subset_samples(tri3, code == "A Y")
Mother_Tri3 <- otu_table(Mother_Tri3) %>%
  as.data.frame() %>%
  filter(rowSums(across(everything())) != 0)
Mother_Tri3_species <- rownames(Mother_Tri3)

#Mother_Pn12 <- subset_samples(pn12m)
Mother_Pn12 <- subset_samples(pn12m, code == "B Z")
#Mother_Pn12 <- subset_samples(pn12m, code == "A Y")
Mother_Pn12<- otu_table(Mother_Pn12) %>%
  as.data.frame() %>%
  filter(rowSums(across(everything())) != 0)
Mother_Pn12_species <- rownames(Mother_Pn12)

#Mother_Pn56 <- subset_samples(pn56m)
Mother_Pn56 <- subset_samples(pn56m, code == "B Z")
#Mother_Pn56 <- subset_samples(pn56m, code == "A Y")
Mother_Pn56<- otu_table(Mother_Pn56) %>%
  as.data.frame() %>%
  filter(rowSums(across(everything())) != 0)
Mother_Pn56_species <- rownames(Mother_Pn56)

#Infant_Pn12 <- subset_samples(pn12e)
Infant_Pn12 <- subset_samples(pn12e, code == "B Z")
#Infant_Pn12 <- subset_samples(pn12e, code == "A Y")
Infant_Pn12<- otu_table(Infant_Pn12) %>%
  as.data.frame() %>%
  filter(rowSums(across(everything())) != 0)
Infant_Pn12_species <- rownames(Infant_Pn12)

#Infant_Pn56 <- subset_samples(pn56e)
Infant_Pn56 <- subset_samples(pn56e, code == "B Z")
#Infant_Pn56 <- subset_samples(pn56e, code == "A Y")
Infant_Pn56<- otu_table(Infant_Pn56) %>%
  as.data.frame() %>%
  filter(rowSums(across(everything())) != 0)
Infant_Pn56_species <- rownames(Infant_Pn56)


# Combine all species into a single data frame with a presence/absence matrix
species <- unique(c(Mother_Tri2_species, Mother_Tri3_species, Mother_Pn12_species, Mother_Pn56_species, Infant_Pn12_species, Infant_Pn56_species))
df <- data.frame(
  species = species,
  Mother_Tri2 = as.numeric(species %in% Mother_Tri2_species),
  Mother_Tri3 = as.numeric(species %in% Mother_Tri3_species),
  Mother_Pn12 = as.numeric(species %in% Mother_Pn12_species),
  Mother_Pn56 = as.numeric(species %in% Mother_Pn56_species),
  Infant_Pn12 = as.numeric(species %in% Infant_Pn12_species),
  Infant_Pn56 = as.numeric(species %in% Infant_Pn56_species)
)

# Create an UpSet plot
upset <- upset(
  df, 
  sets = c("Mother_Tri2", "Mother_Tri3", "Mother_Pn12", "Mother_Pn56", "Infant_Pn12", "Infant_Pn56"),
  order.by = "freq",
  main.bar.color = "steelblue",
  sets.bar.color = c("darkred", "#EFC000FF", "#868686FF", "steelblue", "#00A98F", "#6aa84f"),
  keep.order = TRUE
)

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/upset_bep.svg",
  width=7, height=4
  )
upset
dev.off()


```


#### chisq.test
```{r}
# Define the observed values
control_shared <- 165
control_total <- 281
intervention_shared <- 142
intervention_total <- 292

# Create the contingency table
contingency_table <- matrix(c(control_shared, control_total - control_shared,
                              intervention_shared, intervention_total - intervention_shared), 
                            nrow = 2, 
                            byrow = TRUE)

# Set the row and column names for clarity (optional)
rownames(contingency_table) <- c("Control", "Intervention")
colnames(contingency_table) <- c("Shared Species", "Non-Shared Species")

# Perform the chi-square test
test_result <- chisq.test(contingency_table)

# Print the test result
print(test_result)

```



## C. shared species between mat and inf
```{r}

dyads_species <- ps %>%
  tax_glom(taxrank = "species") %>%
  transform_sample_counts(function(x){x/sum(x)}) %>%
  psmelt() %>%
  filter(Abundance > 0)

# First, summarize the data to count distinct values
species_summary <- dyads_species %>%
  filter(timepoint != "incl") %>%
  group_by(OTU, child, code) %>%
  dplyr::summarize(
    n_timepoints = n_distinct(timepoint),
    .groups = 'drop'
  ) %>%
  group_by(OTU, child) %>%
  dplyr::summarize(
    n_codes = n_distinct(code),
    all_codes_have_2_timepoints = all(n_timepoints >= 1),
    .groups = 'drop'
  ) %>%
  group_by(OTU) %>%
  dplyr::summarize(
    n_children = n_distinct(child),
    all_children_have_2_codes_with_2_timepoints = all(n_codes == 1 & all_codes_have_2_timepoints),
    .groups = 'drop'
  ) %>%
  filter(
    n_children == 1,
    all_children_have_2_codes_with_2_timepoints
  )

species_summary <- dyads_species %>%
  filter(timepoint != "incl") %>%
  group_by(species, child, code) %>%
  dplyr::summarize(n_timepoints = n_distinct(timepoint), .groups = 'drop') %>%
  group_by(species, child) %>%
  dplyr::summarize(
    n_codes = n_distinct(code),
    all_codes_have_required_timepoints = ifelse(child == "no", all(n_timepoints >= 3), all(n_timepoints >= 2)),
    .groups = 'drop'
  ) %>%
  group_by(species) %>%
  dplyr::summarize(
    n_children = n_distinct(child),
    all_children_have_required_codes_with_timepoints = all(n_codes == 2 & all_codes_have_required_timepoints),
    .groups = 'drop'
  ) %>%
  filter(
    n_children == 2,
    all_children_have_required_codes_with_timepoints
  )


# Filter original data based on the summarized data
relAbuSpeDyads <- dyads_species %>%
  filter(timepoint != "incl", species %in% species_summary$species)


summary_relAbuSpeDyads <- summarySE(relAbuSpeDyads, measurevar = "Abundance", groupvars = c("species", "timepoint", "code", "child"), na.rm = TRUE) %>%
  mutate(Intervention = if_else(code == "B Z", "Intervention", "Control")) %>%
  mutate(Dyads = if_else(child == "yes", "Infant", "Mother")) %>%
  group_by(species) %>%
  filter(all(N>=5)) %>%
  ungroup()
  

# Calculate the average abundance for each species
avg_abundance <- relAbuSpeDyads %>%
  group_by(species) %>%
  dplyr::summarise(avg_abundance = mean(Abundance, na.rm = TRUE)) %>%
  arrange(desc(avg_abundance))

# Select the top 15 species
top_fifteen_species <- avg_abundance %>%
  filter(species %in% summary_relAbuSpeDyads$species) %>%
  slice_max(order_by = avg_abundance, n=10) %>%
  pull(species)

# Filter the dataset to include only the top 15 species
filtered_data <- summary_relAbuSpeDyads %>%
  filter(species %in% top_fifteen_species) %>%
  mutate(Abundance = Abundance *100) %>%
  mutate(timepoint = factor(timepoint, levels = c("tri3", "pn12", "pn56"))) %>%
  mutate(Dyads = factor(Dyads, levels = c("Mother", "Infant")))%>%
  mutate(se = se*100) %>%
  mutate(sd=sd*100) %>%
  mutate(ci = ci*100)

unique(filtered_data$species)


filtered_data3 <- filtered_data %>%
  filter(species == "Bacteroides fragilis"| species =="Bacteroides fragilis_A" | species =="Escherichia coli" )

SharedInDyadsHigh <- ggplot(filtered_data3, aes(x = timepoint, y = Abundance, color = Intervention, linetype = Dyads, group = interaction(species, Dyads, Intervention))) +
  geom_line(position = position_dodge(0.2)) +  # Add lines for each group
  geom_point(position = position_dodge(0.2)) +  # Add points for each data point
  geom_errorbar(aes(ymin = Abundance - se, ymax = Abundance + se), width = 0.05, position = position_dodge(0.2), alpha = 0.5) +  # Add error bars
  labs(title = "Abundance Across Timepoints by Species",
       x = "Timepoint",
       y = "Relative Abundance (%) (log10)",
       color = "Intervention",
       linetype = "Dyads") +
  scale_y_log10() +  # Apply log10 transformation
  theme_minimal() +
  scale_color_manual(values = c("Intervention" = "steelblue", "Control" = "darkred")) +
  theme(legend.position = "right",  # Improve readability of x-axis labels
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.line = element_line(color = "black"),  # Add axis lines
        axis.ticks = element_line(color = "black"),  # Add axis ticks
        axis.text = element_text(color = "black"),  # Add axis text
        axis.title = element_text(color = "black")) +  # Add axis titles
  facet_wrap(~ species, nrow = 1, strip.position = "bottom", scales = "free_x")

SharedInDyadsHigh

svglite::svglite(
  filename="/Users/lishideng/Library/CloudStorage/OneDrive-SharedLibraries-UGent/BioLizard - General/manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/SharedInDyadsHigh_Jan17.svg",
  width=5.5, height=2.5
  )
SharedInDyadsHigh
dev.off()


filtered_data4 <- filtered_data %>%
  filter(species != "Bacteroides fragilis" & species !="Bacteroides fragilis_A" & species !="Escherichia coli")

SharedInDyadsLow <- ggplot(filtered_data4, aes(x = timepoint, y = Abundance, color = Intervention, linetype = Dyads, group = interaction(species, Dyads, Intervention))) +
  geom_line(position = position_dodge(0.2)) +  # Add lines for each group
  geom_point(position = position_dodge(0.2)) +  # Add points for each data point
  geom_errorbar(aes(ymin = Abundance - se, ymax = Abundance + se), width = 0.05, position = position_dodge(0.2), alpha = 0.5) +  # Add error bars
  labs(title = "Abundance Across Timepoints by Species",
       x = "Timepoint",
       y = "Relative Abundance (%) (log10)",
       color = "Intervention",
       linetype = "Dyads") +
  scale_y_log10() +  # Apply log10 transformation
  theme_minimal() +
  scale_color_manual(values = c("Intervention" = "steelblue", "Control" = "darkred")) +
  theme(legend.position = "right",  # Improve readability of x-axis labels
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.line = element_line(color = "black"),  # Add axis lines
        axis.ticks = element_line(color = "black"),  # Add axis ticks
        axis.text = element_text(color = "black"),  # Add axis text
        axis.title = element_text(color = "black")) +  # Add axis titles
  facet_wrap(~ species, nrow = 1, strip.position = "bottom", scales = "free_x")

SharedInDyadsLow

svglite::svglite(
  filename="/Users/lishideng/Library/CloudStorage/OneDrive-SharedLibraries-UGent/BioLizard - General/manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/SharedInDyadsLow_Jan17.svg",
  width=10, height=2.5
  )
SharedInDyadsLow
dev.off()


```

## D. top in mother
```{r}

mother_species <- mother_ps %>%
  tax_glom(taxrank = "species") %>%
  transform_sample_counts(function(x){x/sum(x)}) %>%
  psmelt() %>%
  filter(Abundance > 0)
  
relAbuSpeThreeMo <- mother_species %>%
  filter(timepoint != "incl") %>%
  dplyr::group_by(OTU) %>%
  filter(n_distinct(timepoint) == 3 & n_distinct(code) ==2) %>%
  mutate(timepoint = factor(timepoint, levels = desired_order)) %>%
  dplyr::ungroup()

summary_relAbuSpeThreeMo <- summarySE(relAbuSpeThreeMo, measurevar = "Abundance", groupvars = c("species", "timepoint", "code"), na.rm = TRUE) %>%
  mutate(Intervention = if_else(code == "B Z", "Intervention", "Control")) %>%
  group_by(species) %>%
  filter(all(N>5)) %>%
  ungroup()
  

# Calculate the average abundance for each species
avg_abundance <- relAbuSpeThreeMo %>%
  group_by(species) %>%
  dplyr::summarise(avg_abundance = mean(Abundance, na.rm = TRUE)) %>%
  arrange(desc(avg_abundance))

# Select the top 15 species
top_fifteen_species <- avg_abundance %>%
  filter(species %in% summary_relAbuSpeThreeMo$species) %>%
  slice_max(order_by = avg_abundance, n=10) %>%
  pull(species)

# Filter the dataset to include only the top 15 species
filtered_data <- summary_relAbuSpeThreeMo %>%
  filter(species %in% top_fifteen_species) %>%
  mutate(Abundance = Abundance *100) %>%
  mutate(se = se*100) %>%
  mutate(sd=sd*100) %>%
  mutate(ci = ci*100)

plotMoThreeBEPIFA <- ggplot(filtered_data, aes(x = timepoint, y = Abundance, color = Intervention)) +
  geom_errorbar(aes(ymin = Abundance - se, ymax = Abundance +se), width = 0.05, position = position_dodge(0.05), alpha = 0.5) +
  geom_line(aes(group = interaction(species, Intervention), color = Intervention), position = position_dodge(0.05) )+
  geom_point(position = position_dodge(0.05)) +
  facet_wrap(~ species, scales = "free_y", ncol = 5) +
  theme_bw() +
  labs(color = "Intervention", y = "Relative Abundance, %", x = "Timepoint" ) +
  scale_color_manual(values = c("Intervention" = "steelblue", "Control" = "darkred")) +
  expand_limits(y = 0) +  # Ensure y-axis starts at 0
  theme(panel.spacing = unit(0.5, "lines"),
        strip.text = element_text(size = 8))

  plotMoThreeBEPIFA

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/NotsharedSpeciesThreeMoBEPIFA2.svg",
  width=10, height=3
  )
plotMoThreeBEPIFA
dev.off()

```


### mixed effect

```{r}

relAbuSpeMo <- relAbuSpeThreeMo %>%
  filter(timepoint != "incl") %>%
  #group_by(OTU, idbs, code) %>%
  #filter(n_distinct(Sample) == 3) %>%
  mutate(Intervention = if_else(code == "B Z", "2", "1")) %>%
  mutate(timepoint = factor(timepoint, levels = desired_order)) %>%
  mutate(Timepoint = if_else(timepoint == "tri3", "1",
                             if_else(timepoint == "pn12", "2", "3"))) %>%
  #ungroup() %>%
  select(species, Sample, idbs, Intervention, Timepoint, Abundance)

library(lmerTest)
library(emmeans)

Mo_fragilis <- relAbuSpeMo %>% filter(species == "Bacteroides fragilis")
Mo_fragilis_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Mo_fragilis)
summary(Mo_fragilis_model)
confint(Mo_fragilis_model, level = 0.95, method = "profile")
p_values <- summary(Mo_fragilis_model)$coefficients[, "Pr(>|t|)"]
adjusted_p_values <- round(p.adjust(p_values, method = "BH"),3)
adjusted_p_values


Mo_c941 <- relAbuSpeMo %>% filter(species == "C941 sp004557565")
Mo_c941_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Mo_c941)
summary(Mo_c941_model)

Mo_copri <- relAbuSpeMo %>% filter(species == "Prevotella copri") 
Mo_copri_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Mo_copri)
summary(Mo_copri_model)

Mo_sp002251295 <- relAbuSpeMo %>% filter(species == "Prevotella sp002251295") 
Mo_sp002251295_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Mo_sp002251295)
summary(Mo_sp002251295_model)

Mo_sp000434515 <- relAbuSpeMo %>% filter(species == "Prevotella sp000434515") 
Mo_sp000434515_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Mo_sp000434515)
summary(Mo_sp000434515_model)

Mo_sp002299635 <- relAbuSpeMo %>% filter(species == "Prevotella sp002299635") 
Mo_sp002299635_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Mo_sp002299635)
summary(Mo_sp002299635_model)
confint(Mo_sp002299635_model, level = 0.95, method = "profile")
p_values <- summary(Mo_sp002299635_model)$coefficients[, "Pr(>|t|)"]
adjusted_p_values <- round(p.adjust(p_values, method = "BH"),3)
adjusted_p_values

Mo_sp900546535 <- relAbuSpeMo %>% filter(species == "Prevotella sp900546535") 
Mo_sp900546535_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Mo_sp900546535)
summary(Mo_sp900546535_model)

Mo_sp900551275 <- relAbuSpeMo %>% filter(species == "Prevotella sp900551275") 
Mo_sp900551275_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Mo_sp900551275)
summary(Mo_sp900551275_model)

Mo_rc9 <- relAbuSpeMo %>% filter(species == "RC9 sp000433355") 
Mo_rc9_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Mo_rc9 )
summary(Mo_rc9_model)
confint(Mo_rc9_model, level = 0.95, method = "profile")
p_values <- summary(Mo_rc9_model)$coefficients[, "Pr(>|t|)"]
adjusted_p_values <- round(p.adjust(p_values, method = "BH"),3)
adjusted_p_values


Mo_unk <- relAbuSpeMo %>% filter(species == "Unknown_Prevotella") 
Mo_unk_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Mo_unk)
summary(Mo_unk_model)


```


## D. top in infant
```{r}

library(Rmisc) 
infant_species <- infant_ps %>%
  tax_glom(taxrank = "species") %>%
  transform_sample_counts(function(x){x/sum(x)}) %>%
  psmelt() %>%
  filter(Abundance > 0)
  
relAbuSpeInf <- infant_species %>%
  group_by(OTU) %>%
  filter(n_distinct(timepoint) == 2 & n_distinct(code) == 2) %>%
  mutate(timepoint = factor(timepoint, levels = desired_order)) %>%
  ungroup()

summary_relAbuSpeInf <- summarySE(relAbuSpeInf, measurevar = "Abundance", groupvars = c("species", "timepoint", "code"), na.rm = TRUE)%>%
  mutate(Intervention = if_else(code == "B Z", "Intervention", "Control"))%>%
  group_by(species) %>%
  filter(all(N>5)) %>%
  ungroup()

avg_abundance <- relAbuSpeInf %>%
  group_by(species) %>%
  dplyr::summarise(avg_abundance = mean(Abundance, na.rm = TRUE)) %>%
  arrange(desc(avg_abundance))

# Select the top 15 species
top_fifteen_species <- avg_abundance %>%
  filter(species %in% summary_relAbuSpeInf$species) %>%
  slice_max(order_by = avg_abundance, n=10) %>%
  pull(species)

# Filter the dataset to include only the top 15 species
filtered_data <- summary_relAbuSpeInf %>%
  filter(species %in% top_fifteen_species) %>%
  mutate(Abundance = Abundance *100) %>%
  mutate(se = se*100) %>%
  mutate(sd=sd*100) %>%
  mutate(ci = ci*100) 

plotInfBEPIFA <- ggplot(filtered_data, aes(x = timepoint, y = Abundance, color = Intervention)) +
  geom_errorbar(aes(ymin = Abundance - se, ymax = Abundance + se), linetype = "dashed", width = 0.05, position = position_dodge(0.05), alpha = 0.5) +
  geom_line(aes(group = interaction(species, Intervention)), linetype = "dashed", position = position_dodge(0.05)) +  # Set linetype globally to "dashed"
  geom_point(position = position_dodge(0.05)) +
  facet_wrap(~ species, scales = "free_y", ncol = 5) +
  theme_bw() +
  labs(color = "Intervention", y = "Relative Abundance, %", x = "Timepoint") +
  scale_color_manual(values = c("Intervention" = "steelblue", "Control" = "darkred")) +
  expand_limits(y = 0.0) + # Ensure y-axis starts at 0
  theme(panel.spacing = unit(0.5, "lines"),
        strip.text = element_text(size = 8))
  
plotInfBEPIFA

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/NotsharedSpeciesInfBEPIFA2.svg",
  width=10, height=2.8
  )
plotInfBEPIFA
dev.off()

```


### mixed effect

```{r}

relAbuSpeInf2 <- relAbuSpeInf %>%
  #group_by(OTU, idbs, code) %>%
  #filter(n_distinct(Sample) == 2) %>%
  mutate(Intervention = if_else(code == "B Z", "2", "1")) %>%
  mutate(timepoint = factor(timepoint, levels = desired_order)) %>%
  mutate(Timepoint = if_else(timepoint == "pn12", "1", "2")) %>%
  #ungroup() %>%
  select(species, Sample, idbs, Intervention, Timepoint, Abundance)

Inf_fragilis <- relAbuSpeInf2 %>% filter(species == "Bacteroides fragilis_A")
Inf_fragilis_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Inf_fragilis)
summary(Inf_fragilis_model)

Inf_uniformis <- relAbuSpeInf2 %>% filter(species == "Bacteroides uniformis")
Inf_uniformis_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Inf_uniformis)
summary(Inf_uniformis_model)

Inf_bifidum <- relAbuSpeInf2 %>% filter(species == "Bifidobacterium bifidum") 
Inf_bifidum_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Inf_bifidum)
summary(Inf_bifidum_model)

Inf_breve <- relAbuSpeInf2 %>% filter(species == "Bifidobacterium breve") 
Inf_breve_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Inf_breve)
summary(Inf_breve_model)
confint(Inf_breve_model, level = 0.95, method = "profile")

Inf_infantis <- relAbuSpeInf2 %>% filter(species == "Bifidobacterium infantis") 
Inf_infantis_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Inf_infantis)
summary(Inf_infantis_model)

Inf_longum <- relAbuSpeInf2 %>% filter(species == "Bifidobacterium longum") 
Inf_longum_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Inf_longum)
summary(Inf_longum_model)

Inf_coli <- relAbuSpeInf2 %>% filter(species == "Escherichia coli") 
Inf_coli_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Inf_coli)
summary(Inf_coli_model)

Inf_pneumoniae <- relAbuSpeInf2 %>% filter(species == "Klebsiella pneumoniae") 
Inf_pneumoniae_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Inf_pneumoniae)
summary(Inf_pneumoniae_model)

Inf_uli <- relAbuSpeInf2 %>% filter(species == "Parolsenella uli_B") 
Inf_uli_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Inf_uli)
summary(Inf_uli_model)

Inf_sp000431435 <- relAbuSpeInf2 %>% filter(species == "Veillonella_A sp000431435") 
Inf_sp000431435_model <- lmer(Abundance ~ Timepoint * Intervention + (1 | idbs), data = Inf_sp000431435)
summary(Inf_sp000431435_model)


```




## E. intra dist ~ days + (1|subject)

### disimilarity and date

```{r}

# Extract the Unifrac distance matrix
unifrac_distances <- as.matrix(phyloseq::distance(ps, method = "unifrac"))

# Create a data frame from the distance matrix
dist_df <- as.data.frame(as.table(unifrac_distances))
colnames(dist_df) <- c("Sample1", "Sample2", "dist")

# Convert Sample1 and Sample2 columns to character
dist_df <- dist_df %>%
  mutate(Sample1 = as.character(Sample1),
         Sample2 = as.character(Sample2))

# Function to extract timepoint from sample names
extract_timepoint <- function(sample_name) {
  parts <- strsplit(sample_name, "_")[[1]]
  timepoint <- parts[1]
  timepoint_numeric <- switch(timepoint,
    "incl" = 1,
    "tri3" = 2,
    "pn12" = 3,
    "pn56" = 4,
    NA  # Return NA for unknown timepoints
  )
  return(timepoint_numeric)
}

# Add timepoint columns to dist_df
dist_df <- dist_df %>%
  mutate(timepoint.x = sapply(Sample1, extract_timepoint),
         timepoint.y = sapply(Sample2, extract_timepoint),
         timediff = abs(timepoint.y - timepoint.x)) %>%
  mutate(transformed_dist = log10(-log10(1 - dist))) %>%
  filter(is.finite(transformed_dist))

# Print the first few rows to verify
head(dist_df)

library(lme4)
library(lmerTest)  # For F-tests

# Fit the linear mixed effects model
model <- lmer(transformed_dist ~ timediff + (1 | Sample1), data = dist_df)

# Print summary of the model
summary(model)

confint(model, level = 0.95)

```



#### mother - bep

```{r}

mother_bep <- subset_samples(mother, code=="B Z")
unifrac_mo_bep <- as.matrix(phyloseq::distance(mother_bep, method = "unifrac"))

# Create a data frame from the distance matrix
dist_df1 <- as.data.frame(as.table(unifrac_mo_bep))
colnames(dist_df1) <- c("Sample1", "Sample2", "dist")

# Convert Sample1 and Sample2 columns to character
dist_df1 <- dist_df1 %>%
  mutate(Sample1 = as.character(Sample1),
         Sample2 = as.character(Sample2))

# Function to extract timepoint from sample names
extract_timepoint <- function(sample_name) {
  parts <- strsplit(sample_name, "_")[[1]]
  timepoint <- parts[1]
  timepoint_numeric <- switch(timepoint,
    "incl" = 1,
    "tri3" = 2,
    "pn12" = 3,
    "pn56" = 4,
    NA  # Return NA for unknown timepoints
  )
  return(timepoint_numeric)
}

# Add timepoint columns to dist_df
dist_df1 <- dist_df1 %>%
  mutate(timepoint.x = sapply(Sample1, extract_timepoint),
         timepoint.y = sapply(Sample2, extract_timepoint),
         timediff = abs(timepoint.y - timepoint.x)) %>%
  mutate(transformed_dist = log10(-log10(1 - dist))) %>%
  filter(is.finite(transformed_dist)) %>%
  mutate(group = "Mother-intervention")

# Print the first few rows to verify
head(dist_df1)

library(lme4)
library(lmerTest)  # For F-tests

# Fit the linear mixed effects model
model1 <- lmer(transformed_dist ~ timediff + (1 | Sample1), data = dist_df1)

# Print summary of the model
summary(model1)

```

#### mother - ifa

```{r}

mother_ifa <- subset_samples(mother, code=="A Y")
unifrac_mo_ifa <- as.matrix(phyloseq::distance(mother_ifa, method = "unifrac"))

# Create a data frame from the distance matrix
dist_df2 <- as.data.frame(as.table(unifrac_mo_ifa))
colnames(dist_df2) <- c("Sample1", "Sample2", "dist")

# Convert Sample1 and Sample2 columns to character
dist_df2 <- dist_df2 %>%
  mutate(Sample1 = as.character(Sample1),
         Sample2 = as.character(Sample2))

# Function to extract timepoint from sample names
extract_timepoint <- function(sample_name) {
  parts <- strsplit(sample_name, "_")[[1]]
  timepoint <- parts[1]
  timepoint_numeric <- switch(timepoint,
    "incl" = 1,
    "tri3" = 2,
    "pn12" = 3,
    "pn56" = 4,
    NA  # Return NA for unknown timepoints
  )
  return(timepoint_numeric)
}

# Add timepoint columns to dist_df
dist_df2 <- dist_df2 %>%
  mutate(timepoint.x = sapply(Sample1, extract_timepoint),
         timepoint.y = sapply(Sample2, extract_timepoint),
         timediff = abs(timepoint.y - timepoint.x)) %>%
  mutate(transformed_dist = log10(-log10(1 - dist))) %>%
  filter(is.finite(transformed_dist)) %>%
  mutate(group = "Mother-control")

# Print the first few rows to verify
head(dist_df2)

library(lme4)
library(lmerTest)  # For F-tests

# Fit the linear mixed effects model
model2 <- lmer(transformed_dist ~ timediff + (1 | Sample1), data = dist_df2)

# Print summary of the model
summary(model2)

```


#### infant - bep

```{r}

infant_bep <- subset_samples(infant, code=="B Z")
unifrac_inf_bep <- as.matrix(phyloseq::distance(infant_bep, method = "unifrac"))

# Create a data frame from the distance matrix
dist_df3 <- as.data.frame(as.table(unifrac_inf_bep))
colnames(dist_df3) <- c("Sample1", "Sample2", "dist")

# Convert Sample1 and Sample2 columns to character
dist_df3 <- dist_df3 %>%
  mutate(Sample1 = as.character(Sample1),
         Sample2 = as.character(Sample2))

# Function to extract timepoint from sample names
extract_timepoint <- function(sample_name) {
  parts <- strsplit(sample_name, "_")[[1]]
  timepoint <- parts[1]
  timepoint_numeric <- switch(timepoint,
    "incl" = 1,
    "tri3" = 2,
    "pn12" = 3,
    "pn56" = 4,
    NA  # Return NA for unknown timepoints
  )
  return(timepoint_numeric)
}

# Add timepoint columns to dist_df
dist_df3 <- dist_df3 %>%
  mutate(timepoint.x = sapply(Sample1, extract_timepoint),
         timepoint.y = sapply(Sample2, extract_timepoint),
         timediff = abs(timepoint.y - timepoint.x)) %>%
  mutate(transformed_dist = log10(-log10(1 - dist))) %>%
  filter(is.finite(transformed_dist)) %>%
    mutate(group = "Infant-intervention")


# Print the first few rows to verify
head(dist_df3)

library(lme4)
library(lmerTest)  # For F-tests

# Fit the linear mixed effects model
model3 <- lmer(transformed_dist ~ timediff + (1 | Sample1), data = dist_df3)

# Print summary of the model
summary(model3)

```

#### infant - ifa

```{r}

infant_ifa <- subset_samples(infant, code=="A Y")
unifrac_inf_ifa <- as.matrix(phyloseq::distance(infant_ifa, method = "unifrac"))

# Create a data frame from the distance matrix
dist_df4 <- as.data.frame(as.table(unifrac_inf_ifa))
colnames(dist_df4) <- c("Sample1", "Sample2", "dist")

# Convert Sample1 and Sample2 columns to character
dist_df4 <- dist_df4 %>%
  mutate(Sample1 = as.character(Sample1),
         Sample2 = as.character(Sample2))

# Function to extract timepoint from sample names
extract_timepoint <- function(sample_name) {
  parts <- strsplit(sample_name, "_")[[1]]
  timepoint <- parts[1]
  timepoint_numeric <- switch(timepoint,
    "incl" = 1,
    "tri3" = 2,
    "pn12" = 3,
    "pn56" = 4,
    NA  # Return NA for unknown timepoints
  )
  return(timepoint_numeric)
}

# Add timepoint columns to dist_df
dist_df4 <- dist_df4 %>%
  mutate(timepoint.x = sapply(Sample1, extract_timepoint),
         timepoint.y = sapply(Sample2, extract_timepoint),
         timediff = abs(timepoint.y - timepoint.x)) %>%
  mutate(transformed_dist = log10(-log10(1 - dist))) %>%
  filter(is.finite(transformed_dist)) %>%
      mutate(group = "Infant-control")


# Print the first few rows to verify
head(dist_df4)

library(lme4)
library(lmerTest)  # For F-tests

# Fit the linear mixed effects model
model4 <- lmer(transformed_dist ~ timediff + (1 | Sample1), data = dist_df4)

# Print summary of the model
summary(model4)

```

#### reformate tables 
```{r}
incl_se1g_date <- incl_se1g %>%
  mutate(
    incl_se1g_dadi = na_if(incl_se1g_dadi, ""),
    incl_sa10ab_dtdi = na_if(incl_sa10ab_dtdi, ""),
    incl_sa20ab_dadi = na_if(incl_sa20ab_dadi, ""),
    incl_sapl_dati = na_if(incl_sapl_dati, "")
  ) %>%
  mutate(date = coalesce(substr(incl_se1g_dadi, 1, 8), substr(incl_sa10ab_dtdi, 1, 8), substr(incl_sa20ab_dadi, 1, 8), substr(incl_sapl_dati, 1, 8))) %>%
  mutate(sampleid = paste0("incl_se1g_", incl_idbs)) %>%
  dplyr::select(sampleid, date)

tri3_se1g_date <- tri3_se1g %>%
  mutate(
    tri3_se1g_dadi = na_if(tri3_se1g_dadi, ""),
    tri3_sa10ab_dtdi = na_if(tri3_sa10ab_dtdi, ""),
    tri3_sa20ab_dadi = na_if(tri3_sa20ab_dadi, ""),
    tri3_sapl_dati = na_if(tri3_sapl_dati, "")
  ) %>%
  mutate(date = coalesce(substr(tri3_se1g_dadi, 1, 8), substr(tri3_sa10ab_dtdi, 1, 8), substr(tri3_sa20ab_dadi, 1, 8), substr(tri3_sapl_dati, 1, 8))) %>%
  mutate(sampleid = paste0("tri3_se1g_", tri3_idbs)) %>%
  dplyr::select(sampleid, date)

pn12m_se1g_date <- pn12m_se1g %>%
  mutate(
    pn12m_se1g_dadi = na_if(pn12m_se1g_dadi, ""),
    pn12m_sapl_dati = na_if(pn12m_sapl_dati, "")
  ) %>%
  mutate(date = coalesce(substr(pn12m_se1g_dadi, 1, 8),  substr(pn12m_sapl_dati, 1, 8))) %>%
  mutate(sampleid = paste0("pn12_se1g_", pn12m_idbs)) %>%
  dplyr::select(sampleid, date)

pn56m_se1g_date <- pn56m_se1g %>%
  mutate(
    pn56m_se1g_dadi = na_if(pn56m_se1g_dadi, ""),
    pn56m_sa10ab_dtdi = na_if(pn56m_sa10ab_dtdi, ""),
    pn56m_sa20ab_dadi = na_if(pn56m_sa20ab_dadi, ""),
    pn56m_sapl_dati = na_if(pn56m_sapl_dati, "")
  ) %>%
  mutate(date = coalesce(substr(pn56m_se1g_dadi, 1, 8), substr(pn56m_sa10ab_dtdi, 1, 8), substr(pn56m_sa20ab_dadi, 1, 8), substr(pn56m_sapl_dati, 1, 8))) %>%
  mutate(sampleid = paste0("pn56_se1g_", pn56m_idbs)) %>%
  dplyr::select(sampleid, date)

pn12e_se1g_date <- pn12e_se1g %>%
  mutate(
    pn12e_se1g_dadi = na_if(pn12e_se1g_dadi, ""),
    pn12e_sa10ab_dtdi = na_if(pn12e_sa10ab_dtdi, ""),
    pn12e_sa20ab_dati = na_if(pn12e_sa20ab_dati, ""),
  ) %>%
  mutate(date = coalesce(substr(pn12e_se1g_dadi, 1, 8), substr(pn12e_sa10ab_dtdi, 1, 8), substr(pn12e_sa20ab_dati, 1, 8))) %>%
  mutate(sampleid = paste0("pn12_se1g_", pn12e_idbs, "_e")) %>%
  dplyr::select(sampleid, date)

pn56e_se1g_date <- pn56e_se1g %>%
  mutate(
    pn56e_se1g_dadi = na_if(pn56e_se1g_dadi, ""),
    pn56e_sa10ab_dtdi = na_if(pn56e_sa10ab_dtdi, ""),
    pn56e_sa20ab_dati = na_if(pn56e_sa20ab_dati, ""),
  ) %>%
  mutate(date = coalesce(substr(pn56e_se1g_dadi, 1, 8), substr(pn56e_sa10ab_dtdi, 1, 8), substr(pn56e_sa20ab_dati, 1, 8))) %>%
  mutate(sampleid = paste0("pn56_se1g_", pn56e_idbs, "_e")) %>%
  dplyr::select(sampleid, date)


combined_se1g_date <- bind_rows(
  incl_se1g_date,
  tri3_se1g_date,
  pn12e_se1g_date,
  pn56e_se1g_date,
  pn12m_se1g_date,
  pn56m_se1g_date
)

dist_df_combined <- rbind(dist_df1, dist_df2, dist_df3, dist_df4)
# Convert the date column to Date type if it's not already
combined_se1g_date <- combined_se1g_date %>%
  mutate(date = as.Date(date))

# Print the structure of combined_se1g_date to check column names
print(str(combined_se1g_date))

# Merge datasets to get date1 and date2
dist_df_combined <- dist_df_combined %>%
  # Merge to get date1
  left_join(
    combined_se1g_date %>%
      dplyr::select(sampleid, date) %>%
      mutate(date1 = date) %>%
      dplyr::select(-date),  # Exclude original date column
    by = c("Sample1" = "sampleid")
  ) %>%
  # Merge to get date2
  left_join(
    combined_se1g_date %>%
      dplyr::select(sampleid, date) %>%
      mutate(date2 = date) %>%
      dplyr::select(-date),  # Exclude original date column
    by = c("Sample2" = "sampleid")
  )

# Print the structure of dist_df_combined to check if date1 and date2 are added
print(str(dist_df_combined))

# Ensure date columns are in Date format and calculate the absolute difference
dist_df_combined <- dist_df_combined %>%
  mutate(
    date1 = as.Date(date1),
    date2 = as.Date(date2),
    # Calculate the absolute difference in days
    date_diff = abs(as.numeric(difftime(date1, date2, units = "days")))
  )

# Extract idbs from Sample1 and Sample2
dist_df_combined <- dist_df_combined %>%
  mutate(idbs1 = sub(".*_(\\d+)", "\\1", Sample1),
         idbs2 = sub(".*_(\\d+)", "\\1", Sample2))

# Filter to keep only intra-subject distances
dist_df_combined_intra <- dist_df_combined %>%
  filter(idbs1 == idbs2) %>%
  drop_na()
```



```{r}

# Initialize a list to store results
results <- list()

# Fit the model for each unique group level
for (group_level in unique(dist_df_combined_intra$group)) {
  # Filter the data for the current group level
  group_data <- dist_df_combined_intra %>%
    filter(group == group_level)
  
  # Check if there are enough data points to fit the model
  if (nrow(group_data) > 1) {
    # Rescale variables
    group_data <- group_data %>%
      mutate(
        transformed_dist_scaled = scale(transformed_dist, center = TRUE, scale = TRUE),
        date_diff_scaled = scale(date_diff, center = TRUE, scale = TRUE)
      )
    
    # Fit the mixed-effects model with tryCatch to handle potential errors
    model <- tryCatch(
      lmer(transformed_dist_scaled ~ date_diff_scaled + (1 | idbs1), data = group_data),
      error = function(e) NULL
    )
    
    if (!is.null(model)) {
      # Extract beta coefficients and CI using broom.mixed
      model_summary <- tidy(model, conf.int = TRUE, conf.level = 0.95)
      
      # Filter for date_diff_scaled coefficient
      beta_coefficient <- model_summary %>%
        filter(term == "date_diff_scaled") %>%
        dplyr::select(estimate, conf.low, conf.high)
      
      # Store the result in the list
      results[[group_level]] <- beta_coefficient
    } else {
      warning(paste("Model did not converge for group:", group_level))
    }
  }
}

# Convert results list to a data frame
results_df <- bind_rows(
  lapply(names(results), function(group) {
    data.frame(
      group = group,
      beta_coefficient = results[[group]]$estimate,
      conf_int_lower = results[[group]]$conf.low,
      conf_int_upper = results[[group]]$conf.high
    )
  })
)

# Print the results data frame
print(results_df)

intra_dist_plot <- ggplot(results_df, aes(x = beta_coefficient, y = group, color = group, shape = group)) +
  geom_point(size = 4) +
  geom_errorbarh(aes(xmin = `conf_int_lower`, xmax = `conf_int_upper`), height = 0.2) +
  scale_color_manual(values = c("#00A98F", "#C00000", "skyblue", "#f1c232")) +  # Customize colors
  scale_shape_manual(values = c(16, 17, 18, 15)) +  # Customize shapes
  labs(x = "Beta Coefficient Estimates", y = "") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 12)) +
  xlim(0, 0.7)# Customize y-axis text
intra_dist_plot

#svglite::svglite(
#  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/intra_dist_plot.svg",
#  width=6, height=2
#  )
#intra_dist_plot
#dev.off()

model <- lmer(transformed_dist ~ date_diff*group + (1 | idbs1), data = dist_df_combined_intra)

summary(model)

dist_df_combined_intra$group <- relevel(factor(dist_df_combined_intra$group), ref = "Infant-intervention")

```

## F. line plot 
line plot show the change of distance

```{r}

dist_df_combined_intra_mo <- dist_df_combined_intra %>%
  filter(group == "Mother-intervention" | group == "Mother-control") %>%
  mutate(Intervention = if_else(group == "Mother-intervention", "Intervention", "Control")) %>%
  mutate(tp1 = str_extract(Sample1, "^[^_]+"),  # Extracts the first part before "_"
         tp2 = str_extract(Sample2, "^[^_]+")) %>%
  filter((tp1 == "incl" & tp2 == "tri3") |
         (tp1 == "tri3" & tp2 == "pn12") |
         (tp1 == "pn12" & tp2 == "pn56")) %>%
  mutate(
    timediff = case_when(  # Use case_when for readability
      tp2 == "tri3" ~ "1",
      tp2 == "pn12" ~ "2",
      tp2 == "pn56" ~ "3",
      TRUE ~ NA_character_  # Ensure no missing values are introduced
    )
  )

# Calculate summary statistics including SE and CI
summary_stats_mo <- dist_df_combined_intra_mo %>%
  group_by(Intervention, timediff) %>%
  dplyr::summarise(
    mean_dist = mean(dist, na.rm = TRUE),
    sd_dist = sd(dist, na.rm = TRUE),
    n = n(),  # Number of observations
    se = sd_dist / sqrt(n),  # Standard Error
    ci = se * qt(0.975, df = n - 1)  # 95% Confidence Interval
  )

# Print the summarized data
print(summary_stats_mo)

# Create the plot with error bars
dist_plot_mo <- ggplot(summary_stats_mo, aes(x = timediff, y = mean_dist, color = Intervention, group = Intervention)) +
  geom_line(size = 1) +
  geom_point(size = 3, position = position_dodge(0.05)) +
  geom_errorbar(aes(ymin = mean_dist - se, ymax = mean_dist + se), width = 0.2, alpha = 0.5, position = position_dodge(0.05)) +
  labs(x = "Time point Interval ", y = "Transformed Distance", color = "Group") +
scale_color_manual(values = c("Intervention" = "steelblue", "Control" = "darkred")) +  theme_bw() +
  theme(axis.text.y = element_text(size = 12),
        axis.text.x = element_text(size = 12),
        legend.title = element_text(size = 12),
        legend.text = element_text(size = 10)) 

# Print the plot
print(dist_plot_mo)

#svglite::svglite(
#  filename="/Users/lishideng/Library/CloudStorage/OneDrive-SharedLibraries-UGent/BioLizard - General/manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/dist_plot_mo_feb14.svg", width=4.2, height=3)
#dist_plot_mo
#dev.off()

dist_df_combined_intra_mo_model <- lmer(dist ~ timediff * group + (1 | idbs1), data = dist_df_combined_intra_mo)
summary(dist_df_combined_intra_mo_model)

anova(dist_df_combined_intra_mo_model, type = "III")  # Type III ANOVA (useful for interactions)


```


```{r}

dist_df_combined_intra_inf <- dist_df_combined_intra %>%
  filter(group == "Infant-intervention" | group == "Infant-control") %>%
  mutate(Intervention = if_else(group == "Infant-intervention", "Intervention", "Control")) %>%
    mutate(tp1 = str_extract(Sample1, "^[^_]+"),  # Extracts the first part before "_"
         tp2 = str_extract(Sample2, "^[^_]+")) %>%
  filter(
         (tp1 == "pn12" & tp2 == "pn56"))


# Calculate summary statistics including SE and CI
summary_stats_inf <- dist_df_combined_intra_inf %>%
  group_by(Intervention, timediff) %>%
  dplyr::summarise(
    mean_dist = mean(dist, na.rm = TRUE),
    sd_dist = sd(dist, na.rm = TRUE),
    n = n(),  # Number of observations
    se = sd_dist / sqrt(n),  # Standard Error
    ci = se * qt(0.975, df = n - 1)  # 95% Confidence Interval
  )

# Print the summarized data
print(summary_stats_inf)

# Create the plot with error bars
dist_plot_inf <- ggplot(summary_stats_inf, aes(x = timediff, y = mean_dist, color = Intervention, group = Intervention)) +
  geom_line(size = 1, linetype = "dashed") +
  geom_point(size = 3, position = position_dodge(0.05)) +
  geom_errorbar(aes(ymin = mean_dist - se, ymax = mean_dist + se), linetype = "dashed", width = 0.2, alpha = 0.5, position = position_dodge(0.05)) +
  labs(x = "Time Point Interval", y = "UniFrac Distance", color = "Group") +
scale_color_manual(values = c("Intervention" = "steelblue", "Control" = "darkred")) +  theme_bw() +
  theme(axis.text.y = element_text(size = 12),
        axis.text.x = element_text(size = 12),
        legend.title = element_text(size = 12),
        legend.text = element_text(size = 10)) # Optionally ensure y-axis starts at 0

# Print the plot
print(dist_plot_inf)

#svglite::svglite(
#  filename="/Users/lishideng/Library/CloudStorage/OneDrive-SharedLibraries-UGent/BioLizard - General/manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/dist_plot_inf_Jan17.svg", width=3.5, height=3)
#dist_plot_inf
#dev.off()

model7 <- lmer(transformed_dist ~ date_diff*group + (1 | Sample1), data = dist_df_combined)
summary(model7)


dist_df_combined <- rbind(dist_df1, dist_df2, dist_df3, dist_df4)
model5 <- lmer(transformed_dist ~ timediff*group + (1 | Sample1), data = dist_df_combined)
summary(model5)

# Relevel the group factor so that "Mother-control" is the reference level
dist_df_combined$group <- relevel(factor(dist_df_combined$group), ref = "Mother-control")
model6 <- lmer(transformed_dist ~ timediff * group + (1 | Sample1), data = dist_df_combined)
summary(model6)

# Define a function to extract the required values from a model
extract_model_stats <- function(model, model_name) {
  model_summary <- summary(model)
  conf_intervals <- confint(model, level = 0.95)
  
  data.frame(
    Model = model_name,
    Effect = "timediff",
    Estimate = model_summary$coefficients["timediff", "Estimate"],
    `2.5% CI` = conf_intervals["timediff", "2.5 %"],
    `97.5% CI` = conf_intervals["timediff", "97.5 %"],
    `P-value` = model_summary$coefficients["timediff", "Pr(>|t|)"]
  )
}

# Apply the function to all models
results_list <- list(
  extract_model_stats(model1, "model1"),
  extract_model_stats(model2, "model2"),
  extract_model_stats(model3, "model3"),
  extract_model_stats(model4, "model4")
)

# Combine all results into one data frame
all_results <- do.call(rbind, results_list)
all_results <- all_results %>%
  mutate(Model = case_when(
    Model == "model1" ~ "Mother - intervention",
    Model == "model2" ~ "Mother - control",
    Model == "model3" ~ "Infant - intervention",
    Model == "model4" ~ "Infant - control"
  ))
ggplot(all_results, aes(x = Estimate, y = Model, color = Model, shape = Model)) +
  geom_point(size = 4) +
  geom_errorbarh(aes(xmin = `X2.5..CI`, xmax = `X97.5..CI`), height = 0.2) +
  scale_color_manual(values = c("#00A98F", "#C00000", "skyblue", "#f1c232")) +  # Customize colors
  scale_shape_manual(values = c(16, 17, 18, 15)) +  # Customize shapes
  labs(x = "Beta Coefficient Estimates", y = "") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 12))  # Customize y-axis text


```


# Figure 4. co-occurrence
## incl 
```{r}
# subset data
incl_bep <- subset_samples(incl, code=="B Z")
incl_ifa <- subset_samples(incl, code=="A Y")

# filter data
filtered_incl_bep <- filter_taxa(incl_bep, function(x) sum(x > 0.01) > (0.05 * length(x)), TRUE)
filtered_incl_ifa <- filter_taxa(incl_ifa, function(x) sum(x > 0.01) > (0.05 * length(x)), TRUE)
otu_data_filtered_incl_bep <- otu_table(filtered_incl_bep)
otu_data_filtered_incl_ifa <- otu_table(filtered_incl_ifa)

# model: time-consuming
spiec_result_incl_bep <- spiec.easi(filtered_incl_bep, method = "mb", lambda.min.ratio = 1e-2, nlambda = 20)
spiec_result_incl_ifa <- spiec.easi(filtered_incl_ifa, method = "mb", lambda.min.ratio = 1e-2, nlambda = 20)

# plot and density/diameter
adj_matrix_incl_bep <- as.matrix(symBeta(getOptBeta(spiec_result_incl_bep), mode = "maxabs"))
adj_matrix_incl_bep[adj_matrix_incl_bep < 0] <- 0
net_incl_bep <- graph_from_adjacency_matrix(adj_matrix_incl_bep, mode = "undirected", weighted = TRUE, diag = FALSE)
plot(net_incl_bep, vertex.label = NA, vertex.size = 5, edge.width = E(net_incl_bep)$weight * 5, edge.color = "gray50")
degree_values <- degree(net_incl_bep)
betweenness_values <- betweenness(net_incl_bep)
closeness_values <- closeness(net_incl_bep)

adj_matrix_incl_ifa <- as.matrix(symBeta(getOptBeta(spiec_result_incl_ifa), mode = "maxabs"))
adj_matrix_incl_ifa[adj_matrix_incl_ifa < 0] <- 0
net_incl_ifa <- graph_from_adjacency_matrix(adj_matrix_incl_ifa, mode = "undirected", weighted = TRUE, diag = FALSE)
plot(net_incl_ifa, vertex.label = NA, vertex.size = 5, edge.width = E(net_incl_ifa)$weight * 5, edge.color = "gray50")
degree_values <- degree(net_incl_ifa)
betweenness_values <- betweenness(net_incl_ifa)
closeness_values <- closeness(net_incl_ifa)

#  Community Detection
louvain_communities <- cluster_louvain(net_incl_bep)
plot(louvain_communities, net_incl_bep, vertex.size = 5, vertex.label = NA, main = "Louvain Community Detection")

louvain_communities <- cluster_louvain(net_incl_ifa)
plot(louvain_communities, net_incl_ifa, vertex.size = 5, vertex.label = NA, main = "Louvain Community Detection")

```

```{r}
# Community Detection and Layout for B Z Group
louvain_communities_bep <- cluster_louvain(net_incl_bep)
layout_bep <- layout_with_fr(net_incl_bep, niter = 1000)  # Increase iterations for better distribution
layout_bep_scaled <- layout_bep * 1000  # Multiply the coordinates by a factor to increase spacing between nodes

# Plot the Louvain communities with scaled layout
svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/coOccurrence_incl_bep.svg",
  width=10, height=10
)

plot(louvain_communities_bep, net_incl_bep,
     layout = layout_bep_scaled,        # Apply the scaled layout for longer edges
     vertex.size = degree(net_incl_bep) * 0.2,  # Node size proportional to degree (number of connections)
     vertex.label = NA,                  # No vertex labels
     edge.width = E(net_incl_bep)$weight * 1, # Edge width proportional to weight
     edge.color = "gray",
     mark.groups = NULL,
     main = "Louvain Community Detection (B Z Group - Longer Edges)")

dev.off()


# Community Detection and Layout for A Y Group
louvain_communities_ifa <- cluster_louvain(net_incl_ifa)
layout_ifa <- layout_with_fr(net_incl_ifa, niter = 1000)  # Increase iterations for better distribution
layout_ifa_scaled <- layout_ifa * 1000  # Multiply the coordinates by a factor to increase spacing between nodes

# Plot the Louvain communities with scaled layout
svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/coOccurrence_incl_ifa.svg",
  width=10, height=10
)

plot(louvain_communities_ifa, net_incl_ifa,
     layout = layout_ifa,               # Apply the more spread-out layout
     vertex.size = degree(net_incl_ifa) * 0.2,  # Node size proportional to degree (number of connections)
     vertex.label = NA,                  # No vertex labels
     edge.width = E(net_incl_ifa)$weight * 1, # Edge width proportional to weight
     edge.color = "gray",
     mark.groups = NULL,
     main = "Louvain Community Detection (A Y Group - Larger Layout)")
dev.off()

```



## tri3 
```{r}
# subset data
tri3_bep <- subset_samples(tri3, code=="B Z")
tri3_ifa <- subset_samples(tri3, code=="A Y")

# filter data
filtered_tri3_bep <- filter_taxa(tri3_bep, function(x) sum(x > 0.01) > (0.05 * length(x)), TRUE)
filtered_tri3_ifa <- filter_taxa(tri3_ifa, function(x) sum(x > 0.01) > (0.05 * length(x)), TRUE)
otu_data_filtered_tri3_bep <- otu_table(filtered_tri3_bep)
otu_data_filtered_tri3_ifa <- otu_table(filtered_tri3_ifa)

# model: time-consuming
spiec_result_tri3_bep <- spiec.easi(filtered_tri3_bep, method = "mb", lambda.min.ratio = 1e-2, nlambda = 20)
spiec_result_tri3_ifa <- spiec.easi(filtered_tri3_ifa, method = "mb", lambda.min.ratio = 1e-2, nlambda = 20)

# plot and density/diameter
adj_matrix_tri3_bep <- as.matrix(symBeta(getOptBeta(spiec_result_tri3_bep), mode = "maxabs"))
adj_matrix_tri3_bep[adj_matrix_tri3_bep < 0] <- 0
net_tri3_bep <- graph_from_adjacency_matrix(adj_matrix_tri3_bep, mode = "undirected", weighted = TRUE, diag = FALSE)
degree_values <- degree(net_tri3_bep)
betweenness_values <- betweenness(net_tri3_bep)
closeness_values <- closeness(net_tri3_bep)

adj_matrix_tri3_ifa <- as.matrix(symBeta(getOptBeta(spiec_result_tri3_ifa), mode = "maxabs"))
adj_matrix_tri3_ifa[adj_matrix_tri3_ifa < 0] <- 0
net_tri3_ifa <- graph_from_adjacency_matrix(adj_matrix_tri3_ifa, mode = "undirected", weighted = TRUE, diag = FALSE)
degree_values <- degree(net_tri3_ifa)
betweenness_values <- betweenness(net_tri3_ifa)
closeness_values <- closeness(net_tri3_ifa)

#  Community Detection
louvain_communities <- cluster_louvain(net_tri3_bep)
plot(louvain_communities, net_tri3_bep, vertex.size = 5, vertex.label = NA, main = "Louvain Community Detection")

louvain_communities <- cluster_louvain(net_tri3_ifa)
plot(louvain_communities, net_tri3_ifa, vertex.size = 5, vertex.label = NA, main = "Louvain Community Detection")

```

```{r}
# Community Detection and Layout for B Z Group
louvain_communities_bep <- cluster_louvain(net_tri3_bep)
layout_bep <- layout_with_fr(net_tri3_bep, niter = 1000)  # Increase iterations for better distribution
layout_bep_scaled <- layout_bep * 1000  # Multiply the coordinates by a factor to increase spacing between nodes

# Plot the Louvain communities with scaled layout
svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/coOccurrence_tri3_bep.svg",
  width=10, height=10
)

plot(louvain_communities_bep, net_tri3_bep,
     layout = layout_bep_scaled,        # Apply the scaled layout for longer edges
     vertex.size = degree(net_tri3_bep) * 0.2,  # Node size proportional to degree (number of connections)
     vertex.label = NA,                  # No vertex labels
     edge.width = E(net_tri3_bep)$weight * 1, # Edge width proportional to weight
     edge.color = "gray",
     mark.groups = NULL,
     main = "Louvain Community Detection (B Z Group - Longer Edges)")

dev.off()


# Community Detection and Layout for A Y Group
louvain_communities_ifa <- cluster_louvain(net_tri3_ifa)
layout_ifa <- layout_with_fr(net_tri3_ifa, niter = 1000)  # Increase iterations for better distribution
layout_ifa_scaled <- layout_ifa * 1000  # Multiply the coordinates by a factor to increase spacing between nodes

# Plot the Louvain communities with scaled layout
svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/coOccurrence_tri3_ifa.svg",
  width=10, height=10
)

plot(louvain_communities_ifa, net_tri3_ifa,
     layout = layout_ifa,               # Apply the more spread-out layout
     vertex.size = degree(net_tri3_ifa) * 0.2,  # Node size proportional to degree (number of connections)
     vertex.label = NA,                  # No vertex labels
     edge.width = E(net_tri3_ifa)$weight * 1, # Edge width proportional to weight
     edge.color = "gray",
     mark.groups = NULL,
     main = "Louvain Community Detection (A Y Group - Larger Layout)")
dev.off()

```



## pn12m 
```{r}
# subset data
pn12m_bep <- subset_samples(pn12m, code=="B Z")
pn12m_ifa <- subset_samples(pn12m, code=="A Y")

# filter data
filtered_pn12m_bep <- filter_taxa(pn12m_bep, function(x) sum(x > 0.01) > (0.05 * length(x)), TRUE)
filtered_pn12m_ifa <- filter_taxa(pn12m_ifa, function(x) sum(x > 0.01) > (0.05 * length(x)), TRUE)
otu_data_filtered_pn12m_bep <- otu_table(filtered_pn12m_bep)
otu_data_filtered_pn12m_ifa <- otu_table(filtered_pn12m_ifa)

# model: time-consuming
spiec_result_pn12m_bep <- spiec.easi(filtered_pn12m_bep, method = "mb", lambda.min.ratio = 1e-2, nlambda = 20)
spiec_result_pn12m_ifa <- spiec.easi(filtered_pn12m_ifa, method = "mb", lambda.min.ratio = 1e-2, nlambda = 20)

# plot and density/diameter
adj_matrix_pn12m_bep <- as.matrix(symBeta(getOptBeta(spiec_result_pn12m_bep), mode = "maxabs"))
adj_matrix_pn12m_bep[adj_matrix_pn12m_bep < 0] <- 0
net_pn12m_bep <- graph_from_adjacency_matrix(adj_matrix_pn12m_bep, mode = "undirected", weighted = TRUE, diag = FALSE)
degree_values <- degree(net_pn12m_bep)
betweenness_values <- betweenness(net_pn12m_bep)
closeness_values <- closeness(net_pn12m_bep)

adj_matrix_pn12m_ifa <- as.matrix(symBeta(getOptBeta(spiec_result_pn12m_ifa), mode = "maxabs"))
adj_matrix_pn12m_ifa[adj_matrix_pn12m_ifa < 0] <- 0
net_pn12m_ifa <- graph_from_adjacency_matrix(adj_matrix_pn12m_ifa, mode = "undirected", weighted = TRUE, diag = FALSE)
degree_values <- degree(net_pn12m_ifa)
betweenness_values <- betweenness(net_pn12m_ifa)
closeness_values <- closeness(net_pn12m_ifa)

#  Community Detection
louvain_communities <- cluster_louvain(net_pn12m_bep)
plot(louvain_communities, net_pn12m_bep, vertex.size = 5, vertex.label = NA, main = "Louvain Community Detection")

louvain_communities <- cluster_louvain(net_pn12m_ifa)
plot(louvain_communities, net_pn12m_ifa, vertex.size = 5, vertex.label = NA, main = "Louvain Community Detection")

```

```{r}
# Community Detection and Layout for B Z Group
louvain_communities_bep <- cluster_louvain(net_pn12m_bep)
layout_bep <- layout_with_fr(net_pn12m_bep, niter = 1000)  # Increase iterations for better distribution
layout_bep_scaled <- layout_bep * 1000  # Multiply the coordinates by a factor to increase spacing between nodes

# Plot the Louvain communities with scaled layout
svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/coOccurrence_pn12m_bep.svg",
  width=10, height=10
)

plot(louvain_communities_bep, net_pn12m_bep,
     layout = layout_bep_scaled,        # Apply the scaled layout for longer edges
     vertex.size = degree(net_pn12m_bep) * 0.2,  # Node size proportional to degree (number of connections)
     vertex.label = NA,                  # No vertex labels
     edge.width = E(net_pn12m_bep)$weight * 1, # Edge width proportional to weight
     edge.color = "gray",
     mark.groups = NULL,
     main = "Louvain Community Detection (B Z Group - Longer Edges)")

dev.off()


# Community Detection and Layout for A Y Group
louvain_communities_ifa <- cluster_louvain(net_pn12m_ifa)
layout_ifa <- layout_with_fr(net_pn12m_ifa, niter = 1000)  # Increase iterations for better distribution
layout_ifa_scaled <- layout_ifa * 1000  # Multiply the coordinates by a factor to increase spacing between nodes

# Plot the Louvain communities with scaled layout
svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/coOccurrence_pn12m_ifa.svg",
  width=10, height=10
)

plot(louvain_communities_ifa, net_pn12m_ifa,
     layout = layout_ifa,               # Apply the more spread-out layout
     vertex.size = degree(net_pn12m_ifa) * 0.2,  # Node size proportional to degree (number of connections)
     vertex.label = NA,                  # No vertex labels
     edge.width = E(net_pn12m_ifa)$weight * 1, # Edge width proportional to weight
     edge.color = "gray",
     mark.groups = NULL,
     main = "Louvain Community Detection (A Y Group - Larger Layout)")
dev.off()

```


## pn56m 
```{r}
# subset data
pn56m_bep <- subset_samples(pn56m, code=="B Z")
pn56m_ifa <- subset_samples(pn56m, code=="A Y")

# filter data
filtered_pn56m_bep <- filter_taxa(pn56m_bep, function(x) sum(x > 0.01) > (0.05 * length(x)), TRUE)
filtered_pn56m_ifa <- filter_taxa(pn56m_ifa, function(x) sum(x > 0.01) > (0.05 * length(x)), TRUE)
otu_data_filtered_pn56m_bep <- otu_table(filtered_pn56m_bep)
otu_data_filtered_pn56m_ifa <- otu_table(filtered_pn56m_ifa)

# model: time-consuming
spiec_result_pn56m_bep <- spiec.easi(filtered_pn56m_bep, method = "mb", lambda.min.ratio = 1e-2, nlambda = 20)
spiec_result_pn56m_ifa <- spiec.easi(filtered_pn56m_ifa, method = "mb", lambda.min.ratio = 1e-2, nlambda = 20)

# plot and density/diameter
adj_matrix_pn56m_bep <- as.matrix(symBeta(getOptBeta(spiec_result_pn56m_bep), mode = "maxabs"))
adj_matrix_pn56m_bep[adj_matrix_pn56m_bep < 0] <- 0
net_pn56m_bep <- graph_from_adjacency_matrix(adj_matrix_pn56m_bep, mode = "undirected", weighted = TRUE, diag = FALSE)
degree_values <- degree(net_pn56m_bep)
betweenness_values <- betweenness(net_pn56m_bep)
closeness_values <- closeness(net_pn56m_bep)

adj_matrix_pn56m_ifa <- as.matrix(symBeta(getOptBeta(spiec_result_pn56m_ifa), mode = "maxabs"))
adj_matrix_pn56m_ifa[adj_matrix_pn56m_ifa < 0] <- 0
net_pn56m_ifa <- graph_from_adjacency_matrix(adj_matrix_pn56m_ifa, mode = "undirected", weighted = TRUE, diag = FALSE)
degree_values <- degree(net_pn56m_ifa)
betweenness_values <- betweenness(net_pn56m_ifa)
closeness_values <- closeness(net_pn56m_ifa)

#  Community Detection
louvain_communities <- cluster_louvain(net_pn56m_bep)
plot(louvain_communities, net_pn56m_bep, vertex.size = 5, vertex.label = NA, main = "Louvain Community Detection")

louvain_communities <- cluster_louvain(net_pn56m_ifa)
plot(louvain_communities, net_pn56m_ifa, vertex.size = 5, vertex.label = NA, main = "Louvain Community Detection")

```

```{r}
# Community Detection and Layout for B Z Group
louvain_communities_bep <- cluster_louvain(net_pn56m_bep)
layout_bep <- layout_with_fr(net_pn56m_bep, niter = 1000)  # Increase iterations for better distribution
layout_bep_scaled <- layout_bep * 1000  # Multiply the coordinates by a factor to increase spacing between nodes

# Plot the Louvain communities with scaled layout
svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/coOccurrence_pn56m_bep.svg",
  width=10, height=10
)

plot(louvain_communities_bep, net_pn56m_bep,
     layout = layout_bep_scaled,        # Apply the scaled layout for longer edges
     vertex.size = degree(net_pn56m_bep) * 0.2,  # Node size proportional to degree (number of connections)
     vertex.label = NA,                  # No vertex labels
     edge.width = E(net_pn56m_bep)$weight * 1, # Edge width proportional to weight
     edge.color = "gray",
     mark.groups = NULL,
     main = "Louvain Community Detection (B Z Group - Longer Edges)")

dev.off()


# Community Detection and Layout for A Y Group
louvain_communities_ifa <- cluster_louvain(net_pn56m_ifa)
layout_ifa <- layout_with_fr(net_pn56m_ifa, niter = 1000)  # Increase iterations for better distribution
layout_ifa_scaled <- layout_ifa * 1000  # Multiply the coordinates by a factor to increase spacing between nodes

# Plot the Louvain communities with scaled layout
svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/coOccurrence_pn56m_ifa.svg",
  width=10, height=10
)

plot(louvain_communities_ifa, net_pn56m_ifa,
     layout = layout_ifa,               # Apply the more spread-out layout
     vertex.size = degree(net_pn56m_ifa) * 0.2,  # Node size proportional to degree (number of connections)
     vertex.label = NA,                  # No vertex labels
     edge.width = E(net_pn56m_ifa)$weight * 1, # Edge width proportional to weight
     edge.color = "gray",
     mark.groups = NULL,
     main = "Louvain Community Detection (A Y Group - Larger Layout)")
dev.off()

```


## pn12e 
```{r}
# subset data
pn12e_bep <- subset_samples(pn12e, code=="B Z")
pn12e_ifa <- subset_samples(pn12e, code=="A Y")

# filter data
filtered_pn12e_bep <- filter_taxa(pn12e_bep, function(x) sum(x > 0.01) > (0.05 * length(x)), TRUE)
filtered_pn12e_ifa <- filter_taxa(pn12e_ifa, function(x) sum(x > 0.01) > (0.05 * length(x)), TRUE)
otu_data_filtered_pn12e_bep <- otu_table(filtered_pn12e_bep)
otu_data_filtered_pn12e_ifa <- otu_table(filtered_pn12e_ifa)

# model: time-consuming
spiec_result_pn12e_bep <- spiec.easi(filtered_pn12e_bep, method = "mb", lambda.min.ratio = 1e-2, nlambda = 20)
spiec_result_pn12e_ifa <- spiec.easi(filtered_pn12e_ifa, method = "mb", lambda.min.ratio = 1e-2, nlambda = 20)

# plot and density/diameter
adj_matrix_pn12e_bep <- as.matrix(symBeta(getOptBeta(spiec_result_pn12e_bep), mode = "maxabs"))
adj_matrix_pn12e_bep[adj_matrix_pn12e_bep < 0] <- 0
net_pn12e_bep <- graph_from_adjacency_matrix(adj_matrix_pn12e_bep, mode = "undirected", weighted = TRUE, diag = FALSE)
degree_values <- degree(net_pn12e_bep)
betweenness_values <- betweenness(net_pn12e_bep)
closeness_values <- closeness(net_pn12e_bep)

adj_matrix_pn12e_ifa <- as.matrix(symBeta(getOptBeta(spiec_result_pn12e_ifa), mode = "maxabs"))
adj_matrix_pn12e_ifa[adj_matrix_pn12e_ifa < 0] <- 0
net_pn12e_ifa <- graph_from_adjacency_matrix(adj_matrix_pn12e_ifa, mode = "undirected", weighted = TRUE, diag = FALSE)
degree_values <- degree(net_pn12e_ifa)
betweenness_values <- betweenness(net_pn12e_ifa)
closeness_values <- closeness(net_pn12e_ifa)

#  Community Detection
louvain_communities <- cluster_louvain(net_pn12e_bep)
plot(louvain_communities, net_pn12e_bep, vertex.size = 5, vertex.label = NA, main = "Louvain Community Detection")

louvain_communities <- cluster_louvain(net_pn12e_ifa)
plot(louvain_communities, net_pn12e_ifa, vertex.size = 5, vertex.label = NA, main = "Louvain Community Detection")

```

```{r}
# Community Detection and Layout for B Z Group
louvain_communities_bep <- cluster_louvain(net_pn12e_bep)
layout_bep <- layout_with_fr(net_pn12e_bep, niter = 1000)  # Increase iterations for better distribution
layout_bep_scaled <- layout_bep * 1000  # Multiply the coordinates by a factor to increase spacing between nodes

# Plot the Louvain communities with scaled layout
svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/coOccurrence_pn12e_bep.svg",
  width=5, height=5
)

plot(louvain_communities_bep, net_pn12e_bep,
     layout = layout_bep_scaled,        # Apply the scaled layout for longer edges
     vertex.size = degree(net_pn12e_bep) * 0.2,  # Node size proportional to degree (number of connections)
     vertex.label = NA,                  # No vertex labels
     edge.width = E(net_pn12e_bep)$weight * 2, # Edge width proportional to weight
     edge.color = "gray",
     main = "Louvain Community Detection (B Z Group - Longer Edges)")

dev.off()


# Community Detection and Layout for A Y Group
louvain_communities_ifa <- cluster_louvain(net_pn12e_ifa)
layout_ifa <- layout_with_fr(net_pn12e_ifa, niter = 1000)  # Increase iterations for better distribution
layout_ifa_scaled <- layout_ifa * 1000  # Multiply the coordinates by a factor to increase spacing between nodes

# Plot the Louvain communities with scaled layout
svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/coOccurrence_pn12e_ifa.svg",
  width=5, height=5
)

plot(louvain_communities_ifa, net_pn12e_ifa,
     layout = layout_ifa,               # Apply the more spread-out layout
     vertex.size = degree(net_pn12e_ifa) * 0.2,  # Node size proportional to degree (number of connections)
     vertex.label = NA,                  # No vertex labels
     edge.width = E(net_pn12e_ifa)$weight * 2, # Edge width proportional to weight
     edge.color = "gray",
     main = "Louvain Community Detection (A Y Group - Larger Layout)")
dev.off()

```

## pn56e 
```{r}
# subset data
pn56e_bep <- subset_samples(pn56e, code=="B Z")
pn56e_ifa <- subset_samples(pn56e, code=="A Y")

# filter data
filtered_pn56e_bep <- filter_taxa(pn56e_bep, function(x) sum(x > 0.01) > (0.05 * length(x)), TRUE)
filtered_pn56e_ifa <- filter_taxa(pn56e_ifa, function(x) sum(x > 0.01) > (0.05 * length(x)), TRUE)
otu_data_filtered_pn56e_bep <- otu_table(filtered_pn56e_bep)
otu_data_filtered_pn56e_ifa <- otu_table(filtered_pn56e_ifa)

# model: time-consuming
spiec_result_pn56e_bep <- spiec.easi(filtered_pn56e_bep, method = "mb", lambda.min.ratio = 1e-2, nlambda = 20)
spiec_result_pn56e_ifa <- spiec.easi(filtered_pn56e_ifa, method = "mb", lambda.min.ratio = 1e-2, nlambda = 20)

# plot and density/diameter
adj_matrix_pn56e_bep <- as.matrix(symBeta(getOptBeta(spiec_result_pn56e_bep), mode = "maxabs"))
adj_matrix_pn56e_bep[adj_matrix_pn56e_bep < 0] <- 0
net_pn56e_bep <- graph_from_adjacency_matrix(adj_matrix_pn56e_bep, mode = "undirected", weighted = TRUE, diag = FALSE)
degree_values <- degree(net_pn56e_bep)
betweenness_values <- betweenness(net_pn56e_bep)
closeness_values <- closeness(net_pn56e_bep)

adj_matrix_pn56e_ifa <- as.matrix(symBeta(getOptBeta(spiec_result_pn56e_ifa), mode = "maxabs"))
adj_matrix_pn56e_ifa[adj_matrix_pn56e_ifa < 0] <- 0
net_pn56e_ifa <- graph_from_adjacency_matrix(adj_matrix_pn56e_ifa, mode = "undirected", weighted = TRUE, diag = FALSE)
degree_values <- degree(net_pn56e_ifa)
betweenness_values <- betweenness(net_pn56e_ifa)
closeness_values <- closeness(net_pn56e_ifa)

#  Community Detection
louvain_communities <- cluster_louvain(net_pn56e_bep)
plot(louvain_communities, net_pn56e_bep, vertex.size = 5, vertex.label = NA, main = "Louvain Community Detection")

louvain_communities <- cluster_louvain(net_pn56e_ifa)
plot(louvain_communities, net_pn56e_ifa, vertex.size = 5, vertex.label = NA, main = "Louvain Community Detection")

```

```{r}
# Community Detection and Layout for B Z Group
louvain_communities_bep <- cluster_louvain(net_pn56e_bep)
layout_bep <- layout_with_fr(net_pn56e_bep, niter = 1000)  # Increase iterations for better distribution
layout_bep_scaled <- layout_bep * 1000  # Multiply the coordinates by a factor to increase spacing between nodes

# Plot the Louvain communities with scaled layout
svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/coOccurrence_pn56e_bep.svg",
  width=5, height=5
)

plot(louvain_communities_bep, net_pn56e_bep,
     layout = layout_bep_scaled,        # Apply the scaled layout for longer edges
     vertex.size = degree(net_pn56e_bep) * 0.2,  # Node size proportional to degree (number of connections)
     vertex.label = NA,                  # No vertex labels
     edge.width = E(net_pn56e_bep)$weight * 2, # Edge width proportional to weight
     edge.color = "gray",
     main = "Louvain Community Detection (B Z Group - Longer Edges)")

dev.off()


# Community Detection and Layout for A Y Group
louvain_communities_ifa <- cluster_louvain(net_pn56e_ifa)
layout_ifa <- layout_with_fr(net_pn56e_ifa, niter = 1000)  # Increase iterations for better distribution
layout_ifa_scaled <- layout_ifa * 1000  # Multiply the coordinates by a factor to increase spacing between nodes

# Plot the Louvain communities with scaled layout
svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/coOccurrence_pn56e_ifa.svg",
  width=5, height=5
)

plot(louvain_communities_ifa, net_pn56e_ifa,
     layout = layout_ifa,               # Apply the more spread-out layout
     vertex.size = degree(net_pn56e_ifa) * 0.2,  # Node size proportional to degree (number of connections)
     vertex.label = NA,                  # No vertex labels
     edge.width = E(net_pn56e_ifa)$weight * 2, # Edge width proportional to weight
     edge.color = "gray",
     main = "Louvain Community Detection (A Y Group - Larger Layout)")
dev.off()

```


## permutation analysis
### nodes, edges, coefficient, density, diameter, and community
```{r}
vcount(net_incl_bep)
ecount(net_incl_bep)
weighted_clustering_bep <- transitivity(net_incl_bep, type = "global")
cat("Weighted Clustering Coefficient (BEP):", weighted_clustering_bep, "\n")
cat("Network Density: ", edge_density(net_incl_bep), "\n")
cat("Network Diameter: ", diameter(net_incl_bep), "\n")
louvain_communities <- cluster_louvain(net_incl_bep)
cat("Louvain communities:", length(louvain_communities), "\n")
cat("Modularity (Louvain):", modularity(louvain_communities), "\n")
cat("Sizes of Louvain communities:", sizes(louvain_communities), "\n")
avg_path_length_bep <- average.path.length(net_incl_bep, directed = FALSE, unconnected = TRUE)
cat("Average Path Length (Intervention Group):", avg_path_length_bep, "\n")


vcount(net_incl_ifa)
ecount(net_incl_ifa)
weighted_clustering_ifa <- transitivity(net_incl_ifa, type = "global")
cat("Weighted Clustering Coefficient (IFA):", weighted_clustering_ifa, "\n")
cat("Network Density: ", edge_density(net_incl_ifa), "\n")
cat("Network Diameter: ", diameter(net_incl_ifa), "\n")
louvain_communities <- cluster_louvain(net_incl_ifa)
cat("Louvain communities:", length(louvain_communities), "\n")
cat("Modularity (Louvain):", modularity(louvain_communities), "\n")
cat("Sizes of Louvain communities:", sizes(louvain_communities), "\n")
avg_path_length_bep <- average.path.length(net_incl_ifa, directed = FALSE, unconnected = TRUE)
cat("Average Path Length (Intervention Group):", avg_path_length_ifa, "\n")

vcount(net_tri3_bep)
ecount(net_tri3_bep)
weighted_clustering_bep <- transitivity(net_tri3_bep, type = "global")
cat("Weighted Clustering Coefficient (BEP):", weighted_clustering_bep, "\n")
cat("Network Density: ", edge_density(net_tri3_bep), "\n")
cat("Network Diameter: ", diameter(net_tri3_bep), "\n")
louvain_communities <- cluster_louvain(net_tri3_bep)
cat("Louvain communities:", length(louvain_communities), "\n")
cat("Modularity (Louvain):", modularity(louvain_communities), "\n")
cat("Sizes of Louvain communities:", sizes(louvain_communities), "\n")
avg_path_length_bep <- average.path.length(net_tri3_bep, directed = FALSE, unconnected = TRUE)
cat("Average Path Length (Intervention Group):", avg_path_length_bep, "\n")

vcount(net_tri3_ifa)
ecount(net_tri3_ifa)
weighted_clustering_ifa <- transitivity(net_tri3_ifa, type = "global")
cat("Weighted Clustering Coefficient (IFA):", weighted_clustering_ifa, "\n")
cat("Network Density: ", edge_density(net_tri3_ifa), "\n")
cat("Network Diameter: ", diameter(net_tri3_ifa), "\n")
louvain_communities <- cluster_louvain(net_tri3_ifa)
cat("Louvain communities:", length(louvain_communities), "\n")
cat("Modularity (Louvain):", modularity(louvain_communities), "\n")
cat("Sizes of Louvain communities:", sizes(louvain_communities), "\n")
avg_path_length_bep <- average.path.length(net_tri3_ifa, directed = FALSE, unconnected = TRUE)
cat("Average Path Length (Intervention Group):", avg_path_length_ifa, "\n")

vcount(net_pn12m_bep)
ecount(net_pn12m_bep)
weighted_clustering_bep <- transitivity(net_pn12m_bep, type = "global")
cat("Weighted Clustering Coefficient (BEP):", weighted_clustering_bep, "\n")
cat("Network Density: ", edge_density(net_pn12m_bep), "\n")
cat("Network Diameter: ", diameter(net_pn12m_bep), "\n")
louvain_communities <- cluster_louvain(net_pn12m_bep)
cat("Louvain communities:", length(louvain_communities), "\n")
cat("Modularity (Louvain):", modularity(louvain_communities), "\n")
cat("Sizes of Louvain communities:", sizes(louvain_communities), "\n")
avg_path_length_bep <- average.path.length(net_pn12m_bep, directed = FALSE, unconnected = TRUE)
cat("Average Path Length (Intervention Group):", avg_path_length_bep, "\n")

vcount(net_pn12m_ifa)
ecount(net_pn12m_ifa)
weighted_clustering_ifa <- transitivity(net_pn12m_ifa, type = "global")
cat("Weighted Clustering Coefficient (IFA):", weighted_clustering_ifa, "\n")
cat("Network Density: ", edge_density(net_pn12m_ifa), "\n")
cat("Network Diameter: ", diameter(net_pn12m_ifa), "\n")
louvain_communities <- cluster_louvain(net_pn12m_ifa)
cat("Louvain communities:", length(louvain_communities), "\n")
cat("Modularity (Louvain):", modularity(louvain_communities), "\n")
cat("Sizes of Louvain communities:", sizes(louvain_communities), "\n")
avg_path_length_bep <- average.path.length(net_pn12m_ifa, directed = FALSE, unconnected = TRUE)
cat("Average Path Length (Intervention Group):", avg_path_length_ifa, "\n")

vcount(net_pn56m_bep)
ecount(net_pn56m_bep)
weighted_clustering_bep <- transitivity(net_pn56m_bep, type = "global")
cat("Weighted Clustering Coefficient (BEP):", weighted_clustering_bep, "\n")
cat("Network Density: ", edge_density(net_pn56m_bep), "\n")
cat("Network Diameter: ", diameter(net_pn56m_bep), "\n")
louvain_communities <- cluster_louvain(net_pn56m_bep)
cat("Louvain communities:", length(louvain_communities), "\n")
cat("Modularity (Louvain):", modularity(louvain_communities), "\n")
cat("Sizes of Louvain communities:", sizes(louvain_communities), "\n")
avg_path_length_bep <- average.path.length(net_pn56m_bep, directed = FALSE, unconnected = TRUE)
cat("Average Path Length (Intervention Group):", avg_path_length_bep, "\n")

vcount(net_pn56m_ifa)
ecount(net_pn56m_ifa)
weighted_clustering_ifa <- transitivity(net_pn56m_ifa, type = "global")
cat("Weighted Clustering Coefficient (IFA):", weighted_clustering_ifa, "\n")
cat("Network Density: ", edge_density(net_pn56m_ifa), "\n")
cat("Network Diameter: ", diameter(net_pn56m_ifa), "\n")
louvain_communities <- cluster_louvain(net_pn56m_ifa)
cat("Louvain communities:", length(louvain_communities), "\n")
cat("Modularity (Louvain):", modularity(louvain_communities), "\n")
cat("Sizes of Louvain communities:", sizes(louvain_communities), "\n")
avg_path_length_bep <- average.path.length(net_pn56m_ifa, directed = FALSE, unconnected = TRUE)
cat("Average Path Length (Intervention Group):", avg_path_length_ifa, "\n")

vcount(net_pn12e_bep)
ecount(net_pn12e_bep)
weighted_clustering_bep <- transitivity(net_pn12e_bep, type = "global")
cat("Weighted Clustering Coefficient (BEP):", weighted_clustering_bep, "\n")
cat("Network Density: ", edge_density(net_pn12e_bep), "\n")
cat("Network Diameter: ", diameter(net_pn12e_bep), "\n")
louvain_communities <- cluster_louvain(net_pn12e_bep)
cat("Louvain communities:", length(louvain_communities), "\n")
cat("Modularity (Louvain):", modularity(louvain_communities), "\n")
cat("Sizes of Louvain communities:", sizes(louvain_communities), "\n")
avg_path_length_bep <- average.path.length(net_pn12e_bep, directed = FALSE, unconnected = TRUE)
cat("Average Path Length (Intervention Group):", avg_path_length_bep, "\n")

vcount(net_pn12e_ifa)
ecount(net_pn12e_ifa)
weighted_clustering_ifa <- transitivity(net_pn12e_ifa, type = "global")
cat("Weighted Clustering Coefficient (IFA):", weighted_clustering_ifa, "\n")
cat("Network Density: ", edge_density(net_pn12e_ifa), "\n")
cat("Network Diameter: ", diameter(net_pn12e_ifa), "\n")
louvain_communities <- cluster_louvain(net_pn12e_ifa)
cat("Louvain communities:", length(louvain_communities), "\n")
cat("Modularity (Louvain):", modularity(louvain_communities), "\n")
cat("Sizes of Louvain communities:", sizes(louvain_communities), "\n")
avg_path_length_bep <- average.path.length(net_pn12e_ifa, directed = FALSE, unconnected = TRUE)
cat("Average Path Length (Intervention Group):", avg_path_length_ifa, "\n")

vcount(net_pn56e_bep)
ecount(net_pn56e_bep)
weighted_clustering_bep <- transitivity(net_pn56e_bep, type = "global")
cat("Weighted Clustering Coefficient (BEP):", weighted_clustering_bep, "\n")
cat("Network Density: ", edge_density(net_pn56e_bep), "\n")
cat("Network Diameter: ", diameter(net_pn56e_bep), "\n")
louvain_communities <- cluster_louvain(net_pn56e_bep)
cat("Louvain communities:", length(louvain_communities), "\n")
cat("Modularity (Louvain):", modularity(louvain_communities), "\n")
cat("Sizes of Louvain communities:", sizes(louvain_communities), "\n")
avg_path_length_bep <- average.path.length(net_pn56e_bep, directed = FALSE, unconnected = TRUE)
cat("Average Path Length (Intervention Group):", avg_path_length_bep, "\n")

vcount(net_pn56e_ifa)
ecount(net_pn56e_ifa)
weighted_clustering_ifa <- transitivity(net_pn56e_ifa, type = "global")
cat("Weighted Clustering Coefficient (IFA):", weighted_clustering_ifa, "\n")
cat("Network Density: ", edge_density(net_pn56e_ifa), "\n")
cat("Network Diameter: ", diameter(net_pn56e_ifa), "\n")
louvain_communities <- cluster_louvain(net_pn56e_ifa)
cat("Louvain communities:", length(louvain_communities), "\n")
cat("Modularity (Louvain):", modularity(louvain_communities), "\n")
cat("Sizes of Louvain communities:", sizes(louvain_communities), "\n")
avg_path_length_bep <- average.path.length(net_pn56e_ifa, directed = FALSE, unconnected = TRUE)
cat("Average Path Length (Intervention Group):", avg_path_length_ifa, "\n")

```


### density comparison
```{r}
# Function to calculate network density
calculate_density <- function(network) {
  return(edge_density(network))
}

# Function to perform permutation test between two networks
permutation_test_density <- function(net_bep, net_ifa, num_permutations = 1000) {
  # Calculate the observed density difference
  density_bep <- calculate_density(net_bep)
  density_ifa <- calculate_density(net_ifa)
  observed_difference <- density_bep - density_ifa
  
  # Initialize vector to store permuted differences
  permuted_differences <- numeric(num_permutations)
  
  # Combine networks' edges
  all_edges <- cbind(get.edgelist(net_bep), rep("BEP", ecount(net_bep)))  # Label for BEP edges
  all_edges <- rbind(all_edges, cbind(get.edgelist(net_ifa), rep("IFA", ecount(net_ifa))))  # Label for IFA edges
  
  # Perform permutations
  for (i in 1:num_permutations) {
    # Randomly shuffle the labels (BEP and IFA) of the edges
    permuted_labels <- sample(all_edges[, 3])  # Shuffle the labels
    
    # Recreate permuted BEP and IFA networks
    permuted_bep_edges <- all_edges[permuted_labels == "BEP", 1:2]
    permuted_ifa_edges <- all_edges[permuted_labels == "IFA", 1:2]
    
    # Create graphs for permuted groups
    permuted_net_bep <- graph_from_edgelist(as.matrix(permuted_bep_edges), directed = FALSE)
    permuted_net_ifa <- graph_from_edgelist(as.matrix(permuted_ifa_edges), directed = FALSE)
    
    # Calculate the permuted density difference
    permuted_density_bep <- calculate_density(permuted_net_bep)
    permuted_density_ifa <- calculate_density(permuted_net_ifa)
    permuted_differences[i] <- permuted_density_bep - permuted_density_ifa
  }
  
  # Calculate p-value
  p_value <- mean(abs(permuted_differences) >= abs(observed_difference))
  
  # Return results
  return(list(
    observed_difference = observed_difference,
    permuted_differences = permuted_differences,
    p_value = p_value
  ))
}


# Run the permutation test for network density comparison between BEP and IFA
set.seed(123)
result <- permutation_test_density(net_incl_bep, net_incl_ifa)
cat("Observed Density Difference (BEP - IFA):", result$observed_difference, "\n")
cat("Permutation Test P-value:", result$p_value, "\n")

result <- permutation_test_density(net_tri3_bep, net_tri3_ifa)
cat("Observed Density Difference (BEP - IFA):", result$observed_difference, "\n")
cat("Permutation Test P-value:", result$p_value, "\n")

result <- permutation_test_density(net_pn12m_bep, net_pn12m_ifa)
cat("Observed Density Difference (BEP - IFA):", result$observed_difference, "\n")
cat("Permutation Test P-value:", result$p_value, "\n")

result <- permutation_test_density(net_pn56m_bep, net_pn56m_ifa)
cat("Observed Density Difference (BEP - IFA):", result$observed_difference, "\n")
cat("Permutation Test P-value:", result$p_value, "\n")

result <- permutation_test_density(net_pn12e_bep, net_pn12e_ifa)
cat("Observed Density Difference (BEP - IFA):", result$observed_difference, "\n")
cat("Permutation Test P-value:", result$p_value, "\n")

result <- permutation_test_density(net_pn56e_bep, net_pn56e_ifa)
cat("Observed Density Difference (BEP - IFA):", result$observed_difference, "\n")
cat("Permutation Test P-value:", result$p_value, "\n")


```

### diameter comparison
```{r}
# Function to calculate network diameter
calculate_diameter <- function(network) {
  return(diameter(network))
}

# Function to perform permutation test between two networks for diameter
permutation_test_diameter <- function(net_bep, net_ifa, num_permutations = 1000) {
  # Calculate the observed diameter difference
  diameter_bep <- calculate_diameter(net_bep)
  diameter_ifa <- calculate_diameter(net_ifa)
  observed_difference <- diameter_bep - diameter_ifa
  
  # Initialize vector to store permuted differences
  permuted_differences <- numeric(num_permutations)
  
  # Combine networks' edges
  all_edges <- cbind(get.edgelist(net_bep), rep("BEP", ecount(net_bep)))  # Label for BEP edges
  all_edges <- rbind(all_edges, cbind(get.edgelist(net_ifa), rep("IFA", ecount(net_ifa))))  # Label for IFA edges
  
  # Perform permutations
  for (i in 1:num_permutations) {
    # Randomly shuffle the labels (BEP and IFA) of the edges
    permuted_labels <- sample(all_edges[, 3])  # Shuffle the labels
    
    # Recreate permuted BEP and IFA networks
    permuted_bep_edges <- all_edges[permuted_labels == "BEP", 1:2]
    permuted_ifa_edges <- all_edges[permuted_labels == "IFA", 1:2]
    
    # Create graphs for permuted groups
    permuted_net_bep <- graph_from_edgelist(as.matrix(permuted_bep_edges), directed = FALSE)
    permuted_net_ifa <- graph_from_edgelist(as.matrix(permuted_ifa_edges), directed = FALSE)
    
    # Calculate the permuted diameter difference
    permuted_diameter_bep <- calculate_diameter(permuted_net_bep)
    permuted_diameter_ifa <- calculate_diameter(permuted_net_ifa)
    permuted_differences[i] <- permuted_diameter_bep - permuted_diameter_ifa
  }
  
  # Calculate p-value
  p_value <- mean(abs(permuted_differences) >= abs(observed_difference))
  
  # Return results
  return(list(
    observed_difference = observed_difference,
    permuted_differences = permuted_differences,
    p_value = p_value
  ))
}

# Run the permutation test for network diameter comparison between BEP and IFA
set.seed(123)
result_diameter <- permutation_test_diameter(net_incl_bep, net_incl_ifa)
cat("Observed Diameter Difference (BEP - IFA):", result_diameter$observed_difference, "\n")
cat("Permutation Test P-value:", result_diameter$p_value, "\n")

result_diameter <- permutation_test_diameter(net_tri3_bep, net_tri3_ifa)
cat("Observed Diameter Difference (BEP - IFA):", result_diameter$observed_difference, "\n")
cat("Permutation Test P-value:", result_diameter$p_value, "\n")

result_diameter <- permutation_test_diameter(net_pn12m_bep, net_pn12m_ifa)
cat("Observed Diameter Difference (BEP - IFA):", result_diameter$observed_difference, "\n")
cat("Permutation Test P-value:", result_diameter$p_value, "\n")

result_diameter <- permutation_test_diameter(net_pn56m_bep, net_pn56m_ifa)
cat("Observed Diameter Difference (BEP - IFA):", result_diameter$observed_difference, "\n")
cat("Permutation Test P-value:", result_diameter$p_value, "\n")

result_diameter <- permutation_test_diameter(net_pn12e_bep, net_pn12e_ifa)
cat("Observed Diameter Difference (BEP - IFA):", result_diameter$observed_difference, "\n")
cat("Permutation Test P-value:", result_diameter$p_value, "\n")

result_diameter <- permutation_test_diameter(net_pn56e_bep, net_pn56e_ifa)
cat("Observed Diameter Difference (BEP - IFA):", result_diameter$observed_difference, "\n")
cat("Permutation Test P-value:", result_diameter$p_value, "\n")


```

### modularity comparison
```{r}
# Function to calculate modularity
calculate_modularity <- function(network) {
  # Perform Louvain community detection
  louvain_communities <- cluster_louvain(network)
  # Calculate modularity of the community structure
  return(modularity(louvain_communities))
}

# Function to perform permutation test for modularity between two networks
permutation_test_modularity <- function(net_bep, net_ifa, num_permutations = 10000) {
  # Calculate the observed modularity difference
  modularity_bep <- calculate_modularity(net_bep)
  modularity_ifa <- calculate_modularity(net_ifa)
  observed_difference <- modularity_bep - modularity_ifa
  
  # Initialize vector to store permuted differences
  permuted_differences <- numeric(num_permutations)
  
  # Combine networks' edges
  all_edges <- cbind(get.edgelist(net_bep), rep("BEP", ecount(net_bep)))  # Label for BEP edges
  all_edges <- rbind(all_edges, cbind(get.edgelist(net_ifa), rep("IFA", ecount(net_ifa))))  # Label for IFA edges
  
  # Perform permutations
  for (i in 1:num_permutations) {
    # Randomly shuffle the labels (BEP and IFA) of the edges
    permuted_labels <- sample(all_edges[, 3])  # Shuffle the labels
    
    # Recreate permuted BEP and IFA networks
    permuted_bep_edges <- all_edges[permuted_labels == "BEP", 1:2]
    permuted_ifa_edges <- all_edges[permuted_labels == "IFA", 1:2]
    
    # Create graphs for permuted groups
    permuted_net_bep <- graph_from_edgelist(as.matrix(permuted_bep_edges), directed = FALSE)
    permuted_net_ifa <- graph_from_edgelist(as.matrix(permuted_ifa_edges), directed = FALSE)
    
    # Calculate the permuted modularity difference
    permuted_modularity_bep <- calculate_modularity(permuted_net_bep)
    permuted_modularity_ifa <- calculate_modularity(permuted_net_ifa)
    permuted_differences[i] <- permuted_modularity_bep - permuted_modularity_ifa
  }
  
  # Calculate p-value
  p_value <- mean(abs(permuted_differences) >= abs(observed_difference))
  
  # Return results
  return(list(
    observed_difference = observed_difference,
    permuted_differences = permuted_differences,
    p_value = p_value
  ))
}

# Run the permutation test for modularity comparison between BEP and IFA
set.seed(123)
result_modularity <- permutation_test_modularity(net_incl_bep, net_incl_ifa)
cat("Observed Modularity Difference (BEP - IFA):", result_modularity$observed_difference, "\n")
cat("Permutation Test P-value:", result_modularity$p_value, "\n")

result_modularity <- permutation_test_modularity(net_tri3_bep, net_tri3_ifa)
cat("Observed Modularity Difference (BEP - IFA):", result_modularity$observed_difference, "\n")
cat("Permutation Test P-value:", result_modularity$p_value, "\n")

result_modularity <- permutation_test_modularity(net_pn12m_bep, net_pn12m_ifa)
cat("Observed Modularity Difference (BEP - IFA):", result_modularity$observed_difference, "\n")
cat("Permutation Test P-value:", result_modularity$p_value, "\n")

result_modularity <- permutation_test_modularity(net_pn56m_bep, net_pn56m_ifa)
cat("Observed Modularity Difference (BEP - IFA):", result_modularity$observed_difference, "\n")
cat("Permutation Test P-value:", result_modularity$p_value, "\n")

result_modularity <- permutation_test_modularity(net_pn12e_bep, net_pn12e_ifa)
cat("Observed Modularity Difference (BEP - IFA):", result_modularity$observed_difference, "\n")
cat("Permutation Test P-value:", result_modularity$p_value, "\n")

result_modularity <- permutation_test_modularity(net_pn56e_bep, net_pn56e_ifa)
cat("Observed Modularity Difference (BEP - IFA):", result_modularity$observed_difference, "\n")
cat("Permutation Test P-value:", result_modularity$p_value, "\n")

```

### clustering coefficient
```{r}
# Function to calculate global clustering coefficient (transitivity)
calculate_clustering <- function(network) {
  return(transitivity(network, type = "global"))
}

# Function to perform permutation test for clustering coefficient between two networks
permutation_test_clustering <- function(net_bep, net_ifa, num_permutations = 1000) {
  # Calculate the observed clustering coefficient difference
  clustering_bep <- calculate_clustering(net_bep)
  clustering_ifa <- calculate_clustering(net_ifa)
  observed_difference <- clustering_bep - clustering_ifa
  
  # Initialize vector to store permuted differences
  permuted_differences <- numeric(num_permutations)
  
  # Combine networks' edges
  all_edges <- cbind(get.edgelist(net_bep), rep("BEP", ecount(net_bep)))  # Label for BEP edges
  all_edges <- rbind(all_edges, cbind(get.edgelist(net_ifa), rep("IFA", ecount(net_ifa))))  # Label for IFA edges
  
  # Perform permutations
  for (i in 1:num_permutations) {
    # Randomly shuffle the labels (BEP and IFA) of the edges
    permuted_labels <- sample(all_edges[, 3])  # Shuffle the labels
    
    # Recreate permuted BEP and IFA networks
    permuted_bep_edges <- all_edges[permuted_labels == "BEP", 1:2]
    permuted_ifa_edges <- all_edges[permuted_labels == "IFA", 1:2]
    
    # Create graphs for permuted groups
    permuted_net_bep <- graph_from_edgelist(as.matrix(permuted_bep_edges), directed = FALSE)
    permuted_net_ifa <- graph_from_edgelist(as.matrix(permuted_ifa_edges), directed = FALSE)
    
    # Calculate the permuted clustering coefficient difference
    permuted_clustering_bep <- calculate_clustering(permuted_net_bep)
    permuted_clustering_ifa <- calculate_clustering(permuted_net_ifa)
    permuted_differences[i] <- permuted_clustering_bep - permuted_clustering_ifa
  }
  
  # Calculate p-value
  p_value <- mean(abs(permuted_differences) >= abs(observed_difference))
  
  # Return results
  return(list(
    observed_difference = observed_difference,
    permuted_differences = permuted_differences,
    p_value = p_value
  ))
}

# Run the permutation test for clustering coefficient comparison between BEP and IFA
set.seed(123)
result_clustering <- permutation_test_clustering(net_incl_bep, net_incl_ifa)
cat("Observed Clustering Coefficient Difference (BEP - IFA):", result_clustering$observed_difference, "\n")
cat("Permutation Test P-value:", result_clustering$p_value, "\n")

result_clustering <- permutation_test_clustering(net_tri3_bep, net_tri3_ifa)
cat("Observed Clustering Coefficient Difference (BEP - IFA):", result_clustering$observed_difference, "\n")
cat("Permutation Test P-value:", result_clustering$p_value, "\n")

result_clustering <- permutation_test_clustering(net_pn12m_bep, net_pn12m_ifa)
cat("Observed Clustering Coefficient Difference (BEP - IFA):", result_clustering$observed_difference, "\n")
cat("Permutation Test P-value:", result_clustering$p_value, "\n")

result_clustering <- permutation_test_clustering(net_pn56m_bep, net_pn56m_ifa)
cat("Observed Clustering Coefficient Difference (BEP - IFA):", result_clustering$observed_difference, "\n")
cat("Permutation Test P-value:", result_clustering$p_value, "\n")

result_clustering <- permutation_test_clustering(net_pn12e_bep, net_pn12e_ifa)
cat("Observed Clustering Coefficient Difference (BEP - IFA):", result_clustering$observed_difference, "\n")
cat("Permutation Test P-value:", result_clustering$p_value, "\n")

result_clustering <- permutation_test_clustering(net_pn56e_bep, net_pn56e_ifa)
cat("Observed Clustering Coefficient Difference (BEP - IFA):", result_clustering$observed_difference, "\n")
cat("Permutation Test P-value:", result_clustering$p_value, "\n")

```

### average path length
```{r}
library(igraph)

# Function to calculate the average path length for a network
calculate_avg_path_length <- function(network) {
  return(average.path.length(network, directed = FALSE, unconnected = TRUE))
}

# Function to perform permutation test for average path length
permutation_test_path_length <- function(net_bep, net_ifa, num_permutations = 1000) {
  # Calculate the observed average path length difference
  avg_path_length_bep <- calculate_avg_path_length(net_bep)
  avg_path_length_ifa <- calculate_avg_path_length(net_ifa)
  observed_difference <- avg_path_length_bep - avg_path_length_ifa
  
  # Initialize a vector to store permuted differences
  permuted_differences <- numeric(num_permutations)
  
  # Combine networks' edges
  all_edges <- cbind(get.edgelist(net_bep), rep("BEP", ecount(net_bep)))  # Label for BEP edges
  all_edges <- rbind(all_edges, cbind(get.edgelist(net_ifa), rep("IFA", ecount(net_ifa))))  # Label for IFA edges
  
  # Perform permutations
  for (i in 1:num_permutations) {
    # Randomly shuffle the group labels of the edges
    permuted_labels <- sample(all_edges[, 3])  # Shuffle the labels
    
    # Recreate permuted BEP and IFA networks
    permuted_bep_edges <- all_edges[permuted_labels == "BEP", 1:2]
    permuted_ifa_edges <- all_edges[permuted_labels == "IFA", 1:2]
    
    # Create graphs for permuted groups
    permuted_net_bep <- graph_from_edgelist(as.matrix(permuted_bep_edges), directed = FALSE)
    permuted_net_ifa <- graph_from_edgelist(as.matrix(permuted_ifa_edges), directed = FALSE)
    
    # Calculate the permuted average path length difference
    permuted_avg_path_length_bep <- calculate_avg_path_length(permuted_net_bep)
    permuted_avg_path_length_ifa <- calculate_avg_path_length(permuted_net_ifa)
    permuted_differences[i] <- permuted_avg_path_length_bep - permuted_avg_path_length_ifa
  }
  
  # Calculate p-value
  p_value <- mean(abs(permuted_differences) >= abs(observed_difference))
  
  # Return results
  return(list(
    observed_difference = observed_difference,
    permuted_differences = permuted_differences,
    p_value = p_value
  ))
}

# Run the permutation test for average path length
set.seed(123)
result_path_length <- permutation_test_path_length(net_incl_bep, net_incl_ifa, num_permutations = 1000)
cat("Observed Average Path Length Difference (BEP - IFA):", result_path_length$observed_difference, "\n")
cat("Permutation Test P-value:", result_path_length$p_value, "\n")

result_path_length <- permutation_test_path_length(net_tri3_bep, net_tri3_ifa, num_permutations = 1000)
cat("Observed Average Path Length Difference (BEP - IFA):", result_path_length$observed_difference, "\n")
cat("Permutation Test P-value:", result_path_length$p_value, "\n")

result_path_length <- permutation_test_path_length(net_pn12m_bep, net_pn12m_ifa, num_permutations = 1000)
cat("Observed Average Path Length Difference (BEP - IFA):", result_path_length$observed_difference, "\n")
cat("Permutation Test P-value:", result_path_length$p_value, "\n")

result_path_length <- permutation_test_path_length(net_pn56m_bep, net_pn56m_ifa, num_permutations = 1000)
cat("Observed Average Path Length Difference (BEP - IFA):", result_path_length$observed_difference, "\n")
cat("Permutation Test P-value:", result_path_length$p_value, "\n")

result_path_length <- permutation_test_path_length(net_pn12e_bep, net_pn12e_ifa, num_permutations = 1000)
cat("Observed Average Path Length Difference (BEP - IFA):", result_path_length$observed_difference, "\n")
cat("Permutation Test P-value:", result_path_length$p_value, "\n")

result_path_length <- permutation_test_path_length(net_pn56e_bep, net_pn56e_ifa, num_permutations = 1000)
cat("Observed Average Path Length Difference (BEP - IFA):", result_path_length$observed_difference, "\n")
cat("Permutation Test P-value:", result_path_length$p_value, "\n")

```



# Figure 5. ANCOMBC 
##  A. Separated TP
### incl
```{r}
ancomincl <- ancombc2(data=incl, 
                        assay_name = "counts", tax_level = "species",
                        fix_formula = "code", rand_formula = NULL,
                        p_adj_method = "BH", 
                        prv_cut = 0.1,
                        lib_cut = 1000, s0_perc = 0.05,
                        group = "code", struc_zero = TRUE, neg_lb = TRUE,
                        alpha = 0.05, n_cl = 2, verbose = TRUE,
                        global = F, pairwise = F, dunnet = F, trend = F,
                        iter_control = list(tol=1e-2, max_iter=1, verbose=TRUE),
                        em_control = list(tol=1e-5, max_iter=100),
                        lme_control = lme4::lmerControl(),
                        mdfdr_control = list(fwer_ctrl_method = "BH", B=1000),
                        trend_control = list(contrast=list(matrix(c(1,0,-1,1),
                                                                  nrow = 2,
                                                                  byrow = TRUE)),
                                             node=list(2),
                                             solver="ECOS",
                                             B=10))

incl_primres = ancomincl$res

mother_incl <- incl_primres |>
  mutate(`q_codeB Z`=-log10(`q_codeB Z`)) |>
  mutate(color=if_else(`lfc_codeB Z`>0, "steelblue", "darkred")) |>
  mutate(label=if_else(abs(`lfc_codeB Z`)>0.5 & `q_codeB Z` >1.3 & `passed_ss_codeB Z` == TRUE, taxon, NA_character_)) |>
  mutate(color=if_else(abs(`lfc_codeB Z`)<0.5 & `q_codeB Z` <1.3, "#000000", color)) |>
  mutate(color=if_else(`passed_ss_codeB Z` == FALSE, "grey55", color)) |>
  ggplot(mapping = aes(`lfc_codeB Z`, `q_codeB Z`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 0.5) +
  geom_vline(linetype = "dotted", xintercept = -0.5) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "Mother - Tri2. - Interv. vs. Contr.",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~P)) +
  theme_bw() + xlim(-2, 2) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
mother_incl

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/mother_incl.svg",
  width=4, height=4
  )
mother_incl
dev.off()
```

### tri3
```{r}
ancomtri3 <- ancombc2(data=tri3, 
                        assay_name = "counts", tax_level = "species",
                        fix_formula = "code", rand_formula = NULL,
                        p_adj_method = "BH", 
                        prv_cut = 0.1,
                        lib_cut = 1000, s0_perc = 0.05,
                        group = "code", struc_zero = TRUE, neg_lb = TRUE,
                        alpha = 0.05, n_cl = 2, verbose = TRUE,
                        global = F, pairwise = F, dunnet = F, trend = F,
                        iter_control = list(tol=1e-2, max_iter=1, verbose=TRUE),
                        em_control = list(tol=1e-5, max_iter=100),
                        lme_control = lme4::lmerControl(),
                        mdfdr_control = list(fwer_ctrl_method = "BH", B=1000),
                        trend_control = list(contrast=list(matrix(c(1,0,-1,1),
                                                                  nrow = 2,
                                                                  byrow = TRUE)),
                                             node=list(2),
                                             solver="ECOS",
                                             B=10))

tri3_primres = ancomtri3$res

mother_tri3 <- tri3_primres |>
  mutate(`q_codeB Z`=-log10(`q_codeB Z`)) |>
  mutate(color=if_else(`lfc_codeB Z`>0, "steelblue", "darkred")) |>
  mutate(label=if_else(abs(`lfc_codeB Z`)>0.5 & `q_codeB Z` >1.3 & `passed_ss_codeB Z` == TRUE, species, NA_character_)) |>
  mutate(color=if_else(abs(`lfc_codeB Z`)<0.5 & `q_codeB Z` <1.3, "#000000", color)) |>
  mutate(color=if_else(`passed_ss_codeB Z` == FALSE, "grey55", color)) |>
  ggplot(mapping = aes(`lfc_codeB Z`, `q_codeB Z`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 0.5) +
  geom_vline(linetype = "dotted", xintercept = -0.5) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "Mother - Tri3. - Interv. vs. Contr.",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~P)) +
  theme_bw() + xlim(-2, 2) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
mother_tri3

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/mother_tri3.svg",
  width=4, height=4
  )
mother_tri3
dev.off()
```

### pn12m
```{r}
ancompn12m <- ancombc2(data=pn12m, 
                        assay_name = "counts", tax_level = "species",
                        fix_formula = "code", rand_formula = NULL,
                        p_adj_method = "BH", 
                        prv_cut = 0.1,
                        lib_cut = 1000, s0_perc = 0.05,
                        group = "code", struc_zero = TRUE, neg_lb = TRUE,
                        alpha = 0.05, n_cl = 2, verbose = TRUE,
                        global = F, pairwise = F, dunnet = F, trend = F,
                        iter_control = list(tol=1e-2, max_iter=1, verbose=TRUE),
                        em_control = list(tol=1e-5, max_iter=100),
                        lme_control = lme4::lmerControl(),
                        mdfdr_control = list(fwer_ctrl_method = "BH", B=1000),
                        trend_control = list(contrast=list(matrix(c(1,0,-1,1),
                                                                  nrow = 2,
                                                                  byrow = TRUE)),
                                             node=list(2),
                                             solver="ECOS",
                                             B=10))

pn12m_primres = ancompn12m$res

mother_pn12m <- pn12m_primres |>
  mutate(`q_codeB Z`=-log10(`q_codeB Z`)) |>
  mutate(color=if_else(`lfc_codeB Z`>0, "steelblue", "darkred")) |>
  mutate(label=if_else(abs(`lfc_codeB Z`)>0.5 & `q_codeB Z` >1.3 & `passed_ss_codeB Z` == TRUE, species, NA_character_)) |>
  mutate(color=if_else(abs(`lfc_codeB Z`)<0.5 & `q_codeB Z` <1.3, "#000000", color)) |>
  mutate(color=if_else(`passed_ss_codeB Z` == FALSE, "grey55", color)) |>
  ggplot(mapping = aes(`lfc_codeB Z`, `q_codeB Z`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 0.5) +
  geom_vline(linetype = "dotted", xintercept = -0.5) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "Mother - PN12. - Interv. vs. Contr.",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~P)) +
  theme_bw() + xlim(-2, 2) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
mother_pn12m

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/mother_pn12m.svg",
  width=4, height=4
  )
mother_pn12m
dev.off()
```

### pn56m
```{r}
ancompn56m <- ancombc2(data=pn56m, 
                        assay_name = "counts", tax_level = "species",
                        fix_formula = "code", rand_formula = NULL,
                        p_adj_method = "BH", 
                        prv_cut = 0.1,
                        lib_cut = 1000, s0_perc = 0.05,
                        group = "code", struc_zero = TRUE, neg_lb = TRUE,
                        alpha = 0.05, n_cl = 2, verbose = TRUE,
                        global = F, pairwise = F, dunnet = F, trend = F,
                        iter_control = list(tol=1e-2, max_iter=1, verbose=TRUE),
                        em_control = list(tol=1e-5, max_iter=100),
                        lme_control = lme4::lmerControl(),
                        mdfdr_control = list(fwer_ctrl_method = "BH", B=1000),
                        trend_control = list(contrast=list(matrix(c(1,0,-1,1),
                                                                  nrow = 2,
                                                                  byrow = TRUE)),
                                             node=list(2),
                                             solver="ECOS",
                                             B=10))

pn56m_primres = ancompn56m$res

mother_pn56m <- pn56m_primres |>
  mutate(`q_codeB Z`=-log10(`q_codeB Z`)) |>
  mutate(color=if_else(`lfc_codeB Z`>0, "steelblue", "darkred")) |>
  mutate(label=if_else(abs(`lfc_codeB Z`)>0.5 & `q_codeB Z` >1.3 & `passed_ss_codeB Z` == TRUE, species, NA_character_)) |>
  mutate(color=if_else(abs(`lfc_codeB Z`)<0.5 & `q_codeB Z` <1.3, "#000000", color)) |>
  mutate(color=if_else(`passed_ss_codeB Z` == FALSE, "grey55", color)) |>
  ggplot(mapping = aes(`lfc_codeB Z`, `q_codeB Z`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 0.5) +
  geom_vline(linetype = "dotted", xintercept = -0.5) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "Mother - PN56. - Interv. vs. Contr.",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~P)) +
  theme_bw() + xlim(-2, 2) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
mother_pn56m

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/mother_pn56m.svg",
  width=4, height=4
  )
mother_pn56m
dev.off()

```

### pn12e
```{r}
ancompn12e <- ancombc2(data=pn12e, 
                        assay_name = "counts", tax_level = "species",
                        fix_formula = "code", rand_formula = NULL,
                        p_adj_method = "BH", 
                        prv_cut = 0.1,
                        lib_cut = 1000, s0_perc = 0.05,
                        group = "code", struc_zero = TRUE, neg_lb = TRUE,
                        alpha = 0.05, n_cl = 2, verbose = TRUE,
                        global = F, pairwise = F, dunnet = F, trend = F,
                        iter_control = list(tol=1e-2, max_iter=1, verbose=TRUE),
                        em_control = list(tol=1e-5, max_iter=100),
                        lme_control = lme4::lmerControl(),
                        mdfdr_control = list(fwer_ctrl_method = "BH", B=1000),
                        trend_control = list(contrast=list(matrix(c(1,0,-1,1),
                                                                  nrow = 2,
                                                                  byrow = TRUE)),
                                             node=list(2),
                                             solver="ECOS",
                                             B=10))

pn12e_primres = ancompn12e$res

infant_pn12e <- pn12e_primres |>
  mutate(`q_codeB Z`=-log10(`q_codeB Z`)) |>
  mutate(color=if_else(`lfc_codeB Z`>0, "steelblue", "darkred")) |>
  mutate(label=if_else(abs(`lfc_codeB Z`)>0.5 & `q_codeB Z` >1.3 & `passed_ss_codeB Z` == TRUE, species, NA_character_)) |>
  mutate(color=if_else(abs(`lfc_codeB Z`)<0.5 & `q_codeB Z` <1.3, "#000000", color)) |>
  mutate(color=if_else(`passed_ss_codeB Z` == FALSE, "grey55", color)) |>
  ggplot(mapping = aes(`lfc_codeB Z`, `q_codeB Z`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 0.5) +
  geom_vline(linetype = "dotted", xintercept = -0.5) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "Infant - PN12. - Interv. vs. Contr.",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~P)) +
  theme_bw() + xlim(-2, 2) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
infant_pn12e

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/infant_pn12e.svg",
  width=4, height=4
  )
infant_pn12e
dev.off()
```

### pn56e
```{r}
ancompn56e <- ancombc2(data=pn56e, 
                        assay_name = "counts", tax_level = "species",
                        fix_formula = "code", rand_formula = NULL,
                        p_adj_method = "BH", 
                        prv_cut = 0.1,
                        lib_cut = 1000, s0_perc = 0.05,
                        group = "code", struc_zero = TRUE, neg_lb = TRUE,
                        alpha = 0.05, n_cl = 2, verbose = TRUE,
                        global = F, pairwise = F, dunnet = F, trend = F,
                        iter_control = list(tol=1e-2, max_iter=1, verbose=TRUE),
                        em_control = list(tol=1e-5, max_iter=100),
                        lme_control = lme4::lmerControl(),
                        mdfdr_control = list(fwer_ctrl_method = "BH", B=1000),
                        trend_control = list(contrast=list(matrix(c(1,0,-1,1),
                                                                  nrow = 2,
                                                                  byrow = TRUE)),
                                             node=list(2),
                                             solver="ECOS",
                                             B=10))

pn56e_primres = ancompn56e$res

infant_pn56e <- pn56e_primres |>
  mutate(`q_codeB Z`=-log10(`q_codeB Z`)) |>
  mutate(color=if_else(`lfc_codeB Z`>0, "steelblue", "darkred")) |>
  mutate(label=if_else(abs(`lfc_codeB Z`)>0.5 & `q_codeB Z` >1.3 & `passed_ss_codeB Z` == TRUE, species, NA_character_)) |>
  mutate(color=if_else(abs(`lfc_codeB Z`)<0.5 & `q_codeB Z` <1.3, "#000000", color)) |>
  mutate(color=if_else(`passed_ss_codeB Z` == FALSE, "grey55", color)) |>
  ggplot(mapping = aes(`lfc_codeB Z`, `q_codeB Z`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 0.5) +
  geom_vline(linetype = "dotted", xintercept = -0.5) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "Mother - PN56. - Interv. vs. Contr.",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~P)) +
  theme_bw() + xlim(-2, 2) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
infant_pn56e

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/infant_pn56e.svg",
  width=4, height=4
  )
infant_pn56e
dev.off()


```



## B. Combined TP

### mother 4 TP

```{r}

mother_all_repeated_ancombc <- ancombc2(data=mother_ps, 
                        assay_name = "counts", tax_level = "species",
                        fix_formula = "code + timepoint", rand_formula = "(1 | idbs)",
                        p_adj_method = "BH", pseudo_sens = TRUE,
                        prv_cut = 0.1,
                        lib_cut = 1000, s0_perc = 0.05,
                        group = "timepoint", struc_zero = TRUE, neg_lb = TRUE,
                        alpha = 0.05, n_cl = 2, verbose = TRUE,
                        global = TRUE, pairwise = TRUE, dunnet = TRUE, trend = TRUE,
                        iter_control = list(tol=1e-2, max_iter=20, verbose=TRUE),
                        em_control = list(tol=1e-5, max_iter=100),
                        lme_control = lme4::lmerControl(),
                        mdfdr_control = list(fwer_ctrl_method = "BH", B=1000),
                        trend_control = list(contrast=list(matrix(c(1,-1,0,0,1,-1,-1,0,1), 
                                                                  nrow = 3, byrow = TRUE)),
                                             node=list(2),
                                             solver="ECOS",
                                             B=10)) 
mother_all_repeated_ancombc_output <- mother_all_repeated_ancombc$res

```


### mother 3 TP

```{r}

# mother dataset excluding incl
mother_ps@sam_data$timepoint <- factor(mother_ps@sam_data$timepoint, levels = c("incl", "tri3", "pn12", "pn56"))

mother_repeated <- subset_samples(mother_ps, timepoint !="incl") 

mother_repeated_ancombc <- ancombc2(data=mother_repeated, 
                        assay_name = "counts", tax_level = "species",
                        fix_formula = "code + timepoint", rand_formula = "(1 | idbs)",
                        p_adj_method = "BH", pseudo_sens = TRUE,
                        prv_cut = 0.1,
                        lib_cut = 1000, s0_perc = 0.05,
                        group = "timepoint", struc_zero = TRUE, neg_lb = TRUE,
                        alpha = 0.05, n_cl = 2, verbose = TRUE,
                        global = TRUE, pairwise = TRUE, dunnet = TRUE, trend = TRUE,
                        iter_control = list(tol=1e-2, max_iter=20, verbose=TRUE),
                        em_control = list(tol=1e-5, max_iter=100),
                        lme_control = lme4::lmerControl(),
                        mdfdr_control = list(fwer_ctrl_method = "BH", B=1000),
                        trend_control = list(contrast=list(matrix(c(1,0,-1,1),
                                                                  nrow = 2,
                                                                  byrow = TRUE)),
                                             node=list(2),
                                             solver="ECOS",
                                             B=10)) 
# ancombc output
mother_repeated_ancombc_output <- mother_repeated_ancombc$res %>%
  mutate_if(is.numeric, function(x) round(x, 2))
mother_repeated_ancombc_output$taxon[mother_repeated_ancombc_output$`diff_codeB Z`==TRUE] #0
mother_repeated_ancombc_output$taxon[mother_repeated_ancombc_output$diff_timepointpn56==TRUE & mother_repeated_ancombc_output$passed_ss_timepointpn56==TRUE] 
mother_repeated_ancombc_output$taxon[mother_repeated_ancombc_output$diff_timepointpn12==TRUE & mother_repeated_ancombc_output$passed_ss_timepointpn12==TRUE]

# further look into each
mother_repeated_ancombc_res_global <- mother_repeated_ancombc$res_global %>%
  mutate_if(is.numeric, function(x) round(x, 2))

mother_repeated_ancombc_res_pair <- mother_repeated_ancombc$res_pair 
#write.csv(mother_repeated_ancombc_res_pair, "../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/mother_repeated_ancom.csv")

mother_repeated_ancombc_res_pair <- merge(mother_repeated_ancombc_res_pair, taxonomy, by.x = "taxon", by.y = "species", all.x = TRUE)
library(openxlsx)
write.xlsx(mother_repeated_ancombc_res_pair, "../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/mother_repeated_ancom.xlsx")

mother_repeated_ancombc_res_trend <- mother_repeated_ancombc$res_trend %>%
  mutate_if(is.numeric, function(x) round(x, 2))



mother_repeatedTri3PN12 <- mother_repeated_ancombc_res_pair |>
  mutate(`q_timepointpn12`=-log10(`q_timepointpn12`)) |>
  mutate(color=if_else(`lfc_timepointpn12`>0, "steelblue", "darkred")) |>
  mutate(label=if_else(abs(`lfc_timepointpn12`)>0.5 & `q_timepointpn12` >1.3 & `passed_ss_timepointpn12` == TRUE, taxon, NA_character_)) |>
  mutate(color=if_else(abs(`lfc_timepointpn12`)<0.5 & `q_timepointpn12` <1.3, "#000000", color)) |>
  mutate(color=if_else(`passed_ss_timepointpn12` == FALSE, "grey55", color)) |>
  ggplot(mapping = aes(`lfc_timepointpn12`, `q_timepointpn12`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 0.5) +
  geom_vline(linetype = "dotted", xintercept = -0.5) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "Mother - PN12 vs. Tri3",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~P)) +
  theme_bw() + xlim(-2, 2) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
mother_repeatedTri3PN12

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/mother_repeatedTri3PN12.svg",
  width=4, height=4
  )
mother_repeatedTri3PN12
dev.off()

mother_repeatedTri3PN56 <- mother_repeated_ancombc_res_pair |>
  mutate(`q_timepointpn56`=-log10(`q_timepointpn56`)) |>
  mutate(color=if_else(`lfc_timepointpn56`>0, "steelblue", "darkred")) |>
  mutate(label=if_else(abs(`lfc_timepointpn56`)>0.5 & `q_timepointpn56` >1.3 & `passed_ss_timepointpn56` == TRUE, taxon, NA_character_)) |>
  mutate(color=if_else(abs(`lfc_timepointpn56`)<0.5 & `q_timepointpn56` <1.3, "#000000", color)) |>
  mutate(color=if_else(`passed_ss_timepointpn56` == FALSE, "grey55", color)) |>
  ggplot(mapping = aes(`lfc_timepointpn56`, `q_timepointpn56`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 0.5) +
  geom_vline(linetype = "dotted", xintercept = -0.5) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "Mother - PN56 vs. Tri3",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~P)) +
  theme_bw() + xlim(-2, 2) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
mother_repeatedTri3PN56

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/mother_repeatedTri3PN56.svg",
  width=4, height=4
  )
mother_repeatedTri3PN56
dev.off()

mother_repeatedPN12PN56 <- mother_repeated_ancombc_res_pair |>
  mutate(`q_timepointpn56_timepointpn12`=-log10(`q_timepointpn56_timepointpn12`)) |>
  mutate(color=if_else(`lfc_timepointpn56_timepointpn12`>0, "steelblue", "darkred")) |>
  mutate(label=if_else(abs(`lfc_timepointpn56_timepointpn12`)>0.5 & `q_timepointpn56_timepointpn12` >1.3 & `passed_ss_timepointpn56_timepointpn12` == TRUE, taxon, NA_character_)) |>
  mutate(color=if_else(abs(`lfc_timepointpn56_timepointpn12`)<0.5 & `q_timepointpn56_timepointpn12` <1.3, "#000000", color)) |>
  mutate(color=if_else(`passed_ss_timepointpn56_timepointpn12` == FALSE, "grey55", color)) |>
  ggplot(mapping = aes(`lfc_timepointpn56_timepointpn12`, `q_timepointpn56_timepointpn12`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 0.5) +
  geom_vline(linetype = "dotted", xintercept = -0.5) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "Mother - PN56 vs. PN12",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~P)) +
  theme_bw() + xlim(-2, 2) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
mother_repeatedPN12PN56

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/mother_repeatedPN12PN56.svg",
  width=4, height=4
  )
mother_repeatedPN12PN56
dev.off()

mother_repeatedBEP <- mother_repeated_ancombc_output |>
  mutate(`q_codeB Z`=-log10(`q_codeB Z`)) |>
  mutate(color=if_else(`lfc_codeB Z`>0, "steelblue", "darkred")) |>
  mutate(label=if_else(abs(`lfc_codeB Z`)>0.5 & `q_codeB Z` >1.3 & `passed_ss_codeB Z` == TRUE, taxon, NA_character_)) |>
  mutate(color=if_else(abs(`lfc_codeB Z`)<0.5 & `q_codeB Z` <1.3, "#000000", color)) |>
  mutate(color=if_else(`passed_ss_codeB Z` == FALSE, "grey55", color)) |>
  ggplot(mapping = aes(`lfc_codeB Z`, `q_codeB Z`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 0.5) +
  geom_vline(linetype = "dotted", xintercept = -0.5) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "Mother - Interv. vs. Contr.",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~P)) +
  theme_bw() + xlim(-2, 2) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
mother_repeatedBEP

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/mother_repeatedBEP.svg",
  width=4, height=4
  )
mother_repeatedBEP
dev.off()


```


### infant


```{r}

# infant
infant_repeated_ancombc <- ancombc2(data=infant_ps, 
                        assay_name = "counts", tax_level = "species",
                        fix_formula = "code + timepoint", rand_formula = "(1 | idbs)",
                        p_adj_method = "BH", pseudo_sens = TRUE,
                        prv_cut = 0.1,
                        lib_cut = 1000, s0_perc = 0.05,
                        group = "timepoint", struc_zero = TRUE, neg_lb = TRUE,
                        alpha = 0.05, n_cl = 2, verbose = TRUE,
                        global = TRUE, pairwise = TRUE, dunnet = TRUE, trend = TRUE,
                        iter_control = list(tol=1e-2, max_iter=20, verbose=TRUE),
                        em_control = list(tol=1e-5, max_iter=100),
                        lme_control = lme4::lmerControl(),
                        mdfdr_control = list(fwer_ctrl_method = "BH", B=1000),
                        trend_control = list(contrast=list(matrix(c(1,0,-1,1),
                                                                  nrow = 2,
                                                                  byrow = TRUE)),
                                             node=list(2),
                                             solver="ECOS",
                                             B=10)) 
# ancombc output
infant_repeated_ancombc_output <- infant_repeated_ancombc$res
infant_repeated_ancombc_output$taxon[infant_repeated_ancombc_output$`diff_codeB Z`==TRUE] #0
infant_repeated_ancombc_output$taxon[infant_repeated_ancombc_output$diff_timepointpn56==TRUE] #"Escherichia coli"
infant_repeated_ancombc_output$taxon[infant_repeated_ancombc_output$diff_timepointpn56==TRUE & infant_repeated_ancombc_output$passed_ss_timepointpn56==TRUE] #"Escherichia coli"

infant_repeated_ancombc_output <- merge(infant_repeated_ancombc_output, taxonomy, by.x = "taxon", by.y = "species", all.x = TRUE)

library(openxlsx)
write.xlsx(infant_repeated_ancombc_output, "../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/infant_repeated_ancombc_output.xlsx")


infant_repeatedPN12PN56 <- infant_repeated_ancombc_output |>
  mutate(`q_timepointpn56`=-log10(`q_timepointpn56`)) |>
  mutate(color=if_else(`lfc_timepointpn56`>0, "steelblue", "darkred")) |>
  mutate(label=if_else(abs(`lfc_timepointpn56`)>0.5 & `q_timepointpn56` >1.3 & `passed_ss_timepointpn56` == TRUE, taxon, NA_character_)) |>
  mutate(color=if_else(abs(`lfc_timepointpn56`)<0.5 & `q_timepointpn56` <1.3, "#000000", color)) |>
  mutate(color=if_else(`passed_ss_timepointpn56` == FALSE, "grey55", color)) |>
  ggplot(mapping = aes(`lfc_timepointpn56`, `q_timepointpn56`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 0.5) +
  geom_vline(linetype = "dotted", xintercept = -0.5) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "Infant - PN56 vs. PN12",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~P)) +
  theme_bw() + xlim(-2, 2) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
infant_repeatedPN12PN56

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/infant_repeatedPN12PN56.svg",
  width=4, height=4
  )
infant_repeatedPN12PN56
dev.off()


infant_repeatedBEP <- infant_repeated_ancombc_output |>
  mutate(`q_codeB Z`=-log10(`q_codeB Z`)) |>
  mutate(color=if_else(`lfc_codeB Z`>0, "steelblue", "darkred")) |>
  mutate(label=if_else(abs(`lfc_codeB Z`)>0.5 & `q_codeB Z` >1.3 & `passed_ss_codeB Z` == TRUE, taxon, NA_character_)) |>
  mutate(color=if_else(abs(`lfc_codeB Z`)<0.5 & `q_codeB Z` <1.3, "#000000", color)) |>
  mutate(color=if_else(`passed_ss_codeB Z` == FALSE, "grey55", color)) |>
  ggplot(mapping = aes(`lfc_codeB Z`, `q_codeB Z`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 0.5) +
  geom_vline(linetype = "dotted", xintercept = -0.5) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "Infant - Interv. vs. Contr.",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~P)) +
  theme_bw() + xlim(-2, 2) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
mother_repeatedBEP

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/infant_repeatedBEP.svg",
  width=4, height=4
  )
infant_repeatedBEP
dev.off()

```


## C. Boxplots
```{r}

my_colors2 <- c("Intervention" = "steelblue", "Control" = "darkred")

incl_species <- incl %>%
  tax_glom(taxrank = "species") %>%
  transform_sample_counts(function(x){x/sum(x)}) %>%
  psmelt() %>%
  filter(Abundance > 0)

relAbuSpe_incl <- incl_species %>%
  dplyr::select(species, idbs, code, Abundance) %>%
  filter(species %in% c( "CAG-103 sp000432375")) %>%
  dplyr::mutate(Abundance=round(Abundance*100, 2)) %>%
  dplyr::mutate(Intervention = if_else(code== "B Z", "Intervention", "Control"))

relAbuSpe_incl_plot <- relAbuSpe_incl %>%
  ggplot(aes(x = Intervention, y = Abundance, color = Intervention)) +
  geom_boxplot() + # Suppresses the outliers in boxplot
  geom_jitter(width = 0.2, size = 1.5, alpha = 0.6) + # Adds jitter to the points
  facet_wrap(~ species, scales = "free_y", ncol = 7) + # Facets by species with independent y-axes
  scale_color_manual(values = my_colors2) + # Uses your custom color palette
  theme_minimal() + # Clean theme
  #labs(x = "Intervention", y = "Abundance", color = "Intervention") + # Labels
  #theme(axis.text.x = element_text(hjust = 1)) +
  theme(
        #panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.border = element_rect(color = "black", fill = NA),
        axis.text = element_text(size = 12),     # Set axis text font size
        axis.title = element_blank(),    # Set axis title font size
        plot.title = element_blank(), legend.position = "None",
        strip.background = element_rect(fill = "grey", color = "black"), # Grey background for facet titles
        strip.text = element_text(color = "white", face = "bold") )    # Set plot title font size

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/box_relAbuSpe_incl.svg",
  width=2.5, height=3
  )
relAbuSpe_incl_plot
dev.off()


tri3_species <- tri3 %>%
  tax_glom(taxrank = "species") %>%
  transform_sample_counts(function(x){x/sum(x)}) %>%
  psmelt() %>%
  filter(Abundance > 0)

relAbuSpe_tri3 <- tri3_species %>%
  select(species, idbs, code, Abundance) %>%
  filter(species %in% c( "CAG-245 sp000434195", "Unknown_UMGS882", "Unknown_UBA1436", "Unknown_Lachnospiraceae")) %>%
  mutate(Abundance=round(Abundance*100, 2)) %>%
  mutate(Intervention = if_else(code== "B Z", "Intervention", "Control"))

relAbuSpe_tri3_plot <- relAbuSpe_tri3 %>%
  ggplot(aes(x = Intervention, y = Abundance, color = Intervention)) +
  geom_boxplot() + # Suppresses the outliers in boxplot
  geom_jitter(width = 0.2, size = 1.5, alpha = 0.6) + # Adds jitter to the points
  facet_wrap(~ species, scales = "free_y", ncol = 7) + # Facets by species with independent y-axes
  scale_color_manual(values = my_colors2) + # Uses your custom color palette
  theme_minimal() + # Clean theme
  labs(x = "Intervention", y = "Abundance", color = "Intervention") + # Labels
  #theme(axis.text.x = element_text(hjust = 1)) +
  theme(
        #panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.border = element_rect(color = "black", fill = NA),
        axis.text = element_text(size = 12),     # Set axis text font size
        axis.title = element_blank(),    # Set axis title font size
        plot.title = element_blank(), legend.position = "None",
        strip.background = element_rect(fill = "grey", color = "black"), # Grey background for facet titles
        strip.text = element_text(color = "white", face = "bold") )    # Set plot title font size


svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/box_relAbuSpe_tri3.svg",
  width=10, height=3
  )
relAbuSpe_tri3_plot
dev.off()

pn12m_species <- pn12m %>%
  tax_glom(taxrank = "species") %>%
  transform_sample_counts(function(x){x/sum(x)}) %>%
  psmelt() %>%
  filter(Abundance > 0)

relAbuSpe_pn12m <- pn12m_species %>%
  select(species, idbs, code, Abundance) %>%
  filter(species %in% c( "CAG-452 sp000434035", "Unknown_Agathobacter", "Ruminococcus_C sp000433635", "CAG-533 sp900551105", "Agathobacter sp900546625")) %>%
  mutate(Abundance=round(Abundance*100, 2)) %>%
  mutate(Intervention = if_else(code== "B Z", "Intervention", "Control"))

relAbuSpe_pn12m_plot <- relAbuSpe_pn12m %>%
  ggplot(aes(x = Intervention, y = Abundance, color = Intervention)) +
  geom_boxplot() + # Suppresses the outliers in boxplot
  geom_jitter(width = 0.2, size = 1.5, alpha = 0.6) + # Adds jitter to the points
  facet_wrap(~ species, scales = "free_y", ncol = 7) + # Facets by species with independent y-axes
  scale_color_manual(values = my_colors2) + # Uses your custom color palette
  theme_minimal() + # Clean theme
  #labs(x = "Intervention", y = "Abundance", color = "Intervention") + # Labels
  #theme(axis.text.x = element_text(hjust = 1)) +
  theme(
        #panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.border = element_rect(color = "black", fill = NA),
        axis.text = element_text(size = 12),     # Set axis text font size
        axis.title = element_blank(),    # Set axis title font size
        plot.title = element_blank(), legend.position = "None",
        strip.background = element_rect(fill = "grey", color = "black"), # Grey background for facet titles
        strip.text = element_text(color = "white", face = "bold") )    # Set plot title font size


svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/box_relAbuSpe_pn12m.svg",
  width=12.5, height=3
  )
relAbuSpe_pn12m_plot
dev.off()

pn56m_species <- pn56m %>%
  tax_glom(taxrank = "species") %>%
  transform_sample_counts(function(x){x/sum(x)}) %>%
  psmelt() %>%
  filter(Abundance > 0)

relAbuSpe_pn56m <- pn56m_species %>%
  select(species, idbs, code, Abundance) %>%
  filter(species %in% c( "Bacteroides fragilis", "Unknown_UMGS874", "UBA1232 sp900544815", "Unknown_Veillonella", "Zag111 sp003258735", "Phascolarctobacterium_A sp900551335")) %>%
  mutate(Abundance=round(Abundance*100, 2)) %>%
  mutate(Intervention = if_else(code== "B Z", "Intervention", "Control"))

relAbuSpe_pn56m_plot <- relAbuSpe_pn56m %>%
  ggplot(aes(x = Intervention, y = Abundance, color = Intervention)) +
  geom_boxplot() + # Suppresses the outliers in boxplot
  geom_jitter(width = 0.2, size = 1.5, alpha = 0.6) + # Adds jitter to the points
  facet_wrap(~ species, scales = "free_y", ncol = 7) + # Facets by species with independent y-axes
  scale_color_manual(values = my_colors2) + # Uses your custom color palette
  theme_minimal() + # Clean theme
  #labs(x = "Intervention", y = "Abundance", color = "Intervention") + # Labels
  #theme(axis.text.x = element_text(hjust = 1)) +
  theme(
        #panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.border = element_rect(color = "black", fill = NA),
        axis.text = element_text(size = 12),     # Set axis text font size
        axis.title = element_blank(),    # Set axis title font size
        plot.title = element_blank(), legend.position = "None",
        strip.background = element_rect(fill = "grey", color = "black"), # Grey background for facet titles
        strip.text = element_text(color = "white", face = "bold") )    # Set plot title font size

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/box_relAbuSpe_pn56m.svg",
  width=15, height=3
  )
relAbuSpe_pn56m_plot
dev.off()

```


#Figure 6. GSEA
# Functional analysis
## Pathways
## Formatting
Before we start modelling, we perform a few formatting steps on the data. 
1. We split the data into subsets to analyse separately (because the treatment effect on specific genes will likely be different for mothers and infants and different at each timepoint).
2. We apply two different, common transformations on the data to deal with zero-inflation and variance in sequencing depth, a variance stabilizing transformation (VST) and a total sum scaling (TSS).
3. We combine everything into TreeSummarizedExperiment (TSE) R objects.

### maternal
```{r splitting data maternal, include=FALSE}
# metadata
metadata_mat <- metadata[metadata$child == "no", ]
metadata_mat_incl <-metadata_mat[metadata_mat$timepoint == "incl", ]
metadata_mat_tri3 <-metadata_mat[metadata_mat$timepoint == "tri3", ]
metadata_mat_pn12 <-metadata_mat[metadata_mat$timepoint == "pn12", ]
metadata_mat_pn56 <-metadata_mat[metadata_mat$timepoint == "pn56", ]

# counts
gene_counts_mat <- gene_counts[, rownames(metadata_mat)]
gene_counts_mat <- gene_counts_mat[rowSums(gene_counts_mat) != 0, ]
gene_counts_mat_incl <- gene_counts_mat[, rownames(metadata_mat_incl)]
gene_counts_mat_tri3 <- gene_counts_mat[, rownames(metadata_mat_tri3)]
gene_counts_mat_pn12 <- gene_counts_mat[, rownames(metadata_mat_pn12)]
gene_counts_mat_pn56 <- gene_counts_mat[, rownames(metadata_mat_pn56)]

# hierarchy
anno_mat <- gene_anno[rownames(gene_counts_mat), ]
phylogenetic_tree_mat <- TreeSummarizedExperiment::toTree(
  data = anno_mat
  )

# fix order
metadata_mat_incl <- metadata_mat_incl[colnames(gene_counts_mat_incl), ]
metadata_mat_tri3 <- metadata_mat_tri3[colnames(gene_counts_mat_tri3), ]
metadata_mat_pn12 <- metadata_mat_pn12[colnames(gene_counts_mat_pn12), ]
metadata_mat_pn56 <- metadata_mat_pn56[colnames(gene_counts_mat_pn56), ]
```


### infant
```{r transforming infant, include=FALSE}
# TSS
scaled_counts_inf_pn12 <- data.frame(
  scaled_integer_transform_counts(prop.table(as.matrix(gene_counts_inf_pn12), 2))
  )
scaled_counts_inf_pn56 <- data.frame(
  scaled_integer_transform_counts(prop.table(as.matrix(gene_counts_inf_pn56), 2))
  )
```

## TSE TreeSummarizedExperiment
### infant
```{r TSE objects infant, include=FALSE}
# TSS
tse_object_scaled_inf_pn12 <- TreeSummarizedExperiment::TreeSummarizedExperiment(
  assays = list(count = as.matrix(scaled_counts_inf_pn12)),
  rowData = anno_inf,
  colData = metadata_inf_pn12,
  rowTree = phylogenetic_tree_inf,
  rowNodeLab = rownames(scaled_counts_inf_pn12)
)
tse_object_scaled_inf_pn56 <- TreeSummarizedExperiment::TreeSummarizedExperiment(
  assays = list(count = as.matrix(scaled_counts_inf_pn56)),
  rowData = anno_inf,
  colData = metadata_inf_pn56,
  rowTree = phylogenetic_tree_inf,
  rowNodeLab = rownames(scaled_counts_inf_pn56)
)
```

### maternal
```{r TSE objects maternal, include=FALSE}
# TSS objects
tse_object_scaled_mat_incl <- TreeSummarizedExperiment::TreeSummarizedExperiment(
  assays = list(count = as.matrix(scaled_counts_mat_incl)),
  rowData = anno_mat,
  colData = metadata_mat_incl,
  rowTree = phylogenetic_tree_mat,
  rowNodeLab = rownames(scaled_counts_mat_incl)
)
tse_object_scaled_mat_tri3 <- TreeSummarizedExperiment::TreeSummarizedExperiment(
  assays = list(count = as.matrix(scaled_counts_mat_tri3)),
  rowData = anno_mat,
  colData = metadata_mat_tri3,
  rowTree = phylogenetic_tree_mat,
  rowNodeLab = rownames(scaled_counts_mat_tri3)
)
tse_object_scaled_mat_pn12 <- TreeSummarizedExperiment::TreeSummarizedExperiment(
  assays = list(count = as.matrix(scaled_counts_mat_pn12)),
  rowData = anno_mat,
  colData = metadata_mat_pn12,
  rowTree = phylogenetic_tree_mat,
  rowNodeLab = rownames(scaled_counts_mat_pn12)
)
tse_object_scaled_mat_pn56 <- TreeSummarizedExperiment::TreeSummarizedExperiment(
  assays = list(count = as.matrix(scaled_counts_mat_pn56)),
  rowData = anno_mat,
  colData = metadata_mat_pn56,
  rowTree = phylogenetic_tree_mat,
  rowNodeLab = rownames(scaled_counts_mat_pn56)
)
```


## Modelling
We can group genes into specific hierarchies based on enzymatic classes, metabolic function etc. and test across these hierarchic levels This can help to find specific functional classes that are differential as a whole where the individual gene would not have showed up, testing all levels increases the false discovery rate and can have issues of multicolinearity. treeclimbR solves this by precomputing the most interesting candidates to test for differential abundance ([Huang et al., 2021](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02368-1)).

```{r volcano plot function, include=FALSE}
#' Create a volcano plot from the outputs of a differential abundance analysis
#'
#' @param results A list containing the DA results as outputted in the
#' `treeclimbR::evalCand` function.
#' @param alpha a numeric alpha level on which to cut off p-values
#' @param effect_cutoff a numeric log fold change on which to cut off log FCs
#' @param color_scheme a list of hex valued colors indicating a scheme.
#' @param title title for the plot to export
#'
#' @examples volcano_plot(results, color_scheme)
#' @export
volcano_plot <- function(results,
                         alpha = 0.1,
                         effect_cutoff = 2,
                         color_scheme, 
                         title = "Differential abundance analysis: Volcano plot") {
  # extract elements
  scores <- results$output

  # color on significance groups and add tooltips
  volcano_plot_df <- scores %>%
    mutate(
      significance_group = dplyr::if_else(
        abs(logFC) < effect_cutoff & PValue > alpha,
        "Non-significant",
        dplyr::if_else(
          abs(logFC) < effect_cutoff,
          "Statistically significant (p-values)",
          dplyr::if_else(
            PValue > alpha,
            "Biologically significant (FC)",
            "Biologically & statistically significant"
          )
        )
      ),
      text = paste0(
        "Taxon: ", names, "\n",
        "p-value: ", PValue, "\n",
        "log FC: ", round(logFC, 4), "\n"
      )
    )

  p <- ggplot2::ggplot(
    volcano_plot_df,
    ggplot2::aes(
      x = logFC, y = -log10(PValue),
      color = significance_group,
      text = text
    )
  ) +
    ggplot2::geom_point(size = 3, alpha = 0.8) +
    ggplot2::theme_bw() +
    ggplot2::scale_color_manual(
      values = c(
        "Non-significant" = "#9EA0A5",
        "Biologically significant (FC)" = color_scheme[1],
        "Statistically significant (p-values)" = color_scheme[3],
        "Biologically & statistically significant" = color_scheme[2]
      )
    ) +
    ggplot2::labs(
      title = title,
      x = "Fold change (Log2)",
      y = "p-value (-Log10)"
    ) +
    ggplot2::theme(
      plot.title = ggplot2::element_text(face = "bold"),
      axis.text.x = ggplot2::element_text(vjust = 0.4, hjust = 1),
      axis.text.y = ggplot2::element_text(vjust = 0.4, hjust = 1),
      legend.title = ggplot2::element_blank(),
      legend.position = "right",
      text = ggplot2::element_text(size = 10)
    ) +
    ggplot2::geom_vline(
      xintercept = c(-effect_cutoff, effect_cutoff),
      linetype = "dashed"
    ) +
    ggplot2::geom_hline(
      yintercept = -log10(alpha),
      linetype = "dashed"
    )

  return(plotly::ggplotly(p, tooltip = "text"))
}
```


### Mothers

#### incl
```{r treeclimbR maternal incl TSS, include=FALSE}
res_scaled_mat_incl <- treeclimbR::runDA(
  TSE = tse_object_scaled_mat_incl,
  feature_on_row = TRUE,
  assay = 1,
  design_terms = "code"
  )
res_table_scaled_mat_incl <- treeclimbR::nodeResult(
  object = res_scaled_mat_incl,
  n = Inf
  )
candidates_scaled_mat_incl <- treeclimbR::getCand(
  tree = TreeSummarizedExperiment::rowTree(tse_object_scaled_mat_incl),
  score_data = res_table_scaled_mat_incl,
  node_column = "node",
  p_column = "PValue",
  sign_column = "logFC"
  )
da_results_scaled_mat_incl <- treeclimbR::evalCand(
  tree = TreeSummarizedExperiment::rowTree(tse_object_scaled_mat_incl),
  type = "single",
  levels = candidates_scaled_mat_incl$candidate_list,
  limit_rej = 0.1,
  score_data = res_table_scaled_mat_incl,
  node_column = "node",
  p_column = "PValue",
  sign_column = "logFC",
  use_pseudo_leaf = FALSE
  )

da_results_scaled_mat_incl$output$names <- TreeSummarizedExperiment::convertNode(
  phylogenetic_tree_mat,
  da_results_scaled_mat_incl$output$node
  )
```


##### Volcano plot
```{r plot volcano maternal incl TSS, echo=FALSE}
# permutation based False Discovery Proportion estimation
fdr_thresh_scaled_mat_incl <- permFDP::permFDP.adjust.threshold(
  pVals = da_results_scaled_mat_incl$output$PValue,
  threshold = 0.1,
  myDesign = ifelse(metadata_mat_incl[colnames(scaled_counts_mat_incl), "code"] == "A Y", 1, 2),
  intOnly = scaled_counts_mat_incl[da_results_scaled_mat_incl$output$names, ],
  nPerms = 999
)
# plot
p_scaled_mat_incl <- volcano_plot(
  da_results_scaled_mat_incl, 
  alpha = fdr_thresh_scaled_mat_incl, 
  color_scheme = COLOR_SCHEME, 
  title = "Differential functions at incl - Mother (TSS transformed)")
p_scaled_mat_incl

# adjusted p-values plot
gene_daa_incl <- da_results_scaled_mat_incl$output
DT::datatable(gene_daa_incl)

gene_volcano_incl <- gene_daa_incl |>
  mutate(`adj.p`=-log10(`adj.p`)) |>
  mutate(label=if_else(abs(`logFC`)>2, names, NA_character_)) |>
  mutate(label=if_else(`adj.p` >1.3, names, NA_character_)) |>
  mutate(color=if_else(`logFC`>0, "#FF9E4A", "#729ECE")) |>
  mutate(color=if_else(is.na(label), "#000000", color)) |>
  ggplot(mapping = aes(`logFC`, `adj.p`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 2) +
  geom_vline(linetype = "dotted", xintercept = -2) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "treeclimbR, maternal samples at incl, gene counts",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~Q)) +
  theme_bw()
gene_volcano_incl

```


##### GSEA
```{r GSEA maternal incl TSS, echo=FALSE}
tmp <- data.frame(da_results_scaled_mat_incl$output[, c("logFC", "names")])
# the DA lists also include CARD, CAZyme, and HMO genes, but we only have mappings for KEGG atm
kegg_genes_tmp <- tmp$names %in% gene_anno[gene_anno$database == "KEGG", "anno"]
tmp <- tmp[kegg_genes_tmp, ]
gene_list_scaled_mat_incl <- tmp[,1]
names(gene_list_scaled_mat_incl) <- as.character(tmp[, 2])
gene_list_scaled_mat_incl <- sort(gene_list_scaled_mat_incl, decreasing = TRUE)

kk_scaled_mat_incl <- clusterProfiler::GSEA(
  gene_list_scaled_mat_incl, 
  TERM2GENE = pathway_to_gene,  
  TERM2NAME = pathway_to_description,
  pvalueCutoff = 0.1,
  minGSSize = 1
  )
DT::datatable(kk_scaled_mat_incl@result)
vis <- nrow(kk_scaled_mat_incl@result) > 0 # only run plots if there are results

enrich_incl <- enrichplot::dotplot(
  kk_scaled_mat_incl, 
  x = "enrichmentScore", 
  title = "Differential pathways at incl - Mother"
  ) + xlim(-1, 1)
enrich_incl
enrichplot::ridgeplot(kk_scaled_mat_incl)

```


#### tri3
```{r treeclimbR maternal tri3 TSS, include=FALSE}
res_scaled_mat_tri3 <- treeclimbR::runDA(
  TSE = tse_object_scaled_mat_tri3,
  feature_on_row = TRUE,
  assay = 1,
  design_terms = "code"
  )
res_table_scaled_mat_tri3 <- treeclimbR::nodeResult(
  object = res_scaled_mat_tri3,
  n = Inf
  )
candidates_scaled_mat_tri3 <- treeclimbR::getCand(
  tree = TreeSummarizedExperiment::rowTree(tse_object_scaled_mat_tri3),
  score_data = res_table_scaled_mat_tri3,
  node_column = "node",
  p_column = "PValue",
  sign_column = "logFC"
  )
da_results_scaled_mat_tri3 <- treeclimbR::evalCand(
  tree = TreeSummarizedExperiment::rowTree(tse_object_scaled_mat_tri3),
  type = "single",
  levels = candidates_scaled_mat_tri3$candidate_list,
  limit_rej = 0.1,
  score_data = res_table_scaled_mat_tri3,
  node_column = "node",
  p_column = "PValue",
  sign_column = "logFC",
  use_pseudo_leaf = FALSE
  )

da_results_scaled_mat_tri3$output$names <- TreeSummarizedExperiment::convertNode(
  phylogenetic_tree_mat,
  da_results_scaled_mat_tri3$output$node
  )
```

##### Volcano plot
```{r plot volcano maternal tri3 TSS, echo=FALSE}
# permutation based False Discovery Proportion estimation
fdr_thresh_scaled_mat_tri3 <- permFDP::permFDP.adjust.threshold(
  pVals = da_results_scaled_mat_tri3$output$PValue,
  threshold = 0.1,
  myDesign = ifelse(metadata_mat_tri3[colnames(scaled_counts_mat_tri3), "code"] == "A Y", 1, 2),
  intOnly = scaled_counts_mat_tri3[da_results_scaled_mat_tri3$output$names, ],
  nPerms = 999
)
# plot
p_scaled_mat_tri3 <- volcano_plot(
  da_results_scaled_mat_tri3, 
  alpha = fdr_thresh_scaled_mat_tri3, 
  color_scheme = COLOR_SCHEME, 
  title = "Differential functions at tri3 - Mother (TSS transformed)")
p_scaled_mat_tri3

# adjusted p-values plot
gene_daa_tri3 <- da_results_scaled_mat_tri3$output
DT::datatable(gene_daa_tri3)

gene_volcano_tri3 <- gene_daa_tri3 |>
  mutate(`adj.p`=-log10(`adj.p`)) |>
  mutate(label=if_else(abs(`logFC`)>2, names, NA_character_)) |>
  mutate(label=if_else(`adj.p` >1.3, names, NA_character_)) |>
  mutate(color=if_else(`logFC`>0, "#FF9E4A", "#729ECE")) |>
  mutate(color=if_else(is.na(label), "#000000", color)) |>
  ggplot(mapping = aes(`logFC`, `adj.p`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 2) +
  geom_vline(linetype = "dotted", xintercept = -2) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "treeclimbR, maternal samples at tri3, gene counts",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~Q)) +
  theme_bw()
gene_volcano_tri3

```


##### GSEA
```{r GSEA maternal tri3 TSS, echo=FALSE}
tmp <- data.frame(da_results_scaled_mat_tri3$output[, c("logFC", "names")])
# the DA lists also include CARD, CAZyme, and HMO genes, but we only have mappings for KEGG atm
kegg_genes_tmp <- tmp$names %in% gene_anno[gene_anno$database == "KEGG", "anno"]
tmp <- tmp[kegg_genes_tmp, ]
gene_list_scaled_mat_tri3 <- tmp[,1]
names(gene_list_scaled_mat_tri3) <- as.character(tmp[, 2])
gene_list_scaled_mat_tri3 <- sort(gene_list_scaled_mat_tri3, decreasing = TRUE)

kk_scaled_mat_tri3 <- clusterProfiler::GSEA(
  gene_list_scaled_mat_tri3, 
  TERM2GENE = pathway_to_gene,  
  TERM2NAME = pathway_to_description,
  pvalueCutoff = 0.1,
  minGSSize = 1
  )
DT::datatable(kk_scaled_mat_tri3@result)
vis <- nrow(kk_scaled_mat_tri3@result) > 0 # only run plots if there are results

enrich_tri3 <- enrichplot::dotplot(
  kk_scaled_mat_tri3, 
  x = "enrichmentScore", 
  title = "Differential pathways at tri3 - Mother"
  ) + xlim(-1, 1)

enrichplot::ridgeplot(kk_scaled_mat_tri3)

```


#### pn12m
```{r treeclimbR maternal pn12m TSS, include=FALSE}
res_scaled_mat_pn12 <- treeclimbR::runDA(
  TSE = tse_object_scaled_mat_pn12,
  feature_on_row = TRUE,
  assay = 1,
  design_terms = "code"
  )
res_table_scaled_mat_pn12 <- treeclimbR::nodeResult(
  object = res_scaled_mat_pn12,
  n = Inf
  )
candidates_scaled_mat_pn12 <- treeclimbR::getCand(
  tree = TreeSummarizedExperiment::rowTree(tse_object_scaled_mat_pn12),
  score_data = res_table_scaled_mat_pn12,
  node_column = "node",
  p_column = "PValue",
  sign_column = "logFC"
  )
da_results_scaled_mat_pn12 <- treeclimbR::evalCand(
  tree = TreeSummarizedExperiment::rowTree(tse_object_scaled_mat_pn12),
  type = "single",
  levels = candidates_scaled_mat_pn12$candidate_list,
  limit_rej = 0.1,
  score_data = res_table_scaled_mat_pn12,
  node_column = "node",
  p_column = "PValue",
  sign_column = "logFC",
  use_pseudo_leaf = FALSE
  )

da_results_scaled_mat_pn12$output$names <- TreeSummarizedExperiment::convertNode(
  phylogenetic_tree_mat,
  da_results_scaled_mat_pn12$output$node
  )
```

##### Volcano plot
```{r plot volcano maternal pn12 TSS, echo=FALSE}
# permutation based False Discovery Proportion estimation
fdr_thresh_scaled_mat_pn12 <- permFDP::permFDP.adjust.threshold(
  pVals = da_results_scaled_mat_pn12$output$PValue,
  threshold = 0.1,
  myDesign = ifelse(metadata_mat_pn12[colnames(scaled_counts_mat_pn12), "code"] == "A Y", 1, 2),
  intOnly = scaled_counts_mat_pn12[da_results_scaled_mat_pn12$output$names, ],
  nPerms = 999
)
# plot
p_scaled_mat_pn12 <- volcano_plot(
  da_results_scaled_mat_pn12, 
  alpha = fdr_thresh_scaled_mat_pn12, 
  color_scheme = COLOR_SCHEME, 
  title = "Differential functions at pn12 - Mother (TSS transformed)")
p_scaled_mat_pn12

# adjusted p-values plot
gene_daa_pn12m <- da_results_scaled_mat_pn12$output
DT::datatable(gene_daa_pn12m)

gene_volcano_pn12m <- gene_daa_pn12m |>
  mutate(`adj.p`=-log10(`adj.p`)) |>
  mutate(label=if_else(abs(`logFC`)>2, names, NA_character_)) |>
  mutate(label=if_else(`adj.p` >1.3, names, NA_character_)) |>
  mutate(color=if_else(`logFC`>0, "#FF9E4A", "#729ECE")) |>
  mutate(color=if_else(is.na(label), "#000000", color)) |>
  ggplot(mapping = aes(`logFC`, `adj.p`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 2) +
  geom_vline(linetype = "dotted", xintercept = -2) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "treeclimbR, maternal samples at pn12m, gene counts",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~Q)) +
  theme_bw()
gene_volcano_pn12m

```


##### GSEA
```{r GSEA maternal pn12 TSS, echo=FALSE}
tmp <- data.frame(da_results_scaled_mat_pn12$output[, c("logFC", "names")])
# the DA lists also include CARD, CAZyme, and HMO genes, but we only have mappings for KEGG atm
kegg_genes_tmp <- tmp$names %in% gene_anno[gene_anno$database == "KEGG", "anno"]
tmp <- tmp[kegg_genes_tmp, ]
gene_list_scaled_mat_pn12 <- tmp[,1]
names(gene_list_scaled_mat_pn12) <- as.character(tmp[, 2])
gene_list_scaled_mat_pn12 <- sort(gene_list_scaled_mat_pn12, decreasing = TRUE)

kk_scaled_mat_pn12 <- clusterProfiler::GSEA(
  gene_list_scaled_mat_pn12, 
  TERM2GENE = pathway_to_gene,  
  TERM2NAME = pathway_to_description,
  pvalueCutoff = 0.1,
  minGSSize = 1
  )
DT::datatable(kk_scaled_mat_pn12@result)
vis <- nrow(kk_scaled_mat_pn12@result) > 0 # only run plots if there are results

enrich_pn12m <- enrichplot::dotplot(
  kk_scaled_mat_pn12, 
  x = "enrichmentScore", 
  title = "Differential pathways at pn12 - Mother"
  ) + xlim(-1, 1)
enrichplot::ridgeplot(kk_scaled_mat_pn12)

```


#### pn56m
```{r treeclimbR maternal pn56m TSS, include=FALSE}
res_scaled_mat_pn56 <- treeclimbR::runDA(
  TSE = tse_object_scaled_mat_pn56,
  feature_on_row = TRUE,
  assay = 1,
  design_terms = "code"
  )
res_table_scaled_mat_pn56 <- treeclimbR::nodeResult(
  object = res_scaled_mat_pn56,
  n = Inf
  )
candidates_scaled_mat_pn56 <- treeclimbR::getCand(
  tree = TreeSummarizedExperiment::rowTree(tse_object_scaled_mat_pn56),
  score_data = res_table_scaled_mat_pn56,
  node_column = "node",
  p_column = "PValue",
  sign_column = "logFC"
  )
da_results_scaled_mat_pn56 <- treeclimbR::evalCand(
  tree = TreeSummarizedExperiment::rowTree(tse_object_scaled_mat_pn56),
  type = "single",
  levels = candidates_scaled_mat_pn56$candidate_list,
  limit_rej = 0.1,
  score_data = res_table_scaled_mat_pn56,
  node_column = "node",
  p_column = "PValue",
  sign_column = "logFC",
  use_pseudo_leaf = FALSE
  )

da_results_scaled_mat_pn56$output$names <- TreeSummarizedExperiment::convertNode(
  phylogenetic_tree_mat,
  da_results_scaled_mat_pn56$output$node
  )
```

##### Volcano plot
```{r plot volcano maternal pn56 TSS, echo=FALSE}
# permutation based False Discovery Proportion estimation
fdr_thresh_scaled_mat_pn56 <- permFDP::permFDP.adjust.threshold(
  pVals = da_results_scaled_mat_pn56$output$PValue,
  threshold = 0.1,
  myDesign = ifelse(metadata_mat_pn56[colnames(scaled_counts_mat_pn56), "code"] == "A Y", 1, 2),
  intOnly = scaled_counts_mat_pn56[da_results_scaled_mat_pn56$output$names, ],
  nPerms = 999
)
# plot
p_scaled_mat_pn56 <- volcano_plot(
  da_results_scaled_mat_pn56, 
  alpha = fdr_thresh_scaled_mat_pn56, 
  color_scheme = COLOR_SCHEME, 
  title = "Differential functions at pn56 - Mother (TSS transformed)")
p_scaled_mat_pn56

# adjusted p-values plot
gene_daa_pn56m <- da_results_scaled_mat_pn56$output
DT::datatable(gene_daa_pn56m)

gene_volcano_pn56m <- gene_daa_pn56m |>
  mutate(`adj.p`=-log10(`adj.p`)) |>
  mutate(label=if_else(abs(`logFC`)>2, names, NA_character_)) |>
  mutate(label=if_else(`adj.p` >1.3, names, NA_character_)) |>
  mutate(color=if_else(`logFC`>0, "#FF9E4A", "#729ECE")) |>
  mutate(color=if_else(is.na(label), "#000000", color)) |>
  ggplot(mapping = aes(`logFC`, `adj.p`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 2) +
  geom_vline(linetype = "dotted", xintercept = -2) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "treeclimbR, maternal samples at pn56m, gene counts",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~Q)) +
  theme_bw()
gene_volcano_pn56m

```


##### GSEA
```{r GSEA maternal pn56 TSS, echo=FALSE}
tmp <- data.frame(da_results_scaled_mat_pn56$output[, c("logFC", "names")])
# the DA lists also include CARD, CAZyme, and HMO genes, but we only have mappings for KEGG atm
kegg_genes_tmp <- tmp$names %in% gene_anno[gene_anno$database == "KEGG", "anno"]
tmp <- tmp[kegg_genes_tmp, ]
gene_list_scaled_mat_pn56 <- tmp[,1]
names(gene_list_scaled_mat_pn56) <- as.character(tmp[, 2])
gene_list_scaled_mat_pn56 <- sort(gene_list_scaled_mat_pn56, decreasing = TRUE)

kk_scaled_mat_pn56 <- clusterProfiler::GSEA(
  gene_list_scaled_mat_pn56, 
  TERM2GENE = pathway_to_gene,  
  TERM2NAME = pathway_to_description,
  pvalueCutoff = 0.1,
  minGSSize = 1
  )
DT::datatable(kk_scaled_mat_pn56@result)
vis <- nrow(kk_scaled_mat_pn56@result) > 0 # only run plots if there are results

enrich_pn56m <- enrichplot::dotplot(
  kk_scaled_mat_pn56, 
  x = "enrichmentScore", 
  title = "Differential pathways at pn56m - Mother"
  ) + xlim(-1, 1)

enrichplot::ridgeplot(kk_scaled_mat_pn56)

```


### Infants

We first look for differentially abundant genes between treatment groups in the infants.

#### pn12e 
```{r treeclimbR infants pn12 TSS, include=FALSE}
res_scaled_inf_pn12 <- treeclimbR::runDA(
  TSE = tse_object_scaled_inf_pn12,
  feature_on_row = TRUE,
  assay = 1,
  design_terms = "code"
  )
res_table_scaled_inf_pn12 <- treeclimbR::nodeResult(
  object = res_scaled_inf_pn12,
  n = Inf
  )
candidates_scaled_inf_pn12 <- treeclimbR::getCand(
  tree = TreeSummarizedExperiment::rowTree(tse_object_scaled_inf_pn12),
  score_data = res_table_scaled_inf_pn12,
  node_column = "node",
  p_column = "PValue",
  sign_column = "logFC"
  )
da_results_scaled_inf_pn12 <- treeclimbR::evalCand(
  tree = TreeSummarizedExperiment::rowTree(tse_object_scaled_inf_pn12),
  type = "single",
  levels = candidates_scaled_inf_pn12$candidate_list,
  limit_rej = 0.1,
  score_data = res_table_scaled_inf_pn12,
  node_column = "node",
  p_column = "PValue",
  sign_column = "logFC",
  use_pseudo_leaf = FALSE
  )

da_results_scaled_inf_pn12$output$names <- TreeSummarizedExperiment::convertNode(
  phylogenetic_tree_inf,
  da_results_scaled_inf_pn12$output$node
  )
```

##### Volcano plot
```{r plot volcano infants pn12 TSS, echo=FALSE}
# permutation based False Discovery Proportion estimation
fdr_thresh_scaled_inf_pn12 <- permFDP::permFDP.adjust.threshold(
  pVals = da_results_scaled_inf_pn12$output$PValue,
  threshold = 0.1,
  myDesign = ifelse(metadata_inf_pn12[colnames(scaled_counts_inf_pn12), "code"] == "A Y", 1, 2),
  intOnly = scaled_counts_inf_pn12[da_results_scaled_inf_pn12$output$names, ],
  nPerms = 999
)
# plot
p_scaled_inf_pn12 <- volcano_plot(
  da_results_scaled_inf_pn12,  
  alpha = fdr_thresh_scaled_inf_pn12,
  color_scheme = COLOR_SCHEME, 
  title = "Differential functions at pn12 - Infant (TSS transformed)")
p_scaled_inf_pn12

# adjusted p-values plot
gene_daa_pn12e <- da_results_scaled_inf_pn12$output
DT::datatable(gene_daa_pn12e)

gene_volcano_pn12e <- gene_daa_pn12e |>
  mutate(`adj.p`=-log10(`adj.p`)) |>
  mutate(label=if_else(abs(`logFC`)>2, names, NA_character_)) |>
  mutate(label=if_else(`adj.p` >1.3, names, NA_character_)) |>
  mutate(color=if_else(`logFC`>0, "#FF9E4A", "#729ECE")) |>
  mutate(color=if_else(is.na(label), "#000000", color)) |>
  ggplot(mapping = aes(`logFC`, `adj.p`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 2) +
  geom_vline(linetype = "dotted", xintercept = -2) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "treeclimbR, infant samples at pn12e, gene counts",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~Q)) +
  theme_bw()
gene_volcano_pn12e

```


##### GSEA
```{r GSEA infant pn12 TSS, echo=FALSE}
tmp <- data.frame(da_results_scaled_inf_pn12$output[, c("logFC", "names")])
# the DA lists also include CARD, CAZyme, and HMO genes, but we only have mappings for KEGG atm
kegg_genes_tmp <- tmp$names %in% gene_anno[gene_anno$database == "KEGG", "anno"]
tmp <- tmp[kegg_genes_tmp, ]
gene_list_scaled_inf_pn12 <- tmp[,1]
names(gene_list_scaled_inf_pn12) <- as.character(tmp[, 2])
gene_list_scaled_inf_pn12 <- sort(gene_list_scaled_inf_pn12, decreasing = TRUE)
#GSEA
kk_scaled_inf_pn12 <- clusterProfiler::GSEA(
  geneList = gene_list_scaled_inf_pn12, 
  minGSSize = 1,
  pvalueCutoff = 0.1,
  TERM2GENE = pathway_to_gene, 
  TERM2NAME = pathway_to_description
  )
# output
DT::datatable(kk_scaled_inf_pn12@result)
vis <- nrow(kk_scaled_inf_pn12@result) > 0 # only run plots if there are results

enrich_pn12e <- enrichplot::dotplot(
  kk_scaled_inf_pn12, 
  x = "enrichmentScore", 
  title = "Differential pathways at pn12e - Infant"
  ) + xlim(-1, 1)

enrichplot::ridgeplot(kk_scaled_inf_pn12)

```

The porphyrin ring is the framework for the heme molecule, the oigment in hemoglobin and red blood cells.


#### pn56e 
```{r treeclimbR infants pn56 TSS, include=FALSE}
res_scaled_inf_pn56 <- treeclimbR::runDA(
  TSE = tse_object_scaled_inf_pn56,
  feature_on_row = TRUE,
  assay = 1,
  design_terms = "code"
  )
res_table_scaled_inf_pn56 <- treeclimbR::nodeResult(
  object = res_scaled_inf_pn56,
  n = Inf
  )
candidates_scaled_inf_pn56 <- treeclimbR::getCand(
  tree = TreeSummarizedExperiment::rowTree(tse_object_scaled_inf_pn56),
  score_data = res_table_scaled_inf_pn56,
  node_column = "node",
  p_column = "PValue",
  sign_column = "logFC"
  )
da_results_scaled_inf_pn56 <- treeclimbR::evalCand(
  tree = TreeSummarizedExperiment::rowTree(tse_object_scaled_inf_pn56),
  type = "single",
  levels = candidates_scaled_inf_pn56$candidate_list,
  limit_rej = 0.1,
  score_data = res_table_scaled_inf_pn56,
  node_column = "node",
  p_column = "PValue",
  sign_column = "logFC",
  use_pseudo_leaf = FALSE
  )

da_results_scaled_inf_pn56$output$names <- TreeSummarizedExperiment::convertNode(
  phylogenetic_tree_inf,
  da_results_scaled_inf_pn56$output$node
  )
```

##### Volcano plot
```{r plot volcano infants pn56 TSS, echo=FALSE}
# permutation based False Discovery Proportion estimation
fdr_thresh_scaled_inf_pn56 <- permFDP::permFDP.adjust.threshold(
  pVals = da_results_scaled_inf_pn56$output$PValue,
  threshold = 0.1,
  myDesign = ifelse(metadata_inf_pn56[colnames(scaled_counts_inf_pn56), "code"] == "A Y", 1, 2),
  intOnly = scaled_counts_inf_pn56[da_results_scaled_inf_pn56$output$names, ],
  nPerms = 999
)
# plot
p_scaled_inf_pn56 <- volcano_plot(
  da_results_scaled_inf_pn56,  
  alpha = fdr_thresh_scaled_inf_pn56,
  color_scheme = COLOR_SCHEME, 
  title = "Differential functions at pn56 - Infant (TSS transformed)")
p_scaled_inf_pn56


# adjusted p-values plot
gene_daa_pn56e <- da_results_scaled_inf_pn56$output
DT::datatable(gene_daa_pn56e)

gene_volcano_pn56e <- gene_daa_pn56e |>
  mutate(`adj.p`=-log10(`adj.p`)) |>
  mutate(label=if_else(abs(`logFC`)>2, names, NA_character_)) |>
  mutate(label=if_else(`adj.p` >1.3, names, NA_character_)) |>
  mutate(color=if_else(`logFC`>0, "#FF9E4A", "#729ECE")) |>
  mutate(color=if_else(is.na(label), "#000000", color)) |>
  ggplot(mapping = aes(`logFC`, `adj.p`, color=I(color), label=label, shape=I(1))) +
  geom_point() +
  geom_hline(linetype = "dotted", yintercept = 1.3) +
  geom_vline(linetype = "dotted", xintercept = 2) +
  geom_vline(linetype = "dotted", xintercept = -2) +
  geom_label_repel(min.segment.length = 0, force = 10, max.overlaps = 20, na.rm = TRUE) +
  labs(title = "treeclimbR, infant samples at pn56e, gene counts",
       x = expression(~Log[2]~FoldChange),
       y = expression(-~Log[10]~Q)) +
  theme_bw()
gene_volcano_pn56e

```


##### GSEA
```{r GSEA infant pn56 TSS, echo=FALSE}
tmp <- data.frame(da_results_scaled_inf_pn56$output[, c("logFC", "names")])
# the DA lists also include CARD, CAZyme, and HMO genes, but we only have mappings for KEGG atm
kegg_genes_tmp <- tmp$names %in% gene_anno[gene_anno$database == "KEGG", "anno"]
tmp <- tmp[kegg_genes_tmp, ]
gene_list_scaled_inf_pn56 <- tmp[,1]
names(gene_list_scaled_inf_pn56) <- as.character(tmp[, 2])
gene_list_scaled_inf_pn56 <- sort(gene_list_scaled_inf_pn56, decreasing = TRUE)
#GSEA
kk_scaled_inf_pn56 <- clusterProfiler::GSEA(
  geneList = gene_list_scaled_inf_pn56, 
  minGSSize = 1,
  pvalueCutoff = 0.1,
  TERM2GENE = pathway_to_gene, 
  TERM2NAME = pathway_to_description
  )
# output
DT::datatable(kk_scaled_inf_pn56@result)
vis <- nrow(kk_scaled_inf_pn56@result) > 0 # only run plots if there are results

enrich_pn56e <- enrichplot::dotplot(
  kk_scaled_inf_pn56, 
  x = "enrichmentScore", 
  title = "Differential pathways at pn56e - Infant"
  ) + xlim(-1, 1)

enrichplot::ridgeplot(kk_scaled_inf_pn56)

```



# Figure 7. HDmediation

## A. mediation analysis against diversity
The metadata used here will be made available through a data-sharing agreement. Please contact carl.lachat@ugent.be for any queries. 

```{r numeric vars}
# continuous variables
covariates_prenatal <- c(
  "m_heightincl", "m_hbincl", "asset1comp_10", "w_nr_jobs",
  "w_age", "m_bmiincl", "mddw_10_ave"
)
covariates_postnatal <- c(covariates_prenatal, "iycf_ebf_age")

# make categorical variables into binary dummy vars
alpha_div_extended_mat_pre[, c(
  "w_marital", "w_language", "w_religion", "w_etnicity", "csps_n"
)] <- lapply(
  alpha_div_extended_mat_pre[, c(
    "w_marital", "w_language", "w_religion", "w_etnicity", "csps_n"
  )],
  factor
)
alpha_div_extended_mat_pre$ch_sex <- dplyr::if_else(
  alpha_div_extended_mat_pre$ch_sex == 1, 0, 1
)
dummified_prenatal <- fastDummies::dummy_columns(
  .data = alpha_div_extended_mat_pre[, c("w_school_level", "w_marital", "w_language", "w_religion", "w_etnicity", "csps_n")],
  remove_selected_columns = TRUE
)
alpha_div_extended_mat_pre <- cbind(
  alpha_div_extended_mat_pre[, !(
    names(alpha_div_extended_mat_pre) %in% c("w_school_level", "w_marital", "w_language", "w_religion", "w_etnicity", "csps_n")
  )],
  dummified_prenatal
)
covariates_prenatal <- c(covariates_prenatal, names(dummified_prenatal))

alpha_div_extended_inf[, c(
  "w_marital", "w_language", "w_religion", "w_etnicity", "csps_n"
)] <- lapply(
  alpha_div_extended_inf[, c(
    "w_marital", "w_language", "w_religion", "w_etnicity", "csps_n"
  )],
  factor
)
alpha_div_extended_inf$ch_sex <- dplyr::if_else(alpha_div_extended_inf$ch_sex == 1, 0, 1)
dummified_postnatal <- fastDummies::dummy_columns(
  .data = alpha_div_extended_inf[, c("w_school_level", "w_marital", "w_language", "w_religion", "w_etnicity", "csps_n")],
  remove_selected_columns = TRUE
)
alpha_div_extended_inf <- cbind(
  alpha_div_extended_inf[, !(
    names(alpha_div_extended_inf) %in% c("w_school_level", "w_marital", "w_language", "w_religion", "w_etnicity", "csps_n")
  )],
  dummified_postnatal
)
covariates_postnatal <- c(covariates_postnatal, names(dummified_postnatal))

# set exposure and outcome types
alpha_div_extended_mat_pre$treatment1_prenatal_n <- as.numeric(alpha_div_extended_mat_pre$treatment1_prenatal_n)
alpha_div_extended_inf$treatment1_prenatal_n <- as.numeric(alpha_div_extended_inf$treatment1_prenatal_n)
alpha_div_extended_mat_pre$Shannon <- as.numeric(alpha_div_extended_mat_pre$Shannon)
alpha_div_extended_inf$Shannon <- as.numeric(alpha_div_extended_inf$Shannon)
alpha_div_extended_mat_pre$c_weight0 <- as.numeric(alpha_div_extended_mat_pre$c_weight0)
alpha_div_extended_inf$c_weight0 <- as.numeric(alpha_div_extended_inf$c_weight0)
alpha_div_extended_mat_pre$c_height0 <- as.numeric(alpha_div_extended_mat_pre$c_height0)
alpha_div_extended_inf$c_height0 <- as.numeric(alpha_div_extended_inf$c_height0)
alpha_div_extended_mat_pre$lbw <- as.numeric(alpha_div_extended_mat_pre$lbw)
alpha_div_extended_inf$lbw <- as.numeric(alpha_div_extended_inf$lbw)
alpha_div_extended_mat_pre$sga_new <- as.numeric(alpha_div_extended_mat_pre$sga_new)
alpha_div_extended_inf$sga_new <- as.numeric(alpha_div_extended_inf$sga_new)
alpha_div_extended_mat_pre$c_haz6 <- as.numeric(alpha_div_extended_mat_pre$c_haz6)
alpha_div_extended_inf$c_haz6 <- as.numeric(alpha_div_extended_inf$c_haz6)
alpha_div_extended_mat_pre$c_stunting6 <- as.numeric(alpha_div_extended_mat_pre$c_stunting6)
alpha_div_extended_inf$c_stunting6 <- as.numeric(alpha_div_extended_inf$c_stunting6)
```

```{r learners}
## for binary output variable
lrn_glm_bin <- sl3::Lrnr_glm$new(family = "binomial")
lrn_ridge_bin <- sl3::Lrnr_glmnet$new(alpha = 0, family = "binomial")
lrn_lasso_bin <- sl3::Lrnr_glmnet$new(alpha = 1, family = "binomial")
lrn_ranger <- sl3::Lrnr_ranger$new()
stack_bin <- sl3::Stack$new(lrn_glm_bin, lrn_ridge_bin, lrn_lasso_bin, lrn_ranger)
sl_bin <- sl3::Lrnr_sl$new(learners = stack_bin, metalearner = Lrnr_nnls$new())
## for continuous output variable
lrn_ridge_contin <- sl3::Lrnr_glmnet$new(alpha=0, family="gaussian")
lrn_lasso_contin <- sl3::Lrnr_glmnet$new(alpha=1, family="gaussian")
lrn_fglm_contin <- sl3::Lrnr_glm_fast$new(family=gaussian())
lrn_ranger <- sl3::Lrnr_ranger$new()
stack_contin <- sl3::Stack$new(lrn_ridge_contin, lrn_lasso_contin, lrn_fglm_contin, lrn_ranger)
sl_contin <- sl3::Lrnr_sl$new(learners = stack_contin, metalearner = Lrnr_nnls$new())
# # SL learners used for continuous data (the nuisance parameter Z)
# enet_contin_learner <- Lrnr_glmnet$new(
#   alpha = 0.5, family = "gaussian", nfolds = 3
# )
# lasso_contin_learner <- Lrnr_glmnet$new(
#   alpha = 1, family = "gaussian", nfolds = 3
# )
# fglm_contin_learner <- Lrnr_glm_fast$new(family = gaussian())
# mean_learner <- Lrnr_mean$new()
# contin_learner_lib <- Stack$new(
#   enet_contin_learner, lasso_contin_learner, fglm_contin_learner, mean_learner
# )
# sl_contin_learner <- Lrnr_sl$new(learners = contin_learner_lib)

# # SL learners used for binary data (nuisance parameters G and E in this case)
# enet_binary_learner <- Lrnr_glmnet$new(
#   alpha = 0.5, family = "binomial", nfolds = 3
# )
# lasso_binary_learner <- Lrnr_glmnet$new(
#   alpha = 1, family = "binomial", nfolds = 3
# )
# fglm_binary_learner <- Lrnr_glm_fast$new(family = binomial())
# binary_learner_lib <- Stack$new(
#   enet_binary_learner, lasso_binary_learner, fglm_binary_learner, mean_learner
# )
# sl_binary_learner <- Lrnr_sl$new(learners = binary_learner_lib)

# # create list for treatment and outcome mechanism regressions
# learner_list <- list(
#   Y = sl_contin_learner,
#   A = sl_binary_learner
# )
```


#### Prenatal microbiome mediating birth weight

##### Combining timepoints

```{r birth weight, warning = F}
w <- alpha_div_extended_mat_pre[, covariates_prenatal]
a <- alpha_div_extended_mat_pre$treatment1_prenatal_n
y <- alpha_div_extended_mat_pre$c_weight0
m <- alpha_div_extended_mat_pre$Shannon

# check missing data
print(sum(is.na(alpha_div_extended_mat_pre)))

tmle_nde_prenatal_weight <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "direct",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_contin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nde_prenatal_weight

tmle_nie_prenatal_weight <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "indirect",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_contin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nie_prenatal_weight
# node_list <- list(
#   W = c("w_age", "m_bmiincl", "mddw_10_ave"), #covariates_prenatal,
#   A = "treatment1_prenatal_n",
#   Z = "Shannon",
#   Y = "c_weight0"
# )
# # handle missing data
# med_process_prenatal <- tmle3::process_missing(
#   alpha_div_extended_mat_pre, node_list
# )
# mediation_prenatal <- med_process_prenatal$data
# node_list <- med_process_prenatal$node_list
# # targeted estimation of NIE
# NIE_tmle_prenatal <- tmle3mediate::tmle_NIE(
#   e_learners = Lrnr_cv$new(lasso_binary_learner, full_fit = TRUE),
#   psi_Z_learners = Lrnr_cv$new(lasso_contin_learner, full_fit = TRUE),
#   max_iter = 5
# )
# NIE_prenatal <- tmle3::tmle3(
#   tmle_spec = NIE_tmle_prenatal,
#   data = mediation_prenatal,
#   node_list = node_list,
#   learner_list = learner_list
# )
# print(NIE_prenatal)
# # targeted estimation of NDE
# NDE_tmle_prenatal <- tmle3mediate::tmle_NDE(
#   e_learners = Lrnr_cv$new(lasso_binary_learner, full_fit = TRUE),
#   psi_Z_learners = Lrnr_cv$new(lasso_contin_learner, full_fit = TRUE),
#   max_iter = 5
# )
# NDE_prenatal <- tmle3::tmle3(
#   tmle_spec = NDE_tmle_prenatal,
#   data = mediation_prenatal,
#   node_list = node_list,
#   learner_list = learner_list
# )
# print(NDE_prenatal)
```

##### Excluding incl

```{r birth weight tri3, warning = F}
alpha_div_extended_mat_incl <- alpha_div_extended_mat_pre[alpha_div_extended_mat_pre$timepoint == "incl", ]
alpha_div_extended_mat_tri3 <- alpha_div_extended_mat_pre[alpha_div_extended_mat_pre$timepoint == "tri3", ]

w <- alpha_div_extended_mat_tri3[, covariates_prenatal]
a <- alpha_div_extended_mat_tri3$treatment1_prenatal_n
y <- alpha_div_extended_mat_tri3$c_weight0
m <- alpha_div_extended_mat_tri3$Shannon

tmle_nde_tri3_weight <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "direct",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_contin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nde_tri3_weight

tmle_nie_tri3_weight <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "indirect",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_contin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nie_tri3_weight
```

#### Prenatal microbiome mediating birth height

```{r birth length, warning = F}
w <- alpha_div_extended_mat_pre[, covariates_prenatal]
a <- alpha_div_extended_mat_pre$treatment1_prenatal_n
y <- alpha_div_extended_mat_pre$c_height0
m <- alpha_div_extended_mat_pre$Shannon

tmle_nde_prenatal_height <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "direct",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_contin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nde_prenatal_height

tmle_nie_prenatal_height <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "indirect",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_contin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nie_prenatal_height
```

##### Excluding incl

```{r birth length tri3, warning = F}
w <- alpha_div_extended_mat_tri3[, covariates_prenatal]
a <- alpha_div_extended_mat_tri3$treatment1_prenatal_n
y <- alpha_div_extended_mat_tri3$c_height0
m <- alpha_div_extended_mat_tri3$Shannon

tmle_nde_tri3_height <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "direct",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_contin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nde_tri3_height

tmle_nie_tri3_height <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "indirect",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_contin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nie_tri3_height
```


#### Prenatal microbiome mediating gestational age at birth

```{r gestational age, warning = F}
w <- alpha_div_extended_mat_pre[, covariates_prenatal]
a <- alpha_div_extended_mat_pre$treatment1_prenatal_n
y <- alpha_div_extended_mat_pre$GAbirthweeks
m <- alpha_div_extended_mat_pre$Shannon

tmle_nde_prenatal_ga <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "direct",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_contin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nde_prenatal_ga

tmle_nie_prenatal_ga <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "indirect",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_contin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nie_prenatal_ga
```

##### Excluding incl

```{r gestational age tri3, warning = F}
w <- alpha_div_extended_mat_tri3[, covariates_prenatal]
a <- alpha_div_extended_mat_tri3$treatment1_prenatal_n
y <- alpha_div_extended_mat_tri3$GAbirthweeks
m <- alpha_div_extended_mat_tri3$Shannon

tmle_nde_tri3_ga <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "direct",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_contin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nde_tri3_ga

tmle_nie_tri3_ga <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "indirect",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_contin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nie_tri3_ga
```


#### Prenatal microbiome mediating low birth weight (LBW)

##### Combining timepoints

```{r low birth weight, warning = F}
w <- alpha_div_extended_mat_pre[, covariates_prenatal]
a <- alpha_div_extended_mat_pre$treatment1_prenatal_n
y <- alpha_div_extended_mat_pre$lbw
m <- alpha_div_extended_mat_pre$Shannon

tmle_nde_prenatal_lbw <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "direct",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_bin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nde_prenatal_lbw

tmle_nie_prenatal_lbw <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "indirect",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_bin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nie_prenatal_lbw
```

##### Excluding incl

```{r low birth weight tri3, warning = F}
w <- alpha_div_extended_mat_tri3[, covariates_prenatal]
a <- alpha_div_extended_mat_tri3$treatment1_prenatal_n
y <- alpha_div_extended_mat_tri3$lbw
m <- alpha_div_extended_mat_tri3$Shannon

tmle_nde_tri3_lbw <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "direct",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_bin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nde_tri3_lbw

tmle_nie_tri3_lbw <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "indirect",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_bin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nie_tri3_lbw
```

#### Prenatal microbiome mediating small for gestational age (SGA)

##### Combining timepoints

```{r small for gestational age, warning = F}
w <- alpha_div_extended_mat_pre[, covariates_prenatal]
a <- alpha_div_extended_mat_pre$treatment1_prenatal_n
y <- alpha_div_extended_mat_pre$sga_new
m <- alpha_div_extended_mat_pre$Shannon

tmle_nde_prenatal_sga <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "direct",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_bin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nde_prenatal_sga

tmle_nie_prenatal_sga <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "indirect",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_bin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nie_prenatal_sga
```

##### Excluding incl

```{r small for gestational age tri3, warning = F}
w <- alpha_div_extended_mat_tri3[, covariates_prenatal]
a <- alpha_div_extended_mat_tri3$treatment1_prenatal_n
y <- alpha_div_extended_mat_tri3$sga_new
m <- alpha_div_extended_mat_tri3$Shannon

tmle_nde_tri3_sga <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "direct",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_bin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nde_tri3_sga

tmle_nie_tri3_sga <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "indirect",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_bin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nie_tri3_sga
```

#### Postnatal microbiome mediating height for age

```{r split postnatal}
alpha_div_extended_inf_pn12 <- alpha_div_extended_inf[alpha_div_extended_inf$timepoint == "pn12", ]
alpha_div_extended_inf_pn56 <- alpha_div_extended_inf[alpha_div_extended_inf$timepoint == "pn56", ]
```

##### Combining timepoints

```{r length for age combined, warning = F}
w <- alpha_div_extended_inf[, covariates_postnatal]
a <- alpha_div_extended_inf$treatment1_prenatal_n
y <- alpha_div_extended_inf$c_haz6
m <- alpha_div_extended_inf$Shannon

# check missing data
print(sum(is.na(alpha_div_extended_inf)))

tmle_nde_postnatal_haz <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "direct",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_contin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nde_postnatal_haz

tmle_nie_postnatal_haz <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "indirect",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_contin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nie_postnatal_haz
```

##### Only pn12

```{r length for age pn12, warning = F}
w <- alpha_div_extended_inf_pn12[, covariates_postnatal]
a <- alpha_div_extended_inf_pn12$treatment1_prenatal_n
y <- alpha_div_extended_inf_pn12$c_haz6
m <- alpha_div_extended_inf_pn12$Shannon

# check missing data
print(sum(is.na(alpha_div_extended_inf_pn12)))

tmle_nde_pn12_haz <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "direct",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_contin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nde_pn12_haz

tmle_nie_pn12_haz <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "indirect",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_contin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nie_pn12_haz
```

##### Only pn56

```{r length for age pn56, warning = F}
w <- alpha_div_extended_inf_pn56[, covariates_postnatal]
a <- alpha_div_extended_inf_pn56$treatment1_prenatal_n
y <- alpha_div_extended_inf_pn56$c_haz6
m <- alpha_div_extended_inf_pn56$Shannon

# check missing data
print(sum(is.na(alpha_div_extended_inf_pn56)))

tmle_nde_pn56_haz <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "direct",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_contin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nde_pn56_haz

tmle_nie_pn56_haz <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "indirect",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_contin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nie_pn56_haz
```

##### Chained mediation

```{r formatting timeseries}
# take data with complete sampling
shannon_div <- phyloseq::estimate_richness(
  ps_object,
  split = TRUE,
  measures = "Shannon"
  )
shannon_div <- merge(
  x=shannon_div,
  y=metadata[, c("idbs", "child", "timepoint")],
  by=0
  )
row.names(shannon_div) <- shannon_div$Row.names
shannon_div$Row.names <- NULL
shannon_div_inf <- shannon_div[shannon_div$child == "yes", ]
shannon_div_inf$child <- NULL
shannon_div_inf_timeseries <- reshape2::dcast(
  data = shannon_div_inf,
  formula = idbs ~ timepoint,
  value.var = "Shannon"
  )

# merge with metadata to split on groups
shannon_div_inf_timeseries <- merge(
  x=shannon_div_inf_timeseries,
  y=unique(
    metadata[, c(
      "idbs", "treatment1_prenatal_n", "m_heightincl", "m_hbincl",
      "w_school_level", "w_marital", "asset1comp_10", "w_language",
      "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age",
      "m_bmiincl", "mddw_10_ave", "ch_sex", "iycf_ebf_age",
      "c_haz6", "c_stunting6"
    )]
  ),
  by="idbs"
  )

# remove missing data (haz6 and stunting6 are missing for the same 9 children, so can be removed together)
shannon_div_inf_timeseries <- shannon_div_inf_timeseries[complete.cases(shannon_div_inf_timeseries), ]

# continuous variables
covariates_postnatal <- c(
  "m_heightincl", "m_hbincl", "asset1comp_10",
  "w_nr_jobs", "w_age", "m_bmiincl", "mddw_10_ave", "iycf_ebf_age"
)

# make categorical variables into binary dummy vars
shannon_div_inf_timeseries[, c(
  "w_marital", "w_language", "w_religion", "w_etnicity", "csps_n"
)] <- lapply(
  shannon_div_inf_timeseries[, c(
    "w_marital", "w_language", "w_religion", "w_etnicity", "csps_n"
  )],
  factor
)
shannon_div_inf_timeseries$ch_sex <- dplyr::if_else(
  shannon_div_inf_timeseries$ch_sex == 1,
  0, 1
)
dummified_postnatal <- fastDummies::dummy_columns(
  .data = shannon_div_inf_timeseries[, c("w_school_level", "w_marital", "w_language", "w_religion", "w_etnicity", "csps_n")],
  remove_selected_columns = TRUE
)
shannon_div_inf_timeseries <- cbind(
  shannon_div_inf_timeseries[, !(
    names(shannon_div_inf_timeseries) %in% c("w_school_level", "w_marital", "w_language", "w_religion", "w_etnicity", "csps_n")
  )],
  dummified_postnatal
)
covariates_postnatal <- c(covariates_postnatal, names(dummified_postnatal))

# set exposure and outcome types
shannon_div_inf_timeseries$treatment1_prenatal_n <- as.numeric(shannon_div_inf_timeseries$treatment1_prenatal_n)
shannon_div_inf_timeseries$pn12 <- as.numeric(shannon_div_inf_timeseries$pn12)
shannon_div_inf_timeseries$pn56 <- as.numeric(shannon_div_inf_timeseries$pn56)
shannon_div_inf_timeseries$c_haz6 <- as.numeric(shannon_div_inf_timeseries$c_haz6)
shannon_div_inf_timeseries$c_stunting6 <- as.numeric(shannon_div_inf_timeseries$c_stunting6)
```

```{r}
#' function to calculate the NDE in a mediation analysis
#' @param df dataframe with variables (possibly split from a bootstrap)
#' @param y_type type of outcome variable (continuous or binary)
#' @return estimated natural direct effect (NDE)
#' 
multimediator_nde <- function(split, y_type = "continuous") {
  stopifnot(y_type %in% c("continuous", "binary"))
  if (y_type == "continuous") {
    message("Continuous outcome variable selected, outputted estimates will be effect sizes.")
  } else {
    message("Binary outcome variable selected, outputted estimates will be odds ratios.")
  }

  # format
  a <- split$a
  m1 <- split$m1
  m2 <- split$m2
  y <- split$y
  covariate_names <- names(split)[!names(split) %in% c("a", "m1", "m2", "y")]
  w <- split[, covariate_names]

  # modelling continuous outcome from exposure, mediators, and covariates
  m_y <- lm(y ~ a + m1 + m2 + a:m1 + a:m2 + m1:m2 + ., data = w)
  theta_a <- coef(m_y)[["a"]]
  theta_am1 <- coef(m_y)[["a:m1"]]
  theta_am2 <- coef(m_y)[["a:m2"]]

  # modelling continuous mediators from exposure
  m_m1 <- lm(m1 ~ a + ., data = w)
  m_m2 <- lm(m2 ~ a + ., data = w)

  # predicting counterfactual mediator values
  a0_counterfactual_m1 <- predict(
    m_m1,
    data.frame(cbind("a" = rep(0, length(a)), w))
  )
  a0_counterfactual_m2 <- predict(
    m_m2,
    data.frame(cbind("a" = rep(0, length(a)), w))
  )
  a1_counterfactual_m1 <- predict(
    m_m1,
    data.frame(cbind("a" = rep(0, length(a)), w))
  )
  a1_counterfactual_m2 <- predict(
    m_m2,
    data.frame(cbind("a" = rep(0, length(a)), w))
  )

  # predicting mediator interaction assuming a constant covariance matrix
  a0_counterfactual_m1m2 <- a0_counterfactual_m1 * a0_counterfactual_m2 + cov(m_m1$residuals, m_m2$residuals)
  a1_counterfactual_m1m2 <- a1_counterfactual_m1 * a1_counterfactual_m2 + cov(m_m1$residuals, m_m2$residuals)

  return(theta_a + (theta_am1 * mean(a0_counterfactual_m1) + theta_am2 * mean(a0_counterfactual_m2)))
}


#' function to calculate the NIE in a mediation analysis
#' @param df dataframe with variables (possibly split from a bootstrap)
#' @param y_type type of outcome variable (continuous or binary)
#' @return estimated natural direct effect (NDE)
#' 
multimediator_nie <- function(split, y_type = "continuous") {
  stopifnot(y_type %in% c("continuous", "binary"))
  if (y_type == "continuous") {
    message("Continuous outcome variable selected, outputted estimates will be effect sizes.")
  } else {
    message("Binary outcome variable selected, outputted estimates will be odds ratios.")
  }
  
  # format
  a <- split$a
  m1 <- split$m1
  m2 <- split$m2
  y <- split$y
  covariate_names <- names(split)[!names(split) %in% c("a", "m1", "m2", "y")]
  w <- split[, covariate_names]

  # modelling continuous outcome from exposure, mediators, and covariates
  m_y <- lm(y ~ a + m1 + m2 + a:m1 + a:m2 + m1:m2 + ., data = w)
  theta_m1 <- coef(m_y)[["m1"]]
  theta_m2 <- coef(m_y)[["m2"]]
  theta_am1 <- coef(m_y)[["a:m1"]]
  theta_am2 <- coef(m_y)[["a:m2"]]
  tau_m1m2 <- coef(m_y)[["m1:m2"]]

  # modelling continuous mediators from exposure
  m_m1 <- lm(m1 ~ a + ., data = w)
  beta_m1_a <- coef(m_m1)[["a"]]
  m_m2 <- lm(m2 ~ a + ., data = w)
  beta_m2_a <- coef(m_m2)[["a"]]

  # predicting counterfactual mediator values
  a0_counterfactual_m1 <- predict(
    m_m1,
    data.frame(cbind("a" = rep(0, length(a)), w))
  )
  a0_counterfactual_m2 <- predict(
    m_m2,
    data.frame(cbind("a" = rep(0, length(a)), w))
  )
  a1_counterfactual_m1 <- predict(
    m_m1,
    data.frame(cbind("a" = rep(0, length(a)), w))
  )
  a1_counterfactual_m2 <- predict(
    m_m2,
    data.frame(cbind("a" = rep(0, length(a)), w))
  )

  # predicting mediator interaction assuming a constant covariance matrix
  a0_counterfactual_m1m2 <- a0_counterfactual_m1 * a0_counterfactual_m2 + cov(m_m1$residuals, m_m2$residuals)
  a1_counterfactual_m1m2 <- a1_counterfactual_m1 * a1_counterfactual_m2 + cov(m_m1$residuals, m_m2$residuals)

  return(((theta_m1 + theta_am1)*beta_m1_a + (theta_m2 + theta_am2)*beta_m2_a) + (tau_m1m2*(mean(a1_counterfactual_m1m2) - mean(a0_counterfactual_m1m2))))
}
```

```{r length for age chained}
w <- shannon_div_inf_timeseries[, covariates_postnatal]
a <- shannon_div_inf_timeseries$treatment1_prenatal_n
y <- shannon_div_inf_timeseries$c_haz6
m1 <- shannon_div_inf_timeseries$pn12
m2 <- shannon_div_inf_timeseries$pn56

df <- data.frame(cbind("a" = a, "m1" = m1, "m2" = m2, "y" = y, w))
boots <- rsample::bootstraps(df, times = 3000, apparent = TRUE)

# estimating NDE
nde_estimates <- purrr::map_dbl(
  boots$splits,
  function(x) {
    df <- as.data.frame(x)
    multimediator_nde(df)
  }
)
mm_nde_inf_haz <- list(
  "theta" = mean(nde_estimates),
  "var" = ((1/sqrt(length(a)))*sd(nde_estimates))
  )

# estimating NIE
nie_estimates <- purrr::map_dbl(
  boots$splits,
  function(x) {
    df <- as.data.frame(x)
    multimediator_nie(df)
  }
)
mm_nie_inf_haz <- list(
  "theta" = mean(nie_estimates),
  "var" = ((1/sqrt(length(a)))*sd(nie_estimates))
  )
```

#### Postnatal microbiome mediating stunting

##### Combining timepoints

```{r stunting combined, warning = F}
w <- alpha_div_extended_inf[, covariates_postnatal]
a <- alpha_div_extended_inf$treatment1_prenatal_n
y <- alpha_div_extended_inf$c_stunting6
m <- alpha_div_extended_inf$Shannon

# check missing data
print(sum(is.na(alpha_div_extended_inf)))

tmle_nde_postnatal_stunting <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "direct",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_contin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nde_postnatal_stunting

tmle_nie_postnatal_stunting <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "indirect",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_contin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nie_postnatal_stunting
```

##### Only pn12

```{r stunting pn12, warning = F}
w <- alpha_div_extended_inf_pn12[, covariates_postnatal]
a <- alpha_div_extended_inf_pn12$treatment1_prenatal_n
y <- alpha_div_extended_inf_pn12$c_stunting6
m <- alpha_div_extended_inf_pn12$Shannon

# check missing data
print(sum(is.na(alpha_div_extended_inf_pn12)))

tmle_nde_pn12_stunting <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "direct",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_contin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nde_pn12_stunting

tmle_nie_pn12_stunting <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "indirect",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_contin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nie_pn12_stunting
```

##### Only pn56

```{r stunting pn56, warning = F}
w <- alpha_div_extended_inf_pn56[, covariates_postnatal]
a <- alpha_div_extended_inf_pn56$treatment1_prenatal_n
y <- alpha_div_extended_inf_pn56$c_stunting6
m <- alpha_div_extended_inf_pn56$Shannon

# check missing data
print(sum(is.na(alpha_div_extended_inf_pn56)))

tmle_nde_pn56_stunting <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "direct",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_contin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nde_pn56_stunting

tmle_nie_pn56_stunting <- medoutcon(
  W = w, A = a, Z = NULL, M = m, Y = y,
  effect = "indirect",
  estimator = "tmle",
  g_learners = sl_bin,
  h_learners = sl_contin,
  b_learners = sl_contin,
  r_learners = sl_contin,
  estimator_args = list(cv_folds = 5L, max_iter = 5L, tiltmod_tol = 5)
  )
tmle_nie_pn56_stunting
```


##### Chained mediation

```{r stunting chained, eval = F}
w <- shannon_div_inf_timeseries[, covariates_postnatal]
a <- shannon_div_inf_timeseries$treatment1_prenatal_n
y <- shannon_div_inf_timeseries$c_stunting6
m1 <- shannon_div_inf_timeseries$pn12
m2 <- shannon_div_inf_timeseries$pn56

df <- data.frame(cbind("a" = a, "m1" = m1, "m2" = m2, "y" = y, w))
boots <- rsample::bootstraps(df, times = 3000, apparent = TRUE)

# estimating NDE
nde_estimates <- purrr::map_dbl(
  boots$splits,
  function(x) {
    df <- as.data.frame(x)
    multimediator_nde(df)
  }
)
mm_nde_inf_stunting <- list(
  "theta" = mean(nde_estimates),
  "var" = ((1/sqrt(length(a)))*sd(nde_estimates))
  )

# estimating NIE
nie_estimates <- purrr::map_dbl(
  boots$splits,
  function(x) {
    df <- as.data.frame(x)
    multimediator_nie(df)
  }
)
mm_nie_inf_stunting <- list(
  "theta" = mean(nde_estimates),
  "var" = ((1/sqrt(length(a)))*sd(nde_estimates))
  )
```

#### Combine results

```{r}
tmle_estimates <- list(
  tmle_nde_prenatal_height,
  tmle_nie_prenatal_height,
  tmle_nde_tri3_height,
  tmle_nie_tri3_height,
  tmle_nde_prenatal_weight,
  tmle_nie_prenatal_weight,
  tmle_nde_tri3_weight,
  tmle_nie_tri3_weight,
  tmle_nde_prenatal_ga,
  tmle_nie_prenatal_ga,
  tmle_nde_tri3_ga,
  tmle_nie_tri3_ga,
  tmle_nde_prenatal_lbw,
  tmle_nie_prenatal_lbw,
  tmle_nde_tri3_lbw,
  tmle_nie_tri3_lbw,
  tmle_nde_prenatal_sga,
  tmle_nie_prenatal_sga,
  tmle_nde_tri3_sga,
  tmle_nie_tri3_sga,
  tmle_nde_postnatal_haz,
  tmle_nie_postnatal_haz,
  tmle_nde_pn12_haz,
  tmle_nie_pn12_haz,
  tmle_nde_pn56_haz,
  tmle_nie_pn56_haz,
  mm_nde_inf_haz,
  mm_nie_inf_haz,
  tmle_nde_postnatal_stunting,
  tmle_nie_postnatal_stunting,
  tmle_nde_pn12_stunting,
  tmle_nie_pn12_stunting,
  tmle_nde_pn56_stunting,
  tmle_nie_pn56_stunting#,
  # mm_nde_inf_stunting,
  # mm_nie_inf_stunting
)
effects <- c(
  "height (NDE)", "height (NIE)",
  "height (NDE)", "height (NIE)",
  "weight (NDE)", "weight (NIE)",
  "weight (NDE)", "weight (NIE)",
  "gestational age (NDE)", "gestational age (NIE)",
  "gestational age (NDE)", "gestational age (NIE)",
  "lbw (NDE)", "lbw (NIE)",
  "lbw (NDE)", "lbw (NIE)",
  "sga (NDE)", "sga (NIE)",
  "sga (NDE)", "sga (NIE)",
  "haz (NDE)", "haz (NIE)",
  "haz (NDE)", "haz (NIE)",
  "haz (NDE)", "haz (NIE)",
  "haz (NDE)", "haz (NIE)",
  "stunting (NDE)", "stunting (NIE)",
  "stunting (NDE)", "stunting (NIE)",
  "stunting (NDE)", "stunting (NIE)"#,
  # "stunting (NDE)", "stunting (NIE)"
)
tmle_estimates_df <- data.frame(effect=c(), estimate=c(), std.error=c(), conf.low=c(), conf.high=c())

for (effect_idx in 1:length(tmle_estimates)){
  row <- data.frame(
    "outcome" = effects[effect_idx],
    "estimate" = as.numeric(tmle_estimates[[effect_idx]]$theta),
    "std.error" = sqrt(as.numeric(tmle_estimates[[effect_idx]]$var)),
    "conf.low" = as.numeric(
      tmle_estimates[[effect_idx]]$theta
    ) - 1.96 * sqrt(as.numeric(tmle_estimates[[effect_idx]]$var)),
    "conf.high" = as.numeric(
      tmle_estimates[[effect_idx]]$theta
    ) + 1.96 * sqrt(as.numeric(tmle_estimates[[effect_idx]]$var))
  )
  tmle_estimates_df <- rbind(tmle_estimates_df, row)
}
tmle_estimates_df <- cbind(
  "treatment" = rep("BEP", nrow(tmle_estimates_df)),
  # "mediator" = rep("maternal microbiome diversity", nrow(tmle_estimates_df)),
  "mediator" = c(rep("maternal microbiome diversity", 20), rep("infant microbiome diversity", 14)),
  "timepoint" = c(
    rep("incl-tri3 (combined)", 2), rep("tri3", 2), rep("incl-tri3 (combined)", 2), rep("tri3", 2),
    rep("incl-tri3 (combined)", 2), rep("tri3", 2), rep("incl-tri3 (combined)", 2), rep("tri3", 2),
    rep("incl-tri3 (combined)", 2), rep("tri3", 2),
    rep("pn12-pn56 (combined)", 2), rep("pn12", 2), rep("pn56", 2), rep("pn12-pn56 (chained)", 2),
    rep("pn12-pn56 (combined)", 2), rep("pn12", 2), rep("pn56", 2)#, rep("pn12-pn56 (chained)", 2)
  ),
  tmle_estimates_df
)
```




## B-D. mediation on taxonomic data

### load packages
```{r}

source(paste(wd,"/HDmediation/b.R", sep=""))
source(paste(wd,"/HDmediation/c.R", sep=""))
source(paste(wd,"/HDmediation/crossfit.R", sep=""))
source(paste(wd,"/HDmediation/e.R", sep=""))
source(paste(wd,"/HDmediation/folds.R", sep=""))
source(paste(wd,"/HDmediation/g.R", sep=""))
source(paste(wd,"/HDmediation/h_m.R", sep=""))
source(paste(wd,"/HDmediation/h_z.R", sep=""))
source(paste(wd,"/HDmediation/not_transported.R", sep=""))
source(paste(wd,"/HDmediation/npsem.R", sep=""))
source(paste(wd,"/HDmediation/transported.R", sep=""))
source(paste(wd,"/HDmediation/u.R", sep=""))
source(paste(wd,"/HDmediation/utils.R", sep=""))
source(paste(wd,"/HDmediation/v.R", sep=""))
source(paste(wd,"/HDmediation/mediation.R", sep=""))


```

### load function
```{r}
set.seed(123)
sl_library <- c("SL.glm", "SL.mean")

extract_vars <- function(timepoint_ps, outcome_var, taxonomic_level="species", postnatal=FALSE, impute=FALSE){
  stopifnot(taxonomic_level %in% phyloseq::rank_names(timepoint_ps))
  # Extract the exposure
  A <- phyloseq::sample_data(timepoint_ps)$treatment1_prenatal_n
  # Extract the outcome
  Y <- phyloseq::sample_data(timepoint_ps)[[outcome_var]]
  # Extract the mediators
  tax_level_ps <- phyloseq::tax_glom(timepoint_ps, taxonomic_level)
  M <- t(phyloseq::otu_table(tax_level_ps))
  colnames(M) <- phyloseq::tax_table(tax_level_ps)[, taxonomic_level]
  # prevalence filtering (cutoff 20%)
  M <- as.matrix(M[, apply(M, 2, function(col) mean(col != 0)) >= 0.2])
  # extract the covariates
  covars_prenatal <- c(
    "m_heightincl", "m_hbincl", "w_marital", "asset1comp_10", "w_language", "w_religion",
    "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave"
    )
  covars_postnatal <- c(covars_prenatal, "iycf_ebf_age")
  W <- data.frame(phyloseq::sample_data(timepoint_ps)) %>%
    select(
      ifelse(postnatal, covars_postnatal, covars_prenatal)
      )
  if (impute){
    W <- W %>%
      mutate(across(everything(), impute_lowest))
  }
  # take complete cases
  complete_cases <- complete.cases(A, Y, M, W)
  if (sum(complete_cases) != length(A)){
    message("Removed ", length(A) - sum(complete_cases), " observations due to missing values.")
  }
  A <- A[complete_cases]
  Y <- Y[complete_cases]
  M <- M[complete_cases, ]
  W <- W[complete_cases, ]
  # Convert covariates to numeric
  W <- data.frame(lapply(W, function(x) as.numeric(as.character(x)))) %>%
    as.matrix()
  message(
    paste0(
      "Extracted exposure A with ", length(A), " observations.\n",
      "Extracted outcome Y with ", length(Y), " observations.\n",
      "Extracted mediators M with ", dim(M)[1], " observations and ", dim(M)[2], " features.\n",
      "Extracted covariates W with ", dim(W)[1], " observations and ", dim(W)[2], " features."
      )
    )
  vars_df <- data.frame(A, Y, M, W)
  return(
    list(
      df=vars_df,
      mediator_names=names(vars_df)[3:(ncol(vars_df)-ncol(W))],
      covariate_names=names(vars_df)[(ncol(vars_df)-ncol(W)+1):ncol(vars_df)]
    )
  )
}

impute_lowest <- function(x) {
  if (any(!is.na(x))) {
    min_val <- min(x, na.rm = TRUE)
    x[is.na(x)] <- min_val
  }
  return(x)
}
```

### Infant microbiome mediation
#### of the BEP->HAZ6 effect pn12
```{r mat pn12 haz, message=FALSE}
set.seed(123)
library(foreach)
library(doFuture)
vars = extract_vars(
  pn12e,
  "c_haz6",
  taxonomic_level = "species",
  postnatal = TRUE,
  impute = FALSE
  )
mediation_results_pn12_haz <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 3, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_pn12_haz <- rbind(mediation_results_pn12_haz, ans)
}
write.csv(mediation_results_pn12_haz, "./results/tax_mediation/species/mediation_results_pn12_haz.csv")

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_pn12_haz, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_pn12_haz6 <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_pn12_haz$ci_indirect_low, mediation_results_pn12_haz$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_pn12_haz$ci_indirect_high, mediation_results_pn12_haz$ci_direct_high)),
                width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "6m: HAZ", y = "PN12: Infant gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_pn12_haz6

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_pn12_haz6.svg",
  width=5, height=4
  )
plot_pn12_haz6
dev.off()

```

#### of the BEP->HAZ6 effect pn56
```{r mat pn56 haz, message=FALSE}
set.seed(123)
vars = extract_vars(
  pn56e,
  "c_haz6",
  taxonomic_level = "species",
  postnatal = TRUE,
  impute = FALSE
  )
mediation_results_pn56_haz <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_pn56_haz <- rbind(mediation_results_pn56_haz, ans)
}
write.csv(mediation_results_pn56_haz, "./results/tax_mediation/species/mediation_results_pn56_haz.csv")

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_pn56_haz, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_pn56_haz6 <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_pn56_haz$ci_indirect_low, mediation_results_pn56_haz$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_pn56_haz$ci_indirect_high, mediation_results_pn56_haz$ci_direct_high)),
                width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "6m: HAZ", y = "PN56: Infant gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_pn56_haz6

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_pn56_haz6.svg",
  width=5, height=5
  )
plot_pn56_haz6
dev.off()

```

#### of the BEP->stunting effect pn12
```{r mat pn12 stunting, message=FALSE}
set.seed(123)
library(foreach)
library(doFuture)
vars = extract_vars(
  pn12e,
  "c_stunting6",
  taxonomic_level = "species",
  postnatal = TRUE,
  impute = FALSE
  )
mediation_results_pn12_stunting <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "binomial",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_pn12_stunting <- rbind(mediation_results_pn12_stunting, ans)
}
write.csv(mediation_results_pn12_stunting, "./results/tax_mediation/species/mediation_results_pn12_stunting.csv")

library(reshape2)
# Melt the data frame for ggplot
data_melted <- melt(mediation_results_pn12_stunting, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_pn12_stunting6 <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_pn12_stunting$ci_indirect_low, mediation_results_pn12_stunting$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_pn12_stunting$ci_indirect_high, mediation_results_pn12_stunting$ci_direct_high)),
                width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "6m: Stunting", y = "PN12: Infant gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_pn12_stunting6

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_pn12_stunting6.svg",
  width=5, height=4
  )
plot_pn12_stunting6
dev.off()

```

#### of the BEP->stunting effect pn56
```{r mat pn56 stunting, message=FALSE}
set.seed(123)
library(foreach)
library(doFuture)
vars = extract_vars(
  pn56e,
  "c_stunting6",
  taxonomic_level = "species",
  postnatal = TRUE,
  impute = FALSE
  )
mediation_results_pn56_stunting <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "binomial",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_pn56_stunting <- rbind(mediation_results_pn56_stunting, ans)
}
write.csv(mediation_results_pn56_stunting, "./results/tax_mediation/species/mediation_results_pn56_stunting.csv")

library(reshape2)
# Melt the data frame for ggplot
data_melted <- melt(mediation_results_pn56_stunting, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_pn56_stunting6 <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_pn56_stunting$ci_indirect_low, mediation_results_pn56_stunting$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_pn56_stunting$ci_indirect_high, mediation_results_pn56_stunting$ci_direct_high)),
                width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "6m: Stunting", y = "PN12: Infant gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_pn56_stunting6

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_pn56_stunting6.svg",
  width=5, height=4
  )
plot_pn56_stunting6
dev.off()

```


#### of the BEP->WAZ6 effect pn12
```{r mat pn12 waz, message=FALSE}
set.seed(123)
library(foreach)
library(doFuture)
vars = extract_vars(
  pn12e,
  "c_waz6",
  taxonomic_level = "species",
  postnatal = TRUE,
  impute = FALSE
  )
mediation_results_pn12_waz <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 3, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_pn12_waz <- rbind(mediation_results_pn12_waz, ans)
}
write.csv(mediation_results_pn12_waz, "./results/tax_mediation/species/mediation_results_pn12_waz.csv")

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_pn12_waz, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_pn12_waz6 <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_pn12_waz$ci_indirect_low, mediation_results_pn12_waz$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_pn12_waz$ci_indirect_high, mediation_results_pn12_waz$ci_direct_high)),
                width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "6m: WAZ", y = "PN12: Infant gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_pn12_waz6

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_pn12_waz6.svg",
  width=5, height=4
  )
plot_pn12_waz6
dev.off()

```

#### of the BEP->WAZ6 effect pn56
```{r mat pn56 waz, message=FALSE}
set.seed(123)
vars = extract_vars(
  pn56e,
  "c_waz6",
  taxonomic_level = "species",
  postnatal = TRUE,
  impute = FALSE
  )
mediation_results_pn56_waz <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_pn56_waz <- rbind(mediation_results_pn56_waz, ans)
}
write.csv(mediation_results_pn56_waz, "./results/tax_mediation/species/mediation_results_pn56_waz.csv")

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_pn56_waz, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_pn56_waz6 <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_pn56_waz$ci_indirect_low, mediation_results_pn56_waz$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_pn56_waz$ci_indirect_high, mediation_results_pn56_waz$ci_direct_high)),
                width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "6m: WAZ", y = "PN56: Infant gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_pn56_waz6

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_pn56_waz6.svg",
  width=5, height=5
  )
plot_pn56_waz6
dev.off()

```


#### of the BEP->WHZ6 effect pn12
```{r mat pn12 whz, message=FALSE}
set.seed(123)
library(foreach)
library(doFuture)
vars = extract_vars(
  pn12e,
  "c_whz6",
  taxonomic_level = "species",
  postnatal = TRUE,
  impute = FALSE
  )
mediation_results_pn12_whz <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 3, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_pn12_whz <- rbind(mediation_results_pn12_whz, ans)
}
write.csv(mediation_results_pn12_whz, "./results/tax_mediation/species/mediation_results_pn12_whz.csv")

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_pn12_whz, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_pn12_whz6 <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_pn12_whz$ci_indirect_low, mediation_results_pn12_whz$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_pn12_whz$ci_indirect_high, mediation_results_pn12_whz$ci_direct_high)),
                width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "6m: WHZ", y = "PN12: Infant gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_pn12_whz6

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_pn12_whz6.svg",
  width=5, height=4
  )
plot_pn12_whz6
dev.off()

```

#### of the BEP->WHZ6 effect pn56
```{r mat pn56 whz, message=FALSE}
set.seed(123)
vars = extract_vars(
  pn56e,
  "c_whz6",
  taxonomic_level = "species",
  postnatal = TRUE,
  impute = FALSE
  )
mediation_results_pn56_whz <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_pn56_whz <- rbind(mediation_results_pn56_whz, ans)
}
write.csv(mediation_results_pn56_whz, "./results/tax_mediation/species/mediation_results_pn56_whz.csv")

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_pn56_whz, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_pn56_whz6 <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_pn56_whz$ci_indirect_low, mediation_results_pn56_whz$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_pn56_whz$ci_indirect_high, mediation_results_pn56_whz$ci_direct_high)),
                width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "6m: WHZ", y = "PN56: Infant gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_pn56_whz6

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_pn56_whz6.svg",
  width=5, height=5
  )
plot_pn56_whz6
dev.off()

```


#### of the BEP->HAZ3 effect pn12
```{r mat pn12 haz, message=FALSE}
set.seed(123)
library(foreach)
library(doFuture)
vars = extract_vars(
  pn12e,
  "c_haz3",
  taxonomic_level = "species",
  postnatal = TRUE,
  impute = FALSE
  )
mediation_results_pn12_haz <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 3, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_pn12_haz <- rbind(mediation_results_pn12_haz, ans)
}
write.csv(mediation_results_pn12_haz, "./results/tax_mediation/species/mediation_results_pn12_haz3.csv")

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_pn12_haz, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_pn12_haz3 <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_pn12_haz$ci_indirect_low, mediation_results_pn12_haz$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_pn12_haz$ci_indirect_high, mediation_results_pn12_haz$ci_direct_high)),
                width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "3m: HAZ", y = "PN12: Infant gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_pn12_haz3

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_pn12_haz3.svg",
  width=5, height=4
  )
plot_pn12_haz3
dev.off()

```


#### of the BEP->WAZ3 effect pn12
```{r mat pn12 waz, message=FALSE}
set.seed(123)
library(foreach)
library(doFuture)
vars = extract_vars(
  pn12e,
  "c_waz3",
  taxonomic_level = "species",
  postnatal = TRUE,
  impute = FALSE
  )
mediation_results_pn12_waz <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 3, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_pn12_waz <- rbind(mediation_results_pn12_waz, ans)
}
write.csv(mediation_results_pn12_waz, "./results/tax_mediation/species/mediation_results_pn12_waz3.csv")

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_pn12_waz, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_pn12_waz3 <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_pn12_waz$ci_indirect_low, mediation_results_pn12_waz$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_pn12_waz$ci_indirect_high, mediation_results_pn12_waz$ci_direct_high)),
                width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "3m: WAZ", y = "PN12: Infant gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_pn12_waz3

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_pn12_waz3.svg",
  width=5, height=4
  )
plot_pn12_waz3
dev.off()

```


#### of the BEP->WHZ3 effect pn12
```{r mat pn12 whz, message=FALSE}
set.seed(123)
library(foreach)
library(doFuture)
vars = extract_vars(
  pn12e,
  "c_whz3",
  taxonomic_level = "species",
  postnatal = TRUE,
  impute = FALSE
  )

mediation_results_pn12_whz <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 3, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  # Check if the results contain NaNs or missing values
   if (any(is.nan(unlist(ans)))) {
     cat("NaNs produced for mediator:", vars$mediator_names[i], "\n")
   }
  
  ans$mediator <- vars$mediator_names[i]
  mediation_results_pn12_whz <- rbind(mediation_results_pn12_whz, ans)
}
write.csv(mediation_results_pn12_whz, "./results/tax_mediation/species/mediation_results_pn12_whz3.csv")

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_pn12_whz, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_pn12_whz3 <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_pn12_whz$ci_indirect_low, mediation_results_pn12_whz$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_pn12_whz$ci_indirect_high, mediation_results_pn12_whz$ci_direct_high)),
                width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "3m: WHZ", y = "PN12: Infant gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_pn12_whz3

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_pn12_whz3.svg",
  width=5, height=4
  )
plot_pn12_whz3
dev.off()

```


### Maternal microbiome mediation 

#### of the BEP->GA effect
```{r mat incl GA}
set.seed(123)
vars = extract_vars(
  incl,
  "GAbirthweeks",
  taxonomic_level = "species",
  impute = FALSE
  )
mediation_results_incl_GA <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian", folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_incl_GA <- rbind(mediation_results_incl_GA, ans)
}
write.csv(mediation_results_incl_GA, "./results/tax_mediation/species/mediation_results_incl_GA.csv")

mediation_results_incl_GA <- read.csv("./results/tax_mediation/species/mediation_results_incl_GA.csv", row.names = 1)

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_incl_GA, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_incl_GA <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_incl_GA$ci_indirect_low, mediation_results_incl_GA$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_incl_GA$ci_indirect_high, mediation_results_incl_GA$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Gestational age", y = "Tri2: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_incl_GA

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_incl_GA.svg",
  width=5, height=45
  )
plot_incl_GA
dev.off()

filtered_mediation_results_incl_GA <- mediation_results_incl_GA %>%
  filter(direct >=0) %>%
  #filter(sign(ci_indirect_low) == sign(ci_indirect_high)) %>%
  filter(sign(ci_direct_low) == sign(ci_direct_high))

# Melt the data frame for ggplot
data_melted <- melt(filtered_mediation_results_incl_GA, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_incl_GA_filter <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", filtered_mediation_results_incl_GA$ci_indirect_low, filtered_mediation_results_incl_GA$ci_direct_low),
                    xmax = ifelse(variable == "indirect", filtered_mediation_results_incl_GA$ci_indirect_high, filtered_mediation_results_incl_GA$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Gestational age", y = "Tri2: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_incl_GA_filter

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_incl_GA_filter2.svg",
  width=6, height=3
  )
plot_incl_GA_filter
dev.off()


```

```{r mat tri3 GA}
set.seed(123)
vars = extract_vars(
  tri3,
  "GAbirthweeks",
  taxonomic_level = "species",
  impute = FALSE
  )
mediation_results_tri3_GA <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_tri3_GA <- rbind(mediation_results_tri3_GA, ans)
}
write.csv(mediation_results_tri3_GA, "./results/tax_mediation/species/mediation_results_tri3_GA.csv")

mediation_results_tri3_GA <- read.csv("./results/tax_mediation/species/mediation_results_tri3_GA.csv", row.names = 1)

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_tri3_GA, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_tri3_GA <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_tri3_GA$ci_indirect_low, mediation_results_tri3_GA$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_tri3_GA$ci_indirect_high, mediation_results_tri3_GA$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Gestational age", y = "Tri3: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_tri3_GA

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_tri3_GA.svg",
  width=5, height=45
  )
plot_tri3_GA
dev.off()

filtered_mediation_results_tri3_GA <- mediation_results_tri3_GA %>%
  #filter(direct >=0) %>%
  #filter(indirect >=0) %>%
  filter(sign(ci_indirect_low) == sign(ci_indirect_high)) %>%
  filter(sign(ci_direct_low) == sign(ci_direct_high))

# Melt the data frame for ggplot
data_melted <- melt(filtered_mediation_results_tri3_GA, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_tri3_GA_filter <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", filtered_mediation_results_tri3_GA$ci_indirect_low, filtered_mediation_results_tri3_GA$ci_direct_low),
                    xmax = ifelse(variable == "indirect", filtered_mediation_results_tri3_GA$ci_indirect_high, filtered_mediation_results_tri3_GA$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Gestational age", y = "Tri3: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_tri3_GA_filter

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_tri3_GA_filter2.svg",
  width=6, height=3
  )
plot_tri3_GA_filter
dev.off()

```

#### of the BEP->birth weight effect
```{r mat incl weight}
set.seed(123)
vars = extract_vars(
  incl,
  "c_weight0",
  taxonomic_level = "species",
  impute = FALSE
  )
mediation_results_incl_weight <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_incl_weight <- rbind(mediation_results_incl_weight, ans)
}
write.csv(mediation_results_incl_weight, "./results/tax_mediation/species/mediation_results_incl_weight.csv")

mediation_results_incl_weight <- read.csv("./results/tax_mediation/species/mediation_results_incl_weight.csv", row.names = 1)

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_incl_weight, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_incl_weight <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_incl_weight$ci_indirect_low, mediation_results_incl_weight$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_incl_weight$ci_indirect_high, mediation_results_incl_weight$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Birth Weight", y = "Tri2: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_incl_weight

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_incl_weight.svg",
  width=5, height=45
  )
plot_incl_weight
dev.off()

filtered_mediation_results_incl_weight <- mediation_results_incl_weight %>%
  filter(direct >=0) %>%
  filter(sign(ci_indirect_low) == sign(ci_indirect_high))%>%
  filter(sign(ci_direct_low) == sign(ci_direct_high))

# Melt the data frame for ggplot
data_melted <- melt(filtered_mediation_results_incl_weight, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_incl_weight_filter <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", filtered_mediation_results_incl_weight$ci_indirect_low, filtered_mediation_results_incl_weight$ci_direct_low),
                    xmax = ifelse(variable == "indirect", filtered_mediation_results_incl_weight$ci_indirect_high, filtered_mediation_results_incl_weight$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Birth Weight", y = "Tri2: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_incl_weight_filter

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_incl_weight_filter2.svg",
  width=6, height=3
  )
plot_incl_weight_filter
dev.off()

```

```{r mat tri3 weight}
set.seed(123)
vars = extract_vars(
  tri3,
  "c_weight0",
  taxonomic_level = "species",
  impute = FALSE
  )
mediation_results_tri3_weight <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_tri3_weight <- rbind(mediation_results_tri3_weight, ans)
}
write.csv(mediation_results_tri3_weight, "./results/tax_mediation/species/mediation_results_tri3_weight.csv")

mediation_results_tri3_weight <- read.csv("./results/tax_mediation/species/mediation_results_tri3_weight.csv", row.names = 1)

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_tri3_weight, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_tri3_weight <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_tri3_weight$ci_indirect_low, mediation_results_tri3_weight$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_tri3_weight$ci_indirect_high, mediation_results_tri3_weight$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Birth Weight", y = "Tri3: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_tri3_weight

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_tri3_weight.svg",
  width=5, height=45
  )
plot_tri3_weight
dev.off()

filtered_mediation_results_tri3_weight <- mediation_results_tri3_weight %>%
  filter(direct >=0) %>%
  filter(sign(ci_indirect_low) == sign(ci_indirect_high)) %>%
  filter(sign(ci_direct_low) == sign(ci_direct_high))

# Melt the data frame for ggplot
data_melted <- melt(filtered_mediation_results_tri3_weight, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_tri3_weight_filter <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", filtered_mediation_results_tri3_weight$ci_indirect_low, filtered_mediation_results_tri3_weight$ci_direct_low),
                    xmax = ifelse(variable == "indirect", filtered_mediation_results_tri3_weight$ci_indirect_high, filtered_mediation_results_tri3_weight$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Birth Weight", y = "Tri3: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_tri3_weight_filter

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_tri3_weight_filter2.svg",
  width=6, height=3
  )
plot_tri3_weight_filter
dev.off()

```

#### of the BEP->birth height effect
```{r mat incl height}
set.seed(123)
vars = extract_vars(
  incl,
  "c_height0",
  taxonomic_level = "species",
  impute = FALSE
  )
mediation_results_incl_height <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_incl_height <- rbind(mediation_results_incl_height, ans)
}
write.csv(mediation_results_incl_height, "./results/tax_mediation/species/mediation_results_incl_height.csv")

mediation_results_incl_height <- read.csv("./results/tax_mediation/species/mediation_results_incl_height.csv", row.names = 1)

# Melt the data frame for ggplot
data_melted <- melt(mediation_results_incl_height, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_incl_height <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_incl_height$ci_indirect_low, mediation_results_incl_height$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_incl_height$ci_indirect_high, mediation_results_incl_height$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Birth Height", y = "Tri2: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_incl_height

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_incl_height.svg",
  width=5, height=45
  )
plot_incl_height
dev.off()

filtered_mediation_results_incl_height <- mediation_results_incl_height %>%
  filter(direct >=0) %>%
  filter(sign(ci_indirect_low) == sign(ci_indirect_high)) %>%
  filter(sign(ci_direct_low) == sign(ci_direct_high))

# Melt the data frame for ggplot
data_melted <- melt(filtered_mediation_results_incl_height, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_incl_height_filter <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", filtered_mediation_results_incl_height$ci_indirect_low, filtered_mediation_results_incl_height$ci_direct_low),
                    xmax = ifelse(variable == "indirect", filtered_mediation_results_incl_height$ci_indirect_high, filtered_mediation_results_incl_height$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Birth Height", y = "Tri2: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_incl_height_filter

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_incl_height_filter2.svg",
  width=6, height=3
  )
plot_incl_height_filter
dev.off()

```

```{r mat tri3 height}
set.seed(123)
vars = extract_vars(
  tri3,
  "c_height0",
  taxonomic_level = "species",
  impute = FALSE
  )
mediation_results_tri3_height <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "gaussian",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_tri3_height <- rbind(mediation_results_tri3_height, ans)
}
write.csv(mediation_results_tri3_height, "./results/tax_mediation/species/mediation_results_tri3_height.csv")

mediation_results_tri3_height <- read.csv("./results/tax_mediation/species/mediation_results_tri3_height.csv", row.names = 1)
# Melt the data frame for ggplot
data_melted <- melt(mediation_results_tri3_height, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_tri3_height <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_tri3_height$ci_indirect_low, mediation_results_tri3_height$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_tri3_height$ci_indirect_high, mediation_results_tri3_height$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Birth Height", y = "Tri3: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_tri3_height

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_tri3_height.svg",
  width=5, height=45
  )
plot_tri3_height
dev.off()

filtered_mediation_results_tri3_height <- mediation_results_tri3_height %>%
  filter(direct >=0) %>%
  filter(sign(ci_indirect_low) == sign(ci_indirect_high)) %>%
  filter(sign(ci_direct_low) == sign(ci_direct_high))

# Melt the data frame for ggplot
data_melted <- melt(filtered_mediation_results_tri3_height, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_tri3_height_filter <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", filtered_mediation_results_tri3_height$ci_indirect_low, filtered_mediation_results_tri3_height$ci_direct_low),
                    xmax = ifelse(variable == "indirect", filtered_mediation_results_tri3_height$ci_indirect_high, filtered_mediation_results_tri3_height$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Birth Height", y = "Tri3: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_tri3_height_filter

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_tri3_height_filter2.svg",
  width=6, height=3
  )
plot_tri3_height_filter
dev.off()
```

#### of the BEP->LBW effect
Didn't work for LBW maybe because of the unbalance of 0 and 1 
```{r mat incl lbw}
set.seed(123)
vars = extract_vars(
  incl,
  "lbw",
  taxonomic_level = "species",
  impute = FALSE
  )
mediation_results_incl_lbw <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "binomial",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_incl_lbw <- rbind(mediation_results_incl_lbw, ans)
}
write.csv(mediation_results_incl_lbw, "./results/tax_mediation/species/mediation_results_incl_lbw.csv")
```

```{r mat tri3 lbw}
set.seed(123)
vars = extract_vars(
  tri3,
  "lbw",
  taxonomic_level = "species",
  impute = FALSE
  )
mediation_results_tri3_lbw <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "binomial",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_tri3_lbw <- rbind(mediation_results_tri3_lbw, ans)
}
write.csv(mediation_results_tri3_lbw, "./results/tax_mediation/species/mediation_results_tri3_lbw.csv")
```

 #### of the BEP->SGA effect
```{r mat incl sga}
set.seed(123)
vars = extract_vars(
  incl,
  "sga_new",
  taxonomic_level = "species",
  impute = FALSE
  )
mediation_results_incl_sga <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "binomial",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_incl_sga <- rbind(mediation_results_incl_sga, ans)
}
write.csv(mediation_results_incl_sga, "./results/tax_mediation/species/mediation_results_incl_sga.csv")

mediation_results_incl_sga <- read.csv("./results/tax_mediation/species/mediation_results_incl_sga.csv", row.names = 1)
# Melt the data frame for ggplot
data_melted <- melt(mediation_results_incl_sga, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_incl_sga <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_incl_sga$ci_indirect_low, mediation_results_incl_sga$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_incl_sga$ci_indirect_high, mediation_results_incl_sga$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Small for gestational age", y = "Tri2: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_incl_sga

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_incl_sga.svg",
  width=5, height=45
  )
plot_incl_sga
dev.off()

filtered_mediation_results_incl_sga <- mediation_results_incl_sga %>%
  filter(sign(ci_indirect_low) == sign(ci_indirect_high)) %>%
  filter(sign(ci_direct_low) == sign(ci_direct_high)) %>%
  filter(direct <= 0)

# Melt the data frame for ggplot
data_melted <- melt(filtered_mediation_results_incl_sga, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_incl_sga_filter <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", filtered_mediation_results_incl_sga$ci_indirect_low, filtered_mediation_results_incl_sga$ci_direct_low),
                    xmax = ifelse(variable == "indirect", filtered_mediation_results_incl_sga$ci_indirect_high, filtered_mediation_results_incl_sga$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Small for gestational age", y = "Tri2: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_incl_sga_filter

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_incl_sga_filter2.svg",
  width=6, height=6
  )
plot_incl_sga_filter
dev.off()
```

```{r mat tri3 sga}
set.seed(123)
vars = extract_vars(
  tri3,
  "sga_new",
  taxonomic_level = "species",
  impute = FALSE
  )
mediation_results_tri3_sga <- data.frame()
for (i in seq_along(vars$mediator_names)){
  ans <- mediation(
    data = vars$df,
    A = "A", W = vars$covariate_names,
    Z = vars$mediator_names[-i],
    M = vars$mediator_names[i],
    Y = "Y", S = NULL,
    family = "binomial",
    folds = 1, partial_tmle = TRUE, bounds = NULL,
    learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
    learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
    learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
  )
  ans$mediator <- vars$mediator_names[i]
  mediation_results_tri3_sga <- rbind(mediation_results_tri3_sga, ans)
}
write.csv(mediation_results_tri3_sga, "./results/tax_mediation/species/mediation_results_tri3_sga.csv")

mediation_results_tri3_sga <- read.csv("./results/tax_mediation/species/mediation_results_tri3_sga.csv", row.names = 1)
# Melt the data frame for ggplot
data_melted <- melt(mediation_results_tri3_sga, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_tri3_sga <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", mediation_results_tri3_sga$ci_indirect_low, mediation_results_tri3_sga$ci_direct_low),
                    xmax = ifelse(variable == "indirect", mediation_results_tri3_sga$ci_indirect_high, mediation_results_tri3_sga$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Small for gestational age", y = "Tri3: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_tri3_sga

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_tri3_sga.svg",
  width=5, height=45
  )
plot_tri3_sga
dev.off()

filtered_mediation_results_tri3_sga <- mediation_results_tri3_sga %>%
  filter(sign(ci_indirect_low) == sign(ci_indirect_high)) %>%
  filter(sign(ci_direct_low) == sign(ci_direct_high))

# Melt the data frame for ggplot
data_melted <- melt(filtered_mediation_results_tri3_sga, id.vars = "mediator", measure.vars = c("indirect", "direct"))

# Create the plot
plot_tri3_sga_filter <- ggplot(data_melted, aes(x = value, y = mediator, color = variable)) +
  geom_point(size = 2, position = position_dodge(width = 0.5)) +
  geom_errorbar(aes(xmin = ifelse(variable == "indirect", filtered_mediation_results_tri3_sga$ci_indirect_low, filtered_mediation_results_tri3_sga$ci_direct_low),
                    xmax = ifelse(variable == "indirect", filtered_mediation_results_tri3_sga$ci_indirect_high, filtered_mediation_results_tri3_sga$ci_direct_high)),
               width = 0.5, position = position_dodge(width = 0.5)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkred") +
  labs(x = "Small for gestational age", y = "Tri3: Maternal gut species") +
  scale_color_manual(values = c("steelblue", "#f1c232")) +
  theme(strip.background = element_blank(),
          axis.text = element_text(size = 6),
          strip.text = element_text(size = 8)) +
  theme_bw()

plot_tri3_sga_filter

svglite::svglite(
  filename="../../manuscripts/metagenomic_BEP/tranfer_NatureCommunications/revision1/Figures_tables_revision1/HD_plot_tri3_sga_filter2.svg",
  width=6, height=6
  )
plot_tri3_sga_filter
dev.off()
```


## Other. mediation on gene data
## Formatting 

Before we start modelling, we perform a few formatting steps on the data. 
1. We split the data into subsets to analyse separately (because the treatment effect on specific genes will likely be different for mothers and infants and different at each timepoint).
2. We combine everything into TreeSummarizedExperiment R objects.

```{r splitting data maternal, include=FALSE}
# metadata
metadata_mat <- metadata[metadata$child == "no", ]
metadata_mat_incl <- metadata_mat[metadata_mat$timepoint == "incl", ]
metadata_mat_tri3 <- metadata_mat[metadata_mat$timepoint == "tri3", ]
metadata_mat_pn12 <- metadata_mat[metadata_mat$timepoint == "pn12", ]
metadata_mat_pn56 <- metadata_mat[metadata_mat$timepoint == "pn56", ]

# counts
gene_counts_mat <- gene_counts[, rownames(metadata_mat)]
gene_counts_mat <- gene_counts_mat[rowSums(gene_counts_mat) != 0, ]
gene_counts_mat_incl <- gene_counts_mat[, rownames(metadata_mat_incl)]
gene_counts_mat_tri3 <- gene_counts_mat[, rownames(metadata_mat_tri3)]
gene_counts_mat_pn12 <- gene_counts_mat[, rownames(metadata_mat_pn12)]
gene_counts_mat_pn56 <- gene_counts_mat[, rownames(metadata_mat_pn56)]

# fix order
metadata_mat_incl <- metadata_mat_incl[colnames(gene_counts_mat_incl), ]
metadata_mat_tri3 <- metadata_mat_tri3[colnames(gene_counts_mat_tri3), ]
metadata_mat_pn12 <- metadata_mat_pn12[colnames(gene_counts_mat_pn12), ]
metadata_mat_pn56 <- metadata_mat_pn56[colnames(gene_counts_mat_pn56), ]
```


```{r splitting data infant, include=FALSE}
# metadata
metadata_inf <- metadata[metadata$child == "yes", ]
metadata_inf_pn12 <- metadata_inf[metadata_inf$timepoint == "pn12", ]
metadata_inf_pn56 <- metadata_inf[metadata_inf$timepoint == "pn56", ]

# counts
gene_counts_inf <- gene_counts[, rownames(metadata_inf)]
gene_counts_inf <- gene_counts_inf[rowSums(gene_counts_inf) != 0, ]
gene_counts_inf_pn12 <- gene_counts_inf[, rownames(metadata_inf_pn12)]
gene_counts_inf_pn56 <- gene_counts_inf[, rownames(metadata_inf_pn56)]

# fix order
metadata_inf_pn12 <- metadata_inf_pn12[colnames(gene_counts_inf_pn12), ]
metadata_inf_pn56 <- metadata_inf_pn56[colnames(gene_counts_inf_pn56), ]
```

## functions

```{r functions, echo=FALSE}
run_GSEA <- function(mediation_output, export_path, export_name, top_n = 500, gene_anno=gene_anno, pathway_to_gene=pathway_to_gene, pathway_to_description=pathway_to_description){
  tmp <- data.frame(mediation_output[, c("indirect", "mediator")])
  colnames(tmp) <- c("logFC", "names")
  kegg_genes_tmp <- tmp$names %in% gene_anno[gene_anno$database == "KEGG", "anno"]
  tmp <- tmp[kegg_genes_tmp, ]
  gene_list <- tmp[,1]
  names(gene_list) <- as.character(tmp[, 2])
  gene_list <- sort(gene_list, decreasing = TRUE)
  #GSEA
  kk <- clusterProfiler::GSEA(
    geneList = gene_list,
    minGSSize = 2,
    pvalueCutoff = 0.05,
    TERM2GENE = pathway_to_gene,
    TERM2NAME = pathway_to_description
    )
  # output tables
  enriched_genesets <- base::intersect(
    names(kk@geneSets),
    kk@result$ID
    )
  for (set in enriched_genesets){
    enriched_genes <- as.data.frame(
      na.omit(kk@geneList[unlist(kk@geneSets[set])])
      )
    colnames(enriched_genes) <- "logFC"
    enriched_genes$gene <- rownames(enriched_genes)
    cmap <- c(COLOR_SCHEME[c(3, 1)], "white", COLOR_SCHEME[c(2, 4)])
    lms <- c(-max(abs(enriched_genes$logFC)), max(abs(enriched_genes$logFC)))
    p <- ggplot2::ggplot(enriched_genes, ggplot2::aes(x=1, y=gene, color=logFC)) +
      ggplot2::geom_jitter(size=5) +
      ggplot2::scale_color_gradientn(colors=cmap, limits=lms)
    fname_plot <- paste0(export_path, "figures/", export_name, "_", set, "_enriched_genes.svg")
    ggplot2::ggsave(fname_plot, p)
    fname <- paste0(export_path, export_name, "_", set, "_enriched_genes.csv")
    write.csv(enriched_genes, fname)
  }
  # output plots
  vis <- nrow(kk@result) > 0 # only run plots if there are results
  if (vis) {
    csv_fname <- paste0(export_path, "pathway_enrichment_", export_name, ".csv")
    # svg_fname <- paste0(export_path, "figures/GSEA_", export_name, ".svg")
    write.csv(kk@result, csv_fname)
    # svglite::svglite(svg_fname, width=5, height=6)
    # enrichment_lollipop(kk)
    # dev.off()
    return(kk)
  } else {
    return(NULL)
  }
}

extract_vars <- function(timepoint_metadata, timepoint_genecounts, outcome_var, postnatal=FALSE){
  # Extract the exposure
  A <- timepoint_metadata$treatment1_prenatal_n
  # Extract the outcome
  Y <- timepoint_metadata[[outcome_var]]
  # Extract the mediators
  M <- t(timepoint_genecounts)
  # prevalence filtering (filter columns with less than 20% non-zero values)
  M <- as.matrix(M[, apply(M, 2, function(col) mean(col != 0)) >= 0.2])
  # variance filtering (sort the columns by variance and only take the top n)
  M <- as.matrix(M[, order(apply(M, 2, var), decreasing = TRUE)[1:top_n]])
  # extract the covariates
  if (postnatal){
    W <- timepoint_metadata %>%
      dplyr::select(
        "m_heightincl", "m_hbincl", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave", "iycf_ebf_age"
        )
  } else {
    W <- timepoint_metadata %>%
      dplyr::select(
        "m_heightincl", "m_hbincl", "w_marital", "asset1comp_10", "w_language", "w_religion", "w_etnicity", "w_nr_jobs", "csps_n", "w_age", "m_bmiincl", "mddw_10_ave"
        )
  }
  row.names(W) <- row.names(timepoint_metadata)
  # take complete cases
  complete_cases <- complete.cases(A, Y, M, W)
  if (sum(complete_cases) != length(A)){
    message("Removed ", length(A) - sum(complete_cases), " observations due to missing values.")
  }
  A <- A[complete_cases]
  Y <- Y[complete_cases]
  M <- M[complete_cases, ]
  W <- W[complete_cases, ]
  # Convert covariates to numeric
  W <- data.frame(lapply(W, function(x) as.numeric(as.character(x)))) %>%
    as.matrix()
  message(
    paste0(
      "Extracted exposure A with ", length(A), " observations.\n",
      "Extracted outcome Y with ", length(Y), " observations.\n",
      "Extracted mediators M with ", dim(M)[1], " observations and ", dim(M)[2], " features.\n",
      "Extracted covariates W with ", dim(W)[1], " observations and ", dim(W)[2], " features."
      )
    )
  vars_df <- data.frame(A, Y, M, W)
  return(
    list(
      df=vars_df,
      mediator_names=names(vars_df)[3:(ncol(vars_df)-ncol(W))],
      covariate_names=names(vars_df)[(ncol(vars_df)-ncol(W)+1):ncol(vars_df)]
    )
  )
}
```

## TMLE mediation on genes as mediators using HDmediation
Targeted Maximum Likelihood Estimation (TMLE) is a general approach for constructing an efficient double-robust
semi-parametric substitution estimator of a causal effect parameter.

```{r mediation setup, echo=FALSE}
set.seed(123)
sl_library <- c("SL.glm","SL.mean")
```

### Maternal microbiome mediation 
#### of the BEP->GA effect
```{r mat incl GA}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_incl_GA.csv")) {
  vars <- extract_vars(
    timepoint_metadata = metadata_mat_incl,
    timepoint_genecounts = gene_counts_mat_incl,
    outcome_var = "GAbirthweeks"
    )
  mediation_results_incl_GA <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "gaussian", folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_incl_GA <- rbind(mediation_results_incl_GA, ans)
  }
  write.csv(mediation_results_incl_GA, "./results/gene_mediation/top500/mediation_results_incl_GA.csv")
} else {
  mediation_results_incl_GA <- read.csv("./results/gene_mediation/top500/mediation_results_incl_GA.csv")
}
```

```{r GSEA mat incl GA, echo=FALSE}
kk_incl_GA <- run_GSEA(
  mediation_output = mediation_results_incl_GA,
  export_path = "./results/gene_mediation/top500/",
  export_name = "maternal_incl_GA",
  top_n = 500,
  gene_anno = gene_anno, 
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_incl_GA)){
  DT::datatable(kk_incl_GA@result)
}
```

```{r mat tri3 GA}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_tri3_GA.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_mat_tri3,
    timepoint_genecounts = gene_counts_mat_tri3,
    outcome_var = "GAbirthweeks"
    )
  mediation_results_tri3_GA <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "gaussian",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_tri3_GA <- rbind(mediation_results_tri3_GA, ans)
  }
  write.csv(mediation_results_tri3_GA, "./results/gene_mediation/top500/mediation_results_tri3_GA.csv")
} else {
  mediation_results_tri3_GA <- read.csv("./results/gene_mediation/top500/mediation_results_tri3_GA.csv")
}
```

```{r GSEA mat tri3 GA, echo=FALSE}
kk_tri3_GA <- run_GSEA(
  mediation_output = mediation_results_tri3_GA,
  export_path = "./results/gene_mediation/top500/",
  export_name = "maternal_tri3_GA",
  top_n = 500,
  gene_anno = gene_anno,
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_tri3_GA)){
  DT::datatable(kk_tri3_GA@result)
}
```

#### of the BEP->birth weight effect
```{r mat incl weight}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_incl_weight.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_mat_incl,
    timepoint_genecounts = gene_counts_mat_incl,
    outcome_var = "c_weight0"
    )
  mediation_results_incl_weight <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "gaussian",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_incl_weight <- rbind(mediation_results_incl_weight, ans)
  }
  write.csv(mediation_results_incl_weight, "./results/gene_mediation/top500/mediation_results_incl_weight.csv")
} else {
  mediation_results_incl_weight <- read.csv("./results/gene_mediation/top500/mediation_results_incl_weight.csv")
}
```

```{r GSEA mat incl weight, echo=FALSE}
kk_incl_weight <- run_GSEA(
  mediation_output = mediation_results_incl_weight,
  export_path = "./results/gene_mediation/top500/",
  export_name = "maternal_incl_weight",
  top_n = 500,
  gene_anno = gene_anno, 
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_incl_weight)){
  DT::datatable(kk_incl_weight@result)
}
```

```{r mat tri3 weight}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_tri3_weight.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_mat_tri3,
    timepoint_genecounts = gene_counts_mat_tri3,
    outcome_var = "c_weight0"
    )
  mediation_results_tri3_weight <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "gaussian",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_tri3_weight <- rbind(mediation_results_tri3_weight, ans)
  }
  write.csv(mediation_results_tri3_weight, "./results/gene_mediation/top500/mediation_results_tri3_weight.csv")
} else {
  mediation_results_tri3_weight <- read.csv("./results/gene_mediation/top500/mediation_results_tri3_weight.csv")
}
```

```{r GSEA mat tri3 weight, echo=FALSE}
kk_tri3_weight <- run_GSEA(
  mediation_output = mediation_results_tri3_weight,
  export_path = "./results/gene_mediation/top500/",
  export_name = "maternal_tri3_weight",
  top_n = 500,
  gene_anno = gene_anno,
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_tri3_weight)){
  DT::datatable(kk_tri3_weight@result)
}
```

#### of the BEP->birth height effect
```{r mat incl height}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_incl_height.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_mat_incl,
    timepoint_genecounts = gene_counts_mat_incl,
    outcome_var = "c_height0"
    )
  mediation_results_incl_height <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "gaussian",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_incl_height <- rbind(mediation_results_incl_height, ans)
  }
  write.csv(mediation_results_incl_height, "./results/gene_mediation/top500/mediation_results_incl_height.csv")
} else {
  mediation_results_incl_height <- read.csv("./results/gene_mediation/top500/mediation_results_incl_height.csv")
}
```

```{r GSEA mat incl height, echo=FALSE}
kk_incl_height <- run_GSEA(
  mediation_output = mediation_results_incl_height,
  export_path = "./results/gene_mediation/top500/",
  export_name = "maternal_incl_height",
  top_n = 500,
  gene_anno = gene_anno, 
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_incl_height)){
  DT::datatable(kk_incl_height@result)
}
```

```{r mat tri3 height}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_tri3_height.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_mat_tri3,
    timepoint_genecounts = gene_counts_mat_tri3,
    outcome_var = "c_height0"
    )
  mediation_results_tri3_height <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "gaussian",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_tri3_height <- rbind(mediation_results_tri3_height, ans)
  }
  write.csv(mediation_results_tri3_height, "./results/gene_mediation/top500/mediation_results_tri3_height.csv")
} else {
  mediation_results_tri3_height <- read.csv("./results/gene_mediation/top500/mediation_results_tri3_height.csv")
}
```

```{r GSEA mat tri3 height, echo=FALSE}
kk_tri3_height <- run_GSEA(
  mediation_output = mediation_results_tri3_height,
  export_path = "./results/gene_mediation/top500/",
  export_name = "maternal_tri3_height",
  top_n = 500,
  gene_anno = gene_anno,
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_tri3_height)){
  DT::datatable(kk_tri3_height@result)
}
```

#### of the BEP->LBW effect
```{r mat incl lbw}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_incl_lbw.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_mat_incl,
    timepoint_genecounts = gene_counts_mat_incl,
    outcome_var = "lbw"
    )
  mediation_results_incl_lbw <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "binomial",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_incl_lbw <- rbind(mediation_results_incl_lbw, ans)
  }
  write.csv(mediation_results_incl_lbw, "./results/gene_mediation/top500/mediation_results_incl_lbw.csv")
} else {
  mediation_results_incl_lbw <- read.csv("./results/gene_mediation/top500/mediation_results_incl_lbw.csv")
}
```

```{r GSEA mat incl lbw, echo=FALSE}
kk_incl_lbw <- run_GSEA(
  mediation_output = mediation_results_incl_lbw,
  export_path = "./results/gene_mediation/top500/",
  export_name = "maternal_incl_lbw",
  top_n = 500,
  gene_anno = gene_anno, 
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_incl_lbw)){
  DT::datatable(kk_incl_lbw@result)
}
```

```{r mat tri3 lbw}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_tri3_lbw.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_mat_tri3,
    timepoint_genecounts = gene_counts_mat_tri3,
    outcome_var = "lbw"
    )
  mediation_results_tri3_lbw <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "binomial",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_tri3_lbw <- rbind(mediation_results_tri3_lbw, ans)
  }
  write.csv(mediation_results_tri3_lbw, "./results/gene_mediation/top500/mediation_results_tri3_lbw.csv")
} else {
  mediation_results_tri3_lbw <- read.csv("./results/gene_mediation/top500/mediation_results_tri3_lbw.csv")
}
```

```{r GSEA mat tri3 lbw, echo=FALSE}
kk_tri3_lbw <- run_GSEA(
  mediation_output = mediation_results_tri3_lbw,
  export_path = "./results/gene_mediation/top500/",
  export_name = "maternal_tri3_lbw",
  top_n = 500,
  gene_anno = gene_anno,
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_tri3_lbw)){
  DT::datatable(kk_tri3_lbw@result)
}
```

#### of the BEP->SGA effect
```{r mat incl sga}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_incl_sga.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_mat_incl,
    timepoint_genecounts = gene_counts_mat_incl,
    outcome_var = "sga_new"
    )
  mediation_results_incl_sga <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "binomial",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_incl_sga <- rbind(mediation_results_incl_sga, ans)
  }
  write.csv(mediation_results_incl_sga, "./results/gene_mediation/top500/mediation_results_incl_sga.csv")
} else {
  mediation_results_incl_sga <- read.csv("./results/gene_mediation/top500/mediation_results_incl_sga.csv")
}
```

```{r GSEA mat incl sga, echo=FALSE}
kk_incl_sga <- run_GSEA(
  mediation_output = mediation_results_incl_sga,
  export_path = "./results/gene_mediation/top500/",
  export_name = "maternal_incl_sga",
  top_n = 500,
  gene_anno = gene_anno, 
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_incl_sga)){
  DT::datatable(kk_incl_sga@result)
}
```

```{r mat tri3 sga}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_tri3_sga.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_mat_tri3,
    timepoint_genecounts = gene_counts_mat_tri3,
    outcome_var = "sga_new"
    )
  mediation_results_tri3_sga <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "binomial",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_tri3_sga <- rbind(mediation_results_tri3_sga, ans)
  }
  write.csv(mediation_results_tri3_sga, "./results/gene_mediation/top500/mediation_results_tri3_sga.csv")
} else {
  mediation_results_tri3_sga <- read.csv("./results/gene_mediation/top500/mediation_results_tri3_sga.csv")
}
```

```{r GSEA mat tri3 sga, echo=FALSE}
kk_tri3_sga <- run_GSEA(
  mediation_output = mediation_results_tri3_sga,
  export_path = "./results/gene_mediation/top500/",
  export_name = "maternal_tri3_sga",
  top_n = 500,
  gene_anno = gene_anno,
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_tri3_sga)){
  DT::datatable(kk_tri3_sga@result)
}
```

### Infant microbiome mediation
#### of the BEP->HAZ effect
```{r inf pn12 haz3}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_pn12_haz3.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_inf_pn12,
    timepoint_genecounts = gene_counts_inf_pn12,
    outcome_var = "c_haz3",
    postnatal = TRUE
    )
  mediation_results_pn12_haz3 <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "gaussian",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_pn12_haz3 <- rbind(mediation_results_pn12_haz3, ans)
  }
  write.csv(mediation_results_pn12_haz3, "./results/gene_mediation/top500/mediation_results_pn12_haz3.csv")
} else {
  mediation_results_pn12_haz3 <- read.csv("./results/gene_mediation/top500/mediation_results_pn12_haz3.csv")
}
```

```{r GSEA inf pn12 haz3, echo=FALSE}
kk_inf_pn12_haz3 <- run_GSEA(
  mediation_output = mediation_results_pn12_haz3,
  export_path = "./results/gene_mediation/top500/",
  export_name = "infant_pn12_haz3",
  top_n = 500,
  gene_anno = gene_anno, 
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_inf_pn12_haz3)){
  DT::datatable(kk_inf_pn12_haz3@result)
}
```

```{r inf pn56 haz6}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_pn56_haz6.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_inf_pn56,
    timepoint_genecounts = gene_counts_inf_pn56,
    outcome_var = "c_haz6",
    postnatal = TRUE
    )
  mediation_results_pn56_haz6 <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    print((i/length(vars$mediator_names))*100)
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "gaussian",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_pn56_haz6 <- rbind(mediation_results_pn56_haz6, ans)
  }
  write.csv(mediation_results_pn56_haz6, "./results/gene_mediation/top500/mediation_results_pn56_haz6.csv")
} else {
  mediation_results_pn56_haz6 <- read.csv("./results/gene_mediation/top500/mediation_results_pn56_haz6.csv")
}
```

```{r GSEA inf pn12 haz6, echo=FALSE}
kk_inf_pn56_haz6 <- run_GSEA(
  mediation_output = mediation_results_pn56_haz6,
  export_path = "./results/gene_mediation/top500/",
  export_name = "infant_pn56_haz6",
  top_n = 500,
  gene_anno = gene_anno,
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_inf_pn56_haz6)){
  DT::datatable(kk_inf_pn56_haz6@result)
}
```

#### of the BEP->WAZ effect
```{r inf pn12 waz3}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_pn12_waz3.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_inf_pn12,
    timepoint_genecounts = gene_counts_inf_pn12,
    outcome_var = "c_waz3",
    postnatal = TRUE
    )
  mediation_results_pn12_waz3 <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "gaussian",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_pn12_waz3 <- rbind(mediation_results_pn12_waz3, ans)
  }
  write.csv(mediation_results_pn12_waz3, "./results/gene_mediation/top500/mediation_results_pn12_waz3.csv")
} else {
  mediation_results_pn12_waz3 <- read.csv("./results/gene_mediation/top500/mediation_results_pn12_waz3.csv")
}
```

```{r GSEA inf pn12 waz3, echo=FALSE}
kk_inf_pn12_waz3 <- run_GSEA(
  mediation_output = mediation_results_pn12_waz3,
  export_path = "./results/gene_mediation/top500/",
  export_name = "infant_pn12_waz3",
  top_n = 500,
  gene_anno = gene_anno,
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_inf_pn12_waz3)){
  DT::datatable(kk_inf_pn12_waz3@result)
}
```

```{r inf pn12 waz6}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_pn12_waz6.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_inf_pn12,
    timepoint_genecounts = gene_counts_inf_pn12,
    outcome_var = "c_waz6",
    postnatal = TRUE
    )
  mediation_results_pn12_waz6 <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "gaussian",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_pn12_waz6 <- rbind(mediation_results_pn12_waz6, ans)
  }
  write.csv(mediation_results_pn12_waz6, "./results/gene_mediation/top500/mediation_results_pn12_waz6.csv")
} else {
  mediation_results_pn12_waz6 <- read.csv("./results/gene_mediation/top500/mediation_results_pn12_waz6.csv")
}
```

```{r GSEA inf pn12 waz6, echo=FALSE}
kk_inf_pn12_waz6 <- run_GSEA(
  mediation_output = mediation_results_pn12_waz6,
  export_path = "./results/gene_mediation/top500/",
  export_name = "infant_pn12_waz6",
  top_n = 500,
  gene_anno = gene_anno,
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_inf_pn12_waz6)){
  DT::datatable(kk_inf_pn12_waz6@result)
}
```

```{r inf pn56 waz6}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_pn56_waz6.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_inf_pn56,
    timepoint_genecounts = gene_counts_inf_pn56,
    outcome_var = "c_waz6",
    postnatal = TRUE
    )
  mediation_results_pn56_waz6 <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "gaussian",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_pn56_waz6 <- rbind(mediation_results_pn56_waz6, ans)
  }
  write.csv(mediation_results_pn56_waz6, "./results/gene_mediation/top500/mediation_results_pn56_waz6.csv")
} else {
  mediation_results_pn56_waz6 <- read.csv("./results/gene_mediation/top500/mediation_results_pn56_waz6.csv")
}
```

```{r GSEA inf pn56 waz6, echo=FALSE}
kk_inf_pn56_waz6 <- run_GSEA(
  mediation_output = mediation_results_pn56_waz6,
  export_path = "./results/gene_mediation/top500/",
  export_name = "infant_pn56_waz6",
  top_n = 500,
  gene_anno = gene_anno,
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_inf_pn56_waz6)){
  DT::datatable(kk_inf_pn56_waz6@result)
}
```

#### of the BEP->WHZ effect
```{r mat pn12 whz3}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_pn12_whz3.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_inf_pn12,
    timepoint_genecounts = gene_counts_inf_pn12,
    outcome_var = "c_whz3",
    postnatal = TRUE
    )
  mediation_results_pn12_whz3 <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "gaussian",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_pn12_whz3 <- rbind(mediation_results_pn12_whz3, ans)
  }
  write.csv(mediation_results_pn12_whz3, "./results/gene_mediation/top500/mediation_results_pn12_whz3.csv")
} else {
  mediation_results_pn12_whz3 <- read.csv("./results/gene_mediation/top500/mediation_results_pn12_whz3.csv")
}
```

```{r GSEA inf pn12 whz3, echo=FALSE}
kk_inf_pn12_whz3 <- run_GSEA(
  mediation_output = mediation_results_pn12_whz3,
  export_path = "./results/gene_mediation/top500/",
  export_name = "infant_pn12_whz3",
  top_n = 500,
  gene_anno = gene_anno,
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_inf_pn12_whz3)){
  DT::datatable(kk_inf_pn12_whz3@result)
}
```

```{r inf pn56 whz6}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_pn56_whz6.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_inf_pn56,
    timepoint_genecounts = gene_counts_inf_pn56,
    outcome_var = "c_whz6",
    postnatal = TRUE
    )
  mediation_results_pn56_whz6 <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "gaussian",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_pn56_whz6 <- rbind(mediation_results_pn56_whz6, ans)
  }
  write.csv(mediation_results_pn56_whz6, "./results/gene_mediation/top500/mediation_results_pn56_whz6.csv")
} else {
  mediation_results_pn56_whz6 <- read.csv("./results/gene_mediation/top500/mediation_results_pn56_whz6.csv")
}
```

```{r GSEA inf pn56 whz6, echo=FALSE}
kk_inf_pn56_whz6 <- run_GSEA(
  mediation_output = mediation_results_pn56_whz6,
  export_path = "./results/gene_mediation/top500/",
  export_name = "infant_pn56_whz6",
  top_n = 500,
  gene_anno = gene_anno,
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_inf_pn56_whz6)){
  DT::datatable(kk_inf_pn56_whz6@result)
}
```

#### of the BEP->stunting effect
```{r inf pn12 stunting}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_pn12_stunting.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_inf_pn12,
    timepoint_genecounts = gene_counts_inf_pn12,
    outcome_var = "c_stunting6",
    postnatal = TRUE
    )
  mediation_results_pn12_stunting <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "binomial",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_pn12_stunting <- rbind(mediation_results_pn12_stunting, ans)
  }
  write.csv(mediation_results_pn12_stunting, "./results/gene_mediation/top500/mediation_results_pn12_stunting.csv")
} else {
  mediation_results_pn12_stunting <- read.csv("./results/gene_mediation/top500/mediation_results_pn12_stunting.csv")
}
```

```{r GSEA inf pn12 stunting, echo=FALSE}
kk_inf_pn12_stunting <- run_GSEA(
  mediation_output = mediation_results_pn12_stunting,
  export_path = "./results/gene_mediation/top500/",
  export_name = "infant_pn12_stunting",
  top_n = 500,
  gene_anno = gene_anno,
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_inf_pn12_stunting)){
  DT::datatable(kk_inf_pn12_stunting@result)
}
```

```{r inf pn56 stunting}
set.seed(123)
if (!file.exists("./results/gene_mediation/top500/mediation_results_pn56_stunting.csv")){
  vars <- extract_vars(
    timepoint_metadata = metadata_inf_pn56,
    timepoint_genecounts = gene_counts_inf_pn56,
    outcome_var = "c_stunting",
    postnatal = TRUE
    )
  mediation_results_pn56_stunting <- data.frame()
  for (i in seq_along(vars$mediator_names)){
    ans <- HDmediation::mediation(
      data = vars$df,
      A = "A", W = vars$covariate_names,
      Z = vars$mediator_names[-i],
      M = vars$mediator_names[i],
      Y = "Y", S = NULL,
      family = "binomial",
      folds = 1, partial_tmle = TRUE, bounds = NULL,
      learners_g = sl_library, learners_e = sl_library, learners_c = sl_library,
      learners_b = sl_library, learners_hz = sl_library, learners_u = sl_library,
      learners_ubar = sl_library, learners_v = sl_library, learners_vbar = sl_library
    )
    ans$mediator <- vars$mediator_names[i]
    mediation_results_pn56_stunting <- rbind(mediation_results_pn56_stunting, ans)
  }
  write.csv(mediation_results_pn56_stunting, "./results/gene_mediation/top500/mediation_results_pn56_stunting.csv")
} else {
  mediation_results_pn56_stunting <- read.csv("./results/gene_mediation/top500/mediation_results_pn56_stunting.csv")
}
```

```{r GSEA inf pn56 stunting, echo=FALSE}
kk_inf_pn56_stunting <- run_GSEA(
  mediation_output = mediation_results_pn56_stunting,
  export_path = "./results/gene_mediation/top500/",
  export_name = "infant_pn56_stunting",
  top_n = 500,
  gene_anno = gene_anno,
  pathway_to_gene = pathway_to_gene,
  pathway_to_description = pathway_to_description
  )
if (!is.null(kk_inf_pn56_stunting)){
  DT::datatable(kk_inf_pn56_stunting@result)
}
```


# 🔴 Revision 2
# baseline characteristics for full misame3
the following participant data can only be available after signing a data-shaing agreement. Any requests, please reach out carl.lachat@ugent.be
```{r}
merged_wide <- haven::read_dta("./Merged_wide.dta") # the MISAME-III full dataset

baseline <- haven::read_dta("./baseline.dta")

echo <- haven::read_dta("./echo.dta") %>%
  filter(!is.na(echo_ga_weeks))

compliance_all <- haven::read_dta("./compliance_all.dta")

compliance_corrections <- haven::read_dta("./compliance_corrections.dta")

metadata_for_sharing <- haven::read_dta("./metadata_for_sharing.dta")

merged_wide$d_cdob <- as.Date(merged_wide$d_cdob)

compliance_corrections$d_cdob <- as.Date(compliance_corrections$d_cdob)

compliance_all$date_visite[compliance_all$id_femme == "10005_4" & compliance_all$date_visite == "2020-01-04"] <- as.Date("2020-12-04")


compliance <- compliance_all %>%
  dplyr::select(id_femme, date_enquete, date_visite, supplem_avance, visite_effectue, prise_faf, prise_complement) %>%
  filter(
    !is.na(id_femme),  # Remove rows where id_femme is NA
    !id_femme %in% c(",40040_1", ",40067_1", ".60012_1", ""),
    visite_effectue != 1,
    visite_effectue != 0,
     prise_faf != 0,
     prise_faf != 3,
    prise_complement != 5
) %>%
  group_by(id_femme) %>% 
  slice_min(date_visite) %>%
  ungroup() %>%
  distinct() %>%
  mutate(idwoman = id_femme) %>%
  dplyr::select(idwoman, date_visite)

compliance_c <- compliance_corrections %>%
  dplyr::select(id_femme, date_visite_2, d_incl, visite_effectue, prise_faf, prise_complement, ga_days_incl, ga_weeks_incl, GAbirthdays, d_cdob) %>%
  filter(
    visite_effectue != 1,
    visite_effectue != 0,
     prise_faf != 0,
     prise_faf != 3,
    prise_complement != 5
) %>%
  group_by(id_femme) %>% 
  slice_min(date_visite_2) %>%
  ungroup() %>%
  distinct() %>%
  mutate(idwoman = id_femme) %>%
  dplyr::select(idwoman, date_visite_2, d_incl, ga_days_incl, ga_weeks_incl, GAbirthdays, d_cdob) %>%
  mutate(
    days_of_intervention = as.numeric(difftime(d_cdob, d_incl, units = "days")),
    ga_days_intervention = GAbirthdays - days_of_intervention)

incl_se1g_date_idbs <- incl_se1g_date %>%
  separate(sampleid, into = c("timepoint", "se1g", "idbs"), sep = "_", remove = FALSE) %>%
  mutate(incl_se1g_date = as.Date(date, format = "%y-%m-%d")) %>%
  dplyr::select(idbs, incl_se1g_date)

tri3_se1g_date_idbs <- tri3_se1g_date %>%
  separate(sampleid, into = c("timepoint", "se1g", "idbs"), sep = "_", remove = FALSE) %>%
  mutate(tri3_se1g_date = as.Date(date, format = "%y-%m-%d")) %>%
  dplyr::select(idbs, tri3_se1g_date)

pn12m_se1g_date_idbs <- pn12m_se1g_date %>%
  separate(sampleid, into = c("timepoint", "se1g", "idbs"), sep = "_", remove = FALSE) %>%
  mutate(pn12m_se1g_date = as.Date(date, format = "%y-%m-%d")) %>%
  dplyr::select(idbs, pn12m_se1g_date)

pn56m_se1g_date_idbs <- pn56m_se1g_date %>%
  separate(sampleid, into = c("timepoint", "se1g", "idbs"), sep = "_", remove = FALSE) %>%
  mutate(pn56m_se1g_date = as.Date(date, format = "%y-%m-%d")) %>%
  dplyr::select(idbs, pn56m_se1g_date)

pn12e_se1g_date_idbs <- pn12e_se1g_date %>%
  separate(sampleid, into = c("timepoint", "se1g", "idbs", "e"), sep = "_", remove = FALSE) %>%
  mutate(pn12e_se1g_date = as.Date(date, format = "%y-%m-%d")) %>%
  dplyr::select(idbs, pn12e_se1g_date)

pn56e_se1g_date_idbs <- pn56e_se1g_date %>%
  separate(sampleid, into = c("timepoint", "se1g", "idbs", "e"), sep = "_", remove = FALSE) %>%
  mutate(pn56e_se1g_date = as.Date(date, format = "%y-%m-%d")) %>%
  dplyr::select(idbs, pn56e_se1g_date)

duplicates <- compliance_c %>%
  dplyr::count(idwoman) %>%
  filter(n > 1)

print(duplicates)

compliance_c <- compliance_c %>%
  distinct(idwoman, .keep_all = TRUE)

baseline_full <- merged_wide %>%
  mutate(
    group = case_when(
      gi_treatmentl == 1 & gi_treatment2 == 1 ~ "BEP/BEP",
      gi_treatmentl == 1 & gi_treatment2 == 0 ~ "BEP/IFA",
      gi_treatmentl == 0 & gi_treatment2 == 1 ~ "IFA/BEP",
      gi_treatmentl == 0 & gi_treatment2 == 0 ~ "IFA/IFA",
      TRUE ~ "non")) %>%
  inner_join(baseline %>% dplyr::select(idwoman), by = "idwoman")  %>%
  inner_join(echo %>% dplyr::select(idwoman), by = "idwoman")  %>%
  left_join(compliance_c %>% 
              semi_join(baseline_full, by = "idwoman") %>%
              dplyr::select(idwoman, ga_weeks_incl), by = "idwoman") %>%
  left_join(metadata_for_sharing %>% dplyr::select(idwoman, idbs, code), by = "idwoman") %>%
  left_join(incl_se1g_date_idbs, by = "idbs") %>%
  left_join(tri3_se1g_date_idbs, by = "idbs") %>%
  left_join(pn12m_se1g_date_idbs, by = "idbs") %>%
  left_join(pn56m_se1g_date_idbs, by = "idbs") %>%
  left_join(pn12e_se1g_date_idbs, by = "idbs") %>%
  left_join(pn56e_se1g_date_idbs, by = "idbs") %>%
  mutate(
    days_of_intervention = as.numeric(difftime(d_cdob, incl_date_inclusion, units = "days")),
    ga_days_intervention = d_gabirth_d - days_of_intervention,
    incl_ga_weeks = ga_days_intervention/7,
    ga_weeks_tri2 = as.numeric(difftime(incl_se1g_date, incl_date_inclusion, units = "weeks")),
    ga_weeks_tri3 = as.numeric(difftime(tri3_se1g_date, incl_date_inclusion, units = "weeks")),
    ga_weeks_pn12m = as.numeric(difftime(pn12m_se1g_date, incl_date_inclusion, units = "weeks")),
    ga_weeks_pn56m = as.numeric(difftime(pn56m_se1g_date, incl_date_inclusion, units = "weeks")),
    ga_weeks_pn12e = as.numeric(difftime(pn12e_se1g_date, incl_date_inclusion, units = "weeks")),
    ga_weeks_pn56e = as.numeric(difftime(pn56e_se1g_date, incl_date_inclusion, units = "weeks")))

table(baseline_full$group)

table(baseline_full$gi_csps[baseline_full$group=="BEP/BEP"])
table(baseline_full$gi_csps[baseline_full$group=="BEP/IFA"])
table(baseline_full$gi_csps[baseline_full$group=="IFA/BEP"])
table(baseline_full$gi_csps[baseline_full$group=="IFA/IFA"])

chisq.test(table(baseline_full$gi_csps, baseline_full$group))$p.value

table(baseline_full$gi_csps[baseline_full$code=="B Z"])
table(baseline_full$gi_csps[baseline_full$code=="A Y"])

chisq.test(table(baseline_full$gi_csps, baseline_full$code))$p.value

table(baseline_full$hh_hfias[baseline_full$group=="BEP/BEP"])
table(baseline_full$hh_hfias[baseline_full$group=="BEP/IFA"])
table(baseline_full$hh_hfias[baseline_full$group=="IFA/BEP"])
table(baseline_full$hh_hfias[baseline_full$group=="IFA/IFA"])

chisq.test(table(baseline_full$hh_hfias, baseline_full$group))$p.value

table(baseline_full$hh_hfias[baseline_full$code=="B Z"])
table(baseline_full$hh_hfias[baseline_full$code=="A Y"])

fisher.test(
  table(baseline_full$hh_hfias, baseline_full$code),
  workspace = 2e8  # Increase workspace size (e.g., 200 MB)
)$p.value

mean(baseline_full$incl_hhasset1comp_10[baseline_full$group=="BEP/BEP"])
sd(baseline_full$incl_hhasset1comp_10[baseline_full$group=="BEP/BEP"])

mean(baseline_full$incl_hhasset1comp_10[baseline_full$group=="BEP/IFA"])
sd(baseline_full$incl_hhasset1comp_10[baseline_full$group=="BEP/IFA"])

mean(baseline_full$incl_hhasset1comp_10[baseline_full$group=="IFA/BEP"])
sd(baseline_full$incl_hhasset1comp_10[baseline_full$group=="IFA/BEP"])

mean(baseline_full$incl_hhasset1comp_10[baseline_full$group=="IFA/IFA"])
sd(baseline_full$incl_hhasset1comp_10[baseline_full$group=="IFA/IFA"])

summary(aov(incl_hhasset1comp_10 ~ group, data = baseline_full))[[1]][["Pr(>F)"]][1]

mean(baseline_full$incl_hhasset1comp_10[baseline_full$code=="B Z"], na.rm=TRUE)
sd(baseline_full$incl_hhasset1comp_10[baseline_full$code=="B Z"], na.rm=TRUE)

mean(baseline_full$incl_hhasset1comp_10[baseline_full$code=="A Y"], na.rm=TRUE)
sd(baseline_full$incl_hhasset1comp_10[baseline_full$code=="A Y"], na.rm=TRUE)

summary(aov(incl_hhasset1comp_10 ~ code, data = baseline_full))[[1]][["Pr(>F)"]][1]

mean(baseline_full$incl_wage[baseline_full$group=="BEP/BEP"])
sd(baseline_full$incl_wage[baseline_full$group=="BEP/BEP"])

mean(baseline_full$incl_wage[baseline_full$group=="BEP/IFA"])
sd(baseline_full$incl_wage[baseline_full$group=="BEP/IFA"])

mean(baseline_full$incl_wage[baseline_full$group=="IFA/BEP"])
sd(baseline_full$incl_wage[baseline_full$group=="IFA/BEP"])

mean(baseline_full$incl_wage[baseline_full$group=="IFA/IFA"])
sd(baseline_full$incl_wage[baseline_full$group=="IFA/IFA"])

summary(aov(incl_wage ~ group, data = baseline_full))[[1]][["Pr(>F)"]][1]

mean(baseline_full$incl_wage[baseline_full$code=="B Z"], na.rm=TRUE)
sd(baseline_full$incl_wage[baseline_full$code=="B Z"], na.rm=TRUE)

mean(baseline_full$incl_wage[baseline_full$code=="A Y"], na.rm=TRUE)
sd(baseline_full$incl_wage[baseline_full$code=="A Y"], na.rm=TRUE)

summary(aov(incl_wage ~ code, data = baseline_full))[[1]][["Pr(>F)"]][1]


table(baseline_full$incl_wetnicity[baseline_full$group=="BEP/BEP"])
table(baseline_full$incl_wetnicity[baseline_full$group=="BEP/IFA"])
table(baseline_full$incl_wetnicity[baseline_full$group=="IFA/BEP"])
table(baseline_full$incl_wetnicity[baseline_full$group=="IFA/IFA"])

fisher.test(
  table(baseline_full$incl_wetnicity, baseline_full$group),
  simulate.p.value = TRUE,  # Enable Monte Carlo simulation
  B = 1e6                  # Number of simulations
)$p.value

table(baseline_full$incl_wetnicity[baseline_full$code=="B Z"])
table(baseline_full$incl_wetnicity[baseline_full$code=="A Y"])

fisher.test(
  table(baseline_full$incl_wetnicity, baseline_full$code),
  workspace = 2e8  # Increase workspace size (e.g., 200 MB)
)$p.value


table(baseline_full$incl_wreligion[baseline_full$group=="BEP/BEP"])
table(baseline_full$incl_wreligion[baseline_full$group=="BEP/IFA"])
table(baseline_full$incl_wreligion[baseline_full$group=="IFA/BEP"])
table(baseline_full$incl_wreligion[baseline_full$group=="IFA/IFA"])

chisq.test(table(baseline_full$incl_wreligion, baseline_full$group))$p.value
fisher.test(
  table(baseline_full$incl_wreligion, baseline_full$group),
  simulate.p.value = TRUE,  # Enable Monte Carlo simulation
  B = 1e6                  # Number of simulations
)$p.value

table(baseline_full$incl_wreligion[baseline_full$code=="B Z"])
table(baseline_full$incl_wreligion[baseline_full$code=="A Y"])

chisq.test(table(baseline_full$incl_wreligion, baseline_full$code))$p.value
fisher.test(
  table(baseline_full$incl_wreligion, baseline_full$code),
  simulate.p.value = TRUE,  # Enable Monte Carlo simulation
  B = 1e6                  # Number of simulations
)$p.value


table(baseline_full$incl_wschool_level[baseline_full$group=="BEP/BEP"])
table(baseline_full$incl_wschool_level[baseline_full$group=="BEP/IFA"])
table(baseline_full$incl_wschool_level[baseline_full$group=="IFA/BEP"])
table(baseline_full$incl_wschool_level[baseline_full$group=="IFA/IFA"])

chisq.test(table(baseline_full$incl_wschool_level, baseline_full$group))$p.value
fisher.test(
  table(baseline_full$incl_wschool_level, baseline_full$group),
  simulate.p.value = TRUE,  # Enable Monte Carlo simulation
  B = 1e6                  # Number of simulations
)$p.value


table(baseline_full$incl_wschool_level[baseline_full$code=="B Z"])
table(baseline_full$incl_wschool_level[baseline_full$code=="A Y"])

chisq.test(table(baseline_full$incl_wschool_level, baseline_full$code))$p.value
fisher.test(
  table(baseline_full$incl_wschool_level, baseline_full$code),
  simulate.p.value = TRUE,  # Enable Monte Carlo simulation
  B = 1e6                  # Number of simulations
)$p.value


table(baseline_full$incl_wnr_jobs[baseline_full$group=="BEP/BEP"])
table(baseline_full$incl_wnr_jobs[baseline_full$group=="BEP/IFA"])
table(baseline_full$incl_wnr_jobs[baseline_full$group=="IFA/BEP"])
table(baseline_full$incl_wnr_jobs[baseline_full$group=="IFA/IFA"])

chisq.test(table(baseline_full$incl_wnr_jobs, baseline_full$group))$p.value
fisher.test(
  table(baseline_full$incl_wnr_jobs, baseline_full$group),
  simulate.p.value = TRUE,  # Enable Monte Carlo simulation
  B = 1e6                  # Number of simulations
)$p.value


table(baseline_full$incl_wnr_jobs[baseline_full$code=="B Z"])
table(baseline_full$incl_wnr_jobs[baseline_full$code=="A Y"])

chisq.test(table(baseline_full$incl_wnr_jobs, baseline_full$code))$p.value
fisher.test(
  table(baseline_full$incl_wnr_jobs, baseline_full$code),
  simulate.p.value = TRUE,  # Enable Monte Carlo simulation
  B = 1e6                  # Number of simulations
)$p.value


table(baseline_full$incl_wgravidity[baseline_full$group=="BEP/BEP"])
table(baseline_full$incl_wgravidity[baseline_full$group=="BEP/IFA"])
table(baseline_full$incl_wgravidity[baseline_full$group=="IFA/BEP"])
table(baseline_full$incl_wgravidity[baseline_full$group=="IFA/IFA"])

chisq.test(table(baseline_full$incl_wgravidity, baseline_full$group))$p.value


table(baseline_full$incl_wgravidity[baseline_full$code=="B Z"])
table(baseline_full$incl_wgravidity[baseline_full$code=="A Y"])

chisq.test(table(baseline_full$incl_wgravidity, baseline_full$code))$p.value


mean(baseline_full$w_mean_weightincl[baseline_full$group=="BEP/BEP"])
sd(baseline_full$w_mean_weightincl[baseline_full$group=="BEP/BEP"])

mean(baseline_full$w_mean_weightincl[baseline_full$group=="BEP/IFA"])
sd(baseline_full$w_mean_weightincl[baseline_full$group=="BEP/IFA"])

mean(baseline_full$w_mean_weightincl[baseline_full$group=="IFA/BEP"])
sd(baseline_full$w_mean_weightincl[baseline_full$group=="IFA/BEP"])

mean(baseline_full$w_mean_weightincl[baseline_full$group=="IFA/IFA"])
sd(baseline_full$w_mean_weightincl[baseline_full$group=="IFA/IFA"])

summary(aov(w_mean_weightincl ~ group, data = baseline_full))[[1]][["Pr(>F)"]][1]


mean(baseline_full$w_mean_weightincl[baseline_full$code=="B Z"], na.rm=TRUE)
sd(baseline_full$w_mean_weightincl[baseline_full$code=="B Z"], na.rm=TRUE)

mean(baseline_full$w_mean_weightincl[baseline_full$code=="A Y"], na.rm=TRUE)
sd(baseline_full$w_mean_weightincl[baseline_full$code=="A Y"], na.rm=TRUE)

summary(aov(w_mean_weightincl ~ code, data = baseline_full))[[1]][["Pr(>F)"]][1]


mean(baseline_full$w_mean_heightincl[baseline_full$group=="BEP/BEP"])
sd(baseline_full$w_mean_heightincl[baseline_full$group=="BEP/BEP"])

mean(baseline_full$w_mean_heightincl[baseline_full$group=="BEP/IFA"])
sd(baseline_full$w_mean_heightincl[baseline_full$group=="BEP/IFA"])

mean(baseline_full$w_mean_heightincl[baseline_full$group=="IFA/BEP"])
sd(baseline_full$w_mean_heightincl[baseline_full$group=="IFA/BEP"])

mean(baseline_full$w_mean_heightincl[baseline_full$group=="IFA/IFA"], na.rm=TRUE)
sd(baseline_full$w_mean_heightincl[baseline_full$group=="IFA/IFA"], na.rm=TRUE)

summary(aov(w_mean_heightincl ~ group, data = baseline_full))[[1]][["Pr(>F)"]][1]


mean(baseline_full$w_mean_heightincl[baseline_full$code=="B Z"], na.rm=TRUE)
sd(baseline_full$w_mean_heightincl[baseline_full$code=="B Z"], na.rm=TRUE)

mean(baseline_full$w_mean_heightincl[baseline_full$code=="A Y"], na.rm=TRUE)
sd(baseline_full$w_mean_heightincl[baseline_full$code=="A Y"], na.rm=TRUE)

summary(aov(w_mean_heightincl ~ code, data = baseline_full))[[1]][["Pr(>F)"]][1]


mean(baseline_full$w_bmiincl [baseline_full$group=="BEP/BEP"])
sd(baseline_full$w_bmiincl [baseline_full$group=="BEP/BEP"])

mean(baseline_full$w_bmiincl [baseline_full$group=="BEP/IFA"])
sd(baseline_full$w_bmiincl [baseline_full$group=="BEP/IFA"])

mean(baseline_full$w_bmiincl [baseline_full$group=="IFA/BEP"])
sd(baseline_full$w_bmiincl [baseline_full$group=="IFA/BEP"])

mean(baseline_full$w_bmiincl [baseline_full$group=="IFA/IFA"], na.rm=TRUE)
sd(baseline_full$w_bmiincl [baseline_full$group=="IFA/IFA"], na.rm=TRUE)

summary(aov(w_bmiincl ~ group, data = baseline_full))[[1]][["Pr(>F)"]][1]


mean(baseline_full$w_bmiincl[baseline_full$code=="B Z"], na.rm=TRUE)
sd(baseline_full$w_bmiincl[baseline_full$code=="B Z"], na.rm=TRUE)

mean(baseline_full$w_bmiincl[baseline_full$code=="A Y"], na.rm=TRUE)
sd(baseline_full$w_bmiincl[baseline_full$code=="A Y"], na.rm=TRUE)

summary(aov(w_bmiincl ~ code, data = baseline_full))[[1]][["Pr(>F)"]][1]


mean(baseline_full$w_hbincl[baseline_full$group=="BEP/BEP"])
sd(baseline_full$w_hbincl[baseline_full$group=="BEP/BEP"])

mean(baseline_full$w_hbincl[baseline_full$group=="BEP/IFA"])
sd(baseline_full$w_hbincl[baseline_full$group=="BEP/IFA"])

mean(baseline_full$w_hbincl[baseline_full$group=="IFA/BEP"])
sd(baseline_full$w_hbincl[baseline_full$group=="IFA/BEP"])

mean(baseline_full$w_hbincl[baseline_full$group=="IFA/IFA"])
sd(baseline_full$w_hbincl[baseline_full$group=="IFA/IFA"])

summary(aov(w_hbincl ~ group, data = baseline_full))[[1]][["Pr(>F)"]][1]


mean(baseline_full$w_hbincl[baseline_full$code=="B Z"], na.rm=TRUE)
sd(baseline_full$w_hbincl[baseline_full$code=="B Z"], na.rm=TRUE)

mean(baseline_full$w_hbincl[baseline_full$code=="A Y"], na.rm=TRUE)
sd(baseline_full$w_hbincl[baseline_full$code=="A Y"], na.rm=TRUE)

summary(aov(w_hbincl ~ code, data = baseline_full))[[1]][["Pr(>F)"]][1]


table(baseline_full$w_anemia_incl[baseline_full$group=="BEP/BEP"])
table(baseline_full$w_anemia_incl[baseline_full$group=="BEP/IFA"])
table(baseline_full$w_anemia_incl[baseline_full$group=="IFA/BEP"])
table(baseline_full$w_anemia_incl[baseline_full$group=="IFA/IFA"])

chisq.test(table(baseline_full$w_anemia_incl, baseline_full$group))$p.value


table(baseline_full$w_anemia_incl[baseline_full$code=="B Z"])
table(baseline_full$w_anemia_incl[baseline_full$code=="A Y"])

chisq.test(table(baseline_full$w_anemia_incl, baseline_full$code))$p.value


mean(baseline_full$d_gabirth_wk[baseline_full$group=="BEP/BEP"], na.rm=TRUE)
sd(baseline_full$d_gabirth_wk[baseline_full$group=="BEP/BEP"], na.rm=TRUE)

mean(baseline_full$d_gabirth_wk[baseline_full$group=="BEP/IFA"], na.rm=TRUE)
sd(baseline_full$d_gabirth_wk[baseline_full$group=="BEP/IFA"], na.rm=TRUE)

mean(baseline_full$d_gabirth_wk[baseline_full$group=="IFA/BEP"], na.rm=TRUE)
sd(baseline_full$d_gabirth_wk[baseline_full$group=="IFA/BEP"], na.rm=TRUE)

mean(baseline_full$d_gabirth_wk[baseline_full$group=="IFA/IFA"], na.rm=TRUE)
sd(baseline_full$d_gabirth_wk[baseline_full$group=="IFA/IFA"], na.rm=TRUE)

summary(aov(d_gabirth_wk ~ group, data = baseline_full))[[1]][["Pr(>F)"]][1]


mean(baseline_full$d_gabirth_wk[baseline_full$code=="B Z"], na.rm=TRUE)
sd(baseline_full$d_gabirth_wk[baseline_full$code=="B Z"], na.rm=TRUE)

mean(baseline_full$d_gabirth_wk[baseline_full$code=="A Y"], na.rm=TRUE)
sd(baseline_full$d_gabirth_wk[baseline_full$code=="A Y"], na.rm=TRUE)

summary(aov(d_gabirth_wk ~ code, data = baseline_full))[[1]][["Pr(>F)"]][1]


mean(baseline_full$incl_ga_weeks[baseline_full$group=="BEP/BEP"], na.rm=TRUE)
sd(baseline_full$incl_ga_weeks[baseline_full$group=="BEP/BEP"], na.rm=TRUE)

mean(baseline_full$incl_ga_weeks[baseline_full$group=="BEP/IFA"], na.rm=TRUE)
sd(baseline_full$incl_ga_weeks[baseline_full$group=="BEP/IFA"], na.rm=TRUE)

mean(baseline_full$incl_ga_weeks[baseline_full$group=="IFA/BEP"], na.rm=TRUE)
sd(baseline_full$incl_ga_weeks[baseline_full$group=="IFA/BEP"], na.rm=TRUE)

mean(baseline_full$incl_ga_weeks[baseline_full$group=="IFA/IFA"], na.rm=TRUE)
sd(baseline_full$incl_ga_weeks[baseline_full$group=="IFA/IFA"], na.rm=TRUE)

summary(aov(incl_ga_weeks ~ group, data = baseline_full))[[1]][["Pr(>F)"]][1]


mean(metadata$GAinclweeks[metadata$code=="B Z"], na.rm=TRUE)
sd(metadata$GAinclweeks[metadata$code=="B Z"], na.rm=TRUE)

mean(metadata$GAinclweeks[metadata$code=="A Y"], na.rm=TRUE)
sd(metadata$GAinclweeks[metadata$code=="A Y"], na.rm=TRUE)

summary(aov(GAinclweeks ~ code, data = metadata))[[1]][["Pr(>F)"]][1]

mean(baseline_full$incl_ga_weeks[baseline_full$code=="B Z" | baseline_full$code=="A Y"], na.rm=TRUE)
sd(baseline_full$incl_ga_weeks[baseline_full$code=="B Z" | baseline_full$code=="A Y"], na.rm=TRUE)

mean(baseline_full$incl_ga_weeks[baseline_full$code=="B Z"], na.rm=TRUE)
sd(baseline_full$incl_ga_weeks[baseline_full$code=="B Z"], na.rm=TRUE)

mean(baseline_full$incl_ga_weeks[baseline_full$code=="A Y"], na.rm=TRUE)
sd(baseline_full$incl_ga_weeks[baseline_full$code=="A Y"], na.rm=TRUE)

summary(aov(incl_ga_weeks ~ code, data = baseline_full))[[1]][["Pr(>F)"]][1]

# length of intervention
mean(baseline_full$ga_weeks_tri2[baseline_full$code=="B Z"], na.rm=TRUE)
sd(baseline_full$ga_weeks_tri2[baseline_full$code=="B Z"], na.rm=TRUE)

mean(baseline_full$ga_weeks_tri2[baseline_full$code=="A Y"], na.rm=TRUE)
sd(baseline_full$ga_weeks_tri2[baseline_full$code=="A Y"], na.rm=TRUE)

summary(aov(ga_weeks_tri2 ~ code, data = baseline_full))[[1]][["Pr(>F)"]][1]


mean(baseline_full$ga_weeks_tri3[baseline_full$code=="B Z"], na.rm=TRUE)
sd(baseline_full$ga_weeks_tri3[baseline_full$code=="B Z"], na.rm=TRUE)

mean(baseline_full$ga_weeks_tri3[baseline_full$code=="A Y"], na.rm=TRUE)
sd(baseline_full$ga_weeks_tri3[baseline_full$code=="A Y"], na.rm=TRUE)

summary(aov(ga_weeks_tri3 ~ code, data = baseline_full))[[1]][["Pr(>F)"]][1]


mean(baseline_full$ga_weeks_pn12m[baseline_full$code=="B Z"], na.rm=TRUE)
sd(baseline_full$ga_weeks_pn12m[baseline_full$code=="B Z"], na.rm=TRUE)

mean(baseline_full$ga_weeks_pn12m[baseline_full$code=="A Y"], na.rm=TRUE)
sd(baseline_full$ga_weeks_pn12m[baseline_full$code=="A Y"], na.rm=TRUE)

summary(aov(ga_weeks_pn12m ~ code, data = baseline_full))[[1]][["Pr(>F)"]][1]


mean(baseline_full$ga_weeks_pn56m[baseline_full$code=="B Z"], na.rm=TRUE)
sd(baseline_full$ga_weeks_pn56m[baseline_full$code=="B Z"], na.rm=TRUE)

mean(baseline_full$ga_weeks_pn56m[baseline_full$code=="A Y"], na.rm=TRUE)
sd(baseline_full$ga_weeks_pn56m[baseline_full$code=="A Y"], na.rm=TRUE)

summary(aov(ga_weeks_pn56m ~ code, data = baseline_full))[[1]][["Pr(>F)"]][1]


mean(baseline_full$ga_weeks_pn12e[baseline_full$code=="B Z"], na.rm=TRUE)
sd(baseline_full$ga_weeks_pn12e[baseline_full$code=="B Z"], na.rm=TRUE)

mean(baseline_full$ga_weeks_pn12e[baseline_full$code=="A Y"], na.rm=TRUE)
sd(baseline_full$ga_weeks_pn12e[baseline_full$code=="A Y"], na.rm=TRUE)

summary(aov(ga_weeks_pn12e ~ code, data = baseline_full))[[1]][["Pr(>F)"]][1]


mean(baseline_full$ga_weeks_pn56e[baseline_full$code=="B Z"], na.rm=TRUE)
sd(baseline_full$ga_weeks_pn56e[baseline_full$code=="B Z"], na.rm=TRUE)

mean(baseline_full$ga_weeks_pn56e[baseline_full$code=="A Y"], na.rm=TRUE)
sd(baseline_full$ga_weeks_pn56e[baseline_full$code=="A Y"], na.rm=TRUE)

summary(aov(ga_weeks_pn56e ~ code, data = baseline_full))[[1]][["Pr(>F)"]][1]

```


